<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Example analyses with atlas1006 data • microViz</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../../bootstrap-toc.css">
<script src="../../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><!-- docsearch --><script src="../../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><meta property="og:title" content="Example analyses with atlas1006 data">
<meta property="og:description" content="microViz">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-2YQ669S4V5"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-2YQ669S4V5');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">microViz</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.8.0.9002</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/web-only/phyloseq.html">Working with phyloseq objects</a>
    </li>
    <li>
      <a href="../../articles/web-only/tax-fixing.html">Fixing your tax_table</a>
    </li>
    <li>
      <a href="../../articles/web-only/ordination.html">Ordination plots</a>
    </li>
    <li>
      <a href="../../articles/web-only/ordination-interactive.html">Interactive ordination plots</a>
    </li>
    <li>
      <a href="../../articles/web-only/compositions.html">Visualising microbiota compositions</a>
    </li>
    <li>
      <a href="../../articles/web-only/modelling-taxa.html">Statistical modelling of individual taxa</a>
    </li>
    <li>
      <a href="../../articles/web-only/videos.html">Tutorial video</a>
    </li>
    <li>
      <a href="../../articles/web-only/atlas1006.html">Example analyses with atlas1006 data</a>
    </li>
  </ul>
</li>
<li>
  <a href="../../news/index.html">News</a>
</li>
<li class="dropdown-header">
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/david-barnett/microViz/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="atlas1006_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Example analyses with atlas1006 data</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/david-barnett/microViz/blob/master/vignettes/web-only/atlas1006.Rmd"><code>vignettes/web-only/atlas1006.Rmd</code></a></small>
      <div class="hidden name"><code>atlas1006.Rmd</code></div>

    </div>

    
    
<p>This document shows an example <code>microViz</code> analysis workflow using example data <code>atlas1006</code> from the <code>microbiome</code> package.</p>
<div id="load-necessary-packages" class="section level2">
<h2 class="hasAnchor">
<a href="#load-necessary-packages" class="anchor"></a>Load necessary packages</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com">patchwork</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://dx.plos.org/10.1371/journal.pone.0061217">phyloseq</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://microbiome.github.io/microbiome">microbiome</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://david-barnett.github.io/microViz">microViz</a></span><span class="op">)</span></code></pre></div>
</div>
<div id="get-example-data" class="section level2">
<h2 class="hasAnchor">
<a href="#get-example-data" class="anchor"></a>Get example data</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get some example data</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"atlas1006"</span>, package <span class="op">=</span> <span class="st">"microbiome"</span><span class="op">)</span>

<span class="co"># create a couple of numerical variables (often more useful than character strings)</span>
<span class="va">ps</span> <span class="op">&lt;-</span> <span class="va">atlas1006</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span>
  <span class="fu"><a href="../../reference/ps_mutate.html">ps_mutate</a></span><span class="op">(</span>
    weight <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/recode.html">recode</a></span><span class="op">(</span>
      .x <span class="op">=</span> <span class="va">bmi_group</span>, morbidobese <span class="op">=</span> <span class="fl">5</span>, severeobese <span class="op">=</span> <span class="fl">4</span>, 
      obese <span class="op">=</span> <span class="fl">3</span>, overweight <span class="op">=</span> <span class="fl">2</span>, lean <span class="op">=</span> <span class="fl">1</span>, underweight <span class="op">=</span> <span class="fl">0</span>
    <span class="op">)</span>,
    lean <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html">if_else</a></span><span class="op">(</span><span class="va">weight</span> <span class="op">&lt;</span> <span class="fl">2</span>, true <span class="op">=</span> <span class="fl">1</span>, false <span class="op">=</span> <span class="fl">0</span>, missing <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>,
    female <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html">if_else</a></span><span class="op">(</span><span class="va">sex</span> <span class="op">==</span> <span class="st">"female"</span>, true <span class="op">=</span> <span class="fl">1</span>, false <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>,
    extract_P <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html">if_else</a></span><span class="op">(</span><span class="va">DNA_extraction_method</span> <span class="op">==</span> <span class="st">"p"</span>, true <span class="op">=</span> <span class="fl">1</span>, false <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>
  <span class="op">)</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span>
  <span class="co"># only look at the baseline time point if multiple samples available</span>
  <span class="co"># and drop samples with no DNA extraction method info</span>
  <span class="fu"><a href="../../reference/ps_filter.html">ps_filter</a></span><span class="op">(</span><span class="va">time</span> <span class="op">==</span> <span class="fl">0</span>, <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">DNA_extraction_method</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span>
  <span class="co"># remove the sample data variables about time</span>
  <span class="fu"><a href="../../reference/ps_select.html">ps_select</a></span><span class="op">(</span><span class="op">-</span><span class="va">time</span><span class="op">)</span>

<span class="co"># add a couple of missing values to show how microViz handles missing data</span>
<span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/sample_data-methods.html">sample_data</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span><span class="op">$</span><span class="va">female</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">4</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">NA</span>

<span class="co"># look at phyloseq object description</span>
<span class="va">ps</span>
<span class="co">#&gt; phyloseq-class experiment-level object</span>
<span class="co">#&gt; otu_table()   OTU Table:         [ 125 taxa and 801 samples ]</span>
<span class="co">#&gt; sample_data() Sample Data:       [ 801 samples by 13 sample variables ]</span>
<span class="co">#&gt; tax_table()   Taxonomy Table:    [ 125 taxa by 3 taxonomic ranks ]</span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># this example data has a slightly odd tax_table because it comes from HITChip data (instead of sequencing)</span>
<span class="co"># the taxonomic assignment is done differently, so we need to ensure taxon names are not repeated across columns</span>
<span class="co"># it can otherwise be used the same as sequencing data for this example</span>
<span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/tax_table-methods.html">tax_table</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span>
<span class="co">#&gt; Taxonomy Table:     [3 taxa by 3 taxonomic ranks]:</span>
<span class="co">#&gt;                  Phylum            Family            Genus             </span>
<span class="co">#&gt; Actinomycetaceae "Actinobacteria"  "Actinobacteria"  "Actinomycetaceae"</span>
<span class="co">#&gt; Aerococcus       "Firmicutes"      "Bacilli"         "Aerococcus"      </span>
<span class="co">#&gt; Akkermansia      "Verrucomicrobia" "Verrucomicrobia" "Akkermansia"</span>
<span class="va">ps</span> <span class="op">&lt;-</span> <span class="va">ps</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span> <span class="fu"><a href="../../reference/tax_prepend_ranks.html">tax_prepend_ranks</a></span><span class="op">(</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/tax_table-methods.html">tax_table</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span>
<span class="co">#&gt; Taxonomy Table:     [3 taxa by 3 taxonomic ranks]:</span>
<span class="co">#&gt;                  Phylum               Family               Genus                </span>
<span class="co">#&gt; Actinomycetaceae "P: Actinobacteria"  "F: Actinobacteria"  "G: Actinomycetaceae"</span>
<span class="co">#&gt; Aerococcus       "P: Firmicutes"      "F: Bacilli"         "G: Aerococcus"      </span>
<span class="co">#&gt; Akkermansia      "P: Verrucomicrobia" "F: Verrucomicrobia" "G: Akkermansia"</span>
<span class="co"># replace any uninformative tax_table values</span>
<span class="va">ps</span> <span class="op">&lt;-</span> <span class="va">ps</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span> <span class="fu"><a href="../../reference/tax_fix.html">tax_fix</a></span><span class="op">(</span><span class="op">)</span>

<span class="co"># look at the effect of removing rare Genera, e.g. how many Genera are present in at least 5% of samples?</span>
<span class="va">ps</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span> <span class="fu"><a href="../../reference/tax_filter.html">tax_filter</a></span><span class="op">(</span>min_prevalence <span class="op">=</span> <span class="fl">5</span> <span class="op">/</span> <span class="fl">100</span>, tax_level <span class="op">=</span> <span class="st">"Genus"</span><span class="op">)</span>
<span class="co">#&gt; Proportional min_prevalence given: 0.05 --&gt; min 41/801 samples.</span>
<span class="co">#&gt; phyloseq-class experiment-level object</span>
<span class="co">#&gt; otu_table()   OTU Table:         [ 117 taxa and 801 samples ]</span>
<span class="co">#&gt; sample_data() Sample Data:       [ 801 samples by 13 sample variables ]</span>
<span class="co">#&gt; tax_table()   Taxonomy Table:    [ 117 taxa by 3 taxonomic ranks ]</span>
<span class="co"># we will use this udring other analyses, but not overwrite the original phyloseq object as the</span>
<span class="co"># unfiltered set of taxa would be required if we were performing e.g. alpha diversity analyses</span></code></pre></div>
</div>
<div id="exploring-your-data" class="section level2">
<h2 class="hasAnchor">
<a href="#exploring-your-data" class="anchor"></a>Exploring your data</h2>
<div id="ordination-plots" class="section level3">
<h3 class="hasAnchor">
<a href="#ordination-plots" class="anchor"></a>Ordination plots</h3>
<p>Ordination methods can help you to visualize variation in overall microbial ecosystem composition, and look at whether it might differ markedly between groups, e.g. weight.</p>
<p>Here is one option to try first:</p>
<ol style="list-style-type: decimal">
<li>Filter out rare taxa (e.g. remove Genera not present in at least 5% of samples) - use <code><a href="../../reference/tax_filter.html">tax_filter()</a></code>
</li>
<li>Aggregate the taxa into bacterial families (for example) - use <code><a href="../../reference/tax_agg.html">tax_agg()</a></code>
</li>
<li>Transform the microbial data with the centered-log-ratio transformation - use <code><a href="../../reference/tax_transform.html">tax_transform()</a></code>
</li>
<li>Perform PCA with the CLR-transformed features (equivalent to Aitchison distance PCoA) - use <code><a href="../../reference/ord_calc.html">ord_calc()</a></code>
</li>
<li>Plot the first 2 axes of this PCA ordination, colouring samples by group and adding taxon loading arrows to visualise which taxa generally differ across your samples - use <code><a href="../../reference/ord_plot.html">ord_plot()</a></code>
</li>
<li>Customise the theme of the ggplot as you like and/or add features like ellipses or annotations</li>
</ol>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">clr_pca</span> <span class="op">&lt;-</span> <span class="va">ps</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span>
  <span class="fu"><a href="../../reference/tax_filter.html">tax_filter</a></span><span class="op">(</span>min_prevalence <span class="op">=</span> <span class="fl">5</span> <span class="op">/</span> <span class="fl">100</span>, tax_level <span class="op">=</span> <span class="st">"Genus"</span><span class="op">)</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span>
  <span class="fu"><a href="../../reference/tax_agg.html">tax_agg</a></span><span class="op">(</span><span class="st">"Genus"</span><span class="op">)</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span> <span class="co"># aggregate taxa at Genus level</span>
  <span class="fu"><a href="../../reference/tax_transform.html">tax_transform</a></span><span class="op">(</span><span class="st">"clr"</span><span class="op">)</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span> <span class="co"># centered log ratio transformation</span>
  <span class="fu"><a href="../../reference/ord_calc.html">ord_calc</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span> <span class="co"># Note: PCA is RDA without constraints (&amp; ord_calc uses an RDA method to perform PCA)</span>
<span class="co">#&gt; Proportional min_prevalence given: 0.05 --&gt; min 41/801 samples.</span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">clr_pca</span>
<span class="co">#&gt; ps_extra object - a list with phyloseq and extras:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; phyloseq-class experiment-level object</span>
<span class="co">#&gt; otu_table()   OTU Table:         [ 117 taxa and 801 samples ]</span>
<span class="co">#&gt; sample_data() Sample Data:       [ 801 samples by 13 sample variables ]</span>
<span class="co">#&gt; tax_table()   Taxonomy Table:    [ 117 taxa by 3 taxonomic ranks ]</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; ps_extra info:</span>
<span class="co">#&gt; tax_agg = Genus tax_transform = clr</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; ordination of class: rda cca </span>
<span class="co">#&gt; rda(formula = OTU ~ 1, data = data)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $counts OTU Table: [ 117 taxa and 801 samples ]</span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">clr_pca</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span>
  <span class="fu"><a href="../../reference/ord_plot.html">ord_plot</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="st">"weight"</span>, shape <span class="op">=</span> <span class="st">"DNA_extraction_method"</span>, alpha <span class="op">=</span> <span class="fl">0.7</span>, size <span class="op">=</span> <span class="fl">1.5</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html">scale_colour_viridis_c</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">"inferno"</span>, direction <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>, end <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_shape_manual</a></span><span class="op">(</span>
    values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>o <span class="op">=</span> <span class="st">"square open"</span>, r <span class="op">=</span> <span class="st">"triangle open"</span>, p <span class="op">=</span> <span class="st">"circle"</span><span class="op">)</span>,
    name <span class="op">=</span> <span class="st">"DNA\nextraction\nmethod"</span>
  <span class="op">)</span></code></pre></div>
<p><img src="atlas1006_files/figure-html/visualize%20clr%20pca-1.png" width="700"></p>
</div>
<div id="taxonomic-compositions" class="section level3">
<h3 class="hasAnchor">
<a href="#taxonomic-compositions" class="anchor"></a>Taxonomic compositions?</h3>
<p>Using a PCA ordination allows us reliably draw biplots, showing which taxa are associated with this major variation in sample microbiota composition as represented on axes PC1 and PC2.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pca</span> <span class="op">&lt;-</span> <span class="va">clr_pca</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span>
  <span class="fu"><a href="../../reference/ord_plot.html">ord_plot</a></span><span class="op">(</span>
    colour <span class="op">=</span> <span class="st">"weight"</span>, shape <span class="op">=</span> <span class="st">"DNA_extraction_method"</span>, alpha <span class="op">=</span> <span class="fl">0.7</span>, size <span class="op">=</span> <span class="fl">1</span>,
    plot_taxa <span class="op">=</span> <span class="fl">12</span><span class="op">:</span><span class="fl">1</span>, tax_vec_length <span class="op">=</span> <span class="fl">0.3</span>, 
    taxon_renamer <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_remove.html">str_remove_all</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">"^G: | [ae]t rel."</span><span class="op">)</span>,
    center <span class="op">=</span> <span class="cn">TRUE</span>,
    tax_lab_style <span class="op">=</span> <span class="fu"><a href="../../reference/Ordination-labels.html">tax_lab_style</a></span><span class="op">(</span>
      type <span class="op">=</span> <span class="st">"label"</span>, max_angle <span class="op">=</span> <span class="fl">90</span>, fontface <span class="op">=</span> <span class="st">"bold"</span>, 
      alpha <span class="op">=</span> <span class="fl">0.8</span>, size <span class="op">=</span> <span class="fl">2</span>
    <span class="op">)</span>
  <span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html">scale_colour_viridis_c</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">"inferno"</span>, direction <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>, end <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_shape_manual</a></span><span class="op">(</span>
    values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>o <span class="op">=</span> <span class="st">"square open"</span>, r <span class="op">=</span> <span class="st">"triangle open"</span>, p <span class="op">=</span> <span class="st">"circle"</span><span class="op">)</span>,
    name <span class="op">=</span> <span class="st">"DNA\nextraction\nmethod"</span>
  <span class="op">)</span> <span class="op">+</span>
  <span class="co"># essential for correct label angles</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html">coord_fixed</a></span><span class="op">(</span>clip <span class="op">=</span> <span class="st">"off"</span>, xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1.5</span>, <span class="fl">1.5</span><span class="op">)</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1.5</span>, <span class="fl">1.5</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pca</span></code></pre></div>
<p><img src="atlas1006_files/figure-html/plot%20clr%20pca%20with%20tax%20loading-1.png" width="768"></p>
<p>What about the actual compositions of these 800 samples?</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">iris</span> <span class="op">&lt;-</span> <span class="va">clr_pca</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span>
  <span class="fu"><a href="../../reference/ord_plot_iris.html">ord_plot_iris</a></span><span class="op">(</span>
    tax_level <span class="op">=</span> <span class="st">"Genus"</span>, n_taxa <span class="op">=</span> <span class="fl">15</span>,
    anno_binary <span class="op">=</span> <span class="st">"extract_P"</span>,
    anno_binary_style <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>y <span class="op">=</span> <span class="fl">1.1</span>, size <span class="op">=</span> <span class="fl">0.5</span>, alpha <span class="op">=</span> <span class="fl">0.3</span>, shape <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>,
    taxon_renamer <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_remove.html">str_remove_all</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">"^G: | [ae]t rel."</span><span class="op">)</span>
  <span class="op">)</span>

<span class="va">iris</span></code></pre></div>
<p><img src="atlas1006_files/figure-html/iris%20plot-1.png" width="960"></p>
<p>You can arrange both plots together with patchwork package.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pca</span> <span class="op">/</span> <span class="va">iris</span> </code></pre></div>
<p><img src="atlas1006_files/figure-html/paired%20plots-1.png" width="960"></p>
</div>
</div>
<div id="testing-hypotheses" class="section level2">
<h2 class="hasAnchor">
<a href="#testing-hypotheses" class="anchor"></a>Testing hypotheses</h2>
<div id="permanova" class="section level3">
<h3 class="hasAnchor">
<a href="#permanova" class="anchor"></a>PERMANOVA</h3>
<p>What variables is the overall microbial community variation associated with?</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ps</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span>
  <span class="fu"><a href="../../reference/tax_filter.html">tax_filter</a></span><span class="op">(</span>min_prevalence <span class="op">=</span> <span class="fl">5</span> <span class="op">/</span> <span class="fl">100</span>, tax_level <span class="op">=</span> <span class="st">"Genus"</span><span class="op">)</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span>
  <span class="fu"><a href="../../reference/tax_agg.html">tax_agg</a></span><span class="op">(</span><span class="st">"Genus"</span><span class="op">)</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span>
  <span class="fu"><a href="../../reference/dist_calc.html">dist_calc</a></span><span class="op">(</span><span class="st">"aitchison"</span><span class="op">)</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span>
  <span class="fu"><a href="../../reference/dist_permanova.html">dist_permanova</a></span><span class="op">(</span>
    variables <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"DNA_extraction_method"</span>, <span class="st">"weight"</span>, <span class="st">"female"</span><span class="op">)</span>,
    n_perms <span class="op">=</span> <span class="fl">99</span>, <span class="co"># this is a low number of permutations for speed, you should set more e.g. 9999</span>
    seed <span class="op">=</span> <span class="fl">12345</span>, complete_cases <span class="op">=</span> <span class="cn">TRUE</span>, verbose <span class="op">=</span> <span class="st">"max"</span>
  <span class="op">)</span>
<span class="co">#&gt; Proportional min_prevalence given: 0.05 --&gt; min 41/801 samples.</span>
<span class="co">#&gt; Dropping samples with missings: 17</span>
<span class="co">#&gt; weight has NAs: 15</span>
<span class="co">#&gt; female has NAs: 2</span>
<span class="co">#&gt; 2021-11-13 13:27:21 - Starting PERMANOVA with 99 perms with 1 processes</span>
<span class="co">#&gt; 2021-11-13 13:27:36 - Finished PERMANOVA</span>
<span class="co">#&gt; ps_extra object - a list with phyloseq and extras:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; phyloseq-class experiment-level object</span>
<span class="co">#&gt; otu_table()   OTU Table:         [ 117 taxa and 784 samples ]</span>
<span class="co">#&gt; sample_data() Sample Data:       [ 784 samples by 13 sample variables ]</span>
<span class="co">#&gt; tax_table()   Taxonomy Table:    [ 117 taxa by 3 taxonomic ranks ]</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; ps_extra info:</span>
<span class="co">#&gt; tax_agg = Genus tax_transform = NA</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; aitchison distance matrix of size 784 </span>
<span class="co">#&gt; 10.80016 12.87657 14.29766 13.91425 11.64419 ...</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; permanova:</span>
<span class="co">#&gt; Permutation test for adonis under reduced model</span>
<span class="co">#&gt; Marginal effects of terms</span>
<span class="co">#&gt; Permutation: free</span>
<span class="co">#&gt; Number of permutations: 99</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; vegan::adonis2(formula = formula, data = metadata, permutations = n_perms, by = by, parallel = parall)</span>
<span class="co">#&gt;                        Df SumOfSqs      R2       F Pr(&gt;F)   </span>
<span class="co">#&gt; DNA_extraction_method   2     8847 0.12742 58.3201   0.01 **</span>
<span class="co">#&gt; weight                  1      259 0.00372  3.4087   0.01 **</span>
<span class="co">#&gt; female                  1      317 0.00456  4.1742   0.01 **</span>
<span class="co">#&gt; Residual              779    59084 0.85098                  </span>
<span class="co">#&gt; Total                 783    69431 1.00000                  </span>
<span class="co">#&gt; ---</span>
<span class="co">#&gt; Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></code></pre></div>
</div>
<div id="visualising-significant-predictors" class="section level3">
<h3 class="hasAnchor">
<a href="#visualising-significant-predictors" class="anchor"></a>Visualising significant predictors?</h3>
<p>(Partial) Constrained ordination can be useful to show microbial variation explained by predictors significant in PERMANOVA analyses.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># constraints need to be on the same or similar scales for comparability</span>
<span class="co"># so make binary variables and scale the weight variable</span>
<span class="va">ps</span> <span class="op">&lt;-</span> <span class="va">ps</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span>
  <span class="fu"><a href="../../reference/ps_mutate.html">ps_mutate</a></span><span class="op">(</span>
    wt_scaled <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/scale.html">scale</a></span><span class="op">(</span><span class="va">weight</span>, center <span class="op">=</span> <span class="cn">TRUE</span>, scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>,
    P <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html">if_else</a></span><span class="op">(</span><span class="va">DNA_extraction_method</span> <span class="op">==</span> <span class="st">"p"</span>, <span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span>,
    R <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html">if_else</a></span><span class="op">(</span><span class="va">DNA_extraction_method</span> <span class="op">==</span> <span class="st">"r"</span>, <span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span>,
    O <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html">if_else</a></span><span class="op">(</span><span class="va">DNA_extraction_method</span> <span class="op">==</span> <span class="st">"o"</span>, <span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span>
  <span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">constr_ord</span> <span class="op">&lt;-</span> <span class="va">ps</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span>
  <span class="fu"><a href="../../reference/tax_filter.html">tax_filter</a></span><span class="op">(</span>min_prevalence <span class="op">=</span> <span class="fl">5</span> <span class="op">/</span> <span class="fl">100</span>, tax_level <span class="op">=</span> <span class="st">"Genus"</span><span class="op">)</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span>
  <span class="fu"><a href="../../reference/tax_agg.html">tax_agg</a></span><span class="op">(</span><span class="st">"Genus"</span><span class="op">)</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span>
  <span class="fu"><a href="../../reference/tax_transform.html">tax_transform</a></span><span class="op">(</span><span class="st">"clr"</span><span class="op">)</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span>
  <span class="fu"><a href="../../reference/ord_calc.html">ord_calc</a></span><span class="op">(</span>
    method <span class="op">=</span> <span class="st">"RDA"</span>, constraints <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"female"</span>, <span class="st">"wt_scaled"</span>, <span class="st">"P"</span>, <span class="st">"R"</span>, <span class="st">"O"</span><span class="op">)</span>
  <span class="op">)</span>
<span class="co">#&gt; Proportional min_prevalence given: 0.05 --&gt; min 41/801 samples.</span>
<span class="co">#&gt; Dropping samples with missings: 17</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Centering (mean) and scaling (sd) the constraints and conditions:</span>
<span class="co">#&gt;  female</span>
<span class="co">#&gt;  wt_scaled</span>
<span class="co">#&gt;  P</span>
<span class="co">#&gt;  R</span>
<span class="co">#&gt;  O</span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">constr_ord</span>
<span class="co">#&gt; ps_extra object - a list with phyloseq and extras:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; phyloseq-class experiment-level object</span>
<span class="co">#&gt; otu_table()   OTU Table:         [ 117 taxa and 784 samples ]</span>
<span class="co">#&gt; sample_data() Sample Data:       [ 784 samples by 17 sample variables ]</span>
<span class="co">#&gt; tax_table()   Taxonomy Table:    [ 117 taxa by 3 taxonomic ranks ]</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; ps_extra info:</span>
<span class="co">#&gt; tax_agg = Genus tax_transform = clr</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; ordination of class: rda cca </span>
<span class="co">#&gt; rda(formula = OTU ~ female + wt_scaled + P + R + O, data = data)</span>
<span class="co">#&gt; constraints: female+wt_scaled+P+R+O</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $counts OTU Table: [ 117 taxa and 801 samples ]</span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../../reference/ord_plot.html">ord_plot</a></span><span class="op">(</span><span class="va">constr_ord</span>, plot_taxa <span class="op">=</span> <span class="fl">10</span><span class="op">:</span><span class="fl">1</span>, colour <span class="op">=</span> <span class="st">"DNA_extraction_method"</span>, shape <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_color_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Set2"</span>, name <span class="op">=</span> <span class="st">"DNA\nextraction\nmethod"</span><span class="op">)</span></code></pre></div>
<p><img src="atlas1006_files/figure-html/constrained%20ord%20plot-1.png" width="700"></p>
<p>As DNA extraction method dominates the plot above, we could try “partialing out” the variation explained by DNA extraction method, prior to constraining on the other factors of interest.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ps</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span>
  <span class="fu"><a href="../../reference/tax_filter.html">tax_filter</a></span><span class="op">(</span>min_prevalence <span class="op">=</span> <span class="fl">5</span> <span class="op">/</span> <span class="fl">100</span>, tax_level <span class="op">=</span> <span class="st">"Genus"</span><span class="op">)</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span>
  <span class="fu"><a href="../../reference/tax_agg.html">tax_agg</a></span><span class="op">(</span><span class="st">"Genus"</span><span class="op">)</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span>
  <span class="fu"><a href="../../reference/tax_transform.html">tax_transform</a></span><span class="op">(</span><span class="st">"clr"</span><span class="op">)</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span>
  <span class="fu"><a href="../../reference/ord_calc.html">ord_calc</a></span><span class="op">(</span>
    method <span class="op">=</span> <span class="st">"RDA"</span>, conditions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"P"</span>, <span class="st">"R"</span>, <span class="st">"O"</span><span class="op">)</span>,
    constraints <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"female"</span>, <span class="st">"wt_scaled"</span><span class="op">)</span>
  <span class="op">)</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span>
  <span class="fu"><a href="../../reference/ord_plot.html">ord_plot</a></span><span class="op">(</span>plot_taxa <span class="op">=</span> <span class="fl">10</span><span class="op">:</span><span class="fl">1</span>, colour <span class="op">=</span> <span class="st">"DNA_extraction_method"</span>, shape <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_color_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Set2"</span>, name <span class="op">=</span> <span class="st">"DNA\nextraction\nmethod"</span><span class="op">)</span>
<span class="co">#&gt; Proportional min_prevalence given: 0.05 --&gt; min 41/801 samples.</span>
<span class="co">#&gt; Dropping samples with missings: 17</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Centering (mean) and scaling (sd) the constraints and conditions:</span>
<span class="co">#&gt;  female</span>
<span class="co">#&gt;  wt_scaled</span>
<span class="co">#&gt;  P</span>
<span class="co">#&gt;  R</span>
<span class="co">#&gt;  O</span></code></pre></div>
<p><img src="atlas1006_files/figure-html/partial%20constrained%20ord-1.png" width="700"></p>
</div>
<div id="taxon-models" class="section level3">
<h3 class="hasAnchor">
<a href="#taxon-models" class="anchor"></a>Taxon models</h3>
<p>What are the effects of these factors on individual taxa? Let’s use beta binomial regression models to find out. We will skip fitting models for genera here, only for speed of generating this example. You should include all ranks (which is the default) for the best use of taxatree_plots.</p>
<p>See the taxon modelling article for a more comprehensive look at the taxatree_* family of functions.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/bryandmartin/corncob">corncob</a></span><span class="op">)</span>
<span class="va">tt_models</span> <span class="op">&lt;-</span> <span class="va">ps</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span>
  <span class="fu"><a href="../../reference/tax_filter.html">tax_filter</a></span><span class="op">(</span>min_prevalence <span class="op">=</span> <span class="fl">5</span> <span class="op">/</span> <span class="fl">100</span>, tax_level <span class="op">=</span> <span class="st">"Genus"</span><span class="op">)</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span>
  <span class="fu"><a href="../../reference/taxatree_models.html">taxatree_models</a></span><span class="op">(</span>
    ranks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Phylum"</span>, <span class="st">"Family"</span><span class="op">)</span>, 
    variables <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"female"</span>, <span class="st">"wt_scaled"</span>, <span class="st">"P"</span>, <span class="st">"R"</span><span class="op">)</span>,
    type <span class="op">=</span> <span class="st">"bbdml"</span>, verbose <span class="op">=</span> <span class="st">"max"</span>
  <span class="op">)</span>
<span class="co">#&gt; Proportional min_prevalence given: 0.05 --&gt; min 41/801 samples.</span>
<span class="co">#&gt; 2021-11-13 13:27:45 - modelling at rank: Phylum</span>
<span class="co">#&gt; 2021-11-13 13:27:54 - modelling at rank: Family</span>
<span class="va">tt_stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/taxatree_models2stats.html">taxatree_models2stats</a></span><span class="op">(</span><span class="va">tt_models</span>, param <span class="op">=</span> <span class="st">"mu"</span><span class="op">)</span></code></pre></div>
<p>Visualize the results compactly on a microViz taxonomic association tree.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">tt_stats</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span> 
  <span class="fu"><a href="../../reference/taxatree_plotkey.html">taxatree_plotkey</a></span><span class="op">(</span>
    node_size_range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">8</span><span class="op">)</span>, size_stat <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>mean <span class="op">=</span> <span class="va">mean</span><span class="op">)</span>,
    <span class="va">rank</span> <span class="op">==</span> <span class="st">"Phylum"</span>, 
    taxon_renamer <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_remove.html">str_remove_all</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">"^.: | [ae]t rel."</span><span class="op">)</span>
  <span class="op">)</span></code></pre></div>
<p><img src="atlas1006_files/figure-html/taxa%20tree%20key-1.png" width="700"></p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">tt_stats</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span> 
  <span class="fu"><a href="../../reference/taxatree_plots.html">taxatree_plots</a></span><span class="op">(</span>
    node_size_range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">5</span><span class="op">)</span>, size_stat <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>mean <span class="op">=</span> <span class="va">mean</span><span class="op">)</span>
  <span class="op">)</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span>
    <span class="va">.</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"P"</span>, <span class="st">"R"</span>, <span class="st">"female"</span>, <span class="st">"wt_scaled"</span><span class="op">)</span><span class="op">]</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span> 
  <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html">wrap_plots</a></span><span class="op">(</span><span class="va">.</span>, ncol <span class="op">=</span> <span class="fl">2</span>, guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span></code></pre></div>
<p><img src="atlas1006_files/figure-html/taxatree%20plots-1.png" width="700"></p>
<div id="example-interpretation-illustrative-only" class="section level4">
<h4 class="hasAnchor">
<a href="#example-interpretation-illustrative-only" class="anchor"></a>Example interpretation (illustrative only):</h4>
<ul>
<li>DNA extraction methods P and R are associated with significantly higher levels of Actinobacteria and lower Bacteroides, relative to the reference of extraction method O.</li>
<li>There are associations between weight and various taxa, but these are not as strong as the associations with extraction method.</li>
</ul>
</div>
</div>
</div>
<div id="disclaimer" class="section level2">
<h2 class="hasAnchor">
<a href="#disclaimer" class="anchor"></a>Disclaimer</h2>
<p>This document is intended only to be an example of the kind of analyses and visualization you can do with microViz. The analysis of the atlas1006 data is not intended to be considered theoretically sound or biologically interpretable.</p>
</div>
<div id="session-info" class="section level2">
<h2 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session info</h2>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://r-lib.github.io/sessioninfo/reference/session_info.html">session_info</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; ─ Session info  ──────────────────────────────────────────────────────────────────────────────────</span>
<span class="co">#&gt;  hash: rabbit, person tipping hand: medium-dark skin tone, jeans</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;  setting  value</span>
<span class="co">#&gt;  version  R version 4.1.2 (2021-11-01)</span>
<span class="co">#&gt;  os       Ubuntu 20.04.3 LTS</span>
<span class="co">#&gt;  system   x86_64, linux-gnu</span>
<span class="co">#&gt;  ui       X11</span>
<span class="co">#&gt;  language (EN)</span>
<span class="co">#&gt;  collate  C.UTF-8</span>
<span class="co">#&gt;  ctype    C.UTF-8</span>
<span class="co">#&gt;  tz       UTC</span>
<span class="co">#&gt;  date     2021-11-13</span>
<span class="co">#&gt;  pandoc   2.7.3 @ /usr/bin/ (via rmarkdown)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; ─ Packages ───────────────────────────────────────────────────────────────────────────────────────</span>
<span class="co">#&gt;  package            * version      date (UTC) lib source</span>
<span class="co">#&gt;  ade4                 1.7-18       2021-09-16 [1] Custom</span>
<span class="co">#&gt;  ape                  5.5          2021-04-25 [1] Custom</span>
<span class="co">#&gt;  Biobase              2.54.0       2021-10-26 [1] Bioconductor</span>
<span class="co">#&gt;  BiocGenerics         0.40.0       2021-10-26 [1] Bioconductor</span>
<span class="co">#&gt;  biomformat           1.22.0       2021-10-26 [1] Bioconductor</span>
<span class="co">#&gt;  Biostrings           2.62.0       2021-10-26 [1] Bioconductor</span>
<span class="co">#&gt;  bitops               1.0-7        2021-04-24 [1] Custom</span>
<span class="co">#&gt;  bslib                0.3.1        2021-10-06 [1] Custom</span>
<span class="co">#&gt;  cachem               1.0.6        2021-08-19 [1] Custom</span>
<span class="co">#&gt;  callr                3.7.0        2021-04-20 [1] Custom</span>
<span class="co">#&gt;  cli                  3.1.0        2021-10-27 [1] Custom</span>
<span class="co">#&gt;  cluster              2.1.2        2021-04-17 [2] CRAN (R 4.1.2)</span>
<span class="co">#&gt;  codetools            0.2-18       2020-11-04 [2] CRAN (R 4.1.2)</span>
<span class="co">#&gt;  colorspace           2.0-2        2021-06-24 [1] Custom</span>
<span class="co">#&gt;  corncob            * 0.2.0        2021-03-11 [1] Custom</span>
<span class="co">#&gt;  crayon               1.4.2        2021-10-29 [1] Custom</span>
<span class="co">#&gt;  data.table           1.14.2       2021-09-27 [1] Custom</span>
<span class="co">#&gt;  desc                 1.4.0        2021-09-28 [1] Custom</span>
<span class="co">#&gt;  detectseparation     0.2          2021-04-22 [1] Custom</span>
<span class="co">#&gt;  devtools             2.4.2        2021-06-07 [1] Custom</span>
<span class="co">#&gt;  digest               0.6.28       2021-09-23 [1] Custom</span>
<span class="co">#&gt;  dplyr              * 1.0.7        2021-06-18 [1] Custom</span>
<span class="co">#&gt;  ellipsis             0.3.2        2021-04-29 [1] Custom</span>
<span class="co">#&gt;  evaluate             0.14         2019-05-28 [1] Custom</span>
<span class="co">#&gt;  fansi                0.5.0        2021-05-25 [1] Custom</span>
<span class="co">#&gt;  farver               2.1.0        2021-02-28 [1] Custom</span>
<span class="co">#&gt;  fastmap              1.1.0        2021-01-25 [1] Custom</span>
<span class="co">#&gt;  foreach              1.5.1        2020-10-15 [1] Custom</span>
<span class="co">#&gt;  fs                   1.5.0        2020-07-31 [1] Custom</span>
<span class="co">#&gt;  future               1.23.0       2021-10-31 [1] Custom</span>
<span class="co">#&gt;  future.apply         1.8.1        2021-08-10 [1] Custom</span>
<span class="co">#&gt;  generics             0.1.1        2021-10-25 [1] Custom</span>
<span class="co">#&gt;  GenomeInfoDb         1.30.0       2021-10-26 [1] Bioconductor</span>
<span class="co">#&gt;  GenomeInfoDbData     1.2.7        2021-11-12 [1] Bioconductor</span>
<span class="co">#&gt;  ggforce              0.3.3        2021-03-05 [1] Custom</span>
<span class="co">#&gt;  ggplot2            * 3.3.5        2021-06-25 [1] Custom</span>
<span class="co">#&gt;  ggraph               2.0.5        2021-02-23 [1] Custom</span>
<span class="co">#&gt;  ggrepel              0.9.1        2021-01-15 [1] Custom</span>
<span class="co">#&gt;  ggtext               0.1.1        2020-12-17 [1] Custom</span>
<span class="co">#&gt;  globals              0.14.0       2020-11-22 [1] Custom</span>
<span class="co">#&gt;  glue                 1.5.0        2021-11-07 [1] Custom</span>
<span class="co">#&gt;  graphlayouts         0.7.1        2020-10-26 [1] Custom</span>
<span class="co">#&gt;  gridExtra            2.3          2017-09-09 [1] Custom</span>
<span class="co">#&gt;  gridtext             0.1.4        2020-12-10 [1] Custom</span>
<span class="co">#&gt;  gtable               0.3.0        2019-03-25 [1] Custom</span>
<span class="co">#&gt;  highr                0.9          2021-04-16 [1] Custom</span>
<span class="co">#&gt;  htmltools            0.5.2        2021-08-25 [1] Custom</span>
<span class="co">#&gt;  igraph               1.2.8        2021-11-07 [1] Custom</span>
<span class="co">#&gt;  IRanges              2.28.0       2021-10-26 [1] Bioconductor</span>
<span class="co">#&gt;  iterators            1.0.13       2020-10-15 [1] Custom</span>
<span class="co">#&gt;  jquerylib            0.1.4        2021-04-26 [1] Custom</span>
<span class="co">#&gt;  jsonlite             1.7.2        2020-12-09 [1] Custom</span>
<span class="co">#&gt;  knitr                1.36         2021-09-29 [1] Custom</span>
<span class="co">#&gt;  labeling             0.4.2        2020-10-20 [1] Custom</span>
<span class="co">#&gt;  lattice              0.20-45      2021-09-22 [2] CRAN (R 4.1.2)</span>
<span class="co">#&gt;  lifecycle            1.0.1        2021-09-24 [1] Custom</span>
<span class="co">#&gt;  listenv              0.8.0        2019-12-05 [1] Custom</span>
<span class="co">#&gt;  lpSolveAPI           5.5.2.0-17.7 2020-06-14 [1] Custom</span>
<span class="co">#&gt;  magrittr             2.0.1        2020-11-17 [1] Custom</span>
<span class="co">#&gt;  markdown             1.1          2019-08-07 [1] Custom</span>
<span class="co">#&gt;  MASS                 7.3-54       2021-05-03 [2] CRAN (R 4.1.2)</span>
<span class="co">#&gt;  Matrix               1.3-4        2021-06-01 [2] CRAN (R 4.1.2)</span>
<span class="co">#&gt;  memoise              2.0.0        2021-01-26 [1] Custom</span>
<span class="co">#&gt;  mgcv                 1.8-38       2021-10-06 [2] CRAN (R 4.1.2)</span>
<span class="co">#&gt;  microbiome         * 1.16.0       2021-10-26 [1] Bioconductor</span>
<span class="co">#&gt;  microViz           * 0.8.0.9002   2021-11-13 [1] local</span>
<span class="co">#&gt;  multtest             2.50.0       2021-10-26 [1] Bioconductor</span>
<span class="co">#&gt;  munsell              0.5.0        2018-06-12 [1] Custom</span>
<span class="co">#&gt;  nlme                 3.1-153      2021-09-07 [2] CRAN (R 4.1.2)</span>
<span class="co">#&gt;  numDeriv             2016.8-1.1   2019-06-06 [1] Custom</span>
<span class="co">#&gt;  parallelly           1.28.1       2021-09-09 [1] Custom</span>
<span class="co">#&gt;  patchwork          * 1.1.1        2020-12-17 [1] Custom</span>
<span class="co">#&gt;  permute              0.9-5        2019-03-12 [1] Custom</span>
<span class="co">#&gt;  phyloseq           * 1.38.0       2021-10-26 [1] Bioconductor</span>
<span class="co">#&gt;  pillar               1.6.4        2021-10-18 [1] Custom</span>
<span class="co">#&gt;  pkgbuild             1.2.0        2020-12-15 [1] Custom</span>
<span class="co">#&gt;  pkgconfig            2.0.3        2019-09-22 [1] Custom</span>
<span class="co">#&gt;  pkgdown              1.6.1        2020-09-12 [1] Custom</span>
<span class="co">#&gt;  pkgload              1.2.3        2021-10-13 [1] Custom</span>
<span class="co">#&gt;  plyr                 1.8.6        2020-03-03 [1] Custom</span>
<span class="co">#&gt;  polyclip             1.10-0       2019-03-14 [1] Custom</span>
<span class="co">#&gt;  prettyunits          1.1.1        2020-01-24 [1] Custom</span>
<span class="co">#&gt;  processx             3.5.2        2021-04-30 [1] Custom</span>
<span class="co">#&gt;  ps                   1.6.0        2021-02-28 [1] Custom</span>
<span class="co">#&gt;  purrr                0.3.4        2020-04-17 [1] Custom</span>
<span class="co">#&gt;  R6                   2.5.1        2021-08-19 [1] Custom</span>
<span class="co">#&gt;  ragg                 1.2.0        2021-10-30 [1] Custom</span>
<span class="co">#&gt;  RColorBrewer         1.1-2        2014-12-07 [1] Custom</span>
<span class="co">#&gt;  Rcpp                 1.0.7        2021-07-07 [1] Custom</span>
<span class="co">#&gt;  RCurl                1.98-1.5     2021-09-17 [1] Custom</span>
<span class="co">#&gt;  registry             0.5-1        2019-03-05 [1] Custom</span>
<span class="co">#&gt;  remotes              2.4.1        2021-09-29 [1] Custom</span>
<span class="co">#&gt;  reshape2             1.4.4        2020-04-09 [1] Custom</span>
<span class="co">#&gt;  rhdf5                2.38.0       2021-10-26 [1] Bioconductor</span>
<span class="co">#&gt;  rhdf5filters         1.6.0        2021-10-26 [1] Bioconductor</span>
<span class="co">#&gt;  Rhdf5lib             1.16.0       2021-10-26 [1] Bioconductor</span>
<span class="co">#&gt;  rlang                0.4.12       2021-10-18 [1] Custom</span>
<span class="co">#&gt;  rmarkdown            2.11         2021-09-14 [1] Custom</span>
<span class="co">#&gt;  ROI                  1.0-0        2020-08-31 [1] Custom</span>
<span class="co">#&gt;  ROI.plugin.lpsolve   1.0-1        2021-06-15 [1] Custom</span>
<span class="co">#&gt;  rprojroot            2.0.2        2020-11-15 [1] Custom</span>
<span class="co">#&gt;  Rtsne                0.15         2018-11-10 [1] Custom</span>
<span class="co">#&gt;  S4Vectors            0.32.2       2021-11-07 [1] Bioconductor</span>
<span class="co">#&gt;  sass                 0.4.0        2021-05-12 [1] Custom</span>
<span class="co">#&gt;  scales               1.1.1        2020-05-11 [1] Custom</span>
<span class="co">#&gt;  sessioninfo          1.2.1        2021-11-02 [1] Custom</span>
<span class="co">#&gt;  slam                 0.1-48       2020-12-03 [1] Custom</span>
<span class="co">#&gt;  stringi              1.7.5        2021-10-04 [1] Custom</span>
<span class="co">#&gt;  stringr              1.4.0        2019-02-10 [1] Custom</span>
<span class="co">#&gt;  survival             3.2-13       2021-08-24 [2] CRAN (R 4.1.2)</span>
<span class="co">#&gt;  systemfonts          1.0.3        2021-10-13 [1] Custom</span>
<span class="co">#&gt;  testthat             3.1.0        2021-10-04 [1] Custom</span>
<span class="co">#&gt;  textshaping          0.3.6        2021-10-13 [1] Custom</span>
<span class="co">#&gt;  tibble               3.1.6        2021-11-07 [1] Custom</span>
<span class="co">#&gt;  tidygraph            1.2.0        2020-05-12 [1] Custom</span>
<span class="co">#&gt;  tidyr                1.1.4        2021-09-27 [1] Custom</span>
<span class="co">#&gt;  tidyselect           1.1.1        2021-04-30 [1] Custom</span>
<span class="co">#&gt;  trust                0.1-8        2020-01-10 [1] Custom</span>
<span class="co">#&gt;  tweenr               1.0.2        2021-03-23 [1] Custom</span>
<span class="co">#&gt;  usethis              2.1.3        2021-10-27 [1] Custom</span>
<span class="co">#&gt;  utf8                 1.2.2        2021-07-24 [1] Custom</span>
<span class="co">#&gt;  vctrs                0.3.8        2021-04-29 [1] Custom</span>
<span class="co">#&gt;  vegan                2.5-7        2020-11-28 [1] Custom</span>
<span class="co">#&gt;  VGAM                 1.1-5        2021-01-14 [1] Custom</span>
<span class="co">#&gt;  viridis              0.6.2        2021-10-13 [1] Custom</span>
<span class="co">#&gt;  viridisLite          0.4.0        2021-04-13 [1] Custom</span>
<span class="co">#&gt;  withr                2.4.2        2021-04-18 [1] Custom</span>
<span class="co">#&gt;  xfun                 0.28         2021-11-04 [1] Custom</span>
<span class="co">#&gt;  xml2                 1.3.2        2020-04-23 [1] Custom</span>
<span class="co">#&gt;  XVector              0.34.0       2021-10-26 [1] Bioconductor</span>
<span class="co">#&gt;  yaml                 2.2.1        2020-02-01 [1] Custom</span>
<span class="co">#&gt;  zlibbioc             1.40.0       2021-10-26 [1] Bioconductor</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;  [1] /home/runner/work/_temp/Library</span>
<span class="co">#&gt;  [2] /opt/R/4.1.2/lib/R/library</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; ──────────────────────────────────────────────────────────────────────────────────────────────────</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by David Barnett.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: '22e6f0a75b7b1ffae9bf5bc5b58864d7',
    indexName: 'microViz',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
