<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Visualising compositions • microViz</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../../bootstrap-toc.css">
<script src="../../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><!-- docsearch --><script src="../../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><meta property="og:title" content="Visualising compositions">
<meta property="og:description" content="microViz">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">microViz</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.7.10.9032</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="divider">
    <li class="dropdown-header">Tutorials</li>
    <li>
      <a href="../../articles/web-only/atlas1006.html">Example analyses with atlas1006 data</a>
    </li>
    <li>
      <a href="../../articles/web-only/phyloseq.html">Working with phyloseq objects</a>
    </li>
    <li>
      <a href="../../articles/web-only/tax-fixing.html">Fixing your tax_table</a>
    </li>
    <li>
      <a href="../../articles/web-only/ordination.html">Ordination plots</a>
    </li>
    <li>
      <a href="../../articles/web-only/ordination-interactive.html">Interactive Ordination Plot</a>
    </li>
    <li>
      <a href="../../articles/web-only/compositions.html">Visualising compositions</a>
    </li>
    <li>
      <a href="../../articles/web-only/videos.html">Videos</a>
    </li>
  </ul>
</li>
<li>
  <a href="../../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/david-barnett/microViz/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="compositions_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Visualising compositions</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/david-barnett/microViz/blob/master/vignettes/web-only/compositions.Rmd"><code>vignettes/web-only/compositions.Rmd</code></a></small>
      <div class="hidden name"><code>compositions.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">100</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://david-barnett.github.io/microViz">microViz</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://dx.plos.org/10.1371/journal.pone.0061217">phyloseq</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com">patchwork</a></span><span class="op">)</span> <span class="co"># for arranging groups of plots</span>
<span class="fu">knitr</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/knitr/man/opts_chunk.html">opts_chunk</a></span><span class="op">$</span><span class="fu">set</span><span class="op">(</span>
  fig.height <span class="op">=</span> <span class="fl">6</span>,
  fig.width <span class="op">=</span> <span class="fl">9</span>
<span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get example phyloseq data from corncob package and tidy up</span>
<span class="va">pseq</span> <span class="op">&lt;-</span> <span class="fu">corncob</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/corncob/man/ibd_phylo.html">ibd_phylo</a></span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="../../reference/tax_filter.html">tax_filter</a></span><span class="op">(</span>min_prevalence <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="../../reference/tax_fix.html">tax_fix</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="../../reference/phyloseq_validate.html">phyloseq_validate</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p>The <code>comp_barplot</code> function allows you to visualise the taxonomic compositions of your microbiome samples in a flexible, scalable, group-able, and visually appealing way.</p>
<div id="quick-example-barplot" class="section level2">
<h2 class="hasAnchor">
<a href="#quick-example-barplot" class="anchor"></a>Quick example barplot</h2>
<p>Visualise the top Genera across all the female samples from this inflammatory bowel disease study dataset. The order of the samples is automatically set by their “bray”-curtis dissimilarity.</p>
<p>By default, the top 8 taxa are shown. These taxa are chosen by their total count abundance across all plotted samples.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pseq</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../../reference/ps_filter.html">ps_filter</a></span><span class="op">(</span><span class="va">gender</span> <span class="op">==</span> <span class="st">"female"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../../reference/comp_barplot.html">comp_barplot</a></span><span class="op">(</span>tax_level <span class="op">=</span> <span class="st">"Genus"</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># horizontal bars are often more readable</span>
<span class="co">#&gt; Registered S3 method overwritten by 'seriation':</span>
<span class="co">#&gt;   method         from </span>
<span class="co">#&gt;   reorder.hclust vegan</span></code></pre></div>
<p><img src="compositions_files/figure-html/simple-1.png" width="864"></p>
</div>
<div id="customising-this-barplot" class="section level2">
<h2 class="hasAnchor">
<a href="#customising-this-barplot" class="anchor"></a>Customising this barplot</h2>
<p>The output of comp_barplot can be customised in several ways. See the comment alongside each argument for an explanation of its effect.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pseq</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../../reference/ps_filter.html">ps_filter</a></span><span class="op">(</span><span class="va">gender</span> <span class="op">==</span> <span class="st">"female"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../../reference/comp_barplot.html">comp_barplot</a></span><span class="op">(</span>
    tax_level <span class="op">=</span> <span class="st">"Genus"</span>,
    label <span class="op">=</span> <span class="st">"DiseaseState"</span>, <span class="co"># name an alternative variable to label axes</span>
    n_taxa <span class="op">=</span> <span class="fl">15</span>, <span class="co"># give more taxa unique colours</span>
    merge_other <span class="op">=</span> <span class="cn">FALSE</span>, <span class="co"># split the "other" category to display alpha diversity</span>
    bar_width <span class="op">=</span> <span class="fl">0.7</span>, <span class="co"># reduce the bar width to 70% of one row</span>
    bar_outline_colour <span class="op">=</span> <span class="st">"grey5"</span> <span class="co"># is the default (use NA to remove outlines)</span>
  <span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html">coord_flip</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><img src="compositions_files/figure-html/simple%20customised-1.png" width="864"></p>
<p><strong>Other notes:</strong></p>
<ul>
<li>Dissimilarity is calculated using only the visibly distinct taxa, to optimise sorting for visual similarity. You can change this by setting <code>order_with_all_taxa = TRUE</code>, to always use all taxa for similarity sorting.</li>
<li>The colour palette is important, to allow (adjacent) taxa to be distinguished. The palette microViz uses is generated by the <code>distinct_palette</code> function, which starts with the Paired and Dark2 palettes from ColorBrewer and continues with further distinct colours generated at <a href="http://medialab.github.io/iwanthue/" class="uri">http://medialab.github.io/iwanthue/</a> (all colors, soft k-means).</li>
</ul>
</div>
<div id="averages-faceting-or-grouping" class="section level2">
<h2 class="hasAnchor">
<a href="#averages-faceting-or-grouping" class="anchor"></a>Averages, faceting or grouping?</h2>
<div id="averaging-compositions" class="section level3">
<h3 class="hasAnchor">
<a href="#averaging-compositions" class="anchor"></a>Averaging compositions</h3>
<p>Sometimes, to compare microbial compositions across groups, average compositions are presented. However that “group-averaging” approach hides a lot of within-group variation, as well as any imbalance in group sizes.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pseq</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../../reference/ps_select.html">ps_select</a></span><span class="op">(</span><span class="va">age</span>, <span class="va">DiseaseState</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="co"># avoids lots of phyloseq::merge_samples warnings</span>
  <span class="fu"><a href="../../reference/ps_filter.html">ps_filter</a></span><span class="op">(</span><span class="va">DiseaseState</span> <span class="op">!=</span> <span class="st">"IBDundef"</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/merge_samples-methods.html">merge_samples</a></span><span class="op">(</span>group <span class="op">=</span> <span class="st">"DiseaseState"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../../reference/comp_barplot.html">comp_barplot</a></span><span class="op">(</span>
    tax_level <span class="op">=</span> <span class="st">"Genus"</span>, n_taxa <span class="op">=</span> <span class="fl">12</span>,
    bar_width <span class="op">=</span> <span class="fl">0.8</span>
  <span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="cn">NULL</span>, y <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span>
<span class="co">#&gt; Warning in asMethod(object): NAs introduced by coercion</span></code></pre></div>
<p><img src="compositions_files/figure-html/unnamed-chunk-3-1.png" width="864"></p>
</div>
<div id="faceting" class="section level3">
<h3 class="hasAnchor">
<a href="#faceting" class="anchor"></a>Faceting</h3>
<p>Faceting is where you show each group on a small subplot.</p>
<p>In the plot below can you see that at minority of UC samples have a high abundance of Escherichia/Shigella or Streptococcus. The merged bars above might have misled you into thinking all UC samples had somewhat increased abundances of these taxa.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pseq</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../../reference/ps_filter.html">ps_filter</a></span><span class="op">(</span><span class="va">DiseaseState</span> <span class="op">!=</span> <span class="st">"IBDundef"</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="co"># only one sample in this group</span>
  <span class="co"># convert DiseaseState into ordered factor to control order of facets</span>
  <span class="fu"><a href="../../reference/ps_mutate.html">ps_mutate</a></span><span class="op">(</span>
    DiseaseState <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span>
      <span class="va">DiseaseState</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"UC"</span>, <span class="st">"nonIBD"</span>, <span class="st">"CD"</span><span class="op">)</span>, ordered <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
  <span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="../../reference/comp_barplot.html">comp_barplot</a></span><span class="op">(</span>
    tax_level <span class="op">=</span> <span class="st">"Genus"</span>, n_taxa <span class="op">=</span> <span class="fl">15</span>,
    bar_outline_colour <span class="op">=</span> <span class="cn">NA</span>, facet_by <span class="op">=</span> <span class="st">"DiseaseState"</span>
  <span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html">coord_flip</a></span><span class="op">(</span><span class="op">)</span> </code></pre></div>
<p><img src="compositions_files/figure-html/unnamed-chunk-4-1.png" width="864"></p>
<p>Instead of using the <code>facet_by</code> argument in <code>comp_barplot</code> you can have more control over faceting by doing it yourself afterwards. You can use facet_grid to create row facets.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pseq</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../../reference/ps_filter.html">ps_filter</a></span><span class="op">(</span><span class="va">DiseaseState</span> <span class="op">!=</span> <span class="st">"IBDundef"</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="co"># only one sample in this group</span>
  <span class="co"># convert DiseaseState into ordered factor to control order of facets</span>
  <span class="fu"><a href="../../reference/ps_mutate.html">ps_mutate</a></span><span class="op">(</span>
    DiseaseState <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span>
      <span class="va">DiseaseState</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"UC"</span>, <span class="st">"CD"</span>, <span class="st">"nonIBD"</span><span class="op">)</span>, ordered <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
  <span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="../../reference/comp_barplot.html">comp_barplot</a></span><span class="op">(</span>
    tax_level <span class="op">=</span> <span class="st">"Genus"</span>, n_taxa <span class="op">=</span> <span class="fl">15</span>,
    sample_order <span class="op">=</span> <span class="st">"bray"</span>, bar_outline_colour <span class="op">=</span> <span class="cn">NA</span>,
  <span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html">facet_grid</a></span><span class="op">(</span>
    rows <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/vars.html">vars</a></span><span class="op">(</span><span class="va">DiseaseState</span><span class="op">)</span>, 
    scales <span class="op">=</span> <span class="st">"free"</span>, space <span class="op">=</span> <span class="st">"free"</span> <span class="co"># these options are critically important!</span>
  <span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>
    axis.text.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,
    axis.ticks.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>
  <span class="op">)</span></code></pre></div>
<p><img src="compositions_files/figure-html/unnamed-chunk-5-1.png" width="576"></p>
</div>
<div id="grouping" class="section level3">
<h3 class="hasAnchor">
<a href="#grouping" class="anchor"></a>Grouping</h3>
<p>For even greater control than faceting, <code>comp_barplot</code> allows you to generate separate ggplot objects for each group, whilst maintaining the same taxa colour scheme.</p>
<p>You can assemble these plots into one figure with, for example, the <code>patchwork</code> package, or keep them separate.</p>
<p><em>Note that the ordering of the samples may differ between facet and group_by approaches. In the group_by method, the ordering of the samples by similarity is done separately for each group, whereas in the facet_by method, similarity-based ordering is done with all samples and then the samples are separated by facet afterwards.</em></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">plot_list</span> <span class="op">&lt;-</span> <span class="va">pseq</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="../../reference/ps_filter.html">ps_filter</a></span><span class="op">(</span><span class="va">DiseaseState</span> <span class="op">!=</span> <span class="st">"IBDundef"</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="../../reference/comp_barplot.html">comp_barplot</a></span><span class="op">(</span>
  n_taxa <span class="op">=</span> <span class="fl">15</span>, tax_level <span class="op">=</span> <span class="st">"Genus"</span>, group_by <span class="op">=</span> <span class="st">"DiseaseState"</span>
<span class="op">)</span>

<span class="co"># Plot them side by side with the patchwork package.</span>
<span class="va">patch</span> <span class="op">&lt;-</span> <span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html">wrap_plots</a></span><span class="op">(</span><span class="va">plot_list</span>, nrow <span class="op">=</span> <span class="fl">1</span>, guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span>
<span class="va">patch</span> <span class="op">&amp;</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># make all plots horizontal (note: use &amp; instead of +)</span></code></pre></div>
<p><img src="compositions_files/figure-html/unnamed-chunk-6-1.png" width="864"></p>
<p>Notice how you can theme all plots with the <code><a href="https://rdrr.io/r/base/Logic.html">&amp;</a></code> operator.</p>
<p>See <a href="https://patchwork.data-imaginist.com/index.html" class="uri">https://patchwork.data-imaginist.com/index.html</a> for more examples of arranging multiple plots.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">patch</span> <span class="op">&amp;</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="cn">NULL</span>, y <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span> <span class="op">&amp;</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>
    axis.text.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>,
    legend.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>
  <span class="op">)</span> <span class="op">&amp;</span>
  <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_annotation.html">plot_annotation</a></span><span class="op">(</span>
    title <span class="op">=</span> <span class="st">"Microbial composition across disease groups"</span>,
    caption <span class="op">=</span> <span class="st">"Caption: patchwork is a great package!"</span>,
    theme <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">14</span>, face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span><span class="op">)</span>
  <span class="op">)</span></code></pre></div>
<p><img src="compositions_files/figure-html/unnamed-chunk-7-1.png" width="864"></p>
</div>
</div>
<div id="sorting-the-barplot" class="section level2">
<h2 class="hasAnchor">
<a href="#sorting-the-barplot" class="anchor"></a>Sorting the barplot</h2>
<div id="sorting-by-similarity" class="section level3">
<h3 class="hasAnchor">
<a href="#sorting-by-similarity" class="anchor"></a>Sorting by similarity</h3>
<p>Sorting the samples on compositional barplots by similarity can make patterns in the data much easier to see. Check out this unsorted version of the first barplot in this article.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pseq</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../../reference/ps_filter.html">ps_filter</a></span><span class="op">(</span><span class="va">gender</span> <span class="op">==</span> <span class="st">"female"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../../reference/comp_barplot.html">comp_barplot</a></span><span class="op">(</span>tax_level <span class="op">=</span> <span class="st">"Genus"</span>, sample_order <span class="op">=</span> <span class="st">"default"</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Unsorted barcharts are hard to read!"</span><span class="op">)</span></code></pre></div>
<p><img src="compositions_files/figure-html/unnamed-chunk-8-1.png" width="864"></p>
<p>You can play with the dissimilarity measure (set in <code>sample_order</code> argument) and <code>seriate_method</code> if you like, but the defaults (Bray Curtis and OLO Ward) seem to work pretty well most of the time.</p>
<p>When sorting samples by similarity, the default is to treat the “other” taxa as one group, i.e. when <code>merge_other = TRUE</code> and <code>order_with_all_taxa = FALSE</code>.</p>
<p>If you set <code>order_with_all_taxa = TRUE</code>, samples are sorted BEFORE merging taxa. The resulting sample order is then the same as when <code>merge_other = FALSE</code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pseq</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../../reference/ps_filter.html">ps_filter</a></span><span class="op">(</span><span class="va">gender</span> <span class="op">==</span> <span class="st">"female"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../../reference/comp_barplot.html">comp_barplot</a></span><span class="op">(</span>tax_level <span class="op">=</span> <span class="st">"Genus"</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Samples sorted AFTER merging 'other' taxa"</span><span class="op">)</span></code></pre></div>
<p><img src="compositions_files/figure-html/unnamed-chunk-9-1.png" width="864"></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">pseq</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../../reference/ps_filter.html">ps_filter</a></span><span class="op">(</span><span class="va">gender</span> <span class="op">==</span> <span class="st">"female"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../../reference/comp_barplot.html">comp_barplot</a></span><span class="op">(</span>tax_level <span class="op">=</span> <span class="st">"Genus"</span>, order_with_all_taxa <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Samples sorted BEFORE merging 'other' taxa"</span><span class="op">)</span></code></pre></div>
<p><img src="compositions_files/figure-html/unnamed-chunk-9-2.png" width="864"></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R">

<span class="va">pseq</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../../reference/ps_filter.html">ps_filter</a></span><span class="op">(</span><span class="va">gender</span> <span class="op">==</span> <span class="st">"female"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../../reference/comp_barplot.html">comp_barplot</a></span><span class="op">(</span>tax_level <span class="op">=</span> <span class="st">"Genus"</span>, merge_other <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"'other' taxa not merged"</span><span class="op">)</span></code></pre></div>
<p><img src="compositions_files/figure-html/unnamed-chunk-9-3.png" width="864"></p>
</div>
<div id="sort-by-ordination" class="section level3">
<h3 class="hasAnchor">
<a href="#sort-by-ordination" class="anchor"></a>Sort by ordination</h3>
<p>Coming soon.</p>
</div>
<div id="sort-by-1-taxon" class="section level3">
<h3 class="hasAnchor">
<a href="#sort-by-1-taxon" class="anchor"></a>Sort by 1 taxon</h3>
<p>To study the distribution of a single taxonomic group across your samples, you can use <code>ps_arrange</code> (with the <code>.target</code> argument set to “otu_table”) and the ‘default’ sample_order setting in <code>comp_barplot</code>.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pseq</span> <span class="op">%&gt;%</span> 
   <span class="fu"><a href="../../reference/tax_agg.html">tax_agg</a></span><span class="op">(</span><span class="st">"Phylum"</span><span class="op">)</span> <span class="op">%&gt;%</span> 
   <span class="fu"><a href="../../reference/tax_transform.html">tax_transform</a></span><span class="op">(</span><span class="st">"compositional"</span><span class="op">)</span> <span class="op">%&gt;%</span> 
   <span class="fu"><a href="../../reference/ps_arrange.html">ps_arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html">desc</a></span><span class="op">(</span><span class="va">Firmicutes</span><span class="op">)</span>, .target <span class="op">=</span> <span class="st">"otu_table"</span><span class="op">)</span> <span class="op">%&gt;%</span> 
   <span class="fu"><a href="../../reference/comp_barplot.html">comp_barplot</a></span><span class="op">(</span>tax_level <span class="op">=</span> <span class="st">"Phylum"</span>, sample_order <span class="op">=</span> <span class="st">"default"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html">coord_flip</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; Warning in ps_counts(data = data): otu_table of counts is NOT available!</span>
<span class="co">#&gt; Available otu_table contains non-zero values that are less than 1</span></code></pre></div>
<p><img src="compositions_files/figure-html/unnamed-chunk-10-1.png" width="864"></p>
</div>
<div id="sorting-by-metadata" class="section level3">
<h3 class="hasAnchor">
<a href="#sorting-by-metadata" class="anchor"></a>Sorting by metadata</h3>
<p>Sorting across timepoint groups in another dataset, dietswap.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"dietswap"</span>, package <span class="op">=</span> <span class="st">"microbiome"</span><span class="op">)</span>
<span class="va">ps</span> <span class="op">&lt;-</span> <span class="va">dietswap</span> <span class="op">%&gt;%</span> <span class="fu"><a href="../../reference/ps_filter.html">ps_filter</a></span><span class="op">(</span><span class="va">group</span> <span class="op">==</span> <span class="st">"DI"</span><span class="op">)</span>

<span class="co"># subset to participants/"subjects" with samples for both timepoints</span>
<span class="va">timepoint_groups</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/sample_data-methods.html">sample_data</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/base/split.html">split.data.frame</a></span><span class="op">(</span>f <span class="op">=</span> <span class="va">.</span><span class="op">$</span><span class="va">timepoint.within.group</span><span class="op">)</span>
<span class="va">have_both_timepoints</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/setops.html">intersect</a></span><span class="op">(</span><span class="va">timepoint_groups</span><span class="op">$</span><span class="va">`1`</span><span class="op">$</span><span class="va">subject</span>, <span class="va">timepoint_groups</span><span class="op">$</span><span class="va">`2`</span><span class="op">$</span><span class="va">subject</span><span class="op">)</span>
<span class="va">ps</span> <span class="op">&lt;-</span> <span class="va">ps</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/subset_samples-methods.html">subset_samples</a></span><span class="op">(</span><span class="va">subject</span> <span class="op">%in%</span> <span class="va">have_both_timepoints</span><span class="op">)</span>

<span class="co"># define grouping variable for plotting</span>
<span class="va">ps</span> <span class="op">&lt;-</span> <span class="va">ps</span> <span class="op">%&gt;%</span> <span class="fu"><a href="../../reference/ps_mutate.html">ps_mutate</a></span><span class="op">(</span>
  plot_groups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/interaction.html">interaction</a></span><span class="op">(</span><span class="va">nationality</span>, <span class="va">timepoint.within.group</span><span class="op">)</span>
<span class="op">)</span></code></pre></div>
<p>Grouped by both timepoint and another grouping factor, nationality in this example.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">times_list</span> <span class="op">&lt;-</span> <span class="va">ps</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../../reference/ps_arrange.html">ps_arrange</a></span><span class="op">(</span><span class="va">timepoint.within.group</span>, <span class="va">nationality</span>, <span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html">desc</a></span><span class="op">(</span><span class="va">subject</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../../reference/comp_barplot.html">comp_barplot</a></span><span class="op">(</span>
    tax_level <span class="op">=</span> <span class="st">"Genus"</span>, n_taxa <span class="op">=</span> <span class="fl">10</span>, sample_order <span class="op">=</span> <span class="st">"default"</span>,
    group_by <span class="op">=</span> <span class="st">"plot_groups"</span>, bar_width <span class="op">=</span> <span class="fl">0.7</span>, label <span class="op">=</span> <span class="st">"subject"</span>
  <span class="op">)</span>

<span class="va">times</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html">wrap_plots</a></span><span class="op">(</span>
  <span class="va">times_list</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"AAM.1"</span>, <span class="st">"AAM.2"</span>, <span class="st">"AFR.1"</span>, <span class="st">"AFR.2"</span><span class="op">)</span><span class="op">]</span>,
  byrow <span class="op">=</span> <span class="cn">TRUE</span>, ncol <span class="op">=</span> <span class="fl">2</span>, guides <span class="op">=</span> <span class="st">"collect"</span>, heights <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">4</span><span class="op">)</span>
<span class="op">)</span> <span class="op">&amp;</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">&amp;</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>
    text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>,
    axis.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>
  <span class="op">)</span>

<span class="va">times</span></code></pre></div>
<p><img src="compositions_files/figure-html/unnamed-chunk-11-1.png" width="960"></p>
<p>Same grouping, now showing diversity of taxa within other, with <code>merge_other = FALSE</code></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">times_list</span> <span class="op">&lt;-</span> <span class="va">ps</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../../reference/ps_arrange.html">ps_arrange</a></span><span class="op">(</span><span class="va">timepoint.within.group</span>, <span class="va">nationality</span>, <span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html">desc</a></span><span class="op">(</span><span class="va">subject</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../../reference/comp_barplot.html">comp_barplot</a></span><span class="op">(</span>
    tax_level <span class="op">=</span> <span class="st">"Genus"</span>, n_taxa <span class="op">=</span> <span class="fl">10</span>, sample_order <span class="op">=</span> <span class="st">"default"</span>,
    merge_other <span class="op">=</span> <span class="cn">FALSE</span>, bar_outline_colour <span class="op">=</span> <span class="st">"grey10"</span>,
    group_by <span class="op">=</span> <span class="st">"plot_groups"</span>, bar_width <span class="op">=</span> <span class="fl">0.7</span>, label <span class="op">=</span> <span class="st">"subject"</span>
  <span class="op">)</span>

<span class="va">times</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html">wrap_plots</a></span><span class="op">(</span>
  <span class="va">times_list</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"AAM.1"</span>, <span class="st">"AAM.2"</span>, <span class="st">"AFR.1"</span>, <span class="st">"AFR.2"</span><span class="op">)</span><span class="op">]</span>,
  byrow <span class="op">=</span> <span class="cn">TRUE</span>, ncol <span class="op">=</span> <span class="fl">2</span>, guides <span class="op">=</span> <span class="st">"collect"</span>, heights <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">4</span><span class="op">)</span>
<span class="op">)</span> <span class="op">&amp;</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">&amp;</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>
    text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>,
    axis.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>
  <span class="op">)</span>

<span class="va">times</span></code></pre></div>
<p><img src="compositions_files/figure-html/unnamed-chunk-12-1.png" width="960"></p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/sessioninfo/man/session_info.html">session_info</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; ─ Session info ───────────────────────────────────────────────────────────────────────────────────</span>
<span class="co">#&gt;  setting  value                       </span>
<span class="co">#&gt;  version  R version 4.1.1 (2021-08-10)</span>
<span class="co">#&gt;  os       macOS Catalina 10.15.7      </span>
<span class="co">#&gt;  system   x86_64, darwin17.0          </span>
<span class="co">#&gt;  ui       X11                         </span>
<span class="co">#&gt;  language (EN)                        </span>
<span class="co">#&gt;  collate  en_US.UTF-8                 </span>
<span class="co">#&gt;  ctype    en_US.UTF-8                 </span>
<span class="co">#&gt;  tz       UTC                         </span>
<span class="co">#&gt;  date     2021-09-29                  </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; ─ Packages ───────────────────────────────────────────────────────────────────────────────────────</span>
<span class="co">#&gt;  package          * version     date       lib source        </span>
<span class="co">#&gt;  ade4               1.7-18      2021-09-16 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  ape                5.5         2021-04-25 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  Biobase            2.52.0      2021-05-19 [1] Bioconductor  </span>
<span class="co">#&gt;  BiocGenerics       0.38.0      2021-05-19 [1] Bioconductor  </span>
<span class="co">#&gt;  biomformat         1.20.0      2021-05-19 [1] Bioconductor  </span>
<span class="co">#&gt;  Biostrings         2.60.2      2021-08-05 [1] Bioconductor  </span>
<span class="co">#&gt;  bitops             1.0-7       2021-04-24 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  bslib              0.3.0       2021-09-02 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  cachem             1.0.6       2021-08-19 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  callr              3.7.0       2021-04-20 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  cli                3.0.1       2021-07-17 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  cluster            2.1.2       2021-04-17 [2] CRAN (R 4.1.1)</span>
<span class="co">#&gt;  codetools          0.2-18      2020-11-04 [2] CRAN (R 4.1.1)</span>
<span class="co">#&gt;  colorspace         2.0-2       2021-06-24 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  corncob            0.2.0       2021-03-11 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  crayon             1.4.1       2021-02-08 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  data.table         1.14.2      2021-09-27 [1] CRAN (R 4.1.1)</span>
<span class="co">#&gt;  desc               1.4.0       2021-09-28 [1] CRAN (R 4.1.1)</span>
<span class="co">#&gt;  devtools           2.4.2       2021-06-07 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  digest             0.6.28      2021-09-23 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  dplyr              1.0.7       2021-06-18 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  ellipsis           0.3.2       2021-04-29 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  evaluate           0.14        2019-05-28 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  fansi              0.5.0       2021-05-25 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  farver             2.1.0       2021-02-28 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  fastmap            1.1.0       2021-01-25 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  foreach            1.5.1       2020-10-15 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  fs                 1.5.0       2020-07-31 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  generics           0.1.0       2020-10-31 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  GenomeInfoDb       1.28.4      2021-09-09 [1] Bioconductor  </span>
<span class="co">#&gt;  GenomeInfoDbData   1.2.6       2021-09-23 [1] Bioconductor  </span>
<span class="co">#&gt;  ggplot2          * 3.3.5       2021-06-25 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  glue               1.4.2       2020-08-27 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  gtable             0.3.0       2019-03-25 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  highr              0.9         2021-04-16 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  htmltools          0.5.2       2021-08-25 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  igraph             1.2.6       2020-10-06 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  IRanges            2.26.0      2021-05-19 [1] Bioconductor  </span>
<span class="co">#&gt;  iterators          1.0.13      2020-10-15 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  jquerylib          0.1.4       2021-04-26 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  jsonlite           1.7.2       2020-12-09 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  knitr              1.36        2021-09-29 [1] CRAN (R 4.1.1)</span>
<span class="co">#&gt;  labeling           0.4.2       2020-10-20 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  lattice            0.20-44     2021-05-02 [2] CRAN (R 4.1.1)</span>
<span class="co">#&gt;  lifecycle          1.0.1       2021-09-24 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  magrittr           2.0.1       2020-11-17 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  MASS               7.3-54      2021-05-03 [2] CRAN (R 4.1.1)</span>
<span class="co">#&gt;  Matrix             1.3-4       2021-06-01 [2] CRAN (R 4.1.1)</span>
<span class="co">#&gt;  memoise            2.0.0       2021-01-26 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  mgcv               1.8-36      2021-06-01 [2] CRAN (R 4.1.1)</span>
<span class="co">#&gt;  microbiome         1.14.0      2021-05-19 [1] Bioconductor  </span>
<span class="co">#&gt;  microViz         * 0.7.10.9032 2021-09-29 [1] local         </span>
<span class="co">#&gt;  multtest           2.48.0      2021-05-19 [1] Bioconductor  </span>
<span class="co">#&gt;  munsell            0.5.0       2018-06-12 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  nlme               3.1-152     2021-02-04 [2] CRAN (R 4.1.1)</span>
<span class="co">#&gt;  patchwork        * 1.1.1       2020-12-17 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  permute            0.9-5       2019-03-12 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  phyloseq         * 1.36.0      2021-05-19 [1] Bioconductor  </span>
<span class="co">#&gt;  pillar             1.6.3       2021-09-26 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  pkgbuild           1.2.0       2020-12-15 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  pkgconfig          2.0.3       2019-09-22 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  pkgdown            1.6.1       2020-09-12 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  pkgload            1.2.2       2021-09-11 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  plyr               1.8.6       2020-03-03 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  prettyunits        1.1.1       2020-01-24 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  processx           3.5.2       2021-04-30 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  ps                 1.6.0       2021-02-28 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  purrr              0.3.4       2020-04-17 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  R6                 2.5.1       2021-08-19 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  ragg               1.1.3       2021-06-09 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  Rcpp               1.0.7       2021-07-07 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  RCurl              1.98-1.5    2021-09-17 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  registry           0.5-1       2019-03-05 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  remotes            2.4.1       2021-09-29 [1] CRAN (R 4.1.1)</span>
<span class="co">#&gt;  reshape2           1.4.4       2020-04-09 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  rhdf5              2.36.0      2021-05-19 [1] Bioconductor  </span>
<span class="co">#&gt;  rhdf5filters       1.4.0       2021-05-19 [1] Bioconductor  </span>
<span class="co">#&gt;  Rhdf5lib           1.14.2      2021-07-06 [1] Bioconductor  </span>
<span class="co">#&gt;  rlang              0.4.11      2021-04-30 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  rmarkdown          2.11        2021-09-14 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  rprojroot          2.0.2       2020-11-15 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  Rtsne              0.15        2018-11-10 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  S4Vectors          0.30.1      2021-09-26 [1] Bioconductor  </span>
<span class="co">#&gt;  sass               0.4.0       2021-05-12 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  scales             1.1.1       2020-05-11 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  seriation          1.3.0       2021-06-30 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  sessioninfo        1.1.1       2018-11-05 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  stringi            1.7.4       2021-08-25 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  stringr            1.4.0       2019-02-10 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  survival           3.2-11      2021-04-26 [2] CRAN (R 4.1.1)</span>
<span class="co">#&gt;  systemfonts        1.0.2       2021-05-11 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  testthat           3.0.4       2021-07-01 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  textshaping        0.3.5       2021-06-09 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  tibble             3.1.4       2021-08-25 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  tidyr              1.1.4       2021-09-27 [1] CRAN (R 4.1.1)</span>
<span class="co">#&gt;  tidyselect         1.1.1       2021-04-30 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  TSP                1.1-10      2020-04-17 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  usethis            2.0.1       2021-02-10 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  utf8               1.2.2       2021-07-24 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  vctrs              0.3.8       2021-04-29 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  vegan              2.5-7       2020-11-28 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  withr              2.4.2       2021-04-18 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  xfun               0.26        2021-09-14 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  XVector            0.32.0      2021-05-19 [1] Bioconductor  </span>
<span class="co">#&gt;  yaml               2.2.1       2020-02-01 [1] CRAN (R 4.1.0)</span>
<span class="co">#&gt;  zlibbioc           1.38.0      2021-05-19 [1] Bioconductor  </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [1] /Users/runner/work/_temp/Library</span>
<span class="co">#&gt; [2] /Library/Frameworks/R.framework/Versions/4.1/Resources/library</span></code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by David Barnett.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: '22e6f0a75b7b1ffae9bf5bc5b58864d7',
    indexName: 'microViz',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
