<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Visualising compositions • microViz</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../../favicon.ico">
<link rel="manifest" href="../../site.webmanifest">
<script src="../../lightswitch.js"></script><script src="../../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../../deps/headroom-0.11.0/headroom.min.js"></script><script src="../../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../../deps/search-1.0.0/fuse.min.js"></script><script src="../../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../../pkgdown.js"></script><meta property="og:title" content="Visualising compositions">
<!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-2YQ669S4V5"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-2YQ669S4V5');
</script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../../index.html">microViz</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.13.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../../articles/shao19-analyses.html">Getting Started</a></li>
<li class="nav-item"><a class="nav-link" href="../../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../../articles/web-only/phyloseq.html">Working with phyloseq objects</a></li>
    <li><a class="dropdown-item" href="../../articles/web-only/tax-fixing.html">Fixing your tax_table</a></li>
    <li><a class="dropdown-item" href="../../articles/web-only/ordination.html">Ordination plots</a></li>
    <li><a class="dropdown-item" href="../../articles/web-only/ordination-interactive.html">Interactive ordination plots</a></li>
    <li><a class="dropdown-item" href="../../articles/web-only/compositions.html">Visualising compositions with barplots</a></li>
    <li><a class="dropdown-item" href="../../articles/web-only/heatmaps.html">Heatmaps of microbiome composition or correlation</a></li>
    <li><a class="dropdown-item" href="../../articles/web-only/modelling-taxa.html">Statistical modelling of individual taxa</a></li>
    <li><a class="dropdown-item" href="../../articles/web-only/videos.html">Tutorial video</a></li>
    <li><a class="dropdown-item" href="../../articles/ps_extra-replaced.html">Upgrading to microViz version 0.10.0</a></li>
    <li><a class="dropdown-item" href="../../articles/web-only/atlas1006.html">Example analyses with atlas1006 data</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../../articles/learn/exercises.html">Learn</a></li>
<li class="nav-item"><a class="nav-link" href="../../news/index.html">News</a></li>
<li class="nav-item"><h6 class="dropdown-header" data-toc-skip></h6></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/david-barnett/microViz/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../../logo.png" class="logo" alt=""><h1>Visualising compositions</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/david-barnett/microViz/blob/main/vignettes/web-only/compositions.Rmd" class="external-link"><code>vignettes/web-only/compositions.Rmd</code></a></small>
      <div class="d-none name"><code>compositions.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://david-barnett.github.io/microViz">microViz</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://dx.plos.org/10.1371/journal.pone.0061217" class="external-link">phyloseq</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span> <span class="co"># for arranging groups of plots</span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/knitr/man/opts_chunk.html" class="external-link">opts_chunk</a></span><span class="op">$</span><span class="fu">set</span><span class="op">(</span>fig.height <span class="op">=</span> <span class="fl">6</span>, fig.width <span class="op">=</span> <span class="fl">9</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get example phyloseq data from corncob package and tidy up</span></span>
<span><span class="va">pseq</span> <span class="op">&lt;-</span> <span class="fu">microViz</span><span class="fu">::</span><span class="va"><a href="../../reference/ibd.html">ibd</a></span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/tax_filter.html">tax_filter</a></span><span class="op">(</span>min_prevalence <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/tax_fix.html">tax_fix</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/phyloseq_validate.html">phyloseq_validate</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>The <code>comp_barplot</code> function allows you to visualize the
taxonomic compositions of your microbiome samples in a flexible,
scalable, group-able, and visually appealing way.</p>
<div class="section level2">
<h2 id="quick-example-barplot">Quick example barplot<a class="anchor" aria-label="anchor" href="#quick-example-barplot"></a>
</h2>
<p>Visualize the top Genera across all the female samples from this
inflammatory bowel disease study dataset. The order of the samples is
automatically set by their “bray”-curtis dissimilarity.</p>
<p>By default, the top 8 taxa are shown. These taxa are chosen by their
total count abundance across all plotted samples.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pseq</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/ps_filter.html">ps_filter</a></span><span class="op">(</span><span class="va">gender</span> <span class="op">==</span> <span class="st">"female"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/comp_barplot.html">comp_barplot</a></span><span class="op">(</span>tax_level <span class="op">=</span> <span class="st">"Genus"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># horizontal bars are often more readable</span></span>
<span><span class="co">#&gt; Registered S3 method overwritten by 'seriation':</span></span>
<span><span class="co">#&gt;   method         from </span></span>
<span><span class="co">#&gt;   reorder.hclust vegan</span></span></code></pre></div>
<p><img src="compositions_files/figure-html/simple-1.png" class="r-plt" alt="" width="864"></p>
</div>
<div class="section level2">
<h2 id="customising-the-barplot">Customising the barplot<a class="anchor" aria-label="anchor" href="#customising-the-barplot"></a>
</h2>
<div class="section level3">
<h3 id="comp_barplot-arguments">comp_barplot arguments<a class="anchor" aria-label="anchor" href="#comp_barplot-arguments"></a>
</h3>
<p>The output of comp_barplot can be customised in several ways. See the
comment alongside each argument for an explanation of its effect.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pseq</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/ps_filter.html">ps_filter</a></span><span class="op">(</span><span class="va">gender</span> <span class="op">==</span> <span class="st">"female"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/comp_barplot.html">comp_barplot</a></span><span class="op">(</span></span>
<span>    tax_level <span class="op">=</span> <span class="st">"Genus"</span>,</span>
<span>    label <span class="op">=</span> <span class="st">"DiseaseState"</span>, <span class="co"># name an alternative variable to label axis</span></span>
<span>    n_taxa <span class="op">=</span> <span class="fl">15</span>, <span class="co"># give more taxa unique colours</span></span>
<span>    taxon_renamer <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_replace.html" class="external-link">str_replace_all</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">"_"</span>, <span class="st">" "</span><span class="op">)</span>, <span class="co"># remove underscores</span></span>
<span>    other_name <span class="op">=</span> <span class="st">"Other genera"</span>, <span class="co"># set custom name for the "other" category</span></span>
<span>    merge_other <span class="op">=</span> <span class="cn">FALSE</span>, <span class="co"># split the "Other" category to display alpha diversity</span></span>
<span>    bar_width <span class="op">=</span> <span class="fl">0.7</span>, <span class="co"># reduce the bar width to 70% of one row</span></span>
<span>    bar_outline_colour <span class="op">=</span> <span class="st">"grey5"</span> <span class="co"># is the default (use NA to remove outlines)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="compositions_files/figure-html/unnamed-chunk-3-1.png" class="r-plt" alt="" width="864"></p>
<p><strong>Other notes:</strong></p>
<ul>
<li>Dissimilarity is calculated using only the visibly distinct taxa, to
optimize sorting for visual similarity. You can change this by setting
<code>order_with_all_taxa = TRUE</code>, to always use all taxa for
similarity sorting.</li>
<li>The colour palette is important, to allow (adjacent) taxa to be
distinguished. The palette microViz uses is generated by the
<code>distinct_palette</code> function, which starts with the Paired and
Dark2 palettes from ColorBrewer and continues with further distinct
colours generated at <a href="http://medialab.github.io/iwanthue/" class="external-link uri">http://medialab.github.io/iwanthue/</a> (all colors, soft
k-means).</li>
</ul>
</div>
<div class="section level3">
<h3 id="colour-palette-customisation">Colour palette customisation<a class="anchor" aria-label="anchor" href="#colour-palette-customisation"></a>
</h3>
<p>You can use the <code><a href="../../reference/tax_palette.html">tax_palette()</a></code> function to help create a
custom colour palette for use in (multiple) barplots. Try to ensure you
assign a colour for every taxon you expect to see separately on your
barplot.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myPal</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/tax_palette.html">tax_palette</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="va">pseq</span>, rank <span class="op">=</span> <span class="st">"Genus"</span>, n <span class="op">=</span> <span class="fl">25</span>, pal <span class="op">=</span> <span class="st">"greenArmytage"</span>,</span>
<span>  add <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>Other <span class="op">=</span> <span class="st">"white"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="../../reference/tax_palette_plot.html">tax_palette_plot</a></span><span class="op">(</span><span class="va">myPal</span><span class="op">)</span></span></code></pre></div>
<p><img src="compositions_files/figure-html/unnamed-chunk-4-1.png" class="r-plt" alt="" width="384"></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Override existing values</span></span>
<span><span class="va">myPal</span><span class="op">[</span><span class="st">"Bacteroides"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"grey75"</span></span>
<span><span class="va">myPal</span><span class="op">[</span><span class="st">"Streptococcus"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"black"</span></span>
<span><span class="va">myPal</span><span class="op">[</span><span class="st">"Klebsiella"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"darkorange"</span></span>
<span><span class="co"># Add more values</span></span>
<span><span class="va">myPal</span><span class="op">[</span><span class="st">"Enterobacteriaceae Family"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"hotpink"</span></span>
<span><span class="fu"><a href="../../reference/tax_palette_plot.html">tax_palette_plot</a></span><span class="op">(</span><span class="va">myPal</span><span class="op">)</span></span></code></pre></div>
<p><img src="compositions_files/figure-html/unnamed-chunk-5-1.png" class="r-plt" alt="" width="384"></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pseq</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/ps_filter.html">ps_filter</a></span><span class="op">(</span><span class="va">gender</span> <span class="op">==</span> <span class="st">"female"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/comp_barplot.html">comp_barplot</a></span><span class="op">(</span></span>
<span>    tax_level <span class="op">=</span> <span class="st">"Genus"</span>, palette <span class="op">=</span> <span class="va">myPal</span>,</span>
<span>    n_taxa <span class="op">=</span> <span class="fl">12</span>, other_name <span class="op">=</span> <span class="st">"Other"</span>, merge_other <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="compositions_files/figure-html/unnamed-chunk-6-1.png" class="r-plt" alt="" width="864"></p>
</div>
<div class="section level3">
<h3 id="alternative-taxa-ordering">Alternative taxa ordering<a class="anchor" aria-label="anchor" href="#alternative-taxa-ordering"></a>
</h3>
<p>By default taxa are ordered by overall <code>sum</code> of their
counts across all samples in your dataset. You can sort taxa by another
function, such as <code>prev</code> for prevalence.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pseq</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/ps_filter.html">ps_filter</a></span><span class="op">(</span><span class="va">gender</span> <span class="op">==</span> <span class="st">"male"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/comp_barplot.html">comp_barplot</a></span><span class="op">(</span>tax_level <span class="op">=</span> <span class="st">"Genus"</span>, tax_order <span class="op">=</span> <span class="va">prev</span>, merge_other <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="compositions_files/figure-html/unnamed-chunk-7-1.png" class="r-plt" alt="" width="864"></p>
</div>
<div class="section level3">
<h3 id="customising-taxa-order">Customising taxa order<a class="anchor" aria-label="anchor" href="#customising-taxa-order"></a>
</h3>
<p>It is easy (since microViz version 0.9.6) to set a custom order of
taxa to display with <code>comp_barplot</code>. For example you might be
particularly interested in the abundance of Proteobacteria genera in
your samples, and want them to appear first, regardless of their overall
abundance.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">interestingGenera</span> <span class="op">&lt;-</span> <span class="va">pseq</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/tax_select.html">tax_select</a></span><span class="op">(</span><span class="st">"Proteobacteria"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/tax_top.html">tax_top</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">10</span>, rank <span class="op">=</span> <span class="st">"Genus"</span><span class="op">)</span></span>
<span><span class="va">interestingGenera</span></span>
<span><span class="co">#&gt;  [1] "Escherichia/Shigella"      "Klebsiella"                "Haemophilus"              </span></span>
<span><span class="co">#&gt;  [4] "Sutterella"                "Enterobacteriaceae Family" "Parasutterella"           </span></span>
<span><span class="co">#&gt;  [7] "Cronobacter"               "Comamonadaceae Family"     "Raoultella"               </span></span>
<span><span class="co">#&gt; [10] "Proteus"</span></span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pseq</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/ps_filter.html">ps_filter</a></span><span class="op">(</span><span class="va">gender</span> <span class="op">==</span> <span class="st">"male"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/tax_sort.html">tax_sort</a></span><span class="op">(</span>by <span class="op">=</span> <span class="va">sum</span>, at <span class="op">=</span> <span class="st">"Genus"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> <span class="co"># put other taxa in a reasonable order</span></span>
<span>  <span class="fu"><a href="../../reference/comp_barplot.html">comp_barplot</a></span><span class="op">(</span></span>
<span>    tax_level <span class="op">=</span> <span class="st">"Genus"</span>, n_taxa <span class="op">=</span> <span class="fl">10</span>, merge_other <span class="op">=</span> <span class="cn">FALSE</span>, other_name <span class="op">=</span> <span class="st">"Other"</span>,</span>
<span>    palette <span class="op">=</span> <span class="fu"><a href="../../reference/distinct_palette.html">distinct_palette</a></span><span class="op">(</span><span class="fl">10</span>, pal <span class="op">=</span> <span class="st">"kelly"</span>, add <span class="op">=</span> <span class="st">"grey90"</span><span class="op">)</span>,</span>
<span>    tax_order <span class="op">=</span> <span class="va">interestingGenera</span> <span class="co"># this is the reordering magic</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="compositions_files/figure-html/unnamed-chunk-9-1.png" class="r-plt" alt="" width="864"></p>
</div>
<div class="section level3">
<h3 id="custom-taxa-order-and-colour-palette">Custom taxa order and colour palette<a class="anchor" aria-label="anchor" href="#custom-taxa-order-and-colour-palette"></a>
</h3>
<p>Customising taxa order and colour palette are often needed together.
The pair nicely as you can use the names from a custom palette to fix
the taxa order.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">customPal</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/tax_palette.html">tax_palette</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="va">pseq</span>, rank <span class="op">=</span> <span class="st">"Genus"</span>, pal <span class="op">=</span> <span class="st">"kelly"</span>, n <span class="op">=</span> <span class="fl">20</span>, add <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>Other <span class="op">=</span> <span class="st">"white"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="../../reference/tax_palette_plot.html">tax_palette_plot</a></span><span class="op">(</span><span class="va">customPal</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># I'm interested in Prevotella, so let's put that first</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">customPal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Prevotella"</span>, <span class="fu"><a href="https://generics.r-lib.org/reference/setops.html" class="external-link">setdiff</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">customPal</span><span class="op">)</span>, <span class="st">"Prevotella"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../../reference/tax_palette_plot.html">tax_palette_plot</a></span><span class="op">(</span><span class="va">customPal</span><span class="op">)</span></span></code></pre></div>
<p><img src="compositions_files/figure-html/unnamed-chunk-10-1.png" class="r-plt" alt="" width="288"><img src="compositions_files/figure-html/unnamed-chunk-10-2.png" class="r-plt" alt="" width="288"></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pseq</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/ps_filter.html">ps_filter</a></span><span class="op">(</span><span class="va">gender</span> <span class="op">==</span> <span class="st">"female"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/tax_sort.html">tax_sort</a></span><span class="op">(</span>by <span class="op">=</span> <span class="va">sum</span>, at <span class="op">=</span> <span class="st">"Genus"</span>, tree_warn <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/comp_barplot.html">comp_barplot</a></span><span class="op">(</span></span>
<span>    tax_level <span class="op">=</span> <span class="st">"Genus"</span>, merge_other <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    n_taxa <span class="op">=</span> <span class="fl">12</span>,</span>
<span>    other_name <span class="op">=</span> <span class="st">"Other"</span>, <span class="co"># must match a name in palette</span></span>
<span>    tax_order <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">customPal</span><span class="op">)</span>,</span>
<span>    palette <span class="op">=</span> <span class="va">customPal</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="compositions_files/figure-html/unnamed-chunk-11-1.png" class="r-plt" alt="" width="864"></p>
<p>Let’s look at the abundance of these same genera in a separate
dataset.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"shao19"</span><span class="op">)</span></span>
<span><span class="va">shao19</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/ps_filter.html">ps_filter</a></span><span class="op">(</span><span class="va">family_role</span> <span class="op">==</span> <span class="st">"mother"</span>, <span class="va">age</span> <span class="op">==</span> <span class="fl">38</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/tax_sort.html">tax_sort</a></span><span class="op">(</span>by <span class="op">=</span> <span class="va">sum</span>, at <span class="op">=</span> <span class="st">"genus"</span>, tree_warn <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/comp_barplot.html">comp_barplot</a></span><span class="op">(</span></span>
<span>    tax_level <span class="op">=</span> <span class="st">"genus"</span>, n_taxa <span class="op">=</span> <span class="fl">12</span>, other_name <span class="op">=</span> <span class="st">"Other"</span>, merge_other <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    tax_order <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">customPal</span><span class="op">)</span>, palette <span class="op">=</span> <span class="va">customPal</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in tax_reorder(ps, tax_order = tax_order, tree_warn = FALSE, ignore = other_name): 5 taxa</span></span>
<span><span class="co">#&gt; specified in tax_order are not in phyloseq ps: they are ignored</span></span></code></pre></div>
<p><img src="compositions_files/figure-html/unnamed-chunk-12-1.png" class="r-plt" alt="" width="864"></p>
<p>Not all the genera will be present in the other dataset, so we get a
warning. Notice also that whilst the genus-colour pairings are the same,
the legends are not identical, because the taxa that are not in this 2nd
set don’t appear in its legend.</p>
</div>
<div class="section level3">
<h3 id="merging-phyloseq-objects-to-compare-them">Merging phyloseq objects to compare them<a class="anchor" aria-label="anchor" href="#merging-phyloseq-objects-to-compare-them"></a>
</h3>
<p>The best way to create harmonised barcharts for two (or more)
separate datasets (e.g. your data and a public dataset) is to merge the
datasets. Often the data generation methods (e.g. primer sets) won’t
match, making ASV or species-level merging impossible (or not
advisable).</p>
<p>You can still try merging at a higher taxonomic rank, such as Class,
for comparison.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dataset1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/ps_filter.html">ps_filter</a></span><span class="op">(</span><span class="va">pseq</span>, <span class="va">gender</span> <span class="op">==</span> <span class="st">"female"</span><span class="op">)</span></span>
<span><span class="va">dataset2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/ps_filter.html">ps_filter</a></span><span class="op">(</span><span class="va">shao19</span>, <span class="va">family_role</span> <span class="op">==</span> <span class="st">"mother"</span>, <span class="va">age</span> <span class="op">==</span> <span class="fl">38</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># first make sure the rank_names are the same format (e.g. both lowercase)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">dataset1</span><span class="op">@</span><span class="va">tax_table</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/rank_names.html" class="external-link">rank_names</a></span><span class="op">(</span><span class="va">dataset1</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/chartr.html" class="external-link">tolower</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># remove ranks that are not shared</span></span>
<span><span class="va">dataset1</span> <span class="op">&lt;-</span> <span class="va">dataset1</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../../reference/tax_mutate.html">tax_mutate</a></span><span class="op">(</span>kingdom <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># create a dataset ID variable to distinguish the two datasets after merging</span></span>
<span><span class="va">dataset1</span> <span class="op">&lt;-</span> <span class="va">dataset1</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../../reference/ps_mutate.html">ps_mutate</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="st">"IBD study"</span><span class="op">)</span></span>
<span><span class="va">dataset2</span> <span class="op">&lt;-</span> <span class="va">dataset2</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../../reference/ps_mutate.html">ps_mutate</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="st">"Birth cohort mothers"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># aggregate and merge</span></span>
<span><span class="va">combined</span> <span class="op">&lt;-</span> <span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/merge_phyloseq.html" class="external-link">merge_phyloseq</a></span><span class="op">(</span></span>
<span>  <span class="va">dataset1</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../../reference/tax_agg.html">tax_agg</a></span><span class="op">(</span><span class="st">"class"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../../reference/psExtra-accessors.html">ps_get</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  <span class="va">dataset2</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../../reference/tax_agg.html">tax_agg</a></span><span class="op">(</span><span class="st">"class"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../../reference/psExtra-accessors.html">ps_get</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">combined</span></span>
<span><span class="co">#&gt; phyloseq-class experiment-level object</span></span>
<span><span class="co">#&gt; otu_table()   OTU Table:         [ 28 taxa and 93 samples ]</span></span>
<span><span class="co">#&gt; sample_data() Sample Data:       [ 93 samples by 26 sample variables ]</span></span>
<span><span class="co">#&gt; tax_table()   Taxonomy Table:    [ 28 taxa by 2 taxonomic ranks ]</span></span></code></pre></div>
<p>Now we can plot the two datasets in separate facets of the same
plot.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">combined</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/comp_barplot.html">comp_barplot</a></span><span class="op">(</span><span class="st">"class"</span>, facet_by <span class="op">=</span> <span class="st">"dataset"</span>, n_taxa <span class="op">=</span> <span class="fl">12</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="compositions_files/figure-html/unnamed-chunk-14-1.png" class="r-plt" alt="" width="864"></p>
<p>Or we could plot them as separate plots with the group argument.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plots</span> <span class="op">&lt;-</span> <span class="va">combined</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../../reference/comp_barplot.html">comp_barplot</a></span><span class="op">(</span><span class="st">"class"</span>, n_taxa <span class="op">=</span> <span class="fl">12</span>, group_by <span class="op">=</span> <span class="st">"dataset"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># combine plots with shared legend</span></span>
<span><span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">plots</span>, nrow <span class="op">=</span> <span class="fl">2</span>, guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.ticks.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>, axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="compositions_files/figure-html/unnamed-chunk-15-1.png" class="r-plt" alt="" width="864"></p>
</div>
<div class="section level3">
<h3 id="alphabetical-top-taxa">Alphabetical top taxa<a class="anchor" aria-label="anchor" href="#alphabetical-top-taxa"></a>
</h3>
<p>Sometimes you might prefer the top taxa to be shown in alphabetical
order.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># set up for alphabetical sorting</span></span>
<span><span class="va">topTaxa</span> <span class="op">&lt;-</span> <span class="va">pseq</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/ps_filter.html">ps_filter</a></span><span class="op">(</span><span class="va">gender</span> <span class="op">==</span> <span class="st">"male"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/tax_top.html">tax_top</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">9</span>, rank <span class="op">=</span> <span class="st">"Genus"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># this makes them alphabetical</span></span>
<span></span>
<span><span class="co"># plot with alphabetical sorting</span></span>
<span><span class="va">pseq</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/ps_filter.html">ps_filter</a></span><span class="op">(</span><span class="va">gender</span> <span class="op">==</span> <span class="st">"male"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/tax_sort.html">tax_sort</a></span><span class="op">(</span>by <span class="op">=</span> <span class="va">sum</span>, at <span class="op">=</span> <span class="st">"Genus"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> <span class="co"># this orders all genera by abundance</span></span>
<span>  <span class="fu"><a href="../../reference/comp_barplot.html">comp_barplot</a></span><span class="op">(</span></span>
<span>    tax_order <span class="op">=</span> <span class="va">topTaxa</span>, <span class="co"># this brings the named taxa to the front</span></span>
<span>    tax_level <span class="op">=</span> <span class="st">"Genus"</span>, n_taxa <span class="op">=</span> <span class="fl">9</span>, merge_other <span class="op">=</span> <span class="cn">FALSE</span>, other_name <span class="op">=</span> <span class="st">"Other"</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="compositions_files/figure-html/unnamed-chunk-16-1.png" class="r-plt" alt="" width="864"></p>
<p>Another way to do this is by reordering the names of a custom
palette, like this:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># don't add an "Other" colour for now</span></span>
<span><span class="va">alphaPal</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/tax_palette.html">tax_palette</a></span><span class="op">(</span><span class="va">pseq</span>, pal <span class="op">=</span> <span class="st">"kelly"</span>, rank <span class="op">=</span> <span class="st">"Genus"</span>, n <span class="op">=</span> <span class="fl">12</span>, add <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">alphaPal</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">alphaPal</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># now add the "Other" colour to the end</span></span>
<span><span class="va">alphaPal</span><span class="op">[</span><span class="st">"Other"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"white"</span></span>
<span><span class="fu"><a href="../../reference/tax_palette_plot.html">tax_palette_plot</a></span><span class="op">(</span><span class="va">alphaPal</span><span class="op">)</span></span></code></pre></div>
<p><img src="compositions_files/figure-html/unnamed-chunk-17-1.png" class="r-plt" alt="" width="384"></p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pseq</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/ps_filter.html">ps_filter</a></span><span class="op">(</span><span class="va">gender</span> <span class="op">==</span> <span class="st">"male"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/comp_barplot.html">comp_barplot</a></span><span class="op">(</span></span>
<span>    tax_level <span class="op">=</span> <span class="st">"Genus"</span>, n_taxa <span class="op">=</span> <span class="fl">12</span>, merge_other <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    other_name <span class="op">=</span> <span class="st">"Other"</span>, <span class="co"># must match a name in palette</span></span>
<span>    tax_order <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">alphaPal</span><span class="op">)</span>, palette <span class="op">=</span> <span class="va">alphaPal</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="compositions_files/figure-html/unnamed-chunk-18-1.png" class="r-plt" alt="" width="864"></p>
</div>
<div class="section level3">
<h3 id="custom-hierarchical-sorting-and-palette">Custom hierarchical sorting and palette<a class="anchor" aria-label="anchor" href="#custom-hierarchical-sorting-and-palette"></a>
</h3>
<p>A complex example showing one method for obtaining a hierarchical
colour palette with hues specified by Phylum (or another high rank) and
shades of the hue specified by Family (or another low rank).</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hueRank</span> <span class="op">&lt;-</span> <span class="st">"Phylum"</span></span>
<span><span class="va">hueRankPlural</span> <span class="op">&lt;-</span> <span class="st">"Phyla"</span></span>
<span><span class="va">shadeRank</span> <span class="op">&lt;-</span> <span class="st">"Family"</span></span>
<span></span>
<span><span class="co"># Sort phyloseq at lower, and then higher ranks</span></span>
<span><span class="va">pseq2</span> <span class="op">&lt;-</span> <span class="va">pseq</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/ps_filter.html">ps_filter</a></span><span class="op">(</span><span class="va">gender</span> <span class="op">==</span> <span class="st">"male"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/tax_sort.html">tax_sort</a></span><span class="op">(</span>by <span class="op">=</span> <span class="va">sum</span>, at <span class="op">=</span> <span class="va">shadeRank</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/tax_sort.html">tax_sort</a></span><span class="op">(</span>by <span class="op">=</span> <span class="va">sum</span>, at <span class="op">=</span> <span class="va">hueRank</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/tax_agg.html">tax_agg</a></span><span class="op">(</span>rank <span class="op">=</span> <span class="va">shadeRank</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Specify number of hues and shades desired</span></span>
<span><span class="va">nHues</span> <span class="op">&lt;-</span> <span class="fl">3</span> <span class="co"># "Other" phyla will be shades of grey</span></span>
<span><span class="va">nShades</span> <span class="op">&lt;-</span> <span class="fl">4</span> <span class="co"># "Other" families will be the lightest shade of each hue</span></span>
<span></span>
<span><span class="va">hierarchicalPalInfo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  hue <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="fu"><a href="../../reference/psExtra-accessors.html">tt_get</a></span><span class="op">(</span><span class="va">pseq2</span><span class="op">)</span><span class="op">[</span>, <span class="va">hueRank</span><span class="op">]</span><span class="op">)</span>,</span>
<span>  shade <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="fu"><a href="../../reference/psExtra-accessors.html">tt_get</a></span><span class="op">(</span><span class="va">pseq2</span><span class="op">)</span><span class="op">[</span>, <span class="va">shadeRank</span><span class="op">]</span><span class="op">)</span>,</span>
<span>  counts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/taxa_sums.html" class="external-link">taxa_sums</a></span><span class="op">(</span><span class="fu"><a href="../../reference/psExtra-accessors.html">otu_get</a></span><span class="op">(</span><span class="va">pseq2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">hierarchicalPalInfo</span> <span class="op">&lt;-</span> <span class="va">hierarchicalPalInfo</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>    hue <span class="op">=</span> <span class="fu">forcats</span><span class="fu">::</span><span class="fu"><a href="https://forcats.tidyverse.org/reference/fct_other.html" class="external-link">fct_other</a></span><span class="op">(</span></span>
<span>      f <span class="op">=</span> <span class="va">hue</span>, keep <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">hue</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="va">nHues</span><span class="op">)</span><span class="op">]</span>,</span>
<span>      other_level <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Other"</span>, <span class="va">hueRankPlural</span><span class="op">)</span></span>
<span>    <span class="op">)</span>,</span>
<span>    nChrHue <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nchar.html" class="external-link">nchar</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">hue</span><span class="op">)</span><span class="op">)</span>, padHue <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">nChrHue</span><span class="op">)</span> <span class="op">-</span> <span class="va">nChrHue</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">hue</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>    shade <span class="op">=</span> <span class="fu">forcats</span><span class="fu">::</span><span class="fu"><a href="https://forcats.tidyverse.org/reference/fct_other.html" class="external-link">fct_other</a></span><span class="op">(</span></span>
<span>      f <span class="op">=</span> <span class="va">shade</span>, keep <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">shade</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="va">nShades</span> <span class="op">-</span> <span class="fl">1</span><span class="op">)</span><span class="op">]</span>,</span>
<span>      other_level <span class="op">=</span> <span class="st">"Other"</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">ungroup</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>    nChrShade <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nchar.html" class="external-link">nchar</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">shade</span><span class="op">)</span><span class="op">)</span>, padShade <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">nChrShade</span><span class="op">)</span> <span class="op">-</span> <span class="va">nChrShade</span>,</span>
<span>    Taxa <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">hue</span>, <span class="st">": "</span>, <span class="fu"><a href="https://rdrr.io/r/base/strrep.html" class="external-link">strrep</a></span><span class="op">(</span><span class="st">" "</span>, <span class="va">padHue</span><span class="op">)</span>, <span class="va">shade</span>, <span class="fu"><a href="https://rdrr.io/r/base/strrep.html" class="external-link">strrep</a></span><span class="op">(</span><span class="st">" "</span>, <span class="va">padShade</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hierarchicalPalMatrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span></span>
<span>    X <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span>from <span class="op">=</span> <span class="fl">30</span>, to <span class="op">=</span> <span class="fl">75</span>, length.out <span class="op">=</span> <span class="va">nShades</span><span class="op">)</span>,</span>
<span>    FUN <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">l</span><span class="op">)</span> <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/pal_hue.html" class="external-link">hue_pal</a></span><span class="op">(</span>l <span class="op">=</span> <span class="va">l</span>, h.start <span class="op">=</span> <span class="fl">30</span><span class="op">)</span><span class="op">(</span>n <span class="op">=</span> <span class="va">nHues</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  byrow <span class="op">=</span> <span class="cn">TRUE</span>, ncol <span class="op">=</span> <span class="va">nHues</span></span>
<span><span class="op">)</span></span>
<span><span class="va">hierarchicalPalMatrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">hierarchicalPalMatrix</span>, <span class="fu"><a href="https://rdrr.io/r/grDevices/gray.colors.html" class="external-link">grey.colors</a></span><span class="op">(</span>n <span class="op">=</span> <span class="va">nShades</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">hierarchicalPal</span> <span class="op">&lt;-</span> <span class="va">hierarchicalPalMatrix</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">hierarchicalPalInfo</span><span class="op">$</span><span class="va">Taxa</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../../reference/tax_palette_plot.html">tax_palette_plot</a></span><span class="op">(</span><span class="va">hierarchicalPal</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.y.left <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>family <span class="op">=</span> <span class="st">"mono"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="compositions_files/figure-html/unnamed-chunk-21-1.png" class="r-plt" alt="" width="384"></p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pseq2</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/psExtra-accessors.html">ps_get</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/tax_mutate.html">tax_mutate</a></span><span class="op">(</span><span class="st">"Phylum: Family"</span> <span class="op">=</span> <span class="va">hierarchicalPalInfo</span><span class="op">$</span><span class="va">Taxa</span>, .keep <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/comp_barplot.html">comp_barplot</a></span><span class="op">(</span></span>
<span>    tax_level <span class="op">=</span> <span class="st">"Phylum: Family"</span>, n_taxa <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">hierarchicalPal</span><span class="op">)</span>,</span>
<span>    tax_order <span class="op">=</span> <span class="st">"asis"</span>, palette <span class="op">=</span> <span class="va">hierarchicalPal</span>, bar_width <span class="op">=</span> <span class="fl">0.975</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>family <span class="op">=</span> <span class="st">"mono"</span><span class="op">)</span><span class="op">)</span> <span class="co"># for text alignment</span></span></code></pre></div>
<p><img src="compositions_files/figure-html/unnamed-chunk-22-1.png" class="r-plt" alt="" width="864"></p>
</div>
</div>
<div class="section level2">
<h2 id="averages-faceting-or-grouping">Averages, faceting or grouping?<a class="anchor" aria-label="anchor" href="#averages-faceting-or-grouping"></a>
</h2>
<div class="section level3">
<h3 id="averaging-compositions">Averaging compositions<a class="anchor" aria-label="anchor" href="#averaging-compositions"></a>
</h3>
<p>Sometimes, to compare microbial compositions across groups, average
compositions are presented. However that “group-averaging” approach
hides a lot of within-group variation, as well as any imbalance in group
sizes.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pseq</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/ps_select.html">ps_select</a></span><span class="op">(</span><span class="va">age</span>, <span class="va">DiseaseState</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> <span class="co"># avoids lots of phyloseq::merge_samples warnings</span></span>
<span>  <span class="fu"><a href="../../reference/ps_filter.html">ps_filter</a></span><span class="op">(</span><span class="va">DiseaseState</span> <span class="op">!=</span> <span class="st">"IBDundef"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/merge_samples-methods.html" class="external-link">merge_samples</a></span><span class="op">(</span>group <span class="op">=</span> <span class="st">"DiseaseState"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/comp_barplot.html">comp_barplot</a></span><span class="op">(</span>tax_level <span class="op">=</span> <span class="st">"Genus"</span>, n_taxa <span class="op">=</span> <span class="fl">12</span>, bar_width <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="cn">NULL</span>, y <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in asMethod(object): NAs introduced by coercion</span></span></code></pre></div>
<p><img src="compositions_files/figure-html/unnamed-chunk-23-1.png" class="r-plt" alt="" width="864"></p>
</div>
<div class="section level3">
<h3 id="faceting">Faceting<a class="anchor" aria-label="anchor" href="#faceting"></a>
</h3>
<p>Faceting is where you show each group on a small subplot.</p>
<p>In the plot below can you see that at minority of UC samples have a
high abundance of Escherichia/Shigella or Streptococcus. The merged bars
above might have misled you into thinking all UC samples had somewhat
increased abundances of these taxa.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pseq</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/ps_filter.html">ps_filter</a></span><span class="op">(</span><span class="va">DiseaseState</span> <span class="op">!=</span> <span class="st">"IBDundef"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> <span class="co"># only one sample in this group</span></span>
<span>  <span class="co"># convert DiseaseState into ordered factor to control order of facets</span></span>
<span>  <span class="fu"><a href="../../reference/ps_mutate.html">ps_mutate</a></span><span class="op">(</span></span>
<span>    DiseaseState <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">DiseaseState</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"UC"</span>, <span class="st">"nonIBD"</span>, <span class="st">"CD"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/comp_barplot.html">comp_barplot</a></span><span class="op">(</span></span>
<span>    tax_level <span class="op">=</span> <span class="st">"Genus"</span>, n_taxa <span class="op">=</span> <span class="fl">15</span>,</span>
<span>    bar_outline_colour <span class="op">=</span> <span class="cn">NA</span>, facet_by <span class="op">=</span> <span class="st">"DiseaseState"</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="compositions_files/figure-html/unnamed-chunk-24-1.png" class="r-plt" alt="" width="864"></p>
<p>Instead of using the <code>facet_by</code> argument in
<code>comp_barplot</code> you can have more control over faceting by
doing it yourself afterwards. You can use facet_grid to create row
facets.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pseq</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/ps_filter.html">ps_filter</a></span><span class="op">(</span><span class="va">DiseaseState</span> <span class="op">!=</span> <span class="st">"IBDundef"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> <span class="co"># only one sample in this group</span></span>
<span>  <span class="co"># convert DiseaseState into ordered factor to control order of facets</span></span>
<span>  <span class="fu"><a href="../../reference/ps_mutate.html">ps_mutate</a></span><span class="op">(</span></span>
<span>    DiseaseState <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">DiseaseState</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"UC"</span>, <span class="st">"CD"</span>, <span class="st">"nonIBD"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/comp_barplot.html">comp_barplot</a></span><span class="op">(</span></span>
<span>    tax_level <span class="op">=</span> <span class="st">"Genus"</span>, n_taxa <span class="op">=</span> <span class="fl">15</span>,</span>
<span>    sample_order <span class="op">=</span> <span class="st">"bray"</span>, bar_outline_colour <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html" class="external-link">facet_grid</a></span><span class="op">(</span></span>
<span>    rows <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/vars.html" class="external-link">vars</a></span><span class="op">(</span><span class="va">DiseaseState</span><span class="op">)</span>,</span>
<span>    scales <span class="op">=</span> <span class="st">"free"</span>, space <span class="op">=</span> <span class="st">"free"</span> <span class="co"># these options are critically important!</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>, axis.ticks.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="compositions_files/figure-html/unnamed-chunk-25-1.png" class="r-plt" alt="" width="576"></p>
</div>
<div class="section level3">
<h3 id="grouping">Grouping<a class="anchor" aria-label="anchor" href="#grouping"></a>
</h3>
<p>For even greater control than faceting, <code>comp_barplot</code>
allows you to generate separate ggplot objects for each group, whilst
maintaining the same taxa colour scheme.</p>
<p>You can assemble these plots into one figure with, for example, the
<code>patchwork</code> package, or keep them separate.</p>
<p><em>Note that the ordering of the samples may differ between facet
and group_by approaches. In the group_by method, the ordering of the
samples by similarity is done separately for each group, whereas in the
facet_by method, similarity-based ordering is done with all samples and
then the samples are separated by facet afterwards.</em></p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plot_list</span> <span class="op">&lt;-</span> <span class="va">pseq</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/ps_filter.html">ps_filter</a></span><span class="op">(</span><span class="va">DiseaseState</span> <span class="op">!=</span> <span class="st">"IBDundef"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/comp_barplot.html">comp_barplot</a></span><span class="op">(</span>n_taxa <span class="op">=</span> <span class="fl">15</span>, tax_level <span class="op">=</span> <span class="st">"Genus"</span>, group_by <span class="op">=</span> <span class="st">"DiseaseState"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot them side by side with the patchwork package.</span></span>
<span><span class="va">patch</span> <span class="op">&lt;-</span> <span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">plot_list</span>, nrow <span class="op">=</span> <span class="fl">1</span>, guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span></span>
<span><span class="va">patch</span> <span class="op">&amp;</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># make all plots horizontal (note: use &amp; instead of +)</span></span></code></pre></div>
<p><img src="compositions_files/figure-html/unnamed-chunk-26-1.png" class="r-plt" alt="" width="864"></p>
<p>Notice how you can theme all plots with the <code>&amp;</code>
operator.</p>
<p>See <a href="https://patchwork.data-imaginist.com/index.html" class="external-link uri">https://patchwork.data-imaginist.com/index.html</a> for more
examples of arranging multiple plots.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">patch</span> <span class="op">&amp;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="cn">NULL</span>, y <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>    axis.text.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>,</span>
<span>    legend.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">8</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">&amp;</span></span>
<span>  <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_annotation.html" class="external-link">plot_annotation</a></span><span class="op">(</span></span>
<span>    title <span class="op">=</span> <span class="st">"Microbial composition across disease groups"</span>,</span>
<span>    caption <span class="op">=</span> <span class="st">"Caption: patchwork is a great package!"</span>,</span>
<span>    theme <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">14</span>, face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="compositions_files/figure-html/unnamed-chunk-27-1.png" class="r-plt" alt="" width="864"></p>
</div>
</div>
<div class="section level2">
<h2 id="sorting-the-barplot">Sorting the barplot<a class="anchor" aria-label="anchor" href="#sorting-the-barplot"></a>
</h2>
<div class="section level3">
<h3 id="sorting-by-similarity">Sorting by similarity<a class="anchor" aria-label="anchor" href="#sorting-by-similarity"></a>
</h3>
<p>Sorting the samples on compositional barplots by similarity can make
patterns in the data much easier to see. Check out this unsorted version
of the first barplot in this article.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pseq</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/ps_filter.html">ps_filter</a></span><span class="op">(</span><span class="va">gender</span> <span class="op">==</span> <span class="st">"female"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/comp_barplot.html">comp_barplot</a></span><span class="op">(</span>tax_level <span class="op">=</span> <span class="st">"Genus"</span>, sample_order <span class="op">=</span> <span class="st">"asis"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Unsorted barcharts are hard to read!"</span><span class="op">)</span></span></code></pre></div>
<p><img src="compositions_files/figure-html/unnamed-chunk-28-1.png" class="r-plt" alt="" width="864"></p>
<p>You can play with the dissimilarity measure (set in
<code>sample_order</code> argument) and <code>seriate_method</code> if
you like, but the defaults (Bray Curtis and OLO Ward) seem to work
pretty well most of the time.</p>
<p>When sorting samples by similarity, the default is to treat the
“other” taxa as one group, i.e. when <code>merge_other = TRUE</code> and
<code>order_with_all_taxa = FALSE</code>.</p>
<p>If you set <code>order_with_all_taxa = TRUE</code>, samples are
sorted BEFORE merging taxa. The resulting sample order is then the same
as when <code>merge_other = FALSE</code>.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pseq</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/ps_filter.html">ps_filter</a></span><span class="op">(</span><span class="va">gender</span> <span class="op">==</span> <span class="st">"female"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/comp_barplot.html">comp_barplot</a></span><span class="op">(</span>tax_level <span class="op">=</span> <span class="st">"Genus"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Samples sorted AFTER merging 'other' taxa"</span><span class="op">)</span></span></code></pre></div>
<p><img src="compositions_files/figure-html/unnamed-chunk-29-1.png" class="r-plt" alt="" width="864"></p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pseq</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/ps_filter.html">ps_filter</a></span><span class="op">(</span><span class="va">gender</span> <span class="op">==</span> <span class="st">"female"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/comp_barplot.html">comp_barplot</a></span><span class="op">(</span>tax_level <span class="op">=</span> <span class="st">"Genus"</span>, order_with_all_taxa <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Samples sorted BEFORE merging 'other' taxa"</span><span class="op">)</span></span></code></pre></div>
<p><img src="compositions_files/figure-html/unnamed-chunk-30-1.png" class="r-plt" alt="" width="864"></p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pseq</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/ps_filter.html">ps_filter</a></span><span class="op">(</span><span class="va">gender</span> <span class="op">==</span> <span class="st">"female"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/comp_barplot.html">comp_barplot</a></span><span class="op">(</span>tax_level <span class="op">=</span> <span class="st">"Genus"</span>, merge_other <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"'other' taxa not merged"</span><span class="op">)</span></span></code></pre></div>
<p><img src="compositions_files/figure-html/unnamed-chunk-31-1.png" class="r-plt" alt="" width="864"></p>
</div>
<div class="section level3">
<h3 id="sort-by-ordination">Sort by ordination<a class="anchor" aria-label="anchor" href="#sort-by-ordination"></a>
</h3>
<p>Coming soon.</p>
</div>
<div class="section level3">
<h3 id="sort-by-1-taxon">Sort by 1 taxon<a class="anchor" aria-label="anchor" href="#sort-by-1-taxon"></a>
</h3>
<p>To study the distribution of a single taxonomic group across your
samples, you can use <code>ps_arrange</code> (with the
<code>.target</code> argument set to “otu_table”) and the ‘default’
sample_order setting in <code>comp_barplot</code>.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pseq</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/tax_agg.html">tax_agg</a></span><span class="op">(</span><span class="st">"Phylum"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/tax_transform.html">tax_transform</a></span><span class="op">(</span><span class="st">"compositional"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/ps_arrange.html">ps_arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">Firmicutes</span><span class="op">)</span>, .target <span class="op">=</span> <span class="st">"otu_table"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/comp_barplot.html">comp_barplot</a></span><span class="op">(</span>tax_level <span class="op">=</span> <span class="st">"Phylum"</span>, sample_order <span class="op">=</span> <span class="st">"asis"</span>, counts_warn <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="compositions_files/figure-html/unnamed-chunk-32-1.png" class="r-plt" alt="" width="864"></p>
</div>
<div class="section level3">
<h3 id="sorting-by-time">Sorting by time<a class="anchor" aria-label="anchor" href="#sorting-by-time"></a>
</h3>
<p>Sometimes you have multiple samples from the same individuals/sites
at several timepoints. You must first order the phyloseq by the time or
ordering variable and then set the comp_barplot sample order to “asis”
(as is) to keep this pre-set order. <em>The x argument is available
since microViz 0.9.7, and allows you to specify a variable other than
the sample name as the x aesthetic for the ggplot. This is useful for
when you have some missing samples per timepoint or group, but want to
fix consistent x axes scales across facets with scales =
“fixed”.</em></p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"shao19"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># prepare subset of data</span></span>
<span><span class="va">ps</span> <span class="op">&lt;-</span> <span class="va">shao19</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/ps_mutate.html">ps_mutate</a></span><span class="op">(</span>family_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">family_id</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="co"># take an arbitrary smaller subset of infants and timepoints</span></span>
<span>  <span class="fu"><a href="../../reference/ps_filter.html">ps_filter</a></span><span class="op">(</span><span class="va">family_role</span> <span class="op">==</span> <span class="st">"child"</span>, <span class="va">family_id</span> <span class="op">&lt;</span> <span class="fl">30</span>, <span class="va">infant_age</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">7</span>, <span class="fl">21</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ps</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/ps_mutate.html">ps_mutate</a></span><span class="op">(</span>infant_age <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">infant_age</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">infant_age</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/comp_barplot.html">comp_barplot</a></span><span class="op">(</span></span>
<span>    tax_level <span class="op">=</span> <span class="st">"genus"</span>, sample_order <span class="op">=</span> <span class="st">"asis"</span>, bar_width <span class="op">=</span> <span class="fl">0.7</span>,</span>
<span>    merge_other <span class="op">=</span> <span class="cn">FALSE</span>, x <span class="op">=</span> <span class="st">"infant_age"</span> <span class="co"># x argument available since microViz 0.9.7</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span></span>
<span>    facets <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/vars.html" class="external-link">vars</a></span><span class="op">(</span><span class="va">family_id</span><span class="op">)</span>, labeller <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/as_labeller.html" class="external-link">as_labeller</a></span><span class="op">(</span><span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Fam."</span>, <span class="va">.</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="co"># slightly clearer axes for facets</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Age (days)"</span>, y <span class="op">=</span> <span class="st">"Relative Abundance"</span>, fill <span class="op">=</span> <span class="st">"Genus"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span></span>
<span>    expand <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/expansion.html" class="external-link">expansion</a></span><span class="op">(</span>add <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.1</span><span class="op">)</span><span class="op">)</span>, <span class="co"># axis starts exactly at 0</span></span>
<span>    labels <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/label_percent.html" class="external-link">label_percent</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="compositions_files/figure-html/unnamed-chunk-34-1.png" class="r-plt" alt="" width="672"></p>
<p>You can try to roughly sort subjects by their microbiome composition
by seriating the samples up front before plotting, and then fixing the
<strong>subject</strong> order by mutating subject into a factor,
preserving its unique levels in their sample-sorted order.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ps</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/ps_seriate.html">ps_seriate</a></span><span class="op">(</span>rank <span class="op">=</span> <span class="st">"genus"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/ps_mutate.html">ps_mutate</a></span><span class="op">(</span></span>
<span>    family_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">family_id</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">family_id</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    infant_age <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">infant_age</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">infant_age</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/comp_barplot.html">comp_barplot</a></span><span class="op">(</span></span>
<span>    tax_level <span class="op">=</span> <span class="st">"genus"</span>, sample_order <span class="op">=</span> <span class="st">"asis"</span>, bar_width <span class="op">=</span> <span class="fl">0.7</span>,</span>
<span>    x <span class="op">=</span> <span class="st">"infant_age"</span> <span class="co"># x argument is available since microViz 0.9.7</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span></span>
<span>    facets <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/vars.html" class="external-link">vars</a></span><span class="op">(</span><span class="va">family_id</span><span class="op">)</span>, labeller <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/as_labeller.html" class="external-link">as_labeller</a></span><span class="op">(</span><span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Fam."</span>, <span class="va">.</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="co"># slightly clearer axes for facets</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Age (days)"</span>, y <span class="op">=</span> <span class="st">"Relative Abundance"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span></span>
<span>    expand <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/expansion.html" class="external-link">expansion</a></span><span class="op">(</span>add <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.1</span><span class="op">)</span><span class="op">)</span>, <span class="co"># axis starts exactly at 0</span></span>
<span>    labels <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/label_percent.html" class="external-link">label_percent</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="compositions_files/figure-html/unnamed-chunk-35-1.png" class="r-plt" alt="" width="672"></p>
<p>An alternative way to present grouped repeated samples.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ps</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/ps_arrange.html">ps_arrange</a></span><span class="op">(</span><span class="va">family_id</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/ps_mutate.html">ps_mutate</a></span><span class="op">(</span></span>
<span>    family_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Fam"</span>, <span class="va">family_id</span><span class="op">)</span>, <span class="co"># better labels</span></span>
<span>    family_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">family_id</span>, <span class="fu"><a href="https://rdrr.io/r/base/rev.html" class="external-link">rev</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">family_id</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="co"># fix plot order</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/comp_barplot.html">comp_barplot</a></span><span class="op">(</span></span>
<span>    tax_level <span class="op">=</span> <span class="st">"genus"</span>, bar_width <span class="op">=</span> <span class="fl">0.7</span>, sample_order <span class="op">=</span> <span class="st">"asis"</span>, <span class="co"># don't sort</span></span>
<span>    x <span class="op">=</span> <span class="st">"family_id"</span> <span class="co"># x argument is available since microViz 0.9.7</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span></span>
<span>    facets <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/vars.html" class="external-link">vars</a></span><span class="op">(</span><span class="va">infant_age</span><span class="op">)</span>, labeller <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/as_labeller.html" class="external-link">as_labeller</a></span><span class="op">(</span><span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Age"</span>, <span class="va">.</span>, <span class="st">"days"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    scales <span class="op">=</span> <span class="st">"fixed"</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Family ID"</span>, y <span class="op">=</span> <span class="st">"Relative abundance"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/expansion.html" class="external-link">expansion</a></span><span class="op">(</span>add <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="co"># axis starts exactly at 0</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="co"># slightly clearer axes for facets</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>panel.spacing.x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">6</span>, <span class="st">"mm"</span><span class="op">)</span><span class="op">)</span> <span class="co"># space for non-overlapping axis numbers</span></span></code></pre></div>
<p><img src="compositions_files/figure-html/unnamed-chunk-36-1.png" class="r-plt" alt="" width="768"></p>
</div>
<div class="section level3">
<h3 id="faceting-by-timepoint-and-another-variable">Faceting by timepoint and another variable<a class="anchor" aria-label="anchor" href="#faceting-by-timepoint-and-another-variable"></a>
</h3>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># prepare arbitrary subset of dataset</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"dietswap"</span>, package <span class="op">=</span> <span class="st">"microbiome"</span><span class="op">)</span></span>
<span><span class="va">psD</span> <span class="op">&lt;-</span> <span class="va">dietswap</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../../reference/ps_filter.html">ps_filter</a></span><span class="op">(</span><span class="va">group</span> <span class="op">==</span> <span class="st">"DI"</span><span class="op">)</span></span></code></pre></div>
<p>Grid faceting to separate samples by timepoint and another grouping
variable, whilst keeping subjects’ samples paired.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">psD</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/comp_barplot.html">comp_barplot</a></span><span class="op">(</span></span>
<span>    tax_level <span class="op">=</span> <span class="st">"Genus"</span>, n_taxa <span class="op">=</span> <span class="fl">10</span>, sample_order <span class="op">=</span> <span class="st">"asis"</span>,</span>
<span>    bar_width <span class="op">=</span> <span class="fl">0.7</span>, x <span class="op">=</span> <span class="st">"subject"</span>, <span class="co"># x arg available since microViz 0.9.7</span></span>
<span>    merge_other <span class="op">=</span> <span class="cn">FALSE</span>, bar_outline_colour <span class="op">=</span> <span class="st">"grey25"</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html" class="external-link">facet_grid</a></span><span class="op">(</span></span>
<span>    cols <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/vars.html" class="external-link">vars</a></span><span class="op">(</span><span class="va">timepoint.within.group</span><span class="op">)</span>, rows <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/vars.html" class="external-link">vars</a></span><span class="op">(</span><span class="va">nationality</span><span class="op">)</span>,</span>
<span>    scales <span class="op">=</span> <span class="st">"free_y"</span>, <span class="co"># this only frees y scale per row in grid faceting</span></span>
<span>    space <span class="op">=</span> <span class="st">"free_y"</span> <span class="co"># allows bars to be same size by freeing facet heights</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_discrete.html" class="external-link">scale_x_discrete</a></span><span class="op">(</span>limits <span class="op">=</span> <span class="va">rev</span><span class="op">)</span> <span class="op">+</span> <span class="co"># reverses order: top-down alphabetical</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="compositions_files/figure-html/unnamed-chunk-37-1.png" class="r-plt" alt="" width="864"></p>
<p>You can try to roughly sort subjects by their microbiome composition
by seriating the samples up front before plotting, and then fixing the
<strong>subject</strong> order by mutating subject into a factor,
preserving its unique levels in their sample-sorted order.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">psD</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/ps_seriate.html">ps_seriate</a></span><span class="op">(</span>rank <span class="op">=</span> <span class="st">"Genus"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/ps_mutate.html">ps_mutate</a></span><span class="op">(</span>subject <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">subject</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">subject</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/comp_barplot.html">comp_barplot</a></span><span class="op">(</span></span>
<span>    tax_level <span class="op">=</span> <span class="st">"Genus"</span>, n_taxa <span class="op">=</span> <span class="fl">10</span>, sample_order <span class="op">=</span> <span class="st">"asis"</span>,</span>
<span>    bar_width <span class="op">=</span> <span class="fl">0.7</span>, x <span class="op">=</span> <span class="st">"subject"</span>, <span class="co"># x arg available since microViz 0.9.7</span></span>
<span>    merge_other <span class="op">=</span> <span class="cn">FALSE</span>, bar_outline_colour <span class="op">=</span> <span class="st">"grey25"</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html" class="external-link">facet_grid</a></span><span class="op">(</span></span>
<span>    cols <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/vars.html" class="external-link">vars</a></span><span class="op">(</span><span class="va">timepoint.within.group</span><span class="op">)</span>, rows <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/vars.html" class="external-link">vars</a></span><span class="op">(</span><span class="va">nationality</span><span class="op">)</span>,</span>
<span>    labeller <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labeller.html" class="external-link">labeller</a></span><span class="op">(</span>.cols <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/as_labeller.html" class="external-link">as_labeller</a></span><span class="op">(</span><span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Timepoint"</span>, <span class="va">.</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    scales <span class="op">=</span> <span class="st">"free_y"</span>, <span class="co"># this only frees y scale per row in grid faceting</span></span>
<span>    space <span class="op">=</span> <span class="st">"free_y"</span> <span class="co"># allows bars to be same size by freeing facet heights</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>, panel.spacing.x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">5</span>, <span class="st">"mm"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="compositions_files/figure-html/unnamed-chunk-38-1.png" class="r-plt" alt="" width="864"></p>
<p>Faceted by timepoint and <strong>grouped</strong> by nationality.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">times_list</span> <span class="op">&lt;-</span> <span class="va">psD</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/ps_seriate.html">ps_seriate</a></span><span class="op">(</span>rank <span class="op">=</span> <span class="st">"Genus"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> <span class="co"># can help for approximate subject sorting</span></span>
<span>  <span class="fu"><a href="../../reference/ps_mutate.html">ps_mutate</a></span><span class="op">(</span>subject <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">subject</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">subject</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/comp_barplot.html">comp_barplot</a></span><span class="op">(</span></span>
<span>    tax_level <span class="op">=</span> <span class="st">"Genus"</span>, n_taxa <span class="op">=</span> <span class="fl">11</span>,</span>
<span>    sample_order <span class="op">=</span> <span class="st">"asis"</span>, <span class="co"># don't bother sorting here, will be ignored</span></span>
<span>    group_by <span class="op">=</span> <span class="st">"nationality"</span>, bar_width <span class="op">=</span> <span class="fl">0.7</span>, x <span class="op">=</span> <span class="st">"subject"</span>,</span>
<span>    merge_other <span class="op">=</span> <span class="cn">FALSE</span>, bar_outline_colour <span class="op">=</span> <span class="st">"grey25"</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="va">times_list</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span>guides <span class="op">=</span> <span class="st">"collect"</span>, heights <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">7</span>, <span class="fl">6</span><span class="op">)</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">timepoint.within.group</span>, nrow <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Relative abundance"</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>, panel.spacing.x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">5</span>, <span class="st">"mm"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="compositions_files/figure-html/unnamed-chunk-39-1.png" class="r-plt" alt="" width="768"></p>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://sessioninfo.r-lib.org/reference/session_info.html" class="external-link">session_info</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB; font-weight: bold;">─ Session info ───────────────────────────────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt;  <span style="color: #555555; font-style: italic;">setting </span> <span style="color: #555555; font-style: italic;">value</span></span></span>
<span><span class="co">#&gt;  version  R version 4.5.2 (2025-10-31)</span></span>
<span><span class="co">#&gt;  os       Ubuntu 24.04.3 LTS</span></span>
<span><span class="co">#&gt;  system   x86_64, linux-gnu</span></span>
<span><span class="co">#&gt;  ui       X11</span></span>
<span><span class="co">#&gt;  language en</span></span>
<span><span class="co">#&gt;  collate  C.UTF-8</span></span>
<span><span class="co">#&gt;  ctype    C.UTF-8</span></span>
<span><span class="co">#&gt;  tz       UTC</span></span>
<span><span class="co">#&gt;  date     2026-02-17</span></span>
<span><span class="co">#&gt;  pandoc   3.1.11 @ /opt/hostedtoolcache/pandoc/3.1.11/x64/ (via rmarkdown)</span></span>
<span><span class="co">#&gt;  quarto   NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB; font-weight: bold;">─ Packages ───────────────────────────────────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt;  <span style="color: #555555; font-style: italic;">package     </span> <span style="color: #555555; font-style: italic;">*</span> <span style="color: #555555; font-style: italic;">version </span> <span style="color: #555555; font-style: italic;">date (UTC)</span> <span style="color: #555555; font-style: italic;">lib</span> <span style="color: #555555; font-style: italic;">source</span></span></span>
<span><span class="co">#&gt;  ade4           1.7-23   <span style="color: #555555;">2025-02-14</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ape            5.8-1    <span style="color: #555555;">2024-12-16</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  Biobase        2.70.0   <span style="color: #555555;">2025-10-29</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  BiocGenerics   0.56.0   <span style="color: #555555;">2025-10-29</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  biomformat     1.38.0   <span style="color: #555555;">2025-10-29</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  Biostrings     2.78.0   <span style="color: #555555;">2025-10-29</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  bslib          0.10.0   <span style="color: #555555;">2026-01-26</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ca             0.71.1   <span style="color: #555555;">2020-01-24</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  cachem         1.1.0    <span style="color: #555555;">2024-05-16</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  cli            3.6.5    <span style="color: #555555;">2025-04-23</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  cluster        2.1.8.1  <span style="color: #555555;">2025-03-12</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.5.2)</span></span></span>
<span><span class="co">#&gt;  codetools      0.2-20   <span style="color: #555555;">2024-03-31</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.5.2)</span></span></span>
<span><span class="co">#&gt;  crayon         1.5.3    <span style="color: #555555;">2024-06-20</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  data.table     1.18.2.1 <span style="color: #555555;">2026-01-27</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  desc           1.4.3    <span style="color: #555555;">2023-12-10</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  devtools       2.4.6    <span style="color: #555555;">2025-10-03</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  digest         0.6.39   <span style="color: #555555;">2025-11-19</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  dplyr          1.2.0    <span style="color: #555555;">2026-02-03</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ellipsis       0.3.2    <span style="color: #555555;">2021-04-29</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  evaluate       1.0.5    <span style="color: #555555;">2025-08-27</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  farver         2.1.2    <span style="color: #555555;">2024-05-13</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  fastmap        1.2.0    <span style="color: #555555;">2024-05-15</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  forcats        1.0.1    <span style="color: #555555;">2025-09-25</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  foreach        1.5.2    <span style="color: #555555;">2022-02-02</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  fs             1.6.6    <span style="color: #555555;">2025-04-12</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  generics       0.1.4    <span style="color: #555555;">2025-05-09</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ggplot2      * 4.0.2    <span style="color: #555555;">2026-02-03</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  glue           1.8.0    <span style="color: #555555;">2024-09-30</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  gtable         0.3.6    <span style="color: #555555;">2024-10-25</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  htmltools      0.5.9    <span style="color: #555555;">2025-12-04</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  htmlwidgets    1.6.4    <span style="color: #555555;">2023-12-06</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  igraph         2.2.2    <span style="color: #555555;">2026-02-12</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  IRanges        2.44.0   <span style="color: #555555;">2025-10-29</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  iterators      1.0.14   <span style="color: #555555;">2022-02-05</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  jquerylib      0.1.4    <span style="color: #555555;">2021-04-26</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  jsonlite       2.0.0    <span style="color: #555555;">2025-03-27</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  knitr          1.51     <span style="color: #555555;">2025-12-20</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  labeling       0.4.3    <span style="color: #555555;">2023-08-29</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  lattice        0.22-7   <span style="color: #555555;">2025-04-02</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.5.2)</span></span></span>
<span><span class="co">#&gt;  lifecycle      1.0.5    <span style="color: #555555;">2026-01-08</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  magrittr       2.0.4    <span style="color: #555555;">2025-09-12</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  MASS           7.3-65   <span style="color: #555555;">2025-02-28</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.5.2)</span></span></span>
<span><span class="co">#&gt;  Matrix         1.7-4    <span style="color: #555555;">2025-08-28</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.5.2)</span></span></span>
<span><span class="co">#&gt;  memoise        2.0.1    <span style="color: #555555;">2021-11-26</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  mgcv           1.9-3    <span style="color: #555555;">2025-04-04</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.5.2)</span></span></span>
<span><span class="co">#&gt;  microbiome     1.32.0   <span style="color: #555555;">2025-10-29</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  microViz     * 0.13.0   <span style="color: #555555;">2026-02-17</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">local</span></span></span>
<span><span class="co">#&gt;  multtest       2.66.0   <span style="color: #555555;">2025-10-29</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  nlme           3.1-168  <span style="color: #555555;">2025-03-31</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.5.2)</span></span></span>
<span><span class="co">#&gt;  otel           0.2.0    <span style="color: #555555;">2025-08-29</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  patchwork    * 1.3.2    <span style="color: #555555;">2025-08-25</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  permute        0.9-10   <span style="color: #555555;">2026-02-06</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  phyloseq     * 1.54.0   <span style="color: #555555;">2025-10-29</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  pillar         1.11.1   <span style="color: #555555;">2025-09-17</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  pkgbuild       1.4.8    <span style="color: #555555;">2025-05-26</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  pkgconfig      2.0.3    <span style="color: #555555;">2019-09-22</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  pkgdown        2.2.0    <span style="color: #555555;">2025-11-06</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  pkgload        1.5.0    <span style="color: #555555;">2026-02-03</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  plyr           1.8.9    <span style="color: #555555;">2023-10-02</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  purrr          1.2.1    <span style="color: #555555;">2026-01-09</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  R6             2.6.1    <span style="color: #555555;">2025-02-15</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ragg           1.5.0    <span style="color: #555555;">2025-09-02</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  RColorBrewer   1.1-3    <span style="color: #555555;">2022-04-03</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  Rcpp           1.1.1    <span style="color: #555555;">2026-01-10</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  registry       0.5-1    <span style="color: #555555;">2019-03-05</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  remotes        2.5.0    <span style="color: #555555;">2024-03-17</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  reshape2       1.4.5    <span style="color: #555555;">2025-11-12</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  rhdf5          2.54.1   <span style="color: #555555;">2025-12-04</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  rhdf5filters   1.22.0   <span style="color: #555555;">2025-10-29</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  Rhdf5lib       1.32.0   <span style="color: #555555;">2025-10-29</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  rlang          1.1.7    <span style="color: #555555;">2026-01-09</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  rmarkdown      2.30     <span style="color: #555555;">2025-09-28</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  Rtsne          0.17     <span style="color: #555555;">2023-12-07</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  S4Vectors      0.48.0   <span style="color: #555555;">2025-10-29</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  S7             0.2.1    <span style="color: #555555;">2025-11-14</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  sass           0.4.10   <span style="color: #555555;">2025-04-11</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  scales         1.4.0    <span style="color: #555555;">2025-04-24</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  Seqinfo        1.0.0    <span style="color: #555555;">2025-10-29</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  seriation      1.5.8    <span style="color: #555555;">2025-08-20</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  sessioninfo    1.2.3    <span style="color: #555555;">2025-02-05</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  stringi        1.8.7    <span style="color: #555555;">2025-03-27</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  stringr        1.6.0    <span style="color: #555555;">2025-11-04</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  survival       3.8-3    <span style="color: #555555;">2024-12-17</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.5.2)</span></span></span>
<span><span class="co">#&gt;  systemfonts    1.3.1    <span style="color: #555555;">2025-10-01</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  textshaping    1.0.4    <span style="color: #555555;">2025-10-10</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  tibble         3.3.1    <span style="color: #555555;">2026-01-11</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  tidyr          1.3.2    <span style="color: #555555;">2025-12-19</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  tidyselect     1.2.1    <span style="color: #555555;">2024-03-11</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  TSP            1.2.6    <span style="color: #555555;">2025-11-27</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  usethis        3.2.1    <span style="color: #555555;">2025-09-06</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  vctrs          0.7.1    <span style="color: #555555;">2026-01-23</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  vegan          2.7-2    <span style="color: #555555;">2025-10-08</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  withr          3.0.2    <span style="color: #555555;">2024-10-28</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  xfun           0.56     <span style="color: #555555;">2026-01-18</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  XVector        0.50.0   <span style="color: #555555;">2025-10-29</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  yaml           2.3.12   <span style="color: #555555;">2025-12-10</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #555555;"> [1] /home/runner/work/_temp/Library</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;"> [2] /opt/R/4.5.2/lib/R/site-library</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;"> [3] /opt/R/4.5.2/lib/R/library</span></span></span>
<span><span class="co">#&gt;  <span style="color: #BBBBBB; background-color: #BB0000;">*</span> ── Packages attached to the search path.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB; font-weight: bold;">──────────────────────────────────────────────────────────────────────────────────────────────────</span></span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by David Barnett.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>



  <script data-goatcounter="https://microviz.goatcounter.com/count" async src="//gc.zgo.at/count.js"></script>
</body>
</html>
