<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>microViz annotated heatmaps • microViz</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../../bootstrap-toc.css">
<script src="../../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><!-- docsearch --><script src="../../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><meta property="og:title" content="microViz annotated heatmaps">
<meta property="og:description" content="microViz">
<meta property="og:image" content="https://david-barnett.github.io/microViz/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-2YQ669S4V5"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-2YQ669S4V5');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">microViz</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.10.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/web-only/phyloseq.html">Working with phyloseq objects</a>
    </li>
    <li>
      <a href="../../articles/web-only/tax-fixing.html">Fixing your tax_table</a>
    </li>
    <li>
      <a href="../../articles/web-only/ordination.html">Ordination plots</a>
    </li>
    <li>
      <a href="../../articles/web-only/ordination-interactive.html">Interactive ordination plots</a>
    </li>
    <li>
      <a href="../../articles/web-only/compositions.html">Visualising compositions with barplots</a>
    </li>
    <li>
      <a href="../../articles/web-only/heatmaps.html">Heatmaps of microbiome composition or correlation</a>
    </li>
    <li>
      <a href="../../articles/web-only/modelling-taxa.html">Statistical modelling of individual taxa</a>
    </li>
    <li>
      <a href="../../articles/web-only/videos.html">Tutorial video</a>
    </li>
    <li>
      <a href="../../articles/ps_extra-replaced.html">Upgrading to microViz version 0.10.0</a>
    </li>
    <li>
      <a href="../../articles/web-only/atlas1006.html">Example analyses with atlas1006 data</a>
    </li>
  </ul>
</li>
<li>
  <a href="../../articles/learn/exercises.html">Learn</a>
</li>
<li>
  <a href="../../news/index.html">News</a>
</li>
<li class="dropdown-header">
      </li>
</ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/david-barnett/microViz/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>microViz annotated heatmaps</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/david-barnett/microViz/blob/HEAD/vignettes/web-only/heatmaps.Rmd" class="external-link"><code>vignettes/web-only/heatmaps.Rmd</code></a></small>
      <div class="hidden name"><code>heatmaps.Rmd</code></div>

    </div>

    
    
<p>This article will show you how to plot annotated correlation and
microbial composition heatmaps with microViz.</p>
<div class="section level2">
<h2 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://dx.plos.org/10.1371/journal.pone.0061217" class="external-link">phyloseq</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://david-barnett.github.io/microViz">microViz</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">microViz version 0.10.3 - Copyright (C) 2022 David Barnett</span></span></span>
<span><span class="co">#&gt; <span style="color: #BBBB00;">!</span> Website: https://david-barnett.github.io/microViz</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> Useful?  For citation details, run: `citation("microViz")`</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">✖</span> Silence? `suppressPackageStartupMessages(library(microViz))`</span></span></code></pre></div>
<p>First we’ll get some OTU abundance data from inflammatory bowel
disease patients and controls from the corncob package.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"ibd_phylo"</span>, package <span class="op">=</span> <span class="st">"corncob"</span><span class="op">)</span></span>
<span><span class="va">ibd_phylo</span></span>
<span><span class="co">#&gt; phyloseq-class experiment-level object</span></span>
<span><span class="co">#&gt; otu_table()   OTU Table:         [ 36349 taxa and 91 samples ]</span></span>
<span><span class="co">#&gt; sample_data() Sample Data:       [ 91 samples by 15 sample variables ]</span></span>
<span><span class="co">#&gt; tax_table()   Taxonomy Table:    [ 36349 taxa by 7 taxonomic ranks ]</span></span></code></pre></div>
<p>Remove the mostly unclassified species-level data, drop the rare taxa
and fix the taxonomy of the rest. Also drop patients with unclassified
IBD.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">psq</span> <span class="op">&lt;-</span> <span class="va">ibd_phylo</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/tax_mutate.html">tax_mutate</a></span><span class="op">(</span>Species <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/tax_filter.html">tax_filter</a></span><span class="op">(</span>min_prevalence <span class="op">=</span> <span class="fl">5</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/tax_fix.html">tax_fix</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/ps_filter.html">ps_filter</a></span><span class="op">(</span><span class="va">DiseaseState</span> <span class="op">!=</span> <span class="st">"IBDundef"</span><span class="op">)</span></span>
<span><span class="va">psq</span></span>
<span><span class="co">#&gt; phyloseq-class experiment-level object</span></span>
<span><span class="co">#&gt; otu_table()   OTU Table:         [ 1599 taxa and 90 samples ]</span></span>
<span><span class="co">#&gt; sample_data() Sample Data:       [ 90 samples by 15 sample variables ]</span></span>
<span><span class="co">#&gt; tax_table()   Taxonomy Table:    [ 1599 taxa by 6 taxonomic ranks ]</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="microbiome-heatmaps">Microbiome heatmaps<a class="anchor" aria-label="anchor" href="#microbiome-heatmaps"></a>
</h2>
<p>Visualise the microbial composition of your samples.</p>
<p>The samples and taxa are sorted by similarity. (<em>By default this
uses hierarchical clustering with optimal leaf ordering, using euclidean
distances on the transformed data</em>).</p>
<p>In this example we use a “compositional” transformation, so the Class
abundances are shown as proportions of each sample.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">psq</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/tax_transform.html">tax_transform</a></span><span class="op">(</span><span class="st">"compositional"</span>, rank <span class="op">=</span> <span class="st">"Class"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/comp_heatmap.html">comp_heatmap</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Registered S3 method overwritten by 'seriation':</span></span>
<span><span class="co">#&gt;   method         from </span></span>
<span><span class="co">#&gt;   reorder.hclust vegan</span></span></code></pre></div>
<p><img src="heatmaps_files/figure-html/unnamed-chunk-4-1.png" width="840"></p>
<p>You can easily swap to a symmetrical colour palette for
transformations like “clr” or “standardize”. This is the default
symmetrical palette but you can pick from many.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">psq</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/tax_transform.html">tax_transform</a></span><span class="op">(</span><span class="st">"clr"</span>, rank <span class="op">=</span> <span class="st">"Family"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/comp_heatmap.html">comp_heatmap</a></span><span class="op">(</span>colors <span class="op">=</span> <span class="fu"><a href="../../reference/heat_palette.html">heat_palette</a></span><span class="op">(</span>sym <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, name <span class="op">=</span> <span class="st">"CLR"</span><span class="op">)</span></span></code></pre></div>
<p><img src="heatmaps_files/figure-html/unnamed-chunk-5-1.png" width="840"></p>
<div class="section level3">
<h3 id="annotating-taxa">Annotating taxa<a class="anchor" aria-label="anchor" href="#annotating-taxa"></a>
</h3>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">psq</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/tax_transform.html">tax_transform</a></span><span class="op">(</span><span class="st">"compositional"</span>, rank <span class="op">=</span> <span class="st">"Class"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/comp_heatmap.html">comp_heatmap</a></span><span class="op">(</span>tax_anno <span class="op">=</span> <span class="fu"><a href="../../reference/taxAnnotation.html">taxAnnotation</a></span><span class="op">(</span></span>
<span>    Prev. <span class="op">=</span> <span class="fu"><a href="../../reference/anno_tax_prev.html">anno_tax_prev</a></span><span class="op">(</span>bar_width <span class="op">=</span> <span class="fl">0.3</span>, size <span class="op">=</span> <span class="fu">grid</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">1</span>, <span class="st">"cm"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="heatmaps_files/figure-html/unnamed-chunk-6-1.png" width="840"></p>
</div>
<div class="section level3">
<h3 id="legend-positioning">Legend positioning<a class="anchor" aria-label="anchor" href="#legend-positioning"></a>
</h3>
<p>Positioning the heatmap legend at the bottom is possible. You can
assign the heatmap to a name and then call <code>ComplexHeatmap</code>’s
<code>draw</code> function.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">heat</span> <span class="op">&lt;-</span> <span class="va">psq</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/tax_transform.html">tax_transform</a></span><span class="op">(</span><span class="st">"compositional"</span>, rank <span class="op">=</span> <span class="st">"Class"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/comp_heatmap.html">comp_heatmap</a></span><span class="op">(</span></span>
<span>    tax_anno <span class="op">=</span> <span class="fu"><a href="../../reference/taxAnnotation.html">taxAnnotation</a></span><span class="op">(</span></span>
<span>      Prev. <span class="op">=</span> <span class="fu"><a href="../../reference/anno_tax_prev.html">anno_tax_prev</a></span><span class="op">(</span>bar_width <span class="op">=</span> <span class="fl">0.3</span>, size <span class="op">=</span> <span class="fu">grid</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">1</span>, <span class="st">"cm"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    heatmap_legend_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>      at <span class="op">=</span> <span class="fl">0</span><span class="op">:</span><span class="fl">5</span> <span class="op">/</span> <span class="fl">5</span>,</span>
<span>      direction <span class="op">=</span> <span class="st">"horizontal"</span>, title_position <span class="op">=</span> <span class="st">"leftcenter"</span>, </span>
<span>      legend_width <span class="op">=</span> <span class="fu">grid</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">4</span>, <span class="st">"cm"</span><span class="op">)</span>, grid_height <span class="op">=</span> <span class="fu">grid</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">5</span>, <span class="st">"mm"</span><span class="op">)</span></span>
<span>      <span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/draw-dispatch.html" class="external-link">draw</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">heat</span>, heatmap_legend_side <span class="op">=</span> <span class="st">"bottom"</span>, </span>
<span>  adjust_annotation_extension <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="heatmaps_files/figure-html/unnamed-chunk-7-1.png" width="840"></p>
</div>
<div class="section level3">
<h3 id="annotating-samples">Annotating samples<a class="anchor" aria-label="anchor" href="#annotating-samples"></a>
</h3>
<div class="section level4">
<h4 id="group-membership">Group membership<a class="anchor" aria-label="anchor" href="#group-membership"></a>
</h4>
<p>2 different methods for annotating each sample’s values of
categorical metadata are possible.</p>
<ul>
<li><p><code><a href="../../reference/anno_sample.html">anno_sample()</a></code> cannot have borders around each cell,
but automatically adds a legend.</p></li>
<li><p><code><a href="../../reference/anno_sample_cat.html">anno_sample_cat()</a></code> can have cell borders, but
requires an extra step to draw a legend</p></li>
</ul>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/distinct_palette.html">distinct_palette</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">3</span>, add <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cols</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="../../reference/psExtra-accessors.html">samdat_tbl</a></span><span class="op">(</span><span class="va">psq</span><span class="op">)</span><span class="op">$</span><span class="va">DiseaseState</span><span class="op">)</span></span>
<span></span>
<span><span class="va">psq</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/tax_transform.html">tax_transform</a></span><span class="op">(</span><span class="st">"compositional"</span>, rank <span class="op">=</span> <span class="st">"Class"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/comp_heatmap.html">comp_heatmap</a></span><span class="op">(</span></span>
<span>    tax_anno <span class="op">=</span> <span class="fu"><a href="../../reference/taxAnnotation.html">taxAnnotation</a></span><span class="op">(</span></span>
<span>      Prev. <span class="op">=</span> <span class="fu"><a href="../../reference/anno_tax_prev.html">anno_tax_prev</a></span><span class="op">(</span>bar_width <span class="op">=</span> <span class="fl">0.3</span>, size <span class="op">=</span> <span class="fu">grid</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">1</span>, <span class="st">"cm"</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">)</span>,</span>
<span>    sample_anno <span class="op">=</span> <span class="fu"><a href="../../reference/sampleAnnotation.html">sampleAnnotation</a></span><span class="op">(</span></span>
<span>      State1 <span class="op">=</span> <span class="fu"><a href="../../reference/anno_sample.html">anno_sample</a></span><span class="op">(</span><span class="st">"DiseaseState"</span><span class="op">)</span>,</span>
<span>      col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>State1 <span class="op">=</span> <span class="va">cols</span><span class="op">)</span>, border <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>      State2 <span class="op">=</span> <span class="fu"><a href="../../reference/anno_sample_cat.html">anno_sample_cat</a></span><span class="op">(</span><span class="st">"DiseaseState"</span>, col <span class="op">=</span> <span class="va">cols</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span> </span></code></pre></div>
<p><img src="heatmaps_files/figure-html/unnamed-chunk-8-1.png" width="840"></p>
<p>Let’s try drawing equivalent categorical annotations by two methods.
Both methods can draw annotations with borders and no individual boxes.
This style suits heatmaps with no gridlines
(i.e. <code>grid_col = NA</code>).</p>
<p>In the example below we have suppressed row ordering with
<code>cluster_rows = FALSE</code>, and added spaces between taxa by
splitting at every row with <code>row_split = 1:11</code>, which are
both <code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">ComplexHeatmap::Heatmap()</a></code> arguments.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">psqC</span> <span class="op">&lt;-</span> <span class="va">psq</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../../reference/tax_transform.html">tax_transform</a></span><span class="op">(</span><span class="st">"compositional"</span>, rank <span class="op">=</span> <span class="st">"Class"</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">htmp</span> <span class="op">&lt;-</span> <span class="va">psqC</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/comp_heatmap.html">comp_heatmap</a></span><span class="op">(</span></span>
<span>    grid_col <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>    cluster_rows <span class="op">=</span> <span class="cn">FALSE</span>, row_title <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>    row_split <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/ntaxa-methods.html" class="external-link">ntaxa</a></span><span class="op">(</span><span class="fu"><a href="../../reference/psExtra-accessors.html">ps_get</a></span><span class="op">(</span><span class="va">psqC</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    tax_anno <span class="op">=</span> <span class="fu"><a href="../../reference/taxAnnotation.html">taxAnnotation</a></span><span class="op">(</span></span>
<span>      Prev. <span class="op">=</span> <span class="fu"><a href="../../reference/anno_tax_prev.html">anno_tax_prev</a></span><span class="op">(</span>bar_width <span class="op">=</span> <span class="fl">0.9</span>, size <span class="op">=</span> <span class="fu">grid</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">1</span>, <span class="st">"cm"</span><span class="op">)</span>, border <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span>    <span class="op">)</span>,</span>
<span>    sample_anno <span class="op">=</span> <span class="fu"><a href="../../reference/sampleAnnotation.html">sampleAnnotation</a></span><span class="op">(</span></span>
<span>      <span class="co"># method one</span></span>
<span>      State1 <span class="op">=</span> <span class="fu"><a href="../../reference/anno_sample.html">anno_sample</a></span><span class="op">(</span><span class="st">"DiseaseState"</span><span class="op">)</span>, </span>
<span>      col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>State1 <span class="op">=</span> <span class="va">cols</span><span class="op">)</span>, border <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>      <span class="co"># method two</span></span>
<span>      State2 <span class="op">=</span> <span class="fu"><a href="../../reference/anno_sample_cat.html">anno_sample_cat</a></span><span class="op">(</span></span>
<span>        var <span class="op">=</span> <span class="st">"DiseaseState"</span>, col <span class="op">=</span> <span class="va">cols</span>, box_col <span class="op">=</span> <span class="cn">NA</span>, border_col <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>        legend_title <span class="op">=</span> <span class="st">"State2"</span></span>
<span>      <span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span> </span>
<span><span class="va">htmp</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> <span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/draw-dispatch.html" class="external-link">draw</a></span><span class="op">(</span></span>
<span>  annotation_legend_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">htmp</span>, <span class="st">"AnnoLegends"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="heatmaps_files/figure-html/unnamed-chunk-9-1.png" width="840"></p>
<p>You can also manually draw a legend with the convenience function
<code><a href="../../reference/anno_cat_legend.html">anno_cat_legend()</a></code>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">grid</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grid/grid.newpage.html" class="external-link">grid.newpage</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="../../reference/anno_cat_legend.html">anno_cat_legend</a></span><span class="op">(</span></span>
<span>  col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"a level"</span> <span class="op">=</span> <span class="st">"red"</span>, <span class="st">"another level"</span> <span class="op">=</span> <span class="st">"blue"</span>, c <span class="op">=</span> <span class="st">"white"</span><span class="op">)</span>, </span>
<span>  border <span class="op">=</span> <span class="st">"black"</span>, gap <span class="op">=</span> <span class="fu">grid</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">2</span>, <span class="st">"cm"</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">3</span></span>
<span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span><span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/draw-dispatch.html" class="external-link">draw</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="heatmaps_files/figure-html/unnamed-chunk-10-1.png" width="480"></p>
</div>
</div>
<div class="section level3">
<h3 id="numbering-cells">Numbering cells<a class="anchor" aria-label="anchor" href="#numbering-cells"></a>
</h3>
<p>If you have fewer samples (and taxa) you might like to label the
cells with their values. By default, the raw counts are shown.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">psq</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/tax_transform.html">tax_transform</a></span><span class="op">(</span><span class="st">"compositional"</span>, rank <span class="op">=</span> <span class="st">"Class"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/comp_heatmap.html">comp_heatmap</a></span><span class="op">(</span>samples <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">15</span>, numbers <span class="op">=</span> <span class="fu"><a href="../../reference/heat_numbers.html">heat_numbers</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="heatmaps_files/figure-html/unnamed-chunk-11-1.png" width="840"></p>
<p>You can easily change to showing the same values as the colours by
setting <code>numbers_use_counts = FALSE</code>, and you can/should
change the number of decimals shown too.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">psq</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/tax_transform.html">tax_transform</a></span><span class="op">(</span><span class="st">"compositional"</span>, rank <span class="op">=</span> <span class="st">"Class"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/comp_heatmap.html">comp_heatmap</a></span><span class="op">(</span></span>
<span>    samples <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">15</span>, numbers_use_counts <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    numbers <span class="op">=</span> <span class="fu"><a href="../../reference/heat_numbers.html">heat_numbers</a></span><span class="op">(</span>decimals <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="heatmaps_files/figure-html/unnamed-chunk-12-1.png" width="840"></p>
<p>The numbers can any transformation of counts, irrespective of what
transformations were used for the colours, or seriation.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">psq</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/tax_transform.html">tax_transform</a></span><span class="op">(</span><span class="st">"binary"</span>, undetected <span class="op">=</span> <span class="fl">0</span>, rank <span class="op">=</span> <span class="st">"Class"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/comp_heatmap.html">comp_heatmap</a></span><span class="op">(</span></span>
<span>    samples <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">15</span>, numbers_use_counts <span class="op">=</span> <span class="cn">TRUE</span>, numbers_trans <span class="op">=</span> <span class="st">"compositional"</span>, </span>
<span>    numbers <span class="op">=</span> <span class="fu"><a href="../../reference/heat_numbers.html">heat_numbers</a></span><span class="op">(</span>decimals <span class="op">=</span> <span class="fl">2</span>, col <span class="op">=</span> <span class="st">"white"</span><span class="op">)</span>, </span>
<span>    show_heatmap_legend <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="heatmaps_files/figure-html/unnamed-chunk-13-1.png" width="840"></p>
<p>To demonstrate that coloration, numbering and seriation can all use
different transformations of the original count data, the example below
we specifies seriating the taxa and samples using the same numerical
values used for the numbers transformation, not the colours, which are
just presence/absence!</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">psq</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/tax_transform.html">tax_transform</a></span><span class="op">(</span><span class="st">"binary"</span>, undetected <span class="op">=</span> <span class="fl">0</span>, rank <span class="op">=</span> <span class="st">"Class"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/comp_heatmap.html">comp_heatmap</a></span><span class="op">(</span></span>
<span>    samples <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">15</span>, </span>
<span>    sample_ser_counts <span class="op">=</span> <span class="cn">TRUE</span>, sample_ser_trans <span class="op">=</span> <span class="st">"compositional"</span>,</span>
<span>    tax_ser_counts <span class="op">=</span> <span class="cn">TRUE</span>, tax_ser_trans <span class="op">=</span> <span class="st">"compositional"</span>,</span>
<span>    numbers_use_counts <span class="op">=</span> <span class="cn">TRUE</span>, numbers_trans <span class="op">=</span> <span class="st">"compositional"</span>, </span>
<span>    numbers <span class="op">=</span> <span class="fu"><a href="../../reference/heat_numbers.html">heat_numbers</a></span><span class="op">(</span>decimals <span class="op">=</span> <span class="fl">2</span>, col <span class="op">=</span> <span class="st">"white"</span><span class="op">)</span>, </span>
<span>    show_heatmap_legend <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="heatmaps_files/figure-html/unnamed-chunk-14-1.png" width="840"></p>
</div>
</div>
<div class="section level2">
<h2 id="correlation-heatmaps">Correlation heatmaps<a class="anchor" aria-label="anchor" href="#correlation-heatmaps"></a>
</h2>
<p>Correlation heatmaps can be a nice way to quickly assess patterns of
associations between numerical variables in your dataset, such as
microbial abundances and other metadata.</p>
<p>Let’s make some fake numeric variables to exemplify this.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">111</span><span class="op">)</span> <span class="co"># ensures making same random variables every time!</span></span>
<span><span class="va">psq</span> <span class="op">&lt;-</span> <span class="va">psq</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/ps_arrange.html">ps_arrange</a></span><span class="op">(</span><span class="va">ibd</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/ps_mutate.html">ps_mutate</a></span><span class="op">(</span></span>
<span>    var1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/nsamples-methods.html" class="external-link">nsamples</a></span><span class="op">(</span><span class="va">psq</span><span class="op">)</span>, mean <span class="op">=</span> <span class="fl">10</span>, sd <span class="op">=</span> <span class="fl">3</span><span class="op">)</span>,</span>
<span>    var2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>      <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/nsamples-methods.html" class="external-link">nsamples</a></span><span class="op">(</span><span class="va">psq</span><span class="op">)</span><span class="op">*</span><span class="fl">0.75</span>, mean <span class="op">=</span> <span class="fl">4</span>, sd <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>,</span>
<span>      <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">1</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/nsamples-methods.html" class="external-link">nsamples</a></span><span class="op">(</span><span class="va">psq</span><span class="op">)</span><span class="op">/</span><span class="fl">4</span>, mean <span class="op">=</span> <span class="fl">9</span>, sd <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span>      <span class="op">)</span>,</span>
<span>    var3 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/nsamples-methods.html" class="external-link">nsamples</a></span><span class="op">(</span><span class="va">psq</span><span class="op">)</span>, <span class="fl">2</span>, <span class="fl">10</span><span class="op">)</span>,</span>
<span>    var4 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/nsamples-methods.html" class="external-link">nsamples</a></span><span class="op">(</span><span class="va">psq</span><span class="op">)</span>, mean <span class="op">=</span> <span class="fl">100</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/nsamples-methods.html" class="external-link">nsamples</a></span><span class="op">(</span><span class="va">psq</span><span class="op">)</span><span class="op">:</span><span class="fl">0</span>, sd <span class="op">=</span> <span class="fl">20</span><span class="op">)</span> <span class="op">/</span> <span class="fl">20</span>,</span>
<span>    var5 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/nsamples-methods.html" class="external-link">nsamples</a></span><span class="op">(</span><span class="va">psq</span><span class="op">)</span>, mean <span class="op">=</span> <span class="fl">5</span>, sd <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>,</span>
<span>    var6 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/NegBinomial.html" class="external-link">rnbinom</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/nsamples-methods.html" class="external-link">nsamples</a></span><span class="op">(</span><span class="va">psq</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">75</span><span class="op">/</span><span class="fl">10</span>, mu <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="calculating-correlations">Calculating correlations<a class="anchor" aria-label="anchor" href="#calculating-correlations"></a>
</h3>
<p>By default, the <code>cor_heatmap</code> function will correlate all
taxa to all numerical sample data, using pearson correlation method.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">psq</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/tax_agg.html">tax_agg</a></span><span class="op">(</span><span class="st">"Family"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/cor_heatmap.html">cor_heatmap</a></span><span class="op">(</span>vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"var1"</span>, <span class="st">"var2"</span>, <span class="st">"var3"</span>, <span class="st">"var4"</span>, <span class="st">"var5"</span>, <span class="st">"var6"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="heatmaps_files/figure-html/unnamed-chunk-16-1.png" width="840"></p>
<p>It’s easy to change to a different method, i.e. spearman’s rank
correlation or kendall’s tau, which will be reflected in the legend
title. We will also specify to use only the 15 most abundant taxa, by
maximum count, just to make these tutorial figures a little more
compact!</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">psq</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/tax_agg.html">tax_agg</a></span><span class="op">(</span><span class="st">"Family"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/cor_heatmap.html">cor_heatmap</a></span><span class="op">(</span></span>
<span>    taxa <span class="op">=</span> <span class="fu"><a href="../../reference/tax_top.html">tax_top</a></span><span class="op">(</span><span class="va">psq</span>, <span class="fl">15</span>, by <span class="op">=</span> <span class="va">max</span>, rank <span class="op">=</span> <span class="st">"Family"</span><span class="op">)</span>,</span>
<span>    vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"var"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">)</span>, cor <span class="op">=</span> <span class="st">"spearman"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="heatmaps_files/figure-html/unnamed-chunk-17-1.png" width="840"></p>
<p><em>Older versions of microViz <code>cor_heatmap</code> had a
<code>tax_transform</code> argument. But for flexibility, you must now
transform your taxa <strong>before</strong> passing the psExtra object
to <code>cor_heatmap</code>.</em></p>
<p>Here we have transformed our taxa with the “clr” or
centered-log-ratio transformation prior to correlating. <em>Notice that
the annotations stay on the same scale, as by default the annotation
functions extract the stored counts data from the psExtra input, not the
transformed data.</em></p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">psq</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/tax_agg.html">tax_agg</a></span><span class="op">(</span><span class="st">"Family"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/tax_transform.html">tax_transform</a></span><span class="op">(</span><span class="st">"clr"</span>, zero_replace <span class="op">=</span> <span class="st">"halfmin"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/cor_heatmap.html">cor_heatmap</a></span><span class="op">(</span></span>
<span>    taxa <span class="op">=</span> <span class="fu"><a href="../../reference/tax_top.html">tax_top</a></span><span class="op">(</span><span class="va">psq</span>, <span class="fl">15</span>, by <span class="op">=</span> <span class="va">max</span>, rank <span class="op">=</span> <span class="st">"Family"</span><span class="op">)</span>,</span>
<span>    vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"var"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="heatmaps_files/figure-html/unnamed-chunk-18-1.png" width="840"></p>
<p>Let’s transform and scale the taxon abundances before
correlating.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">psq</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/tax_agg.html">tax_agg</a></span><span class="op">(</span><span class="st">"Family"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/tax_transform.html">tax_transform</a></span><span class="op">(</span><span class="st">"clr"</span>, zero_replace <span class="op">=</span> <span class="st">"halfmin"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/cor_heatmap.html">cor_heatmap</a></span><span class="op">(</span></span>
<span>    taxa <span class="op">=</span> <span class="fu"><a href="../../reference/tax_top.html">tax_top</a></span><span class="op">(</span><span class="va">psq</span>, <span class="fl">15</span>, by <span class="op">=</span> <span class="va">max</span>, rank <span class="op">=</span> <span class="st">"Family"</span><span class="op">)</span>,</span>
<span>    vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"var"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="heatmaps_files/figure-html/unnamed-chunk-19-1.png" width="840"></p>
</div>
<div class="section level3">
<h3 id="taxon-annotations">Taxon annotations<a class="anchor" aria-label="anchor" href="#taxon-annotations"></a>
</h3>
<p>As seen in the previous plots taxa are annotated by default with
prevalence and relative abundance.</p>
<p>You can transform the taxa for the abundance annotation. The
<code>trans</code> and <code>zero_replace</code> arguments are sent to
<code><a href="../../reference/tax_transform.html">tax_transform()</a></code>.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">psq</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/tax_agg.html">tax_agg</a></span><span class="op">(</span><span class="st">"Family"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/cor_heatmap.html">cor_heatmap</a></span><span class="op">(</span></span>
<span>    taxa <span class="op">=</span> <span class="fu"><a href="../../reference/tax_top.html">tax_top</a></span><span class="op">(</span><span class="va">psq</span>, <span class="fl">15</span>, by <span class="op">=</span> <span class="va">max</span>, rank <span class="op">=</span> <span class="st">"Family"</span><span class="op">)</span>,</span>
<span>    vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"var"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">)</span>, </span>
<span>    tax_anno <span class="op">=</span> <span class="fu"><a href="../../reference/taxAnnotation.html">taxAnnotation</a></span><span class="op">(</span></span>
<span>      Prev. <span class="op">=</span> <span class="fu"><a href="../../reference/anno_tax_prev.html">anno_tax_prev</a></span><span class="op">(</span>ylim <span class="op">=</span> <span class="fl">0</span><span class="op">:</span><span class="fl">1</span><span class="op">)</span>, </span>
<span>      Log10. <span class="op">=</span> <span class="fu"><a href="../../reference/anno_tax_box.html">anno_tax_box</a></span><span class="op">(</span>trans <span class="op">=</span> <span class="st">"log10"</span>, zero_replace <span class="op">=</span> <span class="st">"halfmin"</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="heatmaps_files/figure-html/unnamed-chunk-20-1.png" width="840"></p>
<p>You can do multiple transformations and or scaling by supplying a
function, that takes a psExtra or phyloseq object, transforms it, and
returns it.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">psq</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/tax_agg.html">tax_agg</a></span><span class="op">(</span><span class="st">"Family"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/cor_heatmap.html">cor_heatmap</a></span><span class="op">(</span></span>
<span>    taxa <span class="op">=</span> <span class="fu"><a href="../../reference/tax_top.html">tax_top</a></span><span class="op">(</span><span class="va">psq</span>, <span class="fl">15</span>, by <span class="op">=</span> <span class="va">max</span>, rank <span class="op">=</span> <span class="st">"Family"</span><span class="op">)</span>,</span>
<span>    vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"var"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">)</span>, </span>
<span>    tax_anno <span class="op">=</span> <span class="fu"><a href="../../reference/taxAnnotation.html">taxAnnotation</a></span><span class="op">(</span></span>
<span>      Log2 <span class="op">=</span> <span class="fu"><a href="../../reference/anno_tax_density.html">anno_tax_density</a></span><span class="op">(</span></span>
<span>        joyplot_scale <span class="op">=</span> <span class="fl">2</span>, gp <span class="op">=</span> <span class="fu">grid</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grid/gpar.html" class="external-link">gpar</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"black"</span>, alpha <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span>,</span>
<span>        trans <span class="op">=</span> <span class="st">"log2"</span>, zero_replace <span class="op">=</span> <span class="fl">1</span></span>
<span>      <span class="op">)</span>,</span>
<span>      `prop Log2` <span class="op">=</span> <span class="fu"><a href="../../reference/anno_tax_density.html">anno_tax_density</a></span><span class="op">(</span></span>
<span>        joyplot_scale <span class="op">=</span> <span class="fl">1.5</span>, gp <span class="op">=</span> <span class="fu">grid</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grid/gpar.html" class="external-link">gpar</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"black"</span>, alpha <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span>,</span>
<span>        trans <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">ps</span><span class="op">)</span> <span class="op">{</span></span>
<span>          <span class="va">ps</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>            <span class="fu"><a href="../../reference/tax_transform.html">tax_transform</a></span><span class="op">(</span><span class="st">"compositional"</span>, zero_replace <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>            <span class="fu"><a href="../../reference/tax_transform.html">tax_transform</a></span><span class="op">(</span><span class="st">"log2"</span>, chain <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> </span>
<span>        <span class="op">}</span></span>
<span>      <span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="heatmaps_files/figure-html/unnamed-chunk-21-1.png" width="840"></p>
<p>Note that by default the relative abundance is shown only for samples
where the taxon is detected! You can include values for all samples for
all taxa by setting <code>only_detected = FALSE</code>.</p>
<p>Let’s try this with a heatmap-style density plot annotation. We’ll
replace zeroes with ones for an interpretable minimum value on the
plot.</p>
<p>We’ll compare it side-by-side with the default setting of showing
only distribution of values above the detection threshold.</p>
<p>For zero-inflated microbiome data, showing prevalence and “abundance
when detected” often seems like a more informative annotation.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">psq</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/tax_agg.html">tax_agg</a></span><span class="op">(</span><span class="st">"Family"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/cor_heatmap.html">cor_heatmap</a></span><span class="op">(</span></span>
<span>    taxa <span class="op">=</span> <span class="fu"><a href="../../reference/tax_top.html">tax_top</a></span><span class="op">(</span><span class="va">psq</span>, <span class="fl">15</span>, by <span class="op">=</span> <span class="va">max</span>, rank <span class="op">=</span> <span class="st">"Family"</span><span class="op">)</span>,</span>
<span>    vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"var"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">)</span>, </span>
<span>    tax_anno <span class="op">=</span> <span class="fu"><a href="../../reference/taxAnnotation.html">taxAnnotation</a></span><span class="op">(</span></span>
<span>      Prev. <span class="op">=</span> <span class="fu"><a href="../../reference/anno_tax_prev.html">anno_tax_prev</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fu">grid</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">10</span>, <span class="st">"mm"</span><span class="op">)</span>, ylim <span class="op">=</span> <span class="fl">0</span><span class="op">:</span><span class="fl">1</span><span class="op">)</span>, </span>
<span>      All <span class="op">=</span> <span class="fu"><a href="../../reference/anno_tax_density.html">anno_tax_density</a></span><span class="op">(</span></span>
<span>        size <span class="op">=</span> <span class="fu">grid</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">20</span>, <span class="st">"mm"</span><span class="op">)</span>,</span>
<span>        trans <span class="op">=</span> <span class="st">"log10"</span>, zero_replace <span class="op">=</span> <span class="fl">1</span>, </span>
<span>        heatmap_colors <span class="op">=</span> <span class="fu">viridisLite</span><span class="fu">::</span><span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">turbo</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">15</span><span class="op">)</span>,</span>
<span>        type <span class="op">=</span> <span class="st">"heatmap"</span>, only_detected <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>      <span class="op">)</span>,</span>
<span>      Default <span class="op">=</span> <span class="fu"><a href="../../reference/anno_tax_density.html">anno_tax_density</a></span><span class="op">(</span></span>
<span>        size <span class="op">=</span> <span class="fu">grid</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">20</span>, <span class="st">"mm"</span><span class="op">)</span>,</span>
<span>        trans <span class="op">=</span> <span class="st">"log10"</span>, zero_replace <span class="op">=</span> <span class="fl">1</span>, </span>
<span>        heatmap_colors <span class="op">=</span> <span class="fu">viridisLite</span><span class="fu">::</span><span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">turbo</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">15</span><span class="op">)</span>,</span>
<span>        type <span class="op">=</span> <span class="st">"heatmap"</span>, only_detected <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>      <span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="heatmaps_files/figure-html/unnamed-chunk-22-1.png" width="840"></p>
<div class="section level4">
<h4 id="sorting">Sorting<a class="anchor" aria-label="anchor" href="#sorting"></a>
</h4>
<p>By default, rows and columns are sorted using hierarchical clustering
with optimal leaf ordering <code>"OLO_ward"</code>. You can use any
valid method from the <code>seriation</code> package. You can suppress
ordering by using <code>seriation_method = "Identity"</code>. By default
this also suppresses column ordering, so you can set
<code>seriation_method_col = OLO_ward</code> to keep ordering.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">psq</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/tax_agg.html">tax_agg</a></span><span class="op">(</span><span class="st">"Family"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/tax_sort.html">tax_sort</a></span><span class="op">(</span>by <span class="op">=</span> <span class="va">prev</span>, at <span class="op">=</span> <span class="st">"Family"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/cor_heatmap.html">cor_heatmap</a></span><span class="op">(</span></span>
<span>    seriation_method <span class="op">=</span> <span class="st">"Identity"</span>,</span>
<span>    seriation_method_col <span class="op">=</span> <span class="st">"OLO_ward"</span>,</span>
<span>    taxa <span class="op">=</span> <span class="fu"><a href="../../reference/tax_top.html">tax_top</a></span><span class="op">(</span><span class="va">psq</span>, <span class="fl">15</span>, by <span class="op">=</span> <span class="va">max</span>, rank <span class="op">=</span> <span class="st">"Family"</span><span class="op">)</span>,</span>
<span>    vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"var"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">)</span>, </span>
<span>    tax_anno <span class="op">=</span> <span class="fu"><a href="../../reference/taxAnnotation.html">taxAnnotation</a></span><span class="op">(</span></span>
<span>      Prev. <span class="op">=</span> <span class="fu"><a href="../../reference/anno_tax_prev.html">anno_tax_prev</a></span><span class="op">(</span>ylim <span class="op">=</span> <span class="fl">0</span><span class="op">:</span><span class="fl">1</span><span class="op">)</span>, </span>
<span>      CLR <span class="op">=</span> <span class="fu"><a href="../../reference/anno_tax_box.html">anno_tax_box</a></span><span class="op">(</span>trans <span class="op">=</span> <span class="st">"clr"</span>, zero_replace <span class="op">=</span> <span class="st">"halfmin"</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="heatmaps_files/figure-html/unnamed-chunk-23-1.png" width="840"></p>
</div>
<div class="section level4">
<h4 id="taxa-annotation-side">Taxa annotation side<a class="anchor" aria-label="anchor" href="#taxa-annotation-side"></a>
</h4>
<p>You can easily put the taxa annotations on another of the heatmap
with e.g. <code>taxa_side = "left"</code></p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">psq</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/tax_agg.html">tax_agg</a></span><span class="op">(</span><span class="st">"Family"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/tax_sort.html">tax_sort</a></span><span class="op">(</span>by <span class="op">=</span> <span class="va">prev</span>, at <span class="op">=</span> <span class="st">"Family"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/cor_heatmap.html">cor_heatmap</a></span><span class="op">(</span></span>
<span>    seriation_method <span class="op">=</span> <span class="st">"Identity"</span>,</span>
<span>    seriation_method_col <span class="op">=</span> <span class="st">"OLO_ward"</span>,</span>
<span>    taxa_side <span class="op">=</span> <span class="st">"left"</span>,</span>
<span>    taxa <span class="op">=</span> <span class="fu"><a href="../../reference/tax_top.html">tax_top</a></span><span class="op">(</span><span class="va">psq</span>, <span class="fl">15</span>, by <span class="op">=</span> <span class="va">max</span>, rank <span class="op">=</span> <span class="st">"Family"</span><span class="op">)</span>,</span>
<span>    vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"var"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">)</span>,</span>
<span>    tax_anno <span class="op">=</span> <span class="fu"><a href="../../reference/taxAnnotation.html">taxAnnotation</a></span><span class="op">(</span></span>
<span>      Prev. <span class="op">=</span> <span class="fu"><a href="../../reference/anno_tax_prev.html">anno_tax_prev</a></span><span class="op">(</span>ylim <span class="op">=</span> <span class="fl">0</span><span class="op">:</span><span class="fl">1</span><span class="op">)</span>, </span>
<span>      CLR <span class="op">=</span> <span class="fu"><a href="../../reference/anno_tax_box.html">anno_tax_box</a></span><span class="op">(</span>trans <span class="op">=</span> <span class="st">"clr"</span>, zero_replace <span class="op">=</span> <span class="st">"halfmin"</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="heatmaps_files/figure-html/unnamed-chunk-24-1.png" width="840"></p>
<p>Or on the top or bottom is also possible, this will rotate the
heatmap. Remember to swap the seriation method arguments around!</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">psq</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/tax_agg.html">tax_agg</a></span><span class="op">(</span><span class="st">"Family"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/tax_sort.html">tax_sort</a></span><span class="op">(</span>by <span class="op">=</span> <span class="va">prev</span>, at <span class="op">=</span> <span class="st">"Family"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/cor_heatmap.html">cor_heatmap</a></span><span class="op">(</span></span>
<span>    seriation_method_col <span class="op">=</span> <span class="st">"Identity"</span>, <span class="co">#swapped!</span></span>
<span>    seriation_method <span class="op">=</span> <span class="st">"OLO_ward"</span>, <span class="co">#swapped!</span></span>
<span>    taxa_side <span class="op">=</span> <span class="st">"top"</span>,</span>
<span>    taxa <span class="op">=</span> <span class="fu"><a href="../../reference/tax_top.html">tax_top</a></span><span class="op">(</span><span class="va">psq</span>, <span class="fl">15</span>, by <span class="op">=</span> <span class="va">max</span>, rank <span class="op">=</span> <span class="st">"Family"</span><span class="op">)</span>,</span>
<span>    vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"var"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">)</span>,</span>
<span>    tax_anno <span class="op">=</span> <span class="fu"><a href="../../reference/taxAnnotation.html">taxAnnotation</a></span><span class="op">(</span></span>
<span>      Prev. <span class="op">=</span> <span class="fu"><a href="../../reference/anno_tax_prev.html">anno_tax_prev</a></span><span class="op">(</span>ylim <span class="op">=</span> <span class="fl">0</span><span class="op">:</span><span class="fl">1</span><span class="op">)</span>, </span>
<span>      CLR <span class="op">=</span> <span class="fu"><a href="../../reference/anno_tax_box.html">anno_tax_box</a></span><span class="op">(</span>trans <span class="op">=</span> <span class="st">"clr"</span>, zero_replace <span class="op">=</span> <span class="st">"halfmin"</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="heatmaps_files/figure-html/unnamed-chunk-25-1.png" width="600"></p>
</div>
</div>
<div class="section level3">
<h3 id="variable-annotation">Variable annotation<a class="anchor" aria-label="anchor" href="#variable-annotation"></a>
</h3>
<p>As well as annotating the taxa, you can also annotate the
variables.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">psq</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/tax_agg.html">tax_agg</a></span><span class="op">(</span><span class="st">"Family"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/cor_heatmap.html">cor_heatmap</a></span><span class="op">(</span></span>
<span>    taxa <span class="op">=</span> <span class="fu"><a href="../../reference/tax_top.html">tax_top</a></span><span class="op">(</span><span class="va">psq</span>, <span class="fl">15</span>, by <span class="op">=</span> <span class="va">max</span>, rank <span class="op">=</span> <span class="st">"Family"</span><span class="op">)</span>,</span>
<span>    vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"var"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">)</span>, </span>
<span>    var_anno <span class="op">=</span> <span class="fu"><a href="../../reference/varAnnotation.html">varAnnotation</a></span><span class="op">(</span></span>
<span>      Value <span class="op">=</span> <span class="fu"><a href="../../reference/anno_var_box.html">anno_var_box</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>      Zscore <span class="op">=</span> <span class="fu"><a href="../../reference/anno_var_density.html">anno_var_density</a></span><span class="op">(</span>fun <span class="op">=</span> <span class="va">scale</span>, type <span class="op">=</span> <span class="st">"violin"</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="heatmaps_files/figure-html/unnamed-chunk-26-1.png" width="720"></p>
</div>
</div>
<div class="section level2">
<h2 id="other-stuff">Other stuff<a class="anchor" aria-label="anchor" href="#other-stuff"></a>
</h2>
<p>Complicated stuff demonstrated down here, not necessarily useful.</p>
<div class="section level3">
<h3 id="custom-breaks-and-seriation">Custom breaks and seriation<a class="anchor" aria-label="anchor" href="#custom-breaks-and-seriation"></a>
</h3>
<p>Two approaches to custom colour scale breaks. The first way is
better, because the colour scale is interpolated through the default 11
colours, instead of only 5.</p>
<p>Transform data and customise only labels.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">psq</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/tax_transform.html">tax_transform</a></span><span class="op">(</span><span class="st">"compositional"</span>, rank <span class="op">=</span> <span class="st">"Class"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/tax_transform.html">tax_transform</a></span><span class="op">(</span><span class="st">"log10"</span>, zero_replace <span class="op">=</span> <span class="st">"halfmin"</span>, chain <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/comp_heatmap.html">comp_heatmap</a></span><span class="op">(</span></span>
<span>    tax_anno <span class="op">=</span> <span class="fu"><a href="../../reference/taxAnnotation.html">taxAnnotation</a></span><span class="op">(</span></span>
<span>      Prev. <span class="op">=</span> <span class="fu"><a href="../../reference/anno_tax_prev.html">anno_tax_prev</a></span><span class="op">(</span>bar_width <span class="op">=</span> <span class="fl">0.3</span>, size <span class="op">=</span> <span class="fu">grid</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">1</span>, <span class="st">"cm"</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">)</span>,</span>
<span>    heatmap_legend_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>      labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rev.html" class="external-link">rev</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"100%"</span>, <span class="st">" 10%"</span>, <span class="st">"  1%"</span>, <span class="st">" 0.1%"</span>, <span class="st">"0.01%"</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="heatmaps_files/figure-html/unnamed-chunk-27-1.png" width="960"></p>
<p>This alternative way might be helpful in some cases, maybe… It
demonstrates that custom breaks can be set in
<code><a href="../../reference/heat_palette.html">heat_palette()</a></code>.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># seriation transform</span></span>
<span><span class="va">serTrans</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="../../reference/tax_transform.html">tax_transform</a></span><span class="op">(</span><span class="va">x</span>, trans <span class="op">=</span> <span class="st">"log10"</span>, zero_replace <span class="op">=</span> <span class="st">"halfmin"</span>, chain <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">psq</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/tax_transform.html">tax_transform</a></span><span class="op">(</span><span class="st">"compositional"</span>, rank <span class="op">=</span> <span class="st">"Class"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/comp_heatmap.html">comp_heatmap</a></span><span class="op">(</span></span>
<span>    sample_ser_trans <span class="op">=</span> <span class="va">serTrans</span>, tax_ser_trans <span class="op">=</span> <span class="va">serTrans</span>,</span>
<span>    colors <span class="op">=</span> <span class="fu"><a href="../../reference/heat_palette.html">heat_palette</a></span><span class="op">(</span>breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.0001</span>, <span class="fl">0.001</span>, <span class="fl">0.01</span>, <span class="fl">0.1</span>, <span class="fl">1</span><span class="op">)</span>, rev <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>,</span>
<span>    tax_anno <span class="op">=</span> <span class="fu"><a href="../../reference/taxAnnotation.html">taxAnnotation</a></span><span class="op">(</span></span>
<span>      Prev. <span class="op">=</span> <span class="fu"><a href="../../reference/anno_tax_prev.html">anno_tax_prev</a></span><span class="op">(</span>bar_width <span class="op">=</span> <span class="fl">0.3</span>, size <span class="op">=</span> <span class="fu">grid</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">1</span>, <span class="st">"cm"</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">)</span>,</span>
<span>    heatmap_legend_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>at <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.0001</span>, <span class="fl">0.001</span>, <span class="fl">0.01</span>, <span class="fl">0.1</span>, <span class="fl">1</span><span class="op">)</span>, break_dist <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="heatmaps_files/figure-html/unnamed-chunk-28-1.png" width="960"></p>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://r-lib.github.io/sessioninfo/reference/session_info.html" class="external-link">session_info</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB; font-weight: bold;">─ Session info ───────────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt;  <span style="color: #555555; font-style: italic;">setting </span> <span style="color: #555555; font-style: italic;">value</span></span></span>
<span><span class="co">#&gt;  version  R version 4.2.2 (2022-10-31)</span></span>
<span><span class="co">#&gt;  os       Ubuntu 20.04.5 LTS</span></span>
<span><span class="co">#&gt;  system   x86_64, linux-gnu</span></span>
<span><span class="co">#&gt;  ui       X11</span></span>
<span><span class="co">#&gt;  language en</span></span>
<span><span class="co">#&gt;  collate  C.UTF-8</span></span>
<span><span class="co">#&gt;  ctype    C.UTF-8</span></span>
<span><span class="co">#&gt;  tz       UTC</span></span>
<span><span class="co">#&gt;  date     2022-11-18</span></span>
<span><span class="co">#&gt;  pandoc   2.19.2 @ /usr/bin/ (via rmarkdown)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB; font-weight: bold;">─ Packages ───────────────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt;  <span style="color: #555555; font-style: italic;">package         </span> <span style="color: #555555; font-style: italic;">*</span> <span style="color: #555555; font-style: italic;">version </span> <span style="color: #555555; font-style: italic;">date (UTC)</span> <span style="color: #555555; font-style: italic;">lib</span> <span style="color: #555555; font-style: italic;">source</span></span></span>
<span><span class="co">#&gt;  ade4               1.7-20   <span style="color: #555555;">2022-11-01</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ape                5.6-2    <span style="color: #555555;">2022-03-02</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  Biobase            2.56.0   <span style="color: #555555;">2022-04-26</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  BiocGenerics       0.42.0   <span style="color: #555555;">2022-04-26</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  biomformat         1.24.0   <span style="color: #555555;">2022-04-26</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  Biostrings         2.64.1   <span style="color: #555555;">2022-08-18</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  bitops             1.0-7    <span style="color: #555555;">2021-04-24</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  bslib              0.4.1    <span style="color: #555555;">2022-11-02</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ca                 0.71.1   <span style="color: #555555;">2020-01-24</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  cachem             1.0.6    <span style="color: #555555;">2021-08-19</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  callr              3.7.3    <span style="color: #555555;">2022-11-02</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  circlize           0.4.15   <span style="color: #555555;">2022-05-10</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  cli                3.4.1    <span style="color: #555555;">2022-09-23</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  clue               0.3-62   <span style="color: #555555;">2022-10-18</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  cluster            2.1.4    <span style="color: #555555;">2022-08-22</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.2.2)</span></span></span>
<span><span class="co">#&gt;  codetools          0.2-18   <span style="color: #555555;">2020-11-04</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.2.2)</span></span></span>
<span><span class="co">#&gt;  colorspace         2.0-3    <span style="color: #555555;">2022-02-21</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ComplexHeatmap     2.12.1   <span style="color: #555555;">2022-08-09</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  crayon             1.5.2    <span style="color: #555555;">2022-09-29</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  data.table         1.14.6   <span style="color: #555555;">2022-11-16</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  desc               1.4.2    <span style="color: #555555;">2022-09-08</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  devtools           2.4.5    <span style="color: #555555;">2022-10-11</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  digest             0.6.30   <span style="color: #555555;">2022-10-18</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  doParallel         1.0.17   <span style="color: #555555;">2022-02-07</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  dplyr            * 1.0.10   <span style="color: #555555;">2022-09-01</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ellipsis           0.3.2    <span style="color: #555555;">2021-04-29</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  evaluate           0.18     <span style="color: #555555;">2022-11-07</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  fansi              1.0.3    <span style="color: #555555;">2022-03-24</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  fastmap            1.1.0    <span style="color: #555555;">2021-01-25</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  foreach            1.5.2    <span style="color: #555555;">2022-02-02</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  fs                 1.5.2    <span style="color: #555555;">2021-12-08</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  generics           0.1.3    <span style="color: #555555;">2022-07-05</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  GenomeInfoDb       1.32.4   <span style="color: #555555;">2022-09-06</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  GenomeInfoDbData   1.2.8    <span style="color: #555555;">2022-11-14</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  GetoptLong         1.0.5    <span style="color: #555555;">2020-12-15</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ggplot2            3.4.0    <span style="color: #555555;">2022-11-04</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  GlobalOptions      0.1.2    <span style="color: #555555;">2020-06-10</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  glue               1.6.2    <span style="color: #555555;">2022-02-24</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  gtable             0.3.1    <span style="color: #555555;">2022-09-01</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  highr              0.9      <span style="color: #555555;">2021-04-16</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  htmltools          0.5.3    <span style="color: #555555;">2022-07-18</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  htmlwidgets        1.5.4    <span style="color: #555555;">2021-09-08</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  httpuv             1.6.6    <span style="color: #555555;">2022-09-08</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  igraph             1.3.5    <span style="color: #555555;">2022-09-22</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  IRanges            2.30.1   <span style="color: #555555;">2022-08-18</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  iterators          1.0.14   <span style="color: #555555;">2022-02-05</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  jquerylib          0.1.4    <span style="color: #555555;">2021-04-26</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  jsonlite           1.8.3    <span style="color: #555555;">2022-10-21</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  knitr              1.40     <span style="color: #555555;">2022-08-24</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  later              1.3.0    <span style="color: #555555;">2021-08-18</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  lattice            0.20-45  <span style="color: #555555;">2021-09-22</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.2.2)</span></span></span>
<span><span class="co">#&gt;  lifecycle          1.0.3    <span style="color: #555555;">2022-10-07</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  magrittr           2.0.3    <span style="color: #555555;">2022-03-30</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  MASS               7.3-58.1 <span style="color: #555555;">2022-08-03</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.2.2)</span></span></span>
<span><span class="co">#&gt;  Matrix             1.5-1    <span style="color: #555555;">2022-09-13</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.2.2)</span></span></span>
<span><span class="co">#&gt;  matrixStats        0.62.0   <span style="color: #555555;">2022-04-19</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  memoise            2.0.1    <span style="color: #555555;">2021-11-26</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  mgcv               1.8-41   <span style="color: #555555;">2022-10-21</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.2.2)</span></span></span>
<span><span class="co">#&gt;  microbiome         1.18.0   <span style="color: #555555;">2022-04-26</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  microViz         * 0.10.3   <span style="color: #555555;">2022-11-18</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">local</span></span></span>
<span><span class="co">#&gt;  mime               0.12     <span style="color: #555555;">2021-09-28</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  miniUI             0.1.1.1  <span style="color: #555555;">2018-05-18</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  multtest           2.52.0   <span style="color: #555555;">2022-04-26</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  munsell            0.5.0    <span style="color: #555555;">2018-06-12</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  nlme               3.1-160  <span style="color: #555555;">2022-10-10</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.2.2)</span></span></span>
<span><span class="co">#&gt;  permute            0.9-7    <span style="color: #555555;">2022-01-27</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  phyloseq         * 1.40.0   <span style="color: #555555;">2022-04-26</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  pillar             1.8.1    <span style="color: #555555;">2022-08-19</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  pkgbuild           1.3.1    <span style="color: #555555;">2021-12-20</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  pkgconfig          2.0.3    <span style="color: #555555;">2019-09-22</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  pkgdown            2.0.6    <span style="color: #555555;">2022-07-16</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  pkgload            1.3.2    <span style="color: #555555;">2022-11-16</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  plyr               1.8.8    <span style="color: #555555;">2022-11-11</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  png                0.1-7    <span style="color: #555555;">2013-12-03</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  prettyunits        1.1.1    <span style="color: #555555;">2020-01-24</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  processx           3.8.0    <span style="color: #555555;">2022-10-26</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  profvis            0.3.7    <span style="color: #555555;">2020-11-02</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  promises           1.2.0.1  <span style="color: #555555;">2021-02-11</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ps                 1.7.2    <span style="color: #555555;">2022-10-26</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  purrr              0.3.5    <span style="color: #555555;">2022-10-06</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  R6                 2.5.1    <span style="color: #555555;">2021-08-19</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ragg               1.2.4    <span style="color: #555555;">2022-10-24</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  RColorBrewer       1.1-3    <span style="color: #555555;">2022-04-03</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  Rcpp               1.0.9    <span style="color: #555555;">2022-07-08</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  RCurl              1.98-1.9 <span style="color: #555555;">2022-10-03</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  registry           0.5-1    <span style="color: #555555;">2019-03-05</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  remotes            2.4.2    <span style="color: #555555;">2021-11-30</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  reshape2           1.4.4    <span style="color: #555555;">2020-04-09</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  rhdf5              2.40.0   <span style="color: #555555;">2022-04-26</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  rhdf5filters       1.8.0    <span style="color: #555555;">2022-04-26</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  Rhdf5lib           1.18.2   <span style="color: #555555;">2022-05-15</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  rjson              0.2.21   <span style="color: #555555;">2022-01-09</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  rlang              1.0.6    <span style="color: #555555;">2022-09-24</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  rmarkdown          2.18     <span style="color: #555555;">2022-11-09</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  rprojroot          2.0.3    <span style="color: #555555;">2022-04-02</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  Rtsne              0.16     <span style="color: #555555;">2022-04-17</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  S4Vectors          0.34.0   <span style="color: #555555;">2022-04-26</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  sass               0.4.2    <span style="color: #555555;">2022-07-16</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  scales             1.2.1    <span style="color: #555555;">2022-08-20</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  seriation          1.4.0    <span style="color: #555555;">2022-10-21</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  sessioninfo        1.2.2    <span style="color: #555555;">2021-12-06</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  shape              1.4.6    <span style="color: #555555;">2021-05-19</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  shiny              1.7.3    <span style="color: #555555;">2022-10-25</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  stringi            1.7.8    <span style="color: #555555;">2022-07-11</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  stringr            1.4.1    <span style="color: #555555;">2022-08-20</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  survival           3.4-0    <span style="color: #555555;">2022-08-09</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.2.2)</span></span></span>
<span><span class="co">#&gt;  systemfonts        1.0.4    <span style="color: #555555;">2022-02-11</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  textshaping        0.3.6    <span style="color: #555555;">2021-10-13</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  tibble             3.1.8    <span style="color: #555555;">2022-07-22</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  tidyr              1.2.1    <span style="color: #555555;">2022-09-08</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  tidyselect         1.2.0    <span style="color: #555555;">2022-10-10</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  TSP                1.2-1    <span style="color: #555555;">2022-07-14</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  urlchecker         1.0.1    <span style="color: #555555;">2021-11-30</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  usethis            2.1.6    <span style="color: #555555;">2022-05-25</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  utf8               1.2.2    <span style="color: #555555;">2021-07-24</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  vctrs              0.5.1    <span style="color: #555555;">2022-11-16</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  vegan              2.6-4    <span style="color: #555555;">2022-10-11</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  viridisLite        0.4.1    <span style="color: #555555;">2022-08-22</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  withr              2.5.0    <span style="color: #555555;">2022-03-03</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  xfun               0.35     <span style="color: #555555;">2022-11-16</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  xtable             1.8-4    <span style="color: #555555;">2019-04-21</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  XVector            0.36.0   <span style="color: #555555;">2022-04-26</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  yaml               2.3.6    <span style="color: #555555;">2022-10-18</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  zlibbioc           1.42.0   <span style="color: #555555;">2022-04-26</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #555555;"> [1] /home/runner/work/_temp/Library</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;"> [2] /opt/R/4.2.2/lib/R/site-library</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;"> [3] /opt/R/4.2.2/lib/R/library</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB; font-weight: bold;">──────────────────────────────────────────────────────────────────────────────</span></span></span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by David Barnett.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: '22e6f0a75b7b1ffae9bf5bc5b58864d7',
    indexName: 'microViz',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script><script data-goatcounter="https://microviz.goatcounter.com/count" async src="//gc.zgo.at/count.js"></script>
</body>
</html>
