<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Statistical modelling of individual taxa • microViz</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../../bootstrap-toc.css">
<script src="../../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><!-- docsearch --><script src="../../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><meta property="og:title" content="Statistical modelling of individual taxa">
<meta property="og:description" content="microViz">
<meta property="og:image" content="https://david-barnett.github.io/microViz/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-2YQ669S4V5"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-2YQ669S4V5');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">microViz</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.10.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/web-only/phyloseq.html">Working with phyloseq objects</a>
    </li>
    <li>
      <a href="../../articles/web-only/tax-fixing.html">Fixing your tax_table</a>
    </li>
    <li>
      <a href="../../articles/web-only/ordination.html">Ordination plots</a>
    </li>
    <li>
      <a href="../../articles/web-only/ordination-interactive.html">Interactive ordination plots</a>
    </li>
    <li>
      <a href="../../articles/web-only/compositions.html">Visualising compositions with barplots</a>
    </li>
    <li>
      <a href="../../articles/web-only/heatmaps.html">Heatmaps of microbiome composition or correlation</a>
    </li>
    <li>
      <a href="../../articles/web-only/modelling-taxa.html">Statistical modelling of individual taxa</a>
    </li>
    <li>
      <a href="../../articles/web-only/videos.html">Tutorial video</a>
    </li>
    <li>
      <a href="../../articles/ps_extra-replaced.html">Upgrading to microViz version 0.10.0</a>
    </li>
    <li>
      <a href="../../articles/web-only/atlas1006.html">Example analyses with atlas1006 data</a>
    </li>
  </ul>
</li>
<li>
  <a href="../../articles/learn/exercises.html">Learn</a>
</li>
<li>
  <a href="../../news/index.html">News</a>
</li>
<li class="dropdown-header">
      </li>
</ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/david-barnett/microViz/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Statistical modelling of individual taxa</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/david-barnett/microViz/blob/HEAD/vignettes/web-only/modelling-taxa.Rmd" class="external-link"><code>vignettes/web-only/modelling-taxa.Rmd</code></a></small>
      <div class="hidden name"><code>modelling-taxa.Rmd</code></div>

    </div>

    
    
<p>This article will give an example of statistical modelling of the
abundances of individual taxa, and visual presentation of the results
using microViz <em>taxonomic association tree</em> plots.</p>
<div class="section level3">
<h3 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://david-barnett.github.io/microViz">microViz</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">microViz version 0.10.4 - Copyright (C) 2022 David Barnett</span></span></span>
<span><span class="co">#&gt; <span style="color: #BBBB00;">!</span> Website: https://david-barnett.github.io/microViz</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> Useful?  For citation details, run: `citation("microViz")`</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">✖</span> Silence? `suppressPackageStartupMessages(library(microViz))`</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/bryandmartin/corncob" class="external-link">corncob</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span></code></pre></div>
<p>First we’ll get some OTU abundance data from inflammatory bowel
disease patients and controls from the corncob package.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"ibd_phylo"</span><span class="op">)</span></span>
<span><span class="va">ibd_phylo</span></span>
<span><span class="co">#&gt; phyloseq-class experiment-level object</span></span>
<span><span class="co">#&gt; otu_table()   OTU Table:         [ 36349 taxa and 91 samples ]</span></span>
<span><span class="co">#&gt; sample_data() Sample Data:       [ 91 samples by 15 sample variables ]</span></span>
<span><span class="co">#&gt; tax_table()   Taxonomy Table:    [ 36349 taxa by 7 taxonomic ranks ]</span></span></code></pre></div>
<p>We’ll keep only the Ulcerative Colitis and Healthy Control samples,
to simplify the analyses for this example. We’ll also remove the Species
rank information, as most OTUs in this dataset are not assigned to a
species. We’ll also use <code>tax_fix</code> to fill any gaps where the
Genus is unknown, with the family name or whatever higher rank
classification is known.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">phylo</span> <span class="op">&lt;-</span> <span class="va">ibd_phylo</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/ps_filter.html">ps_filter</a></span><span class="op">(</span><span class="va">DiseaseState</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"UC"</span>, <span class="st">"nonIBD"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/tax_mutate.html">tax_mutate</a></span><span class="op">(</span>Species <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/tax_fix.html">tax_fix</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">phylo</span> </span>
<span><span class="co">#&gt; phyloseq-class experiment-level object</span></span>
<span><span class="co">#&gt; otu_table()   OTU Table:         [ 29304 taxa and 67 samples ]</span></span>
<span><span class="co">#&gt; sample_data() Sample Data:       [ 67 samples by 15 sample variables ]</span></span>
<span><span class="co">#&gt; tax_table()   Taxonomy Table:    [ 29304 taxa by 6 taxonomic ranks ]</span></span></code></pre></div>
<p>Let’s have a quick look at the sample data using the
<code>skimr</code> package.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">phylo</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/psExtra-accessors.html">samdat_tbl</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/across.html" class="external-link">across</a></span><span class="op">(</span><span class="fu">where</span><span class="op">(</span><span class="va">is.character</span><span class="op">)</span>, <span class="va">as.factor</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">skimr</span><span class="fu">::</span><span class="fu"><a href="https://docs.ropensci.org/skimr/reference/skim.html" class="external-link">skim</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in sorted_count(x): Variable contains value(s) of "" that have been</span></span>
<span><span class="co">#&gt; converted to "empty".</span></span></code></pre></div>
<table class="table">
<caption>Data summary</caption>
<tbody>
<tr class="odd">
<td align="left">Name</td>
<td align="left">Piped data</td>
</tr>
<tr class="even">
<td align="left">Number of rows</td>
<td align="left">67</td>
</tr>
<tr class="odd">
<td align="left">Number of columns</td>
<td align="left">16</td>
</tr>
<tr class="even">
<td align="left">_______________________</td>
<td align="left"></td>
</tr>
<tr class="odd">
<td align="left">Column type frequency:</td>
<td align="left"></td>
</tr>
<tr class="even">
<td align="left">factor</td>
<td align="left">15</td>
</tr>
<tr class="odd">
<td align="left">numeric</td>
<td align="left">1</td>
</tr>
<tr class="even">
<td align="left">________________________</td>
<td align="left"></td>
</tr>
<tr class="odd">
<td align="left">Group variables</td>
<td align="left">None</td>
</tr>
</tbody>
</table>
<p><strong>Variable type: factor</strong></p>
<table class="table">
<colgroup>
<col width="15%">
<col width="11%">
<col width="15%">
<col width="8%">
<col width="10%">
<col width="38%">
</colgroup>
<thead><tr class="header">
<th align="left">skim_variable</th>
<th align="right">n_missing</th>
<th align="right">complete_rate</th>
<th align="left">ordered</th>
<th align="right">n_unique</th>
<th align="left">top_counts</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">.sample_name</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="left">FALSE</td>
<td align="right">67</td>
<td align="left">003: 1, 004: 1, 005: 1, 009: 1</td>
</tr>
<tr class="even">
<td align="left">sample</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="left">FALSE</td>
<td align="right">67</td>
<td align="left">003: 1, 004: 1, 005: 1, 009: 1</td>
</tr>
<tr class="odd">
<td align="left">gender</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="left">FALSE</td>
<td align="right">2</td>
<td align="left">fem: 36, mal: 31</td>
</tr>
<tr class="even">
<td align="left">DiseaseState</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="left">FALSE</td>
<td align="right">2</td>
<td align="left">UC: 43, non: 24</td>
</tr>
<tr class="odd">
<td align="left">steroids</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="left">FALSE</td>
<td align="right">2</td>
<td align="left">nos: 42, ste: 25</td>
</tr>
<tr class="even">
<td align="left">imsp</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="left">FALSE</td>
<td align="right">2</td>
<td align="left">noi: 44, ims: 23</td>
</tr>
<tr class="odd">
<td align="left">abx</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="left">FALSE</td>
<td align="right">2</td>
<td align="left">noa: 53, abx: 14</td>
</tr>
<tr class="even">
<td align="left">mesalamine</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="left">FALSE</td>
<td align="right">2</td>
<td align="left">nom: 61, mes: 6</td>
</tr>
<tr class="odd">
<td align="left">ibd</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="left">FALSE</td>
<td align="right">2</td>
<td align="left">ibd: 43, non: 24</td>
</tr>
<tr class="even">
<td align="left">activity</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="left">FALSE</td>
<td align="right">5</td>
<td align="left">con: 24, mil: 15, ina: 11, mod: 9</td>
</tr>
<tr class="odd">
<td align="left">active</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="left">FALSE</td>
<td align="right">3</td>
<td align="left">act: 32, con: 24, ina: 11</td>
</tr>
<tr class="even">
<td align="left">race</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="left">FALSE</td>
<td align="right">5</td>
<td align="left">Whi: 54, Bla: 4, Oth: 4, nof: 3</td>
</tr>
<tr class="odd">
<td align="left">fhx</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="left">FALSE</td>
<td align="right">7</td>
<td align="left">nof: 39, fhx: 17, non: 7, lev: 1</td>
</tr>
<tr class="even">
<td align="left">imspLEVEL</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="left">FALSE</td>
<td align="right">7</td>
<td align="left">non: 21, sto: 11, lev: 9, lev: 9</td>
</tr>
<tr class="odd">
<td align="left">SampleType</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="left">FALSE</td>
<td align="right">2</td>
<td align="left">sto: 56, emp: 11</td>
</tr>
</tbody>
</table>
<p><strong>Variable type: numeric</strong></p>
<table class="table">
<colgroup>
<col width="18%">
<col width="13%">
<col width="18%">
<col width="8%">
<col width="6%">
<col width="4%">
<col width="5%">
<col width="5%">
<col width="5%">
<col width="6%">
<col width="8%">
</colgroup>
<thead><tr class="header">
<th align="left">skim_variable</th>
<th align="right">n_missing</th>
<th align="right">complete_rate</th>
<th align="right">mean</th>
<th align="right">sd</th>
<th align="right">p0</th>
<th align="right">p25</th>
<th align="right">p50</th>
<th align="right">p75</th>
<th align="right">p100</th>
<th align="left">hist</th>
</tr></thead>
<tbody><tr class="odd">
<td align="left">age</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">12.09</td>
<td align="right">4.81</td>
<td align="right">3</td>
<td align="right">8.5</td>
<td align="right">12</td>
<td align="right">15</td>
<td align="right">24</td>
<td align="left">▅▇▇▅▂</td>
</tr></tbody>
</table>
<p>Let’s make some sample data variables that are easier to use and
compare in the statistical modelling ahead. We will convert dichotomous
categorical variables into similar binary variables (values: 1 for true,
or 0 for false). We will also scale and center the numeric variable for
age.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">phylo</span> <span class="op">&lt;-</span> <span class="va">phylo</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/ps_mutate.html">ps_mutate</a></span><span class="op">(</span></span>
<span>    UC <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">DiseaseState</span> <span class="op">==</span> <span class="st">"UC"</span>, yes <span class="op">=</span> <span class="fl">1</span>, no <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>,</span>
<span>    female <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">gender</span> <span class="op">==</span> <span class="st">"female"</span>, yes <span class="op">=</span> <span class="fl">1</span>, no <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>,</span>
<span>    antibiotics <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">abx</span> <span class="op">==</span> <span class="st">"abx"</span>, yes <span class="op">=</span> <span class="fl">1</span>, no <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>,</span>
<span>    steroids <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">steroids</span> <span class="op">==</span> <span class="st">"steroids"</span>, yes <span class="op">=</span> <span class="fl">1</span>, no <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>,</span>
<span>    age_scaled <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/scale.html" class="external-link">scale</a></span><span class="op">(</span><span class="va">age</span>, center <span class="op">=</span> <span class="cn">TRUE</span>, scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="modelling-one-taxon-at-a-time">Modelling one taxon at a time<a class="anchor" aria-label="anchor" href="#modelling-one-taxon-at-a-time"></a>
</h3>
<div class="section level4">
<h4 id="tss-log2-linear-regression">TSS log2 linear regression<a class="anchor" aria-label="anchor" href="#tss-log2-linear-regression"></a>
</h4>
<p>We will start by creating a linear regression model for one genus,
<em>Bacteroides</em>. We will transform the count data by first making
it proportions, and then taking the logarithm, with base 2. This is what
MaAsLin2 does by default (except they call the compositional
transformation “Total Sum Scaling (TSS)”). We will replace the zeros
with half the minimum observed abundance proportion (of any taxon)
before log2 transformation.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">parabacteroides_lm</span> <span class="op">&lt;-</span> <span class="va">phylo</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/tax_fix.html">tax_fix</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/tax_transform.html">tax_transform</a></span><span class="op">(</span><span class="st">"compositional"</span>, rank <span class="op">=</span> <span class="st">"Genus"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/tax_transform.html">tax_transform</a></span><span class="op">(</span><span class="st">"log2"</span>, zero_replace <span class="op">=</span> <span class="st">"halfmin"</span>, chain <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/tax_model.html">tax_model</a></span><span class="op">(</span></span>
<span>    type <span class="op">=</span> <span class="st">"lm"</span>, rank <span class="op">=</span> <span class="st">"Genus"</span>, taxa <span class="op">=</span> <span class="st">"Parabacteroides"</span>, </span>
<span>    variables <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"UC"</span>, <span class="st">"female"</span>, <span class="st">"antibiotics"</span>, <span class="st">"steroids"</span>, <span class="st">"age_scaled"</span><span class="op">)</span>,</span>
<span>    return_psx <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>  <span class="op">)</span> </span>
<span><span class="co">#&gt; Modelling: Parabacteroides</span></span>
<span><span class="va">parabacteroides_lm</span><span class="op">$</span><span class="va">Parabacteroides</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Call:</span></span>
<span><span class="co">#&gt; Parabacteroides ~ UC + female + antibiotics + steroids + age_scaled</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Coefficients:</span></span>
<span><span class="co">#&gt; (Intercept)           UC       female  antibiotics     steroids   age_scaled  </span></span>
<span><span class="co">#&gt;    -6.36997     -4.20062     -0.62692      0.07118     -0.95014      0.19781</span></span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">parabacteroides_lm</span><span class="op">$</span><span class="va">Parabacteroides</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Call:</span></span>
<span><span class="co">#&gt; Parabacteroides ~ UC + female + antibiotics + steroids + age_scaled</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Residuals:</span></span>
<span><span class="co">#&gt;    Min     1Q Median     3Q    Max </span></span>
<span><span class="co">#&gt; -7.344 -2.111 -1.308  2.297  7.492 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Coefficients:</span></span>
<span><span class="co">#&gt;             Estimate Std. Error t value Pr(&gt;|t|)    </span></span>
<span><span class="co">#&gt; (Intercept) -6.36997    0.95271  -6.686    8e-09 ***</span></span>
<span><span class="co">#&gt; UC          -4.20062    1.22791  -3.421  0.00112 ** </span></span>
<span><span class="co">#&gt; female      -0.62692    0.87225  -0.719  0.47505    </span></span>
<span><span class="co">#&gt; antibiotics  0.07118    1.18629   0.060  0.95235    </span></span>
<span><span class="co">#&gt; steroids    -0.95014    1.11976  -0.849  0.39946    </span></span>
<span><span class="co">#&gt; age_scaled   0.19781    0.49236   0.402  0.68926    </span></span>
<span><span class="co">#&gt; ---</span></span>
<span><span class="co">#&gt; Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Residual standard error: 3.452 on 61 degrees of freedom</span></span>
<span><span class="co">#&gt; Multiple R-squared:  0.3109, Adjusted R-squared:  0.2545 </span></span>
<span><span class="co">#&gt; F-statistic: 5.505 on 5 and 61 DF,  p-value: 0.0003018</span></span></code></pre></div>
<p>This model suggests <em>Parabacteroides</em> abundances are
significantly lower in Ulcerative Colitis patients than controls, on
average.</p>
</div>
<div class="section level4">
<h4 id="plotting-tss-log2-data">Plotting TSS log2 data<a class="anchor" aria-label="anchor" href="#plotting-tss-log2-data"></a>
</h4>
<p>Let’s boxplot the transformed data to see what this
<em>Parabacteroides</em> association looks like as a crude group
difference.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plot_data</span> <span class="op">&lt;-</span> <span class="va">phylo</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/tax_fix.html">tax_fix</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/tax_transform.html">tax_transform</a></span><span class="op">(</span><span class="st">"compositional"</span>, rank <span class="op">=</span> <span class="st">"Genus"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/tax_transform.html">tax_transform</a></span><span class="op">(</span><span class="st">"log2"</span>, zero_replace <span class="op">=</span> <span class="st">"halfmin"</span>, chain <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/psExtra-accessors.html">ps_get</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/ps_otu2samdat.html">ps_otu2samdat</a></span><span class="op">(</span><span class="st">"Parabacteroides"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> <span class="co"># adds Parabacteroides as sample data!</span></span>
<span>  <span class="fu"><a href="../../reference/psExtra-accessors.html">samdat_tbl</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">plot_data</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">DiseaseState</span>, y <span class="op">=</span> <span class="va">Parabacteroides</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">0.5</span>, colour <span class="op">=</span> <span class="st">"grey35"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html" class="external-link">geom_jitter</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">0.2</span>, alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span></span>
<span>    breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fl">1</span><span class="op">/</span><span class="fl">2</span><span class="op">^</span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="fl">13</span><span class="op">)</span><span class="op">)</span>, </span>
<span>    labels <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fl">100</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fl">2</span><span class="op">^</span><span class="va">x</span>, digits <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>, <span class="st">"%"</span><span class="op">)</span>, </span>
<span>    limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fl">0.00005</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fl">0.25</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="modelling-taxa_files/figure-html/unnamed-chunk-8-1.png" width="480"></p>
</div>
<div class="section level4">
<h4 id="beta-binomial-regression">Beta binomial regression<a class="anchor" aria-label="anchor" href="#beta-binomial-regression"></a>
</h4>
<p>You can also use other regression modelling functions that take a
formula. For example the beta binomial modelling provided in the corncob
package. This approach models both abundance and dispersion, and
directly uses untransformed counts. By default, microViz’s
<code><a href="../../reference/tax_model.html">tax_model()</a></code> will use the same formula for both abundance
and dispersion modelling, but you can override this by setting the
<code>phi.formula</code> argument yourself. See
<code>vignette("corncob-intro", package = "corncob")</code> for more
info on these models.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">parabacteroides_bb</span> <span class="op">&lt;-</span> <span class="va">phylo</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/tax_fix.html">tax_fix</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/tax_model.html">tax_model</a></span><span class="op">(</span></span>
<span>    type <span class="op">=</span> <span class="fu">corncob</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/corncob/man/bbdml.html" class="external-link">bbdml</a></span>, rank <span class="op">=</span> <span class="st">"Genus"</span>, taxa <span class="op">=</span> <span class="st">"Parabacteroides"</span>, </span>
<span>    variables <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"UC"</span>, <span class="st">"female"</span>, <span class="st">"antibiotics"</span>, <span class="st">"steroids"</span>, <span class="st">"age_scaled"</span><span class="op">)</span>,</span>
<span>    return_psx <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>  <span class="op">)</span> </span>
<span><span class="co">#&gt; Modelling: Parabacteroides</span></span>
<span><span class="va">parabacteroides_bb</span><span class="op">$</span><span class="va">Parabacteroides</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Call:</span></span>
<span><span class="co">#&gt; Parabacteroides ~ UC + female + antibiotics + steroids + age_scaled</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Coefficients associated with abundance:</span></span>
<span><span class="co">#&gt;             Estimate Std. Error t value Pr(&gt;|t|)    </span></span>
<span><span class="co">#&gt; (Intercept) -3.22672    0.67443  -4.784 1.33e-05 ***</span></span>
<span><span class="co">#&gt; UC          -1.99769    0.93236  -2.143   0.0366 *  </span></span>
<span><span class="co">#&gt; female      -0.20320    0.65852  -0.309   0.7588    </span></span>
<span><span class="co">#&gt; antibiotics -0.22137    1.56656  -0.141   0.8881    </span></span>
<span><span class="co">#&gt; steroids     0.41015    1.47997   0.277   0.7827    </span></span>
<span><span class="co">#&gt; age_scaled   0.04753    0.39951   0.119   0.9057    </span></span>
<span><span class="co">#&gt; ---</span></span>
<span><span class="co">#&gt; Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Coefficients associated with dispersion:</span></span>
<span><span class="co">#&gt;             Estimate Std. Error t value Pr(&gt;|t|)   </span></span>
<span><span class="co">#&gt; (Intercept) -2.54699    0.85001  -2.996  0.00409 **</span></span>
<span><span class="co">#&gt; UC          -0.93739    1.21342  -0.773  0.44311   </span></span>
<span><span class="co">#&gt; female      -0.12331    0.83238  -0.148  0.88277   </span></span>
<span><span class="co">#&gt; antibiotics -0.16097    1.97520  -0.081  0.93534   </span></span>
<span><span class="co">#&gt; steroids     1.57666    1.87989   0.839  0.40527   </span></span>
<span><span class="co">#&gt; age_scaled  -0.02351    0.52912  -0.044  0.96472   </span></span>
<span><span class="co">#&gt; ---</span></span>
<span><span class="co">#&gt; Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Log-likelihood: -211.25</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="model-all-the-taxa">Model all the taxa!<a class="anchor" aria-label="anchor" href="#model-all-the-taxa"></a>
</h3>
<p>Now we will fit a similar model for almost every taxon at every rank.
The code for <code>taxatree_models</code> is quite similar to
<code>tax_model</code>. However, you might need to run
<code>tax_prepend_ranks</code> to ensure that each taxon at each rank is
always unique. As an example of the problem, Actinobacteria is the name
of both a Phylum and a Class!</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lm_models</span> <span class="op">&lt;-</span> <span class="va">phylo</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/tax_fix.html">tax_fix</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/tax_prepend_ranks.html">tax_prepend_ranks</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="co"># it makes sense to perform the compositional transformation BEFORE filtering</span></span>
<span>  <span class="fu"><a href="../../reference/tax_transform.html">tax_transform</a></span><span class="op">(</span><span class="st">"compositional"</span>, rank <span class="op">=</span> <span class="st">"Genus"</span>, keep_counts <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/tax_filter.html">tax_filter</a></span><span class="op">(</span>min_prevalence <span class="op">=</span> <span class="fl">0.1</span>, undetected <span class="op">=</span> <span class="fl">0</span>, use_counts <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/tax_transform.html">tax_transform</a></span><span class="op">(</span></span>
<span>    trans <span class="op">=</span> <span class="st">"log2"</span>, chain <span class="op">=</span> <span class="cn">TRUE</span>, zero_replace <span class="op">=</span> <span class="st">"halfmin"</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/taxatree_models.html">taxatree_models</a></span><span class="op">(</span></span>
<span>    type <span class="op">=</span> <span class="va">lm</span>, </span>
<span>    ranks <span class="op">=</span> <span class="cn">NULL</span>, <span class="co"># uses every rank available except the first</span></span>
<span>    variables <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"UC"</span>, <span class="st">"female"</span>, <span class="st">"antibiotics"</span>, <span class="st">"steroids"</span>, <span class="st">"age_scaled"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="co">#&gt; Proportional min_prevalence given: 0.1 --&gt; min 7/67 samples.</span></span>
<span><span class="co">#&gt; 2022-11-18 20:04:07 - modelling at rank: Phylum</span></span>
<span><span class="co">#&gt; 2022-11-18 20:04:07 - modelling at rank: Class</span></span>
<span><span class="co">#&gt; 2022-11-18 20:04:08 - modelling at rank: Order</span></span>
<span><span class="co">#&gt; 2022-11-18 20:04:08 - modelling at rank: Family</span></span>
<span><span class="co">#&gt; 2022-11-18 20:04:09 - modelling at rank: Genus</span></span></code></pre></div>
<p>Why filter the taxa? <em>It’s less likely that we are interested in
rare taxa, and models of rare taxon abundances are more likely to be
unreliable. Reducing the the number of taxa modelled also makes
visualising the results easier!</em></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lm_models</span></span>
<span><span class="co">#&gt; psExtra object - a phyloseq object with extra slots:</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; phyloseq-class experiment-level object</span></span>
<span><span class="co">#&gt; otu_table()   OTU Table:         [ 68 taxa and 67 samples ]</span></span>
<span><span class="co">#&gt; sample_data() Sample Data:       [ 67 samples by 19 sample variables ]</span></span>
<span><span class="co">#&gt; tax_table()   Taxonomy Table:    [ 68 taxa by 6 taxonomic ranks ]</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; otu_get(counts = TRUE)        [ 68 taxa and 67 samples ]</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; psExtra info:</span></span>
<span><span class="co">#&gt; tax_agg = "Genus" tax_trans = "compositional&amp;log2" </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; taxatree_models list:</span></span>
<span><span class="co">#&gt; Ranks: Phylum/Class/Order/Family/Genus</span></span></code></pre></div>
<div class="section level4">
<h4 id="getting-stats-from-the-models">Getting stats from the models<a class="anchor" aria-label="anchor" href="#getting-stats-from-the-models"></a>
</h4>
<p>Next we will get a data.frame containing the regression coefficient
estimates, test statistics and corresponding p values from all these
regression models. The function <code><a href="../../reference/models2stats.html">taxatree_models2stats()</a></code> can
do this for any type of model that has a <code><a href="https://generics.r-lib.org/reference/tidy.html" class="external-link">broom::tidy()</a></code>
method, as well as for beta binomial regression models calculated with
the <code>corncob</code> package <code><a href="https://rdrr.io/pkg/corncob/man/bbdml.html" class="external-link">bbdml()</a></code> function.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lm_stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/models2stats.html">taxatree_models2stats</a></span><span class="op">(</span><span class="va">lm_models</span><span class="op">)</span></span>
<span><span class="va">lm_stats</span></span>
<span><span class="co">#&gt; psExtra object - a phyloseq object with extra slots:</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; phyloseq-class experiment-level object</span></span>
<span><span class="co">#&gt; otu_table()   OTU Table:         [ 68 taxa and 67 samples ]</span></span>
<span><span class="co">#&gt; sample_data() Sample Data:       [ 67 samples by 19 sample variables ]</span></span>
<span><span class="co">#&gt; tax_table()   Taxonomy Table:    [ 68 taxa by 6 taxonomic ranks ]</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; otu_get(counts = TRUE)        [ 68 taxa and 67 samples ]</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; psExtra info:</span></span>
<span><span class="co">#&gt; tax_agg = "Genus" tax_trans = "compositional&amp;log2" </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; taxatree_stats dataframe:</span></span>
<span><span class="co">#&gt; 134 taxa at 5 ranks: Phylum, Class, Order, Family, Genus </span></span>
<span><span class="co">#&gt; 5 terms: UC, female, antibiotics, steroids, age_scaled</span></span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lm_stats</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../../reference/psExtra-accessors.html">taxatree_stats_get</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 670 × 8</span></span></span>
<span><span class="co">#&gt;    term        taxon            rank   formula   estim…¹ std.e…² stati…³ p.value</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> UC          P: Firmicutes    Phylum `P: Firm… -<span style="color: #BB0000;">22.3</span>     13.6   -<span style="color: #BB0000;">1.64</span>  0.107  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> female      P: Firmicutes    Phylum `P: Firm… -<span style="color: #BB0000;">11.7</span>      9.66  -<span style="color: #BB0000;">1.21</span>  0.230  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> antibiotics P: Firmicutes    Phylum `P: Firm… -<span style="color: #BB0000;">19.8</span>     13.1   -<span style="color: #BB0000;">1.51</span>  0.137  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> steroids    P: Firmicutes    Phylum `P: Firm… -<span style="color: #BB0000;">32.7</span>     12.4   -<span style="color: #BB0000;">2.63</span>  0.010<span style="text-decoration: underline;">7</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> age_scaled  P: Firmicutes    Phylum `P: Firm…  -<span style="color: #BB0000;">0.600</span>    5.45  -<span style="color: #BB0000;">0.110</span> 0.913  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> UC          P: Bacteroidetes Phylum `P: Bact… -<span style="color: #BB0000;">13.8</span>      4.15  -<span style="color: #BB0000;">3.32</span>  0.001<span style="text-decoration: underline;">51</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> female      P: Bacteroidetes Phylum `P: Bact…  -<span style="color: #BB0000;">1.21</span>     2.95  -<span style="color: #BB0000;">0.410</span> 0.683  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> antibiotics P: Bacteroidetes Phylum `P: Bact…   1.37     4.01   0.342 0.734  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> steroids    P: Bacteroidetes Phylum `P: Bact…  -<span style="color: #BB0000;">4.36</span>     3.79  -<span style="color: #BB0000;">1.15</span>  0.255  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> age_scaled  P: Bacteroidetes Phylum `P: Bact…  -<span style="color: #BB0000;">0.937</span>    1.67  -<span style="color: #BB0000;">0.563</span> 0.576  </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># … with 660 more rows, and abbreviated variable names ¹​estimate, ²​std.error,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   ³​statistic</span></span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="adjusting-p-values">Adjusting p values<a class="anchor" aria-label="anchor" href="#adjusting-p-values"></a>
</h4>
<p>Using the <code><a href="../../reference/taxatree_stats_p_adjust.html">taxatree_stats_p_adjust()</a></code> function, you can
correct for multiple testing / control the false discovery rate or
family-wise error rate.</p>
<p>Instead of applying these adjustment methods across all 88 taxa
models at all ranks, the default behaviour is to control the family-wise
error rate per rank.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lm_stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/taxatree_stats_p_adjust.html">taxatree_stats_p_adjust</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="va">lm_stats</span>, method <span class="op">=</span> <span class="st">"BH"</span>, grouping <span class="op">=</span> <span class="st">"rank"</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># notice the new variable</span></span>
<span><span class="va">lm_stats</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../../reference/psExtra-accessors.html">taxatree_stats_get</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 670 × 9</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Groups:   rank [5]</span></span></span>
<span><span class="co">#&gt;    term        taxon       rank  formula estim…¹ std.e…² stati…³ p.value p.adj…⁴</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> UC          P: Firmicu… Phyl… `P: Fi… -<span style="color: #BB0000;">22.3</span>     13.6   -<span style="color: #BB0000;">1.64</span>  0.107    0.391 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> female      P: Firmicu… Phyl… `P: Fi… -<span style="color: #BB0000;">11.7</span>      9.66  -<span style="color: #BB0000;">1.21</span>  0.230    0.473 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> antibiotics P: Firmicu… Phyl… `P: Fi… -<span style="color: #BB0000;">19.8</span>     13.1   -<span style="color: #BB0000;">1.51</span>  0.137    0.399 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> steroids    P: Firmicu… Phyl… `P: Fi… -<span style="color: #BB0000;">32.7</span>     12.4   -<span style="color: #BB0000;">2.63</span>  0.010<span style="text-decoration: underline;">7</span>   0.125 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> age_scaled  P: Firmicu… Phyl… `P: Fi…  -<span style="color: #BB0000;">0.600</span>    5.45  -<span style="color: #BB0000;">0.110</span> 0.913    0.940 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> UC          P: Bactero… Phyl… `P: Ba… -<span style="color: #BB0000;">13.8</span>      4.15  -<span style="color: #BB0000;">3.32</span>  0.001<span style="text-decoration: underline;">51</span>  0.053<span style="text-decoration: underline;">0</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> female      P: Bactero… Phyl… `P: Ba…  -<span style="color: #BB0000;">1.21</span>     2.95  -<span style="color: #BB0000;">0.410</span> 0.683    0.828 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> antibiotics P: Bactero… Phyl… `P: Ba…   1.37     4.01   0.342 0.734    0.828 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> steroids    P: Bactero… Phyl… `P: Ba…  -<span style="color: #BB0000;">4.36</span>     3.79  -<span style="color: #BB0000;">1.15</span>  0.255    0.495 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> age_scaled  P: Bactero… Phyl… `P: Ba…  -<span style="color: #BB0000;">0.937</span>    1.67  -<span style="color: #BB0000;">0.563</span> 0.576    0.776 </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># … with 660 more rows, and abbreviated variable names ¹​estimate, ²​std.error,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   ³​statistic, ⁴​p.adj.BH.rank</span></span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="plot-all-the-taxatree_stats">Plot all the taxatree_stats!<a class="anchor" aria-label="anchor" href="#plot-all-the-taxatree_stats"></a>
</h3>
<p><code><a href="../../reference/taxatree_plots.html">taxatree_plots()</a></code> allows you to plot statistics
(e.g. point estimates and significance) from all of the taxa models onto
a tree layout. The taxon models are organised by rank, radiating out
from the central root node from e.g. Phyla around the center to Genera
in the outermost ring.</p>
<p><code><a href="../../reference/taxatree_plots.html">taxatree_plots()</a></code> itself returns a list of plots, which
you can arrange into one figure with the <a href="https://patchwork.data-imaginist.com/" class="external-link"><code>patchwork</code></a>
package for example (and/or <a href="https://wilkelab.org/cowplot/articles/plot_grid.html" class="external-link"><code>cowplot</code></a>).</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lm_stats</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../../reference/taxatree_plots.html">taxatree_plots</a></span><span class="op">(</span></span>
<span>  node_size_range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span>, var_renamer <span class="op">=</span> <span class="va">toupper</span></span>
<span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span></span>
<span>    ncol <span class="op">=</span> <span class="fl">2</span>, guides <span class="op">=</span> <span class="st">"collect"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="modelling-taxa_files/figure-html/unnamed-chunk-15-1.png" width="672"></p>
<div class="section level4">
<h4 id="taxatree-key">Taxatree Key<a class="anchor" aria-label="anchor" href="#taxatree-key"></a>
</h4>
<p>But how do we know which taxa are which nodes? We can create a
labelled grey tree with <code>taxatree_plotkey</code>. This labels taxa
based on certain conditions.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span> <span class="co"># label position </span></span>
<span><span class="va">key</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/taxatree_plotkey.html">taxatree_plotkey</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="va">lm_stats</span>, </span>
<span>  taxon_renamer <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_remove.html" class="external-link">str_remove</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">"[PFG]: "</span><span class="op">)</span>,</span>
<span>  <span class="co"># 2 lines of conditions below, for filtering taxa to be labelled</span></span>
<span>  <span class="va">rank</span> <span class="op">==</span> <span class="st">"Phylum"</span> <span class="op">|</span> <span class="va">rank</span> <span class="op">==</span> <span class="st">"Genus"</span> <span class="op">&amp;</span> <span class="va">prevalence</span> <span class="op">&gt;</span> <span class="fl">0.25</span>, </span>
<span>  <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"Kingdom"</span>, <span class="va">taxon</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="co"># add a bit more space for the longer labels by expanding the x axis</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/expansion.html" class="external-link">expansion</a></span><span class="op">(</span>mult <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># all phyla are labelled, and all genera with a prevalence of over 0.2</span></span>
<span><span class="co"># except for any taxa whose names (partly) match "Kingdom" </span></span>
<span><span class="co"># (i.e. an unclassified taxon)</span></span>
<span><span class="va">key</span></span></code></pre></div>
<p><img src="modelling-taxa_files/figure-html/unnamed-chunk-16-1.png" width="576"></p>
<div class="section level5">
<h5 id="key-trees">Key + Trees<a class="anchor" aria-label="anchor" href="#key-trees"></a>
</h5>
<p>Let’s put the key and some of the trees together in one
<code>patchwork</code> figure. Getting the sizing right on these
combined plots can be very tricky! Pay attention to the absolute height
and width of the plot output.</p>
<p><code><a href="https://rdrr.io/pkg/gridExtra/man/arrangeGrob.html" class="external-link">gridExtra::grid.arrange()</a></code> or
<code><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">cowplot::plot_grid()</a></code> are alternatives you can also try.
<code><a href="https://wilkelab.org/cowplot/reference/get_legend.html" class="external-link">cowplot::get_legend()</a></code> can be particularly useful.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">trees</span> <span class="op">&lt;-</span> <span class="va">lm_stats</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/taxatree_plots.html">taxatree_plots</a></span><span class="op">(</span>node_size_range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2.25</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="va">.</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span></span>
<span>    ncol <span class="op">=</span> <span class="fl">2</span>, guides <span class="op">=</span> <span class="st">"collect"</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="va">panel</span> <span class="op">&lt;-</span> <span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">key</span>, <span class="va">trees</span>, nrow <span class="op">=</span> <span class="fl">1</span>, widths <span class="op">=</span> <span class="fl">8</span><span class="op">:</span><span class="fl">7</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">111</span><span class="op">)</span></span>
<span><span class="va">panel</span></span></code></pre></div>
<p><img src="modelling-taxa_files/figure-html/unnamed-chunk-17-1.png" width="1560"></p>
<p>You could save the plot with <code><a href="https://ggplot2.tidyverse.org/reference/ggsave.html" class="external-link">ggsave()</a></code> like this.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">111</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggsave.html" class="external-link">ggsave</a></span><span class="op">(</span><span class="st">"test.png"</span>, <span class="va">panel</span>, width <span class="op">=</span> <span class="fl">13</span>, height <span class="op">=</span> <span class="fl">5.5</span>, dpi <span class="op">=</span> <span class="fl">120</span>, device <span class="op">=</span> <span class="st">"png"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="alternative-label-styling">Alternative label styling<a class="anchor" aria-label="anchor" href="#alternative-label-styling"></a>
</h5>
<p>You can change the default styling of the labels by first suppressing
the automatic drawing of labels with <code>.draw_label = FALSE</code> in
<code><a href="../../reference/taxatree_plotkey.html">taxatree_plotkey()</a></code> and then adding your own custom-style
labels with <code><a href="../../reference/taxatree_plot_labels.html">taxatree_plot_labels()</a></code>. Here we will draw some
yellow labels.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../../reference/taxatree_plotkey.html">taxatree_plotkey</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="va">lm_stats</span>, .draw_label <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">rank</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Phylum"</span>, <span class="st">"Family"</span><span class="op">)</span> <span class="op">&amp;</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"Bacteria"</span>, <span class="va">taxon</span><span class="op">)</span>,</span>
<span>  <span class="va">prevalence</span> <span class="op">&gt;</span> <span class="fl">0.2</span> <span class="op">|</span> <span class="va">rank</span> <span class="op">==</span> <span class="st">"Phylum"</span></span>
<span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/taxatree_plot_labels.html">taxatree_plot_labels</a></span><span class="op">(</span></span>
<span>    taxon_renamer <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_remove.html" class="external-link">str_remove</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">"[PFGO]: "</span><span class="op">)</span>,</span>
<span>    <span class="co"># default fun is ggrepel::geom_text_repel</span></span>
<span>    fun <span class="op">=</span> <span class="fu">ggrepel</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/ggrepel/man/geom_text_repel.html" class="external-link">geom_label_repel</a></span>, </span>
<span>    <span class="co"># arguments modifying label style</span></span>
<span>    size <span class="op">=</span> <span class="fl">2.5</span>, alpha <span class="op">=</span> <span class="fl">0.5</span>, colour <span class="op">=</span> <span class="st">"black"</span>, fill <span class="op">=</span> <span class="st">"gold1"</span>, </span>
<span>    label.size <span class="op">=</span> <span class="fl">0.05</span>, label.r <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">0.05</span>, <span class="st">"lines"</span><span class="op">)</span>, </span>
<span>    label.padding <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">0.15</span>, <span class="st">"lines"</span><span class="op">)</span>, segment.size <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>    <span class="co"># arguments affecting label positioning</span></span>
<span>    box.padding <span class="op">=</span> <span class="fl">0.05</span>, x_nudge <span class="op">=</span> <span class="fl">0.4</span>, y_nudge <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>    hjust <span class="op">=</span> <span class="fl">0.5</span>, seed <span class="op">=</span> <span class="fl">123</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="modelling-taxa_files/figure-html/unnamed-chunk-19-1.png" width="672"></p>
</div>
</div>
<div class="section level4">
<h4 id="directly-labelling-taxa">Directly labelling taxa<a class="anchor" aria-label="anchor" href="#directly-labelling-taxa"></a>
</h4>
<p>You can directly label taxatree_plots too, but it is better to only
do this for a few taxa. You must run <code><a href="../../reference/taxatree_label.html">taxatree_label()</a></code> first
to create a “label” indicator variable.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lm_stats</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/taxatree_label.html">taxatree_label</a></span><span class="op">(</span></span>
<span>    <span class="va">rank</span> <span class="op">==</span> <span class="st">"Genus"</span>, <span class="va">p.value</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">|</span> <span class="va">prevalence</span> <span class="op">&gt;</span> <span class="fl">0.5</span>, <span class="va">estimate</span> <span class="op">&gt;</span> <span class="fl">0</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/taxatree_plots.html">taxatree_plots</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="va">.</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> <span class="co"># show only the first plot</span></span>
<span>  <span class="fu"><a href="../../reference/taxatree_plot_labels.html">taxatree_plot_labels</a></span><span class="op">(</span></span>
<span>    taxon_renamer <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_remove.html" class="external-link">str_remove</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">"G: "</span><span class="op">)</span>,</span>
<span>    fun <span class="op">=</span> <span class="fu">ggrepel</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/ggrepel/man/geom_text_repel.html" class="external-link">geom_label_repel</a></span>, x_nudge <span class="op">=</span> <span class="fl">0.7</span>, hjust <span class="op">=</span> <span class="fl">0.5</span>, size <span class="op">=</span> <span class="fl">2</span></span>
<span>  <span class="op">)</span> </span></code></pre></div>
<p><img src="modelling-taxa_files/figure-html/unnamed-chunk-20-1.png" width="624"></p>
</div>
<div class="section level4">
<h4 id="changing-color-palette">Changing color palette<a class="anchor" aria-label="anchor" href="#changing-color-palette"></a>
</h4>
<p>Choosing another color palette is easy, just name any diverging
palette from colorspace hcl diverging palettes. See your options
below.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">colorspace</span><span class="fu">::</span><span class="fu"><a href="https://colorspace.R-Forge.R-project.org/reference/hcl_palettes.html" class="external-link">hcl_palettes</a></span><span class="op">(</span>type <span class="op">=</span> <span class="st">"diverging"</span>, plot <span class="op">=</span> <span class="cn">TRUE</span>, n <span class="op">=</span> <span class="fl">11</span><span class="op">)</span></span></code></pre></div>
<p><img src="modelling-taxa_files/figure-html/unnamed-chunk-21-1.png" width="288"></p>
<p>By default, the colour scale is transformed by taking the square root
of the absolute values. But you can change this to “identity” to have no
palette transformation.</p>
<p>By default the range of the data is used to set symmetrical limits on
the colour scale, which are the same for all plots in the list produced.
You can set alternative limits. If some data lie outside these limits,
their values will be “squished” to match the max or min limit value.</p>
<p>For finer control of the palette luminance range, you can set custom
values for l1 and l2, e.g. if the extremes are too bright or dark. This
is done by default for the Green-Brown palette.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lm_stats</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/taxatree_label.html">taxatree_label</a></span><span class="op">(</span></span>
<span>    <span class="va">rank</span> <span class="op">==</span> <span class="st">"Genus"</span>, <span class="va">p.value</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">|</span> <span class="va">prevalence</span> <span class="op">&gt;</span> <span class="fl">0.5</span>, <span class="va">estimate</span> <span class="op">&gt;</span> <span class="fl">0</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/taxatree_plots.html">taxatree_plots</a></span><span class="op">(</span></span>
<span>    colour_lims <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">20</span>, <span class="fl">20</span><span class="op">)</span>, colour_trans <span class="op">=</span> <span class="st">"identity"</span>,</span>
<span>    palette <span class="op">=</span> <span class="st">"Blue-Red 3"</span>, l2 <span class="op">=</span> <span class="fl">90</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="va">.</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> <span class="co"># show only the first plot</span></span>
<span>  <span class="fu"><a href="../../reference/taxatree_plot_labels.html">taxatree_plot_labels</a></span><span class="op">(</span></span>
<span>    taxon_renamer <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_remove.html" class="external-link">str_remove</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">"G: "</span><span class="op">)</span>,</span>
<span>    fun <span class="op">=</span> <span class="fu">ggrepel</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/ggrepel/man/geom_text_repel.html" class="external-link">geom_label_repel</a></span>, x_nudge <span class="op">=</span> <span class="fl">0.7</span>, hjust <span class="op">=</span> <span class="fl">0.5</span>, size <span class="op">=</span> <span class="fl">2</span></span>
<span>  <span class="op">)</span> </span></code></pre></div>
<p><img src="modelling-taxa_files/figure-html/unnamed-chunk-22-1.png" width="624"></p>
<p>Palettes like “Berlin” that go through a black midpoint would
probably only make sense with a darker background!</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lm_stats</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/taxatree_label.html">taxatree_label</a></span><span class="op">(</span></span>
<span>    <span class="va">rank</span> <span class="op">==</span> <span class="st">"Genus"</span>, <span class="va">p.value</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">|</span> <span class="va">prevalence</span> <span class="op">&gt;</span> <span class="fl">0.5</span>, <span class="va">estimate</span> <span class="op">&gt;</span> <span class="fl">0</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/taxatree_plots.html">taxatree_plots</a></span><span class="op">(</span></span>
<span>    palette <span class="op">=</span> <span class="st">"Berlin"</span>, colour_lims <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">20</span>, <span class="fl">20</span><span class="op">)</span>, size_guide <span class="op">=</span> <span class="cn">NULL</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="va">.</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> <span class="co"># show only the first plot</span></span>
<span>  <span class="fu"><a href="../../reference/taxatree_plot_labels.html">taxatree_plot_labels</a></span><span class="op">(</span></span>
<span>    taxon_renamer <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_remove.html" class="external-link">str_remove</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">"G: "</span><span class="op">)</span>,</span>
<span>    fun <span class="op">=</span> <span class="fu">ggrepel</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/ggrepel/man/geom_text_repel.html" class="external-link">geom_label_repel</a></span>, </span>
<span>    x_nudge <span class="op">=</span> <span class="fl">0.7</span>, xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1.7</span>, <span class="fl">1.5</span><span class="op">)</span>, hjust <span class="op">=</span> <span class="fl">0.5</span>, size <span class="op">=</span> <span class="fl">2</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>    text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="st">"white"</span><span class="op">)</span>,</span>
<span>    plot.background <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_rect</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"grey30"</span><span class="op">)</span>, </span>
<span>    plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">20</span>, colour <span class="op">=</span> <span class="st">"white"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="modelling-taxa_files/figure-html/unnamed-chunk-23-1.png" width="624"></p>
</div>
<div class="section level4">
<h4 id="sorting-taxa-nodes">Sorting taxa nodes<a class="anchor" aria-label="anchor" href="#sorting-taxa-nodes"></a>
</h4>
<p>If you like, you can sort the nodes by sorting the taxa in the
ps_extra object.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lm_stats</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/tax_sort.html">tax_sort</a></span><span class="op">(</span>by <span class="op">=</span> <span class="st">"prev"</span>, at <span class="op">=</span> <span class="st">"Genus"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/taxatree_plots.html">taxatree_plots</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="va">.</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="co"># show only the first plot</span></span></code></pre></div>
<p><img src="modelling-taxa_files/figure-html/unnamed-chunk-24-1.png" width="480"></p>
<p>You can chain multiple <code><a href="../../reference/tax_sort.html">tax_sort()</a></code> calls together to
fine-tune the order of the nodes on the tree to your own preference.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lm_stats</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/tax_sort.html">tax_sort</a></span><span class="op">(</span>by <span class="op">=</span> <span class="st">"prev"</span>, at <span class="op">=</span> <span class="st">"Family"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/tax_sort.html">tax_sort</a></span><span class="op">(</span>by <span class="op">=</span> <span class="st">"name"</span>, at <span class="op">=</span> <span class="st">"Phylum"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/tax_sort.html">tax_sort</a></span><span class="op">(</span>by <span class="op">=</span> <span class="st">"rev"</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../../reference/taxatree_plots.html">taxatree_plots</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="va">.</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="co"># show only the first plot</span></span></code></pre></div>
<p><img src="modelling-taxa_files/figure-html/unnamed-chunk-25-1.png" width="480"></p>
</div>
<div class="section level4">
<h4 id="plotting-adjusted-p-values">Plotting adjusted p values<a class="anchor" aria-label="anchor" href="#plotting-adjusted-p-values"></a>
</h4>
<p>Remember we made adjusted p values earlier? Let’s plot those instead.
Just to show how it’s done, we’ll also change the symbol used to
identify the significant sites to a cross, and we’ll also relax the
significance threshold to 0.1.</p>
<p>It looks like only the disease state (having ulcerative colitis)
shows any significant associations after this FDR correction.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lm_stats</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/taxatree_plots.html">taxatree_plots</a></span><span class="op">(</span></span>
<span>    sig_stat <span class="op">=</span> <span class="st">"p.adj.BH.rank"</span>, sig_threshold <span class="op">=</span> <span class="fl">0.1</span>, </span>
<span>    sig_shape <span class="op">=</span> <span class="st">"cross"</span>, sig_size <span class="op">=</span> <span class="fl">1.5</span>,</span>
<span>    node_size_range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span>, var_renamer <span class="op">=</span> <span class="va">toupper</span></span>
<span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="va">.</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> <span class="co"># keep only first 4 plots</span></span>
<span>  <span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span></span>
<span>    ncol <span class="op">=</span> <span class="fl">2</span>, guides <span class="op">=</span> <span class="st">"collect"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="modelling-taxa_files/figure-html/unnamed-chunk-26-1.png" width="672"></p>
</div>
<div class="section level4">
<h4 id="plotting-multiple-significance-markers">Plotting multiple significance markers<a class="anchor" aria-label="anchor" href="#plotting-multiple-significance-markers"></a>
</h4>
<p>You can also plot multiple significance markers. You must start with
the strictest threshold. Here we will plot the FDR corrected
significance markers at for p.adj &lt; 0.05 (as thick white crosses) and
then also unadjusted significance markers for p &lt; 0.05 (as outlined
white circles).</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lm_stats</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/taxatree_plots.html">taxatree_plots</a></span><span class="op">(</span></span>
<span>    sig_stat <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"p.adj.BH.rank"</span>, <span class="st">"p.value"</span><span class="op">)</span>, sig_threshold <span class="op">=</span> <span class="fl">0.05</span>, </span>
<span>    sig_shape <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cross"</span>, <span class="st">"circle filled"</span><span class="op">)</span>, sig_colour <span class="op">=</span> <span class="st">"white"</span>,</span>
<span>    sig_size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1.5</span>, <span class="fl">1</span><span class="op">)</span>, sig_stroke <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0.25</span><span class="op">)</span>,</span>
<span>    node_size_range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span>, var_renamer <span class="op">=</span> <span class="va">toupper</span></span>
<span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="va">.</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> <span class="co"># keep only first 4 plots</span></span>
<span>  <span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span></span>
<span>    ncol <span class="op">=</span> <span class="fl">2</span>, guides <span class="op">=</span> <span class="st">"collect"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="modelling-taxa_files/figure-html/unnamed-chunk-27-1.png" width="672"></p>
</div>
</div>
<div class="section level3">
<h3 id="beta-binomial-regression-example">Beta-binomial regression example<a class="anchor" aria-label="anchor" href="#beta-binomial-regression-example"></a>
</h3>
<p>The corncob package provides beta-binomial regression models. See the
paper <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7514055/" class="external-link">here</a>,
and the helpful package vignette:
<code>vignette("corncob-intro", package = "corncob")</code>.</p>
<p>We will filter the taxa more strictly (by a higher prevalence
threshold) before this type of modelling. We do not need to transform
the data, as this approach uses counts.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bb_models</span> <span class="op">&lt;-</span> <span class="va">phylo</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/tax_fix.html">tax_fix</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/tax_prepend_ranks.html">tax_prepend_ranks</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/tax_filter.html">tax_filter</a></span><span class="op">(</span>min_prevalence <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/taxatree_models.html">taxatree_models</a></span><span class="op">(</span></span>
<span>    type <span class="op">=</span> <span class="fu">corncob</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/corncob/man/bbdml.html" class="external-link">bbdml</a></span>, </span>
<span>    ranks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Phylum"</span>, <span class="st">"Class"</span>, <span class="st">"Order"</span>, <span class="st">"Family"</span><span class="op">)</span>,</span>
<span>    variables <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"UC"</span>, <span class="st">"female"</span>, <span class="st">"antibiotics"</span>, <span class="st">"steroids"</span>, <span class="st">"age_scaled"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="co">#&gt; Proportional min_prevalence given: 0.3 --&gt; min 21/67 samples.</span></span>
<span><span class="co">#&gt; 2022-11-18 20:04:42 - modelling at rank: Phylum</span></span>
<span><span class="co">#&gt; 2022-11-18 20:04:42 - modelling at rank: Class</span></span>
<span><span class="co">#&gt; 2022-11-18 20:04:43 - modelling at rank: Order</span></span>
<span><span class="co">#&gt; 2022-11-18 20:04:43 - modelling at rank: Family</span></span>
<span><span class="va">bb_models</span></span>
<span><span class="co">#&gt; psExtra object - a phyloseq object with extra slots:</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; phyloseq-class experiment-level object</span></span>
<span><span class="co">#&gt; otu_table()   OTU Table:         [ 105 taxa and 67 samples ]</span></span>
<span><span class="co">#&gt; sample_data() Sample Data:       [ 67 samples by 19 sample variables ]</span></span>
<span><span class="co">#&gt; tax_table()   Taxonomy Table:    [ 105 taxa by 6 taxonomic ranks ]</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; taxatree_models list:</span></span>
<span><span class="co">#&gt; Ranks: Phylum/Class/Order/Family</span></span></code></pre></div>
<p>When extracting stats from corncob beta-binomial models, you need to
specify which parameter estimate you want, “mu” for differential
abundance, or “phi” for differential variability or overdispersion.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bb_stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/models2stats.html">taxatree_models2stats</a></span><span class="op">(</span><span class="va">bb_models</span>, param <span class="op">=</span> <span class="st">"mu"</span><span class="op">)</span></span>
<span><span class="va">bb_stats</span></span>
<span><span class="co">#&gt; psExtra object - a phyloseq object with extra slots:</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; phyloseq-class experiment-level object</span></span>
<span><span class="co">#&gt; otu_table()   OTU Table:         [ 105 taxa and 67 samples ]</span></span>
<span><span class="co">#&gt; sample_data() Sample Data:       [ 67 samples by 19 sample variables ]</span></span>
<span><span class="co">#&gt; tax_table()   Taxonomy Table:    [ 105 taxa by 6 taxonomic ranks ]</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; taxatree_stats dataframe:</span></span>
<span><span class="co">#&gt; 30 taxa at 4 ranks: Phylum, Class, Order, Family </span></span>
<span><span class="co">#&gt; 5 terms: UC, female, antibiotics, steroids, age_scaled</span></span>
<span><span class="va">bb_stats</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../../reference/psExtra-accessors.html">taxatree_stats_get</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 150 × 9</span></span></span>
<span><span class="co">#&gt;    term        taxon       rank  formula param…¹ estim…² std.e…³ t.sta…⁴ p.value</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> UC          P: Firmicu… Phyl… `P: Fi… mu      -<span style="color: #BB0000;">0.375</span>    0.434 -<span style="color: #BB0000;">0.864</span>    0.392</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> female      P: Firmicu… Phyl… `P: Fi… mu       0.022<span style="text-decoration: underline;">7</span>   0.309  0.073<span style="text-decoration: underline;">4</span>   0.942</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> antibiotics P: Firmicu… Phyl… `P: Fi… mu      -<span style="color: #BB0000;">0.781</span>    0.478 -<span style="color: #BB0000;">1.63</span>     0.108</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> steroids    P: Firmicu… Phyl… `P: Fi… mu      -<span style="color: #BB0000;">0.162</span>    0.426 -<span style="color: #BB0000;">0.380</span>    0.705</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> age_scaled  P: Firmicu… Phyl… `P: Fi… mu       0.294    0.188  1.56     0.124</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> UC          P: Bactero… Phyl… `P: Ba… mu      -<span style="color: #BB0000;">0.814</span>    0.558 -<span style="color: #BB0000;">1.46</span>     0.150</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> female      P: Bactero… Phyl… `P: Ba… mu      -<span style="color: #BB0000;">0.559</span>    0.349 -<span style="color: #BB0000;">1.60</span>     0.115</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> antibiotics P: Bactero… Phyl… `P: Ba… mu      -<span style="color: #BB0000;">0.116</span>    0.706 -<span style="color: #BB0000;">0.164</span>    0.871</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> steroids    P: Bactero… Phyl… `P: Ba… mu      -<span style="color: #BB0000;">0.130</span>    0.656 -<span style="color: #BB0000;">0.197</span>    0.844</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> age_scaled  P: Bactero… Phyl… `P: Ba… mu      -<span style="color: #BB0000;">0.096</span><span style="color: #BB0000; text-decoration: underline;">7</span>   0.264 -<span style="color: #BB0000;">0.366</span>    0.716</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># … with 140 more rows, and abbreviated variable names ¹​parameter, ²​estimate,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   ³​std.error, ⁴​t.statistic</span></span></span></code></pre></div>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bb_stats</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../../reference/taxatree_plots.html">taxatree_plots</a></span><span class="op">(</span></span>
<span>  node_size_range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">4</span><span class="op">)</span>, colour_trans <span class="op">=</span> <span class="st">"identity"</span></span>
<span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="co"># keep only first 4 plots</span></span>
<span>  <span class="va">.</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span></span>
<span>    ncol <span class="op">=</span> <span class="fl">2</span>, guides <span class="op">=</span> <span class="st">"collect"</span></span>
<span>  <span class="op">)</span> </span></code></pre></div>
<p><img src="modelling-taxa_files/figure-html/unnamed-chunk-30-1.png" width="480"></p>
<div class="section level4">
<h4 id="alternative-layouts">Alternative layouts<a class="anchor" aria-label="anchor" href="#alternative-layouts"></a>
</h4>
<p>You do not need to make circular tree plots if you don’t want to!</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">alt_trees</span> <span class="op">&lt;-</span> <span class="va">bb_stats</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/taxatree_plots.html">taxatree_plots</a></span><span class="op">(</span></span>
<span>    node_size_range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">4</span><span class="op">)</span>, circular <span class="op">=</span> <span class="cn">FALSE</span>, colour_trans <span class="op">=</span> <span class="st">"identity"</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="co"># keep only first 4 plots</span></span>
<span>  <span class="va">.</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span></span>
<span>    ncol <span class="op">=</span> <span class="fl">2</span>, guides <span class="op">=</span> <span class="st">"collect"</span></span>
<span>  <span class="op">)</span> <span class="op">&amp;</span> <span class="co"># &amp; is used by patchwork to modify multiple ggplots (instead of +)</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">&amp;</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_reverse</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">alt_trees</span></span></code></pre></div>
<p><img src="modelling-taxa_files/figure-html/unnamed-chunk-31-1.png" width="480"></p>
<p>Let’s add the key for this layout and label it manually with
<code><a href="../../reference/taxatree_plot_labels.html">taxatree_plot_labels()</a></code>.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">alt_tree_key</span> <span class="op">&lt;-</span> <span class="va">bb_stats</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/taxatree_plotkey.html">taxatree_plotkey</a></span><span class="op">(</span>circular <span class="op">=</span> <span class="cn">FALSE</span>, .draw_label <span class="op">=</span> <span class="cn">FALSE</span>, <span class="va">rank</span> <span class="op">==</span> <span class="st">"Family"</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../../reference/taxatree_plot_labels.html">taxatree_plot_labels</a></span><span class="op">(</span></span>
<span>    circular <span class="op">=</span> <span class="cn">FALSE</span>, hjust <span class="op">=</span> <span class="fl">0.5</span>, force <span class="op">=</span> <span class="fl">0</span>, nudge_y <span class="op">=</span> <span class="fl">2</span>, size <span class="op">=</span> <span class="fl">3</span>,</span>
<span>    taxon_renamer <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_remove.html" class="external-link">str_remove</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">"[PFG]: "</span><span class="op">)</span></span>
<span>  <span class="op">)</span>  <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_reverse</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/expansion.html" class="external-link">expansion</a></span><span class="op">(</span>add <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">1.5</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">14</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">alt_tree_key</span>, <span class="va">alt_trees</span>, nrow <span class="op">=</span> <span class="fl">2</span>, heights <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="modelling-taxa_files/figure-html/unnamed-chunk-32-1.png" width="480"></p>
<p>You don’t have to use a regular tree!</p>
<p>Alternative layouts from the igraph package are possible, such as the
Kamada and Kawai spring algorithm (“kk”) or Fruchterman and Reingold
force-directed algorithm (“fr”). You must set a layout_seed number for
these layouts to ensure they are always the same.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bb_stats</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../../reference/taxatree_plots.html">taxatree_plots</a></span><span class="op">(</span></span>
<span>  node_size_range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">4</span><span class="op">)</span>, </span>
<span>  colour_trans <span class="op">=</span> <span class="st">"identity"</span>, layout <span class="op">=</span> <span class="st">"kk"</span>, layout_seed <span class="op">=</span> <span class="fl">321</span></span>
<span><span class="op">)</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="co"># keep only first 4 plots</span></span>
<span>  <span class="va">.</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span> <span class="op"><a href="../../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span></span>
<span>    ncol <span class="op">=</span> <span class="fl">2</span>, guides <span class="op">=</span> <span class="st">"collect"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="modelling-taxa_files/figure-html/unnamed-chunk-33-1.png" width="480"></p>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://r-lib.github.io/sessioninfo/reference/session_info.html" class="external-link">session_info</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB; font-weight: bold;">─ Session info ───────────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt;  <span style="color: #555555; font-style: italic;">setting </span> <span style="color: #555555; font-style: italic;">value</span></span></span>
<span><span class="co">#&gt;  version  R version 4.2.2 (2022-10-31)</span></span>
<span><span class="co">#&gt;  os       Ubuntu 20.04.5 LTS</span></span>
<span><span class="co">#&gt;  system   x86_64, linux-gnu</span></span>
<span><span class="co">#&gt;  ui       X11</span></span>
<span><span class="co">#&gt;  language en</span></span>
<span><span class="co">#&gt;  collate  C.UTF-8</span></span>
<span><span class="co">#&gt;  ctype    C.UTF-8</span></span>
<span><span class="co">#&gt;  tz       UTC</span></span>
<span><span class="co">#&gt;  date     2022-11-18</span></span>
<span><span class="co">#&gt;  pandoc   2.19.2 @ /usr/bin/ (via rmarkdown)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB; font-weight: bold;">─ Packages ───────────────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt;  <span style="color: #555555; font-style: italic;">package           </span> <span style="color: #555555; font-style: italic;">*</span> <span style="color: #555555; font-style: italic;">version     </span> <span style="color: #555555; font-style: italic;">date (UTC)</span> <span style="color: #555555; font-style: italic;">lib</span> <span style="color: #555555; font-style: italic;">source</span></span></span>
<span><span class="co">#&gt;  ade4                 1.7-20       <span style="color: #555555;">2022-11-01</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ape                  5.6-2        <span style="color: #555555;">2022-03-02</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  backports            1.4.1        <span style="color: #555555;">2021-12-13</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  base64enc            0.1-3        <span style="color: #555555;">2015-07-28</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  Biobase              2.56.0       <span style="color: #555555;">2022-04-26</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  BiocGenerics         0.42.0       <span style="color: #555555;">2022-04-26</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  biomformat           1.24.0       <span style="color: #555555;">2022-04-26</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  Biostrings           2.64.1       <span style="color: #555555;">2022-08-18</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  bitops               1.0-7        <span style="color: #555555;">2021-04-24</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  broom                1.0.1        <span style="color: #555555;">2022-08-29</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  bslib                0.4.1        <span style="color: #555555;">2022-11-02</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  cachem               1.0.6        <span style="color: #555555;">2021-08-19</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  callr                3.7.3        <span style="color: #555555;">2022-11-02</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  cli                  3.4.1        <span style="color: #555555;">2022-09-23</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  cluster              2.1.4        <span style="color: #555555;">2022-08-22</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.2.2)</span></span></span>
<span><span class="co">#&gt;  codetools            0.2-18       <span style="color: #555555;">2020-11-04</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.2.2)</span></span></span>
<span><span class="co">#&gt;  colorspace           2.0-3        <span style="color: #555555;">2022-02-21</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  corncob            * 0.3.0        <span style="color: #555555;">2022-09-23</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  crayon               1.5.2        <span style="color: #555555;">2022-09-29</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  data.table           1.14.6       <span style="color: #555555;">2022-11-16</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  desc                 1.4.2        <span style="color: #555555;">2022-09-08</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  detectseparation     0.3          <span style="color: #555555;">2022-08-26</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  devtools             2.4.5        <span style="color: #555555;">2022-10-11</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  digest               0.6.30       <span style="color: #555555;">2022-10-18</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  dplyr              * 1.0.10       <span style="color: #555555;">2022-09-01</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ellipsis             0.3.2        <span style="color: #555555;">2021-04-29</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  evaluate             0.18         <span style="color: #555555;">2022-11-07</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  fansi                1.0.3        <span style="color: #555555;">2022-03-24</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  farver               2.1.1        <span style="color: #555555;">2022-07-06</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  fastmap              1.1.0        <span style="color: #555555;">2021-01-25</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  foreach              1.5.2        <span style="color: #555555;">2022-02-02</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  fs                   1.5.2        <span style="color: #555555;">2021-12-08</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  generics             0.1.3        <span style="color: #555555;">2022-07-05</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  GenomeInfoDb         1.32.4       <span style="color: #555555;">2022-09-06</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  GenomeInfoDbData     1.2.8        <span style="color: #555555;">2022-11-14</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  ggforce              0.4.1        <span style="color: #555555;">2022-10-04</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ggplot2            * 3.4.0        <span style="color: #555555;">2022-11-04</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ggraph               2.1.0        <span style="color: #555555;">2022-10-09</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ggrepel              0.9.2        <span style="color: #555555;">2022-11-06</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  glue                 1.6.2        <span style="color: #555555;">2022-02-24</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  graphlayouts         0.8.3        <span style="color: #555555;">2022-10-20</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  gridExtra            2.3          <span style="color: #555555;">2017-09-09</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  gtable               0.3.1        <span style="color: #555555;">2022-09-01</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  highr                0.9          <span style="color: #555555;">2021-04-16</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  htmltools            0.5.3        <span style="color: #555555;">2022-07-18</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  htmlwidgets          1.5.4        <span style="color: #555555;">2021-09-08</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  httpuv               1.6.6        <span style="color: #555555;">2022-09-08</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  igraph               1.3.5        <span style="color: #555555;">2022-09-22</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  IRanges              2.30.1       <span style="color: #555555;">2022-08-18</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  iterators            1.0.14       <span style="color: #555555;">2022-02-05</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  jquerylib            0.1.4        <span style="color: #555555;">2021-04-26</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  jsonlite             1.8.3        <span style="color: #555555;">2022-10-21</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  knitr                1.41         <span style="color: #555555;">2022-11-18</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.2.2)</span></span></span>
<span><span class="co">#&gt;  labeling             0.4.2        <span style="color: #555555;">2020-10-20</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  later                1.3.0        <span style="color: #555555;">2021-08-18</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  lattice              0.20-45      <span style="color: #555555;">2021-09-22</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.2.2)</span></span></span>
<span><span class="co">#&gt;  lifecycle            1.0.3        <span style="color: #555555;">2022-10-07</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  lpSolveAPI           5.5.2.0-17.9 <span style="color: #555555;">2022-10-20</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  magrittr             2.0.3        <span style="color: #555555;">2022-03-30</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  MASS                 7.3-58.1     <span style="color: #555555;">2022-08-03</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.2.2)</span></span></span>
<span><span class="co">#&gt;  Matrix               1.5-1        <span style="color: #555555;">2022-09-13</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.2.2)</span></span></span>
<span><span class="co">#&gt;  memoise              2.0.1        <span style="color: #555555;">2021-11-26</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  mgcv                 1.8-41       <span style="color: #555555;">2022-10-21</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.2.2)</span></span></span>
<span><span class="co">#&gt;  microbiome           1.18.0       <span style="color: #555555;">2022-04-26</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  microViz           * 0.10.4       <span style="color: #555555;">2022-11-18</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">local</span></span></span>
<span><span class="co">#&gt;  mime                 0.12         <span style="color: #555555;">2021-09-28</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  miniUI               0.1.1.1      <span style="color: #555555;">2018-05-18</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  multtest             2.52.0       <span style="color: #555555;">2022-04-26</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  munsell              0.5.0        <span style="color: #555555;">2018-06-12</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  nlme                 3.1-160      <span style="color: #555555;">2022-10-10</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.2.2)</span></span></span>
<span><span class="co">#&gt;  numDeriv             <span style="color: #BB00BB; font-weight: bold;">2016.8-1.1  </span> <span style="color: #555555;">2019-06-06</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  patchwork            1.1.2        <span style="color: #555555;">2022-08-19</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  permute              0.9-7        <span style="color: #555555;">2022-01-27</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  phyloseq             1.40.0       <span style="color: #555555;">2022-04-26</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  pillar               1.8.1        <span style="color: #555555;">2022-08-19</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  pkgbuild             1.3.1        <span style="color: #555555;">2021-12-20</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  pkgconfig            2.0.3        <span style="color: #555555;">2019-09-22</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  pkgdown              2.0.6        <span style="color: #555555;">2022-07-16</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  pkgload              1.3.2        <span style="color: #555555;">2022-11-16</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  plyr                 1.8.8        <span style="color: #555555;">2022-11-11</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  polyclip             1.10-4       <span style="color: #555555;">2022-10-20</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  prettyunits          1.1.1        <span style="color: #555555;">2020-01-24</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  processx             3.8.0        <span style="color: #555555;">2022-10-26</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  profvis              0.3.7        <span style="color: #555555;">2020-11-02</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  promises             1.2.0.1      <span style="color: #555555;">2021-02-11</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ps                   1.7.2        <span style="color: #555555;">2022-10-26</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  purrr                0.3.5        <span style="color: #555555;">2022-10-06</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  R6                   2.5.1        <span style="color: #555555;">2021-08-19</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ragg                 1.2.4        <span style="color: #555555;">2022-10-24</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  Rcpp                 1.0.9        <span style="color: #555555;">2022-07-08</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  RCurl                1.98-1.9     <span style="color: #555555;">2022-10-03</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  registry             0.5-1        <span style="color: #555555;">2019-03-05</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  remotes              2.4.2        <span style="color: #555555;">2021-11-30</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  repr                 1.1.4        <span style="color: #555555;">2022-01-04</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  reshape2             1.4.4        <span style="color: #555555;">2020-04-09</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  rhdf5                2.40.0       <span style="color: #555555;">2022-04-26</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  rhdf5filters         1.8.0        <span style="color: #555555;">2022-04-26</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  Rhdf5lib             1.18.2       <span style="color: #555555;">2022-05-15</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  rlang                1.0.6        <span style="color: #555555;">2022-09-24</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  rmarkdown            2.18         <span style="color: #555555;">2022-11-09</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ROI                  1.0-0        <span style="color: #555555;">2020-08-31</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ROI.plugin.lpsolve   1.0-1        <span style="color: #555555;">2021-06-15</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  rprojroot            2.0.3        <span style="color: #555555;">2022-04-02</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  Rtsne                0.16         <span style="color: #555555;">2022-04-17</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  S4Vectors            0.34.0       <span style="color: #555555;">2022-04-26</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  sass                 0.4.2        <span style="color: #555555;">2022-07-16</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  scales               1.2.1        <span style="color: #555555;">2022-08-20</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  sessioninfo          1.2.2        <span style="color: #555555;">2021-12-06</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  shiny                1.7.3        <span style="color: #555555;">2022-10-25</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  skimr                2.1.4        <span style="color: #555555;">2022-04-15</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  slam                 0.1-50       <span style="color: #555555;">2022-01-08</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  stringi              1.7.8        <span style="color: #555555;">2022-07-11</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  stringr              1.4.1        <span style="color: #555555;">2022-08-20</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  survival             3.4-0        <span style="color: #555555;">2022-08-09</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.2.2)</span></span></span>
<span><span class="co">#&gt;  systemfonts          1.0.4        <span style="color: #555555;">2022-02-11</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  textshaping          0.3.6        <span style="color: #555555;">2021-10-13</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  tibble               3.1.8        <span style="color: #555555;">2022-07-22</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  tidygraph            1.2.2        <span style="color: #555555;">2022-08-22</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  tidyr                1.2.1        <span style="color: #555555;">2022-09-08</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  tidyselect           1.2.0        <span style="color: #555555;">2022-10-10</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  trust                0.1-8        <span style="color: #555555;">2020-01-10</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  tweenr               2.0.2        <span style="color: #555555;">2022-09-06</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  urlchecker           1.0.1        <span style="color: #555555;">2021-11-30</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  usethis              2.1.6        <span style="color: #555555;">2022-05-25</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  utf8                 1.2.2        <span style="color: #555555;">2021-07-24</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  vctrs                0.5.1        <span style="color: #555555;">2022-11-16</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  vegan                2.6-4        <span style="color: #555555;">2022-10-11</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  VGAM                 1.1-7        <span style="color: #555555;">2022-07-06</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  viridis              0.6.2        <span style="color: #555555;">2021-10-13</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  viridisLite          0.4.1        <span style="color: #555555;">2022-08-22</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  withr                2.5.0        <span style="color: #555555;">2022-03-03</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  xfun                 0.35         <span style="color: #555555;">2022-11-16</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  xtable               1.8-4        <span style="color: #555555;">2019-04-21</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  XVector              0.36.0       <span style="color: #555555;">2022-04-26</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  yaml                 2.3.6        <span style="color: #555555;">2022-10-18</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  zlibbioc             1.42.0       <span style="color: #555555;">2022-04-26</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #555555;"> [1] /home/runner/work/_temp/Library</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;"> [2] /opt/R/4.2.2/lib/R/site-library</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;"> [3] /opt/R/4.2.2/lib/R/library</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB; font-weight: bold;">──────────────────────────────────────────────────────────────────────────────</span></span></span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by David Barnett.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: '22e6f0a75b7b1ffae9bf5bc5b58864d7',
    indexName: 'microViz',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script><script data-goatcounter="https://microviz.goatcounter.com/count" async src="//gc.zgo.at/count.js"></script>
</body>
</html>
