<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Working with phyloseq objects • microViz</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../../bootstrap-toc.css">
<script src="../../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../../pkgdown.css" rel="stylesheet">
<script src="../../pkgdown.js"></script><!-- docsearch --><script src="../../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><meta property="og:title" content="Working with phyloseq objects">
<meta property="og:description" content="microViz">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-2YQ669S4V5"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-2YQ669S4V5');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../../index.html">microViz</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.8.0.9005</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../../articles/web-only/phyloseq.html">Working with phyloseq objects</a>
    </li>
    <li>
      <a href="../../articles/web-only/tax-fixing.html">Fixing your tax_table</a>
    </li>
    <li>
      <a href="../../articles/web-only/ordination.html">Ordination plots</a>
    </li>
    <li>
      <a href="../../articles/web-only/ordination-interactive.html">Interactive ordination plots</a>
    </li>
    <li>
      <a href="../../articles/web-only/compositions.html">Visualising microbiota compositions</a>
    </li>
    <li>
      <a href="../../articles/web-only/modelling-taxa.html">Statistical modelling of individual taxa</a>
    </li>
    <li>
      <a href="../../articles/web-only/videos.html">Tutorial video</a>
    </li>
    <li>
      <a href="../../articles/web-only/atlas1006.html">Example analyses with atlas1006 data</a>
    </li>
  </ul>
</li>
<li>
  <a href="../../news/index.html">News</a>
</li>
<li class="dropdown-header">
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/david-barnett/microViz/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="phyloseq_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Working with phyloseq objects</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/david-barnett/microViz/blob/master/vignettes/web-only/phyloseq.Rmd"><code>vignettes/web-only/phyloseq.Rmd</code></a></small>
      <div class="hidden name"><code>phyloseq.Rmd</code></div>

    </div>

    
    
<div id="intro" class="section level2">
<h2 class="hasAnchor">
<a href="#intro" class="anchor"></a>Intro</h2>
<p>This tutorial will show you how microViz makes working with phyloseq objects easier.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'dplyr'</span>
<span class="co">#&gt; The following objects are masked from 'package:stats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     filter, lag</span>
<span class="co">#&gt; The following objects are masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     intersect, setdiff, setequal, union</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://dx.plos.org/10.1371/journal.pone.0061217">phyloseq</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://david-barnett.github.io/microViz">microViz</a></span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; microViz version 0.8.0.9005 - Copyright (C) 2021 David Barnett</span>
<span class="co">#&gt; * Website: https://david-barnett.github.io/microViz/</span>
<span class="co">#&gt; * Useful? For citation info, run: citation('microViz')</span>
<span class="co">#&gt; * Silence: suppressPackageStartupMessages(library(microViz))</span></code></pre></div>
</div>
<div id="getting-your-data-into-phyloseq" class="section level2">
<h2 class="hasAnchor">
<a href="#getting-your-data-into-phyloseq" class="anchor"></a>Getting your data into phyloseq</h2>
<p>phyloseq objects are probably the most commonly used data format for working with microbiome data in R.</p>
<p>The creator of phyloseq, <a href="https://github.com/joey711">Paul J. McMurdie</a>, explains the structure of phyloseq objects and how to construct them on the <a href="https://joey711.github.io/phyloseq/import-data.html#phyloseq-ize_data_already_in_r">phyloseq website</a>.</p>
<ul>
<li><p><a href="http://biom-format.org/">biom format</a> files can be imported to phyloseq with the <a href="https://joey711.github.io/phyloseq/import-data.html#the_import_family_of_functions"><code>import_biom</code></a> function. Such biom files are generated (or can be) from many processing tools including QIIME <a href="http://qiime.org/tutorials/index.html">1</a> / <a href="https://docs.qiime2.org/2021.8/tutorials/exporting/">2</a>, <a href="http://segatalab.cibio.unitn.it/tools/metaphlan/">MetaPhlAn</a>, and <a href="https://doi.org/10.3389/fgene.2019.01366">NG-Tax</a>.</p></li>
<li><p><a href="https://mothur.org/">mothur</a> output is also directly supported via phyloseq’s <a href="https://joey711.github.io/phyloseq/import-data.html#_import_mothur"><code>import_mothur</code></a> function.</p></li>
<li><p><a href="https://benjjneb.github.io/dada2/">DADA2</a> output can converted to phyloseq according to these <a href="https://benjjneb.github.io/dada2/tutorial.html#bonus-handoff-to-phyloseq">DADA2 handoff instructions</a></p></li>
</ul>
<p>Don’t worry too much about getting all of your sample metadata into your biom file or phyloseq object at the start, as <code><a href="../../reference/ps_join.html">ps_join()</a></code> makes it easy to add sample data later.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># This tutorial will just use some example data </span>
<span class="co"># It is already available as a phyloseq object from the corncob package</span>
<span class="va">example_ps</span> <span class="op">&lt;-</span> <span class="fu">corncob</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/corncob/man/ibd_phylo.html">ibd_phylo</a></span>
<span class="va">example_ps</span>
<span class="co">#&gt; phyloseq-class experiment-level object</span>
<span class="co">#&gt; otu_table()   OTU Table:         [ 36349 taxa and 91 samples ]</span>
<span class="co">#&gt; sample_data() Sample Data:       [ 91 samples by 15 sample variables ]</span>
<span class="co">#&gt; tax_table()   Taxonomy Table:    [ 36349 taxa by 7 taxonomic ranks ]</span></code></pre></div>
<div id="validating-your-phyloseq" class="section level3">
<h3 class="hasAnchor">
<a href="#validating-your-phyloseq" class="anchor"></a>Validating your phyloseq</h3>
<p>phyloseq checks that your sample and taxa names are consistent across the different slots of the phyloseq object. microViz provides <code><a href="../../reference/phyloseq_validate.html">phyloseq_validate()</a></code> to check for and fix other possible problems with your phyloseq that might cause problems in later analyses. It is recommended to run this at the start of your analyses, and fix any problems identified.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">example_ps</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/phyloseq_validate.html">phyloseq_validate</a></span><span class="op">(</span><span class="va">example_ps</span>, remove_undetected <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">#&gt; Short values detected in phyloseq tax_table (nchar&lt;4) :</span>
<span class="co">#&gt; Consider using tax_fix() to make taxa uniquely identifiable</span>
<span class="va">example_ps</span> <span class="op">&lt;-</span> <span class="fu"><a href="../../reference/tax_fix.html">tax_fix</a></span><span class="op">(</span><span class="va">example_ps</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="microviz-and-phyloseq-overview" class="section level2">
<h2 class="hasAnchor">
<a href="#microviz-and-phyloseq-overview" class="anchor"></a>microViz and phyloseq overview</h2>
<p>Once you have a valid phyloseq object, microViz provides several helpful functions for manipulating that object. The names and syntax of some functions will be familiar to users of dplyr, and reading <a href="https://dplyr.tidyverse.org/">the dplyr help pages</a> may be useful for getting the most out of these functions.</p>
<ul>
<li><p>Add a data.frame of metadata to the sample_data slot with <code><a href="../../reference/ps_join.html">ps_join()</a></code></p></li>
<li><p>Compute new sample_data variables with <code><a href="../../reference/ps_mutate.html">ps_mutate()</a></code></p></li>
<li><p>Subset or reorder variables in sample_data with <code><a href="../../reference/ps_select.html">ps_select()</a></code></p></li>
<li><p>Subset samples based on sample_data variables with <code><a href="../../reference/ps_filter.html">ps_filter()</a></code></p></li>
<li>
<p>Reorder samples (can be useful for examining sample data or plotting) with:</p>
<ul>
<li><p><code><a href="../../reference/ps_reorder.html">ps_reorder()</a></code> : manually set sample order</p></li>
<li><p><code><a href="../../reference/ps_arrange.html">ps_arrange()</a></code> : order samples using sample_data variables</p></li>
<li><p><code><a href="../../reference/ps_seriate.html">ps_seriate()</a></code> : order samples according to microbiome similarity</p></li>
</ul>
</li>
<li><p>Remove duplicated/repeated samples with <code><a href="../../reference/ps_dedupe.html">ps_dedupe()</a></code></p></li>
<li><p>Remove samples with missing values in sample_data with <code><a href="../../reference/ps_drop_incomplete.html">ps_drop_incomplete()</a></code></p></li>
</ul>
<p>Check out the examples section on each function’s <a href="https://david-barnett.github.io/microViz/reference/index.html#section-manipulating-sample-data">reference</a> page for extensive usage examples.</p>
</div>
<div id="example-sample-manipulation" class="section level2">
<h2 class="hasAnchor">
<a href="#example-sample-manipulation" class="anchor"></a>Example sample manipulation</h2>
<p>Lets look at the sample data already in our example phyloseq.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># 91 samples, with 15 sample_data variables</span>
<span class="va">example_ps</span> 
<span class="co">#&gt; phyloseq-class experiment-level object</span>
<span class="co">#&gt; otu_table()   OTU Table:         [ 36349 taxa and 91 samples ]</span>
<span class="co">#&gt; sample_data() Sample Data:       [ 91 samples by 15 sample variables ]</span>
<span class="co">#&gt; tax_table()   Taxonomy Table:    [ 36349 taxa by 7 taxonomic ranks ]</span>
<span class="co"># return sample data as a tibble for pretty printing</span>
<span class="fu"><a href="../../reference/ps_extra-accessors.html">samdat_tbl</a></span><span class="op">(</span><span class="va">example_ps</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 91 × 16</span>
<span class="co">#&gt;    .sample_name sample gender   age DiseaseState steroids imsp  abx   mesalamine</span>
<span class="co">#&gt;    &lt;chr&gt;        &lt;chr&gt;  &lt;chr&gt;  &lt;int&gt; &lt;chr&gt;        &lt;chr&gt;    &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;     </span>
<span class="co">#&gt;  1 099A         099-AX female    17 UC           steroids imsp  noabx nomes     </span>
<span class="co">#&gt;  2 199A         199-AX female     5 nonIBD       nostero… noim… noabx nomes     </span>
<span class="co">#&gt;  3 062B         062-BZ male      16 CD           steroids imsp  noabx nomes     </span>
<span class="co">#&gt;  4 194A         194-AZ male      15 UC           steroids noim… noabx nomes     </span>
<span class="co">#&gt;  5 166A         166-AX female    20 UC           nostero… noim… abx   nomes     </span>
<span class="co">#&gt;  6 219A         219-AX female    18 UC           nostero… imsp  abx   nomes     </span>
<span class="co">#&gt;  7 132A         132-AX female    12 CD           steroids imsp  noabx nomes     </span>
<span class="co">#&gt;  8 026A         026-AX male      10 UC           steroids imsp  abx   nomes     </span>
<span class="co">#&gt;  9 102A         102-AZ male       3 nonIBD       nostero… noim… noabx nomes     </span>
<span class="co">#&gt; 10 140A         140-AX female     5 nonIBD       nostero… noim… noabx nomes     </span>
<span class="co">#&gt; # … with 81 more rows, and 7 more variables: ibd &lt;chr&gt;, activity &lt;chr&gt;,</span>
<span class="co">#&gt; #   active &lt;chr&gt;, race &lt;chr&gt;, fhx &lt;chr&gt;, imspLEVEL &lt;chr&gt;, SampleType &lt;chr&gt;</span></code></pre></div>
<p>Maybe you want to only select participants who have IBD (not controls). You can do that by filtering samples based on the values of the sample data variable: ibd</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">example_ps</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span> <span class="fu"><a href="../../reference/ps_filter.html">ps_filter</a></span><span class="op">(</span><span class="va">ibd</span> <span class="op">==</span> <span class="st">"ibd"</span><span class="op">)</span> <span class="co"># it is essential to use `==` , not just one `=`</span>
<span class="co">#&gt; phyloseq-class experiment-level object</span>
<span class="co">#&gt; otu_table()   OTU Table:         [ 24490 taxa and 67 samples ]</span>
<span class="co">#&gt; sample_data() Sample Data:       [ 67 samples by 15 sample variables ]</span>
<span class="co">#&gt; tax_table()   Taxonomy Table:    [ 24490 taxa by 7 taxonomic ranks ]</span>
<span class="co"># notice that taxa that no longer appear in the remaining 67 samples have been removed!</span></code></pre></div>
<p>More complicated filtering rules can be applied. Let’s say you want female IBD patients with “mild” or “severe” activity, who are at least 13 years old.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">partial_ps</span> <span class="op">&lt;-</span> <span class="va">example_ps</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span> 
  <span class="fu"><a href="../../reference/ps_filter.html">ps_filter</a></span><span class="op">(</span>
    <span class="va">gender</span> <span class="op">==</span> <span class="st">"female"</span>,
    <span class="va">activity</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"mild"</span>, <span class="st">"severe"</span><span class="op">)</span>,
    <span class="va">age</span> <span class="op">&gt;=</span> <span class="fl">13</span>
  <span class="op">)</span>
<span class="va">partial_ps</span>
<span class="co">#&gt; phyloseq-class experiment-level object</span>
<span class="co">#&gt; otu_table()   OTU Table:         [ 2600 taxa and 9 samples ]</span>
<span class="co">#&gt; sample_data() Sample Data:       [ 9 samples by 15 sample variables ]</span>
<span class="co">#&gt; tax_table()   Taxonomy Table:    [ 2600 taxa by 7 taxonomic ranks ]</span></code></pre></div>
<p>Let’s have a look at the sample data of these participants. We will also arrange the samples grouped by disease and in descending age order, and select only a few interesting variables to show.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">partial_ps</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span> 
  <span class="fu"><a href="../../reference/ps_arrange.html">ps_arrange</a></span><span class="op">(</span><span class="va">DiseaseState</span>, <span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html">desc</a></span><span class="op">(</span><span class="va">age</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span> 
  <span class="fu"><a href="../../reference/ps_select.html">ps_select</a></span><span class="op">(</span><span class="va">DiseaseState</span>, <span class="va">age</span>, <span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html">matches</a></span><span class="op">(</span><span class="st">"activ"</span><span class="op">)</span>, <span class="va">abx</span><span class="op">)</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span> <span class="co"># selection order is respected</span>
  <span class="fu"><a href="../../reference/ps_extra-accessors.html">samdat_tbl</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># this adds the .sample_name variable</span>
<span class="co">#&gt; # A tibble: 9 × 6</span>
<span class="co">#&gt;   .sample_name DiseaseState   age activity active abx  </span>
<span class="co">#&gt;   &lt;chr&gt;        &lt;chr&gt;        &lt;int&gt; &lt;chr&gt;    &lt;chr&gt;  &lt;chr&gt;</span>
<span class="co">#&gt; 1 195A         CD              13 mild     active noabx</span>
<span class="co">#&gt; 2 164A         UC              21 severe   active abx  </span>
<span class="co">#&gt; 3 166A         UC              20 severe   active abx  </span>
<span class="co">#&gt; 4 038A         UC              19 mild     active noabx</span>
<span class="co">#&gt; 5 114C         UC              19 mild     active noabx</span>
<span class="co">#&gt; 6 099A         UC              17 severe   active noabx</span>
<span class="co">#&gt; 7 009A         UC              15 severe   active noabx</span>
<span class="co">#&gt; 8 120D         UC              14 mild     active noabx</span>
<span class="co">#&gt; 9 186A         UC              14 severe   active abx</span></code></pre></div>
<p>You can also sort sample by microbiome similarity with <code><a href="../../reference/ps_seriate.html">ps_seriate()</a></code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">partial_ps</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span> 
  <span class="fu"><a href="../../reference/tax_agg.html">tax_agg</a></span><span class="op">(</span><span class="st">"Genus"</span><span class="op">)</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span> 
  <span class="fu"><a href="../../reference/ps_seriate.html">ps_seriate</a></span><span class="op">(</span>dist <span class="op">=</span> <span class="st">"bray"</span>, method <span class="op">=</span> <span class="st">"OLO_ward"</span><span class="op">)</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span> <span class="co"># these are the defaults</span>
  <span class="fu"><a href="../../reference/comp_barplot.html">comp_barplot</a></span><span class="op">(</span>tax_level <span class="op">=</span> <span class="st">"Genus"</span>, sample_order <span class="op">=</span> <span class="st">"default"</span>, n_taxa <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>
<span class="co">#&gt; Registered S3 method overwritten by 'seriation':</span>
<span class="co">#&gt;   method         from </span>
<span class="co">#&gt;   reorder.hclust vegan</span></code></pre></div>
<p><img src="phyloseq_files/figure-html/seriate%20and%20plot-1.png" width="700"></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># note that comp_barplot with sample_order = "bray" will run </span>
<span class="co"># this ps_seriate call internally, so you don't have to!</span></code></pre></div>
<p>You can also arrange samples by abundance of one of more microbes using <code><a href="../../reference/ps_arrange.html">ps_arrange()</a></code> with <code>.target = "otu_table"</code>. Arranging by taxon can only be done at the current taxonomic rank, so we will aggregate to Genus level first.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Arranging by decreasing Bacteroides abundance</span>
<span class="va">partial_ps</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span> 
  <span class="fu"><a href="../../reference/tax_agg.html">tax_agg</a></span><span class="op">(</span><span class="st">"Genus"</span><span class="op">)</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span> 
  <span class="fu"><a href="../../reference/ps_arrange.html">ps_arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html">desc</a></span><span class="op">(</span><span class="va">Bacteroides</span><span class="op">)</span>, .target <span class="op">=</span> <span class="st">"otu_table"</span><span class="op">)</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span> 
  <span class="fu"><a href="../../reference/ps_extra-accessors.html">otu_get</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span> <span class="co"># get the otu table</span>
  <span class="va">.</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span> <span class="co"># show only a subset of the otu_table</span>
<span class="co">#&gt; OTU Table:          [6 taxa and 9 samples]</span>
<span class="co">#&gt;                      taxa are columns</span>
<span class="co">#&gt;      Prevotella Clostridium_XlVa Hallella Veillonella Bacteroides Blautia</span>
<span class="co">#&gt; 009A       4480              462        4         250        1542       0</span>
<span class="co">#&gt; 195A          0               40        0           1         514       5</span>
<span class="co">#&gt; 166A        549                7        0          17         170      27</span>
<span class="co">#&gt; 099A          0               60        0           0         132      48</span>
<span class="co">#&gt; 186A          0                7        0           0          69      38</span>
<span class="co">#&gt; 038A          0                0        0         192           0       0</span>
<span class="co">#&gt; 114C          0               44        0           6           0     236</span>
<span class="co">#&gt; 120D          0                0        0           0           0       0</span>
<span class="co">#&gt; 164A          0                0        0           0           0       0</span>

<span class="co"># Plot samples' compositions in this order</span>
<span class="va">partial_ps</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span> 
  <span class="fu"><a href="../../reference/tax_agg.html">tax_agg</a></span><span class="op">(</span><span class="st">"Genus"</span><span class="op">)</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span> 
  <span class="fu"><a href="../../reference/ps_arrange.html">ps_arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html">desc</a></span><span class="op">(</span><span class="va">Bacteroides</span><span class="op">)</span>, .target <span class="op">=</span> <span class="st">"otu_table"</span><span class="op">)</span> <span class="op"><a href="../../reference/%&gt;%.html">%&gt;%</a></span> 
  <span class="fu"><a href="../../reference/comp_barplot.html">comp_barplot</a></span><span class="op">(</span>tax_level <span class="op">=</span> <span class="st">"Genus"</span>, sample_order <span class="op">=</span> <span class="st">"default"</span>, n_taxa <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></code></pre></div>
<p><img src="phyloseq_files/figure-html/arrange%20by%20taxon-1.png" width="700"></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Notice this is sorted by bacteroides counts </span>
<span class="co"># (this doesn't quite match relative abundance % due to sequencing depth variation)</span></code></pre></div>
</div>
<div id="other-notes" class="section level2">
<h2 class="hasAnchor">
<a href="#other-notes" class="anchor"></a>Other notes:</h2>
<div id="ps_filter-vs--phyloseqsubset_samples" class="section level3">
<h3 class="hasAnchor">
<a href="#ps_filter-vs--phyloseqsubset_samples" class="anchor"></a><code>ps_filter()</code> vs. <code>phyloseq::subset_samples()</code>
</h3>
<p>As well as filtering your samples, <code><a href="../../reference/ps_filter.html">ps_filter()</a></code> might also modify the otu_table and tax_table of the phyloseq object (unlike <code><a href="https://rdrr.io/pkg/phyloseq/man/subset_samples-methods.html">phyloseq::subset_samples()</a></code>, which never does this).</p>
<p>Why does it do this?<br>
If you remove many samples from your dataset, often your phyloseq object will be left with taxa that never occur in any of the remaining samples (i.e. total counts of zero). <code><a href="../../reference/ps_filter.html">ps_filter()</a></code> removes those absent taxa by default.</p>
<p>If you don’t want this, you can set the <code>.keep_all_taxa</code> argument to <code>TRUE</code> in <code>ps_filter</code>.</p>
</div>
</div>
<div id="technical-log" class="section level2">
<h2 class="hasAnchor">
<a href="#technical-log" class="anchor"></a>Technical log</h2>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://r-lib.github.io/sessioninfo/reference/session_info.html">session_info</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; ─ Session info  ──────────────────────────────────────────────────────────────</span>
<span class="co">#&gt;  hash: person juggling: medium-light skin tone, Santa Claus: dark skin tone, taxi</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;  setting  value</span>
<span class="co">#&gt;  version  R version 4.1.2 (2021-11-01)</span>
<span class="co">#&gt;  os       Ubuntu 20.04.3 LTS</span>
<span class="co">#&gt;  system   x86_64, linux-gnu</span>
<span class="co">#&gt;  ui       X11</span>
<span class="co">#&gt;  language (EN)</span>
<span class="co">#&gt;  collate  C.UTF-8</span>
<span class="co">#&gt;  ctype    C.UTF-8</span>
<span class="co">#&gt;  tz       UTC</span>
<span class="co">#&gt;  date     2021-11-19</span>
<span class="co">#&gt;  pandoc   2.7.3 @ /usr/bin/ (via rmarkdown)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; ─ Packages ───────────────────────────────────────────────────────────────────</span>
<span class="co">#&gt;  package          * version    date (UTC) lib source</span>
<span class="co">#&gt;  ade4               1.7-18     2021-09-16 [1] Custom</span>
<span class="co">#&gt;  ape                5.5        2021-04-25 [1] Custom</span>
<span class="co">#&gt;  Biobase            2.54.0     2021-10-26 [1] Bioconductor</span>
<span class="co">#&gt;  BiocGenerics       0.40.0     2021-10-26 [1] Bioconductor</span>
<span class="co">#&gt;  biomformat         1.22.0     2021-10-26 [1] Bioconductor</span>
<span class="co">#&gt;  Biostrings         2.62.0     2021-10-26 [1] Bioconductor</span>
<span class="co">#&gt;  bitops             1.0-7      2021-04-24 [1] Custom</span>
<span class="co">#&gt;  bslib              0.3.1      2021-10-06 [1] Custom</span>
<span class="co">#&gt;  cachem             1.0.6      2021-08-19 [1] Custom</span>
<span class="co">#&gt;  callr              3.7.0      2021-04-20 [1] Custom</span>
<span class="co">#&gt;  cli                3.1.0      2021-10-27 [1] Custom</span>
<span class="co">#&gt;  cluster            2.1.2      2021-04-17 [2] CRAN (R 4.1.2)</span>
<span class="co">#&gt;  codetools          0.2-18     2020-11-04 [2] CRAN (R 4.1.2)</span>
<span class="co">#&gt;  colorspace         2.0-2      2021-06-24 [1] Custom</span>
<span class="co">#&gt;  corncob            0.2.0      2021-03-11 [1] Custom</span>
<span class="co">#&gt;  crayon             1.4.2      2021-10-29 [1] Custom</span>
<span class="co">#&gt;  data.table         1.14.2     2021-09-27 [1] Custom</span>
<span class="co">#&gt;  desc               1.4.0      2021-09-28 [1] Custom</span>
<span class="co">#&gt;  devtools           2.4.2      2021-06-07 [1] Custom</span>
<span class="co">#&gt;  digest             0.6.28     2021-09-23 [1] Custom</span>
<span class="co">#&gt;  dplyr            * 1.0.7      2021-06-18 [1] Custom</span>
<span class="co">#&gt;  ellipsis           0.3.2      2021-04-29 [1] Custom</span>
<span class="co">#&gt;  evaluate           0.14       2019-05-28 [1] Custom</span>
<span class="co">#&gt;  fansi              0.5.0      2021-05-25 [1] Custom</span>
<span class="co">#&gt;  farver             2.1.0      2021-02-28 [1] Custom</span>
<span class="co">#&gt;  fastmap            1.1.0      2021-01-25 [1] Custom</span>
<span class="co">#&gt;  foreach            1.5.1      2020-10-15 [1] Custom</span>
<span class="co">#&gt;  fs                 1.5.0      2020-07-31 [1] Custom</span>
<span class="co">#&gt;  generics           0.1.1      2021-10-25 [1] Custom</span>
<span class="co">#&gt;  GenomeInfoDb       1.30.0     2021-10-26 [1] Bioconductor</span>
<span class="co">#&gt;  GenomeInfoDbData   1.2.7      2021-11-12 [1] Bioconductor</span>
<span class="co">#&gt;  ggplot2            3.3.5      2021-06-25 [1] Custom</span>
<span class="co">#&gt;  glue               1.5.0      2021-11-07 [1] Custom</span>
<span class="co">#&gt;  gtable             0.3.0      2019-03-25 [1] Custom</span>
<span class="co">#&gt;  highr              0.9        2021-04-16 [1] Custom</span>
<span class="co">#&gt;  htmltools          0.5.2      2021-08-25 [1] Custom</span>
<span class="co">#&gt;  igraph             1.2.8      2021-11-07 [1] Custom</span>
<span class="co">#&gt;  IRanges            2.28.0     2021-10-26 [1] Bioconductor</span>
<span class="co">#&gt;  iterators          1.0.13     2020-10-15 [1] Custom</span>
<span class="co">#&gt;  jquerylib          0.1.4      2021-04-26 [1] Custom</span>
<span class="co">#&gt;  jsonlite           1.7.2      2020-12-09 [1] Custom</span>
<span class="co">#&gt;  knitr              1.36       2021-09-29 [1] Custom</span>
<span class="co">#&gt;  labeling           0.4.2      2020-10-20 [1] Custom</span>
<span class="co">#&gt;  lattice            0.20-45    2021-09-22 [2] CRAN (R 4.1.2)</span>
<span class="co">#&gt;  lifecycle          1.0.1      2021-09-24 [1] Custom</span>
<span class="co">#&gt;  magrittr           2.0.1      2020-11-17 [1] Custom</span>
<span class="co">#&gt;  MASS               7.3-54     2021-05-03 [2] CRAN (R 4.1.2)</span>
<span class="co">#&gt;  Matrix             1.3-4      2021-06-01 [2] CRAN (R 4.1.2)</span>
<span class="co">#&gt;  memoise            2.0.0      2021-01-26 [1] Custom</span>
<span class="co">#&gt;  mgcv               1.8-38     2021-10-06 [2] CRAN (R 4.1.2)</span>
<span class="co">#&gt;  microbiome         1.16.0     2021-10-26 [1] Bioconductor</span>
<span class="co">#&gt;  microViz         * 0.8.0.9005 2021-11-19 [1] local</span>
<span class="co">#&gt;  multtest           2.50.0     2021-10-26 [1] Bioconductor</span>
<span class="co">#&gt;  munsell            0.5.0      2018-06-12 [1] Custom</span>
<span class="co">#&gt;  nlme               3.1-153    2021-09-07 [2] CRAN (R 4.1.2)</span>
<span class="co">#&gt;  permute            0.9-5      2019-03-12 [1] Custom</span>
<span class="co">#&gt;  phyloseq         * 1.38.0     2021-10-26 [1] Bioconductor</span>
<span class="co">#&gt;  pillar             1.6.4      2021-10-18 [1] Custom</span>
<span class="co">#&gt;  pkgbuild           1.2.0      2020-12-15 [1] Custom</span>
<span class="co">#&gt;  pkgconfig          2.0.3      2019-09-22 [1] Custom</span>
<span class="co">#&gt;  pkgdown            1.6.1      2020-09-12 [1] Custom</span>
<span class="co">#&gt;  pkgload            1.2.3      2021-10-13 [1] Custom</span>
<span class="co">#&gt;  plyr               1.8.6      2020-03-03 [1] Custom</span>
<span class="co">#&gt;  prettyunits        1.1.1      2020-01-24 [1] Custom</span>
<span class="co">#&gt;  processx           3.5.2      2021-04-30 [1] Custom</span>
<span class="co">#&gt;  ps                 1.6.0      2021-02-28 [1] Custom</span>
<span class="co">#&gt;  purrr              0.3.4      2020-04-17 [1] Custom</span>
<span class="co">#&gt;  R6                 2.5.1      2021-08-19 [1] Custom</span>
<span class="co">#&gt;  ragg               1.2.0      2021-10-30 [1] Custom</span>
<span class="co">#&gt;  Rcpp               1.0.7      2021-07-07 [1] Custom</span>
<span class="co">#&gt;  RCurl              1.98-1.5   2021-09-17 [1] Custom</span>
<span class="co">#&gt;  registry           0.5-1      2019-03-05 [1] Custom</span>
<span class="co">#&gt;  remotes            2.4.1      2021-09-29 [1] Custom</span>
<span class="co">#&gt;  reshape2           1.4.4      2020-04-09 [1] Custom</span>
<span class="co">#&gt;  rhdf5              2.38.0     2021-10-26 [1] Bioconductor</span>
<span class="co">#&gt;  rhdf5filters       1.6.0      2021-10-26 [1] Bioconductor</span>
<span class="co">#&gt;  Rhdf5lib           1.16.0     2021-10-26 [1] Bioconductor</span>
<span class="co">#&gt;  rlang              0.4.12     2021-10-18 [1] Custom</span>
<span class="co">#&gt;  rmarkdown          2.11       2021-09-14 [1] Custom</span>
<span class="co">#&gt;  rprojroot          2.0.2      2020-11-15 [1] Custom</span>
<span class="co">#&gt;  Rtsne              0.15       2018-11-10 [1] Custom</span>
<span class="co">#&gt;  S4Vectors          0.32.2     2021-11-07 [1] Bioconductor</span>
<span class="co">#&gt;  sass               0.4.0      2021-05-12 [1] Custom</span>
<span class="co">#&gt;  scales             1.1.1      2020-05-11 [1] Custom</span>
<span class="co">#&gt;  seriation          1.3.1      2021-10-16 [1] Custom</span>
<span class="co">#&gt;  sessioninfo        1.2.1      2021-11-02 [1] Custom</span>
<span class="co">#&gt;  stringi            1.7.5      2021-10-04 [1] Custom</span>
<span class="co">#&gt;  stringr            1.4.0      2019-02-10 [1] Custom</span>
<span class="co">#&gt;  survival           3.2-13     2021-08-24 [2] CRAN (R 4.1.2)</span>
<span class="co">#&gt;  systemfonts        1.0.3      2021-10-13 [1] Custom</span>
<span class="co">#&gt;  testthat           3.1.0      2021-10-04 [1] Custom</span>
<span class="co">#&gt;  textshaping        0.3.6      2021-10-13 [1] Custom</span>
<span class="co">#&gt;  tibble             3.1.6      2021-11-07 [1] Custom</span>
<span class="co">#&gt;  tidyr              1.1.4      2021-09-27 [1] Custom</span>
<span class="co">#&gt;  tidyselect         1.1.1      2021-04-30 [1] Custom</span>
<span class="co">#&gt;  TSP                1.1-11     2021-10-06 [1] Custom</span>
<span class="co">#&gt;  usethis            2.1.3      2021-10-27 [1] Custom</span>
<span class="co">#&gt;  utf8               1.2.2      2021-07-24 [1] Custom</span>
<span class="co">#&gt;  vctrs              0.3.8      2021-04-29 [1] Custom</span>
<span class="co">#&gt;  vegan              2.5-7      2020-11-28 [1] Custom</span>
<span class="co">#&gt;  withr              2.4.2      2021-04-18 [1] Custom</span>
<span class="co">#&gt;  xfun               0.28       2021-11-04 [1] Custom</span>
<span class="co">#&gt;  XVector            0.34.0     2021-10-26 [1] Bioconductor</span>
<span class="co">#&gt;  yaml               2.2.1      2020-02-01 [1] Custom</span>
<span class="co">#&gt;  zlibbioc           1.40.0     2021-10-26 [1] Bioconductor</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;  [1] /home/runner/work/_temp/Library</span>
<span class="co">#&gt;  [2] /opt/R/4.1.2/lib/R/library</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; ──────────────────────────────────────────────────────────────────────────────</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by David Barnett.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: '22e6f0a75b7b1ffae9bf5bc5b58864d7',
    indexName: 'microViz',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
