<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Microbiome Data Analysis and Visualization ‚Ä¢ microViz</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><!-- docsearch --><script src="docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><meta property="og:title" content="Microbiome Data Analysis and Visualization">
<meta property="og:description" content="Extends phyloseq and microbiome with further visualization and data manipulation functionality.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">microViz</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.7.10</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="articles/articles/microViz.html">Get started</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="divider">
    <li class="dropdown-header">Tutorials</li>
    <li>
      <a href="articles/articles/phyloseq.html">Working with phyloseq objects</a>
    </li>
    <li>
      <a href="articles/articles/tax-fixing.html">Fixing your tax_table</a>
    </li>
    <li>
      <a href="articles/articles/ordination.html">Ordination plots</a>
    </li>
    <li>
      <a href="articles/articles/interactive-ordination.html">Interactive Ordination Plot</a>
    </li>
    <li>
      <a href="articles/articles/compositions.html">Visualising compositions</a>
    </li>
    <li>
      <a href="articles/articles/videos.html">Videos</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/david-barnett/microViz/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="microviz-" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#microviz-" class="anchor"></a>microViz <a href="https://david-barnett.github.io/microViz/index.html"><img src="reference/figures/microViz-smaller.png" align="right" height="139"></a>
</h1></div>
<!-- badges: start -->

<!-- badges: end -->
<div id="overview" class="section level2">
<h2 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h2>
<p><span class="emoji" data-emoji="package">üì¶</span> <code>microViz</code> is an R package for analysis and visualization of microbiome sequencing data.</p>
<p><span class="emoji" data-emoji="hammer">üî®</span> <code>microViz</code> functions are intended to be easy to use and flexible.</p>
<p><span class="emoji" data-emoji="microscope">üî¨</span> <code>microViz</code> extends and complements popular microbial ecology packages like <code>phyloseq</code>, <code>vegan</code>, &amp; <code>microbiome</code>.</p>
</div>
<div id="learn-more" class="section level2">
<h2 class="hasAnchor">
<a href="#learn-more" class="anchor"></a>Learn more</h2>
<p><span class="emoji" data-emoji="paperclip">üìé</span> This website is the best place for documentation and examples: <a href="https://david-barnett.github.io/microViz/" class="uri">https://david-barnett.github.io/microViz/</a></p>
<ul>
<li><p><strong>The <a href="https://david-barnett.github.io/microViz/articles/articles/microViz.html">‚ÄúGet started‚Äù</a> page</strong> shows a few example analyses (as does this ReadMe)</p></li>
<li><p><strong>The <a href="https://david-barnett.github.io/microViz/reference/index.html">Reference</a> page</strong> lists all functions and links to help pages and examples</p></li>
<li><p><strong>The <a href="https://david-barnett.github.io/microViz/news/index.html">Changelog</a></strong> describes important changes in new microViz package versions</p></li>
<li>
<p><strong>The Articles pages</strong> give tutorials and further examples</p>
<ul>
<li><p><a href="https://david-barnett.github.io/microViz/articles/articles/phyloseq.html">Working with phyloseq objects</a></p></li>
<li><p><a href="https://david-barnett.github.io/microViz/articles/articles/tax-fixing.html">Fixing your taxa table with tax_fix</a></p></li>
<li><p><a href="https://david-barnett.github.io/microViz/articles/ordination.html">Creating ordination plots</a> (e.g.¬†PCA or PCoA)</p></li>
<li><p><a href="https://david-barnett.github.io/microViz/articles/articles/Interactive-Ordination.html">Interactive ordination plots with ord_explore</a></p></li>
<li><p><a href="https://david-barnett.github.io/microViz/articles/articles/Visualising-compositions.html">Visualising taxonomic compositions with comp_barplot</a></p></li>
<li><p>More coming soon! Post on <a href="https://github.com/david-barnett/microViz/discussions">GitHub discussions</a> if you have questions/requests</p></li>
</ul>
</li>
</ul>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>You can install the latest available microViz package version using the following instructions.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Installing from github requires the devtools package</span>
<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span> 

<span class="co"># You can install the latest versions of phyloseq and microbiome from Bioconductor:</span>
<span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span> 
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"phyloseq"</span>, <span class="st">"microbiome"</span><span class="op">)</span><span class="op">)</span>

<span class="co"># To install the latest "released" version of this package</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span><span class="op">(</span><span class="st">"david-barnett/microViz@0.7.10"</span><span class="op">)</span> <span class="co"># check 0.7.10 is the latest release</span>

<span class="co"># To install the very latest version:</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span><span class="op">(</span><span class="st">"david-barnett/microViz"</span><span class="op">)</span>
<span class="co"># If you encounter a bug please try the latest version &amp; let me know if the bug persists!</span></code></pre></div>
<p><span class="emoji" data-emoji="computer">üíª</span> <strong>Windows users</strong> - will need to have RTools installed so that your computer can build this package (follow instructions here: <a href="http://jtleek.com/modules/01_DataScientistToolbox/02_10_rtools/" class="uri">http://jtleek.com/modules/01_DataScientistToolbox/02_10_rtools/</a>)</p>
<p><strong><span class="emoji" data-emoji="apple">üçé</span> macOS</strong> <strong>users</strong> - might need to install <a href="https://www.xquartz.org/">xquartz</a> to make the heatmaps work (to do this with homebrew, run the following command in your mac‚Äôs Terminal: <code>brew install --cask xquartz</code></p>
<p><span class="emoji" data-emoji="package">üì¶</span> I highly recommend using <a href="https://rstudio.github.io/renv/index.html">renv</a> for managing your R package installations across multiple projects.</p>
<p><span class="emoji" data-emoji="whale">üê≥</span> Alternatively, for Docker users an image with the main branch installed is available at: <a href="https://hub.docker.com/r/barnettdavid/microviz-rocker-verse" class="uri">https://hub.docker.com/r/barnettdavid/microviz-rocker-verse</a></p>
<p><span class="emoji" data-emoji="date">üìÖ</span> microViz is tested to work with R version 4 on Windows, MacOS, and Ubuntu 18 and 20. R version 3.6.* should probably work, but I don‚Äôt formally test this.</p>
</div>
<div id="interactive-ordination-exploration" class="section level2">
<h2 class="hasAnchor">
<a href="#interactive-ordination-exploration" class="anchor"></a>Interactive ordination exploration</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://david-barnett.github.io/microViz">microViz</a></span><span class="op">)</span></code></pre></div>
<p>microViz provides a Shiny app for an easy way to start exploring your microbiome data: all you need is a phyloseq object.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># example data from corncob package</span>
<span class="va">pseq</span> <span class="op">&lt;-</span> <span class="fu">corncob</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/corncob/man/ibd_phylo.html">ibd_phylo</a></span> <span class="op">%&gt;%</span> <span class="fu"><a href="reference/tax_fix.html">tax_fix</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="reference/phyloseq_validate.html">phyloseq_validate</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="reference/ord_explore.html">ord_explore</a></span><span class="op">(</span><span class="va">pseq</span><span class="op">)</span> <span class="co"># gif generated with microViz version 0.7.4 (plays at 1.75x speed)</span></code></pre></div>
<p><img src="articles/articles/images/20210429_ord_explore_x175.gif"></p>
</div>
<div id="example-analyses" class="section level2">
<h2 class="hasAnchor">
<a href="#example-analyses" class="anchor"></a>Example analyses</h2>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://dx.plos.org/10.1371/journal.pone.0061217">phyloseq</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span>
<span class="co">#&gt; Warning: package 'dplyr' was built under R version 4.0.5</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'dplyr'</span>
<span class="co">#&gt; The following objects are masked from 'package:stats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     filter, lag</span>
<span class="co">#&gt; The following objects are masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     intersect, setdiff, setequal, union</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>
<span class="co">#&gt; Warning: package 'ggplot2' was built under R version 4.0.5</span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get some example data</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"dietswap"</span>, package <span class="op">=</span> <span class="st">"microbiome"</span><span class="op">)</span>

<span class="co"># create a couple of numerical variables to use as constraints or conditions</span>
<span class="va">dietswap</span> <span class="op">&lt;-</span> <span class="va">dietswap</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/ps_mutate.html">ps_mutate</a></span><span class="op">(</span>
    weight <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/recode.html">recode</a></span><span class="op">(</span><span class="va">bmi_group</span>, obese <span class="op">=</span> <span class="fl">3</span>, overweight <span class="op">=</span> <span class="fl">2</span>, lean <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>,
    female <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html">if_else</a></span><span class="op">(</span><span class="va">sex</span> <span class="op">==</span> <span class="st">"female"</span>, true <span class="op">=</span> <span class="fl">1</span>, false <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>
  <span class="op">)</span>
<span class="co"># add a couple of missing values to show how microViz handles missing data</span>
<span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/sample_data-methods.html">sample_data</a></span><span class="op">(</span><span class="va">dietswap</span><span class="op">)</span><span class="op">$</span><span class="va">female</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">4</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">NA</span></code></pre></div>
<div id="looking-at-your-data" class="section level3">
<h3 class="hasAnchor">
<a href="#looking-at-your-data" class="anchor"></a>Looking at your data</h3>
<p>You have quite a few samples in your phyloseq object, and would like to visualise their compositions. Perhaps these example data differ across BMI groups?</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dietswap</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/tax_filter.html">tax_filter</a></span><span class="op">(</span>min_prevalence <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/comp_barplot.html">comp_barplot</a></span><span class="op">(</span>
    tax_level <span class="op">=</span> <span class="st">"Genus"</span>,  n_taxa <span class="op">=</span> <span class="fl">12</span>, 
    merge_other <span class="op">=</span> <span class="cn">FALSE</span> 
    <span class="co"># set merge_other = TRUE (the default) to remove outlines from inside "other" category</span>
  <span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="st">"bmi_group"</span>, nrow <span class="op">=</span> <span class="fl">3</span>, scales <span class="op">=</span> <span class="st">"free"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="cn">NULL</span>, y <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>
    axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,
    axis.ticks.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>
  <span class="op">)</span>
<span class="co">#&gt; Registered S3 method overwritten by 'seriation':</span>
<span class="co">#&gt;   method         from </span>
<span class="co">#&gt;   reorder.hclust vegan</span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-2-1.png" width="100%"></p>
</div>
<div id="example-ordination-plot-workflow" class="section level3">
<h3 class="hasAnchor">
<a href="#example-ordination-plot-workflow" class="anchor"></a>Example ordination plot workflow</h3>
<p>Maybe visually inspecting all your samples isn‚Äôt quite what you want. Ordination methods can also help you to visualise if overall microbial ecosystem composition differs markedly between groups, e.g.¬†BMI.</p>
<p>Here is one option as an example:</p>
<ol>
<li>Filter out rare taxa (e.g.¬†remove Genera not present in at least 10% of samples) - use <code><a href="reference/tax_filter.html">tax_filter()</a></code>
</li>
<li>Aggregate the taxa into bacterial families (for example) - use <code><a href="reference/tax_agg.html">tax_agg()</a></code>
</li>
<li>Transform the microbial data with the centre-log-ratio transformation - use <code><a href="reference/tax_transform.html">tax_transform()</a></code>
</li>
<li>Perform PCA with the clr-transformed features (equivalent to aitchison distance PCoA) - use <code><a href="reference/ord_calc.html">ord_calc()</a></code>
</li>
<li>Plot the first 2 axes of this PCA ordination, colouring samples by group and adding taxon loading arrows to visualise which taxa generally differ across your samples - use <code><a href="reference/ord_plot.html">ord_plot()</a></code>
</li>
<li>Customise the theme of the ggplot as you like and/or add features like ellipses or annotations</li>
</ol>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># perform ordination</span>
<span class="va">unconstrained_aitchison_pca</span> <span class="op">&lt;-</span>
  <span class="va">dietswap</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/tax_filter.html">tax_filter</a></span><span class="op">(</span>min_prevalence <span class="op">=</span> <span class="fl">0.1</span>, tax_level <span class="op">=</span> <span class="st">"Genus"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/tax_agg.html">tax_agg</a></span><span class="op">(</span><span class="st">"Family"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/tax_transform.html">tax_transform</a></span><span class="op">(</span><span class="st">"clr"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/ord_calc.html">ord_calc</a></span><span class="op">(</span><span class="op">)</span> 
<span class="co">#&gt; Proportional min_prevalence given: 0.1 --&gt; min 23/222 samples.</span>
<span class="co"># ord_calc will automatically infer you want a "PCA" here</span>
<span class="co"># specify explicitly with method = "PCA", or you can pick another method</span>

<span class="co"># create plot</span>
<span class="va">pca_plot</span> <span class="op">&lt;-</span> <span class="va">unconstrained_aitchison_pca</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/ord_plot.html">ord_plot</a></span><span class="op">(</span>
    plot_taxa <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">6</span>, colour <span class="op">=</span> <span class="st">"bmi_group"</span>, size <span class="op">=</span> <span class="fl">2</span>,
    tax_vec_length <span class="op">=</span> <span class="fl">0.5</span>,
    auto_caption <span class="op">=</span> <span class="cn">TRUE</span>
  <span class="op">)</span>

<span class="co"># customise plot</span>
<span class="va">customised_plot</span> <span class="op">&lt;-</span> <span class="va">pca_plot</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/stat_ellipse.html">stat_ellipse</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>linetype <span class="op">=</span> <span class="va">bmi_group</span>, colour <span class="op">=</span> <span class="va">bmi_group</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_colour_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Set1"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span>

<span class="co"># show plot</span>
<span class="va">customised_plot</span></code></pre></div>
<p><img src="reference/figures/README-ordination-plot-1.png" width="100%"></p>
</div>
<div id="permanova" class="section level3">
<h3 class="hasAnchor">
<a href="#permanova" class="anchor"></a>PERMANOVA</h3>
<p>You visualised your ordinated data in the plot above. Below you can see how to perform a PERMANOVA to test the significance of BMI‚Äôs association with overall microbial composition. This example uses the Family-level aitchison distance to correspond with the plot above.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># calculate distances</span>
<span class="va">aitchison_dists</span> <span class="op">&lt;-</span>
  <span class="va">dietswap</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/tax_filter.html">tax_filter</a></span><span class="op">(</span>min_prevalence <span class="op">=</span> <span class="fl">0.1</span>, tax_level <span class="op">=</span> <span class="st">"Genus"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/tax_agg.html">tax_agg</a></span><span class="op">(</span><span class="st">"Family"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/tax_transform.html">tax_transform</a></span><span class="op">(</span><span class="st">"identity"</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="reference/dist_calc.html">dist_calc</a></span><span class="op">(</span><span class="st">"aitchison"</span><span class="op">)</span>
<span class="co">#&gt; Proportional min_prevalence given: 0.1 --&gt; min 23/222 samples.</span>

<span class="co"># the more permutations you request, the longer it takes</span>
<span class="co"># but also the more stable and precise your p-values become</span>
<span class="va">aitchison_perm</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/dist_permanova.html">dist_permanova</a></span><span class="op">(</span>
  data <span class="op">=</span> <span class="va">aitchison_dists</span>,
  seed <span class="op">=</span> <span class="fl">1234</span>, <span class="co"># for set.seed to ensure reproducibility of random process</span>
  n_perms <span class="op">=</span> <span class="fl">99</span>, <span class="co"># you should use at least 999!</span>
  n_processes <span class="op">=</span> <span class="fl">1</span>,
  variables <span class="op">=</span> <span class="st">"bmi_group + female"</span>
<span class="op">)</span>
<span class="co">#&gt; Dropping samples with missings: 2</span>
<span class="co">#&gt; 2021-07-16 13:26:17 - Starting PERMANOVA with 99 perms with 1 processes</span>
<span class="co">#&gt; 2021-07-16 13:26:17 - Finished PERMANOVA</span>
<span class="co"># view the permanova results</span>
<span class="fu"><a href="reference/ps_extra-accessors.html">perm_get</a></span><span class="op">(</span><span class="va">aitchison_perm</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt;            Df   SumOfSqs         R2        F Pr(&gt;F)</span>
<span class="co">#&gt; bmi_group   2  106.32182 0.04332620 5.008911   0.01</span>
<span class="co">#&gt; female      1   53.22231 0.02168812 5.014697   0.01</span>
<span class="co">#&gt; Residual  216 2292.46559 0.93418103       NA     NA</span>
<span class="co">#&gt; Total     219 2453.98430 1.00000000       NA     NA</span>
<span class="co"># view the info stored about the distance calculation</span>
<span class="fu"><a href="reference/ps_extra-accessors.html">info_get</a></span><span class="op">(</span><span class="va">aitchison_perm</span><span class="op">)</span>
<span class="co">#&gt; ps_extra info:</span>
<span class="co">#&gt; tax_agg = Family </span>
<span class="co">#&gt; tax_transform = identity </span>
<span class="co">#&gt; tax_scale = NA </span>
<span class="co">#&gt; distMethod = aitchison </span>
<span class="co">#&gt; ordMethod = NA </span>
<span class="co">#&gt; constraints = NA </span>
<span class="co">#&gt; conditions = NA</span></code></pre></div>
</div>
<div id="constrained-ordination" class="section level3">
<h3 class="hasAnchor">
<a href="#constrained-ordination" class="anchor"></a>Constrained ordination</h3>
<p>You could visualise the effect of the (numeric/logical) variables in your permanova directly using the ord_plot function with constraints.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">perm2</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/dist_permanova.html">dist_permanova</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">aitchison_dists</span>, variables <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"weight"</span>, <span class="st">"female"</span><span class="op">)</span>, seed <span class="op">=</span> <span class="fl">321</span><span class="op">)</span>
<span class="co">#&gt; Dropping samples with missings: 2</span>
<span class="co">#&gt; 2021-07-16 13:26:17 - Starting PERMANOVA with 999 perms with 1 processes</span>
<span class="co">#&gt; 2021-07-16 13:26:18 - Finished PERMANOVA</span>
<span class="fu"><a href="reference/ps_extra-accessors.html">perm_get</a></span><span class="op">(</span><span class="va">perm2</span><span class="op">)</span>
<span class="co">#&gt; Permutation test for adonis under reduced model</span>
<span class="co">#&gt; Marginal effects of terms</span>
<span class="co">#&gt; Permutation: free</span>
<span class="co">#&gt; Number of permutations: 999</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; vegan::adonis2(formula = formula, data = metadata, permutations = n_perms, by = by, parallel = parall)</span>
<span class="co">#&gt;           Df SumOfSqs      R2      F Pr(&gt;F)    </span>
<span class="co">#&gt; weight     1    56.06 0.02284 5.1927  0.001 ***</span>
<span class="co">#&gt; female     1    55.83 0.02275 5.1714  0.001 ***</span>
<span class="co">#&gt; Residual 217  2342.73 0.95466                  </span>
<span class="co">#&gt; Total    219  2453.98 1.00000                  </span>
<span class="co">#&gt; ---</span>
<span class="co">#&gt; Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span>

<span class="fu"><a href="reference/ord_calc.html">ord_calc</a></span><span class="op">(</span><span class="va">perm2</span>, constraints <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"weight"</span>, <span class="st">"female"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/ord_plot.html">ord_plot</a></span><span class="op">(</span>
    colour <span class="op">=</span> <span class="st">"sex"</span>, size <span class="op">=</span> <span class="fl">2</span>,
    constraint_vec_style <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="st">"black"</span>, size <span class="op">=</span> <span class="fl">1.15</span><span class="op">)</span>,
    constraint_lab_style <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span>
  <span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/stat_ellipse.html">stat_ellipse</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="va">sex</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_color_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Dark2"</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Centering (mean) and scaling (sd) the constraints and conditions:</span>
<span class="co">#&gt;  weight</span>
<span class="co">#&gt;  female</span></code></pre></div>
<p><img src="reference/figures/README-constrained-ord-1.png" width="100%"></p>
</div>
<div id="heatmaps" class="section level3">
<h3 class="hasAnchor">
<a href="#heatmaps" class="anchor"></a>Heatmaps</h3>
<p>microViz heatmaps are powered by <code>ComplexHeatmap</code> and annotated with taxa prevalence and/or abundance.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># set up the data with numerical variables and filter to top taxa</span>
<span class="va">psq</span> <span class="op">&lt;-</span> <span class="va">dietswap</span> <span class="op">%&gt;%</span>
 <span class="fu"><a href="reference/ps_mutate.html">ps_mutate</a></span><span class="op">(</span>
   weight <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/recode.html">recode</a></span><span class="op">(</span><span class="va">bmi_group</span>, obese <span class="op">=</span> <span class="fl">3</span>, overweight <span class="op">=</span> <span class="fl">2</span>, lean <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>,
   female <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html">if_else</a></span><span class="op">(</span><span class="va">sex</span> <span class="op">==</span> <span class="st">"female"</span>, true <span class="op">=</span> <span class="fl">1</span>, false <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>,
   african <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html">if_else</a></span><span class="op">(</span><span class="va">nationality</span> <span class="op">==</span> <span class="st">"AFR"</span>, true <span class="op">=</span> <span class="fl">1</span>, false <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>
 <span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="reference/tax_filter.html">tax_filter</a></span><span class="op">(</span>
    tax_level <span class="op">=</span> <span class="st">"Genus"</span>, min_prevalence <span class="op">=</span> <span class="fl">1</span> <span class="op">/</span> <span class="fl">10</span>, min_sample_abundance <span class="op">=</span> <span class="fl">1</span> <span class="op">/</span> <span class="fl">10</span>
  <span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="reference/tax_transform.html">tax_transform</a></span><span class="op">(</span><span class="st">"identity"</span>, rank <span class="op">=</span> <span class="st">"Genus"</span><span class="op">)</span>
<span class="co">#&gt; Proportional min_prevalence given: 0.1 --&gt; min 23/222 samples.</span>

<span class="co"># randomly select 30 taxa from the 50 most abundant taxa (just for an example)</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span>
<span class="va">taxa</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="fu">microbiome</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/microbiome/man/top_taxa.html">top_taxa</a></span><span class="op">(</span><span class="fu"><a href="reference/ps_extra-accessors.html">ps_get</a></span><span class="op">(</span><span class="va">psq</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">50</span><span class="op">]</span>, size <span class="op">=</span> <span class="fl">30</span><span class="op">)</span>
<span class="co"># actually draw the heatmap</span>
<span class="fu"><a href="reference/cor_heatmap.html">cor_heatmap</a></span><span class="op">(</span><span class="va">psq</span>, <span class="va">taxa</span>, anno_tax <span class="op">=</span> <span class="fu"><a href="reference/heatmap-annotations.html">tax_anno</a></span><span class="op">(</span>undetected <span class="op">=</span> <span class="fl">50</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="reference/figures/README-heatmap-1.png" width="100%"></p>
</div>
</div>
<div id="citation" class="section level2">
<h2 class="hasAnchor">
<a href="#citation" class="anchor"></a>Citation</h2>
<p><span class="emoji" data-emoji="innocent">üòá</span> If you find any part of microViz useful to your work, please consider citing the JOSS article:</p>
<p>Barnett et al., (2021). microViz: an R package for microbiome data visualization and statistics. Journal of Open Source Software, 6(63), 3201, <a href="https://doi.org/10.21105/joss.03201" class="uri">https://doi.org/10.21105/joss.03201</a></p>
</div>
<div id="contributing" class="section level2">
<h2 class="hasAnchor">
<a href="#contributing" class="anchor"></a>Contributing</h2>
<p>Bug reports, questions, suggestions for new features, and other contributions are all welcome. Feel free to create a <a href="https://github.com/david-barnett/microViz/issues">GitHub Issue</a> or write on the <a href="https://github.com/david-barnett/microViz/discussions">Discussions</a> page. Alternatively you could also contact me (David) on Twitter <a href="https://twitter.com/_david_barnett_">@_david_barnett_</a> .</p>
<p>This project is released with a <a href="https://david-barnett.github.io/microViz/CODE_OF_CONDUCT.html">Contributor Code of Conduct</a> and by participating in this project you agree to abide by its terms.</p>
</div>
<div id="session-info" class="section level2">
<h2 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session info</h2>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R version 4.0.3 (2020-10-10)</span>
<span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">#&gt; Running under: Ubuntu 18.04.5 LTS</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas/libblas.so.3</span>
<span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.2.20.so</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt;  [1] LC_CTYPE=en_GB.UTF-8       LC_NUMERIC=C               LC_TIME=en_GB.UTF-8       </span>
<span class="co">#&gt;  [4] LC_COLLATE=en_GB.UTF-8     LC_MONETARY=en_GB.UTF-8    LC_MESSAGES=en_GB.UTF-8   </span>
<span class="co">#&gt;  [7] LC_PAPER=en_GB.UTF-8       LC_NAME=C                  LC_ADDRESS=C              </span>
<span class="co">#&gt; [10] LC_TELEPHONE=C             LC_MEASUREMENT=en_GB.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt; [1] ggplot2_3.3.5   dplyr_1.0.7     phyloseq_1.34.0 microViz_0.7.10 devtools_2.4.0  usethis_2.0.1  </span>
<span class="co">#&gt; [7] pkgdown_1.6.1  </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;   [1] nlme_3.1-149         matrixStats_0.59.0   fs_1.5.0             RColorBrewer_1.1-2  </span>
<span class="co">#&gt;   [5] rprojroot_2.0.2      tools_4.0.3          utf8_1.2.1           R6_2.5.0            </span>
<span class="co">#&gt;   [9] vegan_2.5-7          DBI_1.1.1            BiocGenerics_0.36.1  mgcv_1.8-33         </span>
<span class="co">#&gt;  [13] colorspace_2.0-2     GetoptLong_1.0.5     permute_0.9-5        rhdf5filters_1.2.1  </span>
<span class="co">#&gt;  [17] ade4_1.7-17          withr_2.4.2          tidyselect_1.1.1     prettyunits_1.1.1   </span>
<span class="co">#&gt;  [21] processx_3.5.1       compiler_4.0.3       microbiome_1.12.0    cli_3.0.0           </span>
<span class="co">#&gt;  [25] Biobase_2.50.0       Cairo_1.5-12.2       TSP_1.1-10           desc_1.3.0          </span>
<span class="co">#&gt;  [29] labeling_0.4.2       scales_1.1.1         callr_3.7.0          stringr_1.4.0       </span>
<span class="co">#&gt;  [33] digest_0.6.27        rmarkdown_2.7        XVector_0.30.0       pkgconfig_2.0.3     </span>
<span class="co">#&gt;  [37] htmltools_0.5.1.1    sessioninfo_1.1.1    highr_0.9            fastmap_1.1.0       </span>
<span class="co">#&gt;  [41] GlobalOptions_0.1.2  rlang_0.4.11         shape_1.4.6          farver_2.1.0        </span>
<span class="co">#&gt;  [45] generics_0.1.0       jsonlite_1.7.2       magrittr_2.0.1       biomformat_1.18.0   </span>
<span class="co">#&gt;  [49] Matrix_1.3-2         Rcpp_1.0.7           munsell_0.5.0        S4Vectors_0.28.1    </span>
<span class="co">#&gt;  [53] Rhdf5lib_1.12.1      fansi_0.5.0          ape_5.5              lifecycle_1.0.0     </span>
<span class="co">#&gt;  [57] stringi_1.7.2        yaml_2.2.1           MASS_7.3-53          zlibbioc_1.36.0     </span>
<span class="co">#&gt;  [61] rhdf5_2.34.0         pkgbuild_1.2.0       Rtsne_0.15           plyr_1.8.6          </span>
<span class="co">#&gt;  [65] grid_4.0.3           parallel_4.0.3       crayon_1.4.1         lattice_0.20-41     </span>
<span class="co">#&gt;  [69] Biostrings_2.58.0    splines_4.0.3        multtest_2.46.0      circlize_0.4.13     </span>
<span class="co">#&gt;  [73] magick_2.7.1         ComplexHeatmap_2.6.2 knitr_1.32           ps_1.6.0            </span>
<span class="co">#&gt;  [77] pillar_1.6.1         igraph_1.2.6         rjson_0.2.20         reshape2_1.4.4      </span>
<span class="co">#&gt;  [81] codetools_0.2-18     stats4_4.0.3         pkgload_1.2.1        glue_1.4.2          </span>
<span class="co">#&gt;  [85] corncob_0.2.0        evaluate_0.14        data.table_1.14.0    remotes_2.3.0       </span>
<span class="co">#&gt;  [89] png_0.1-7            vctrs_0.3.8          foreach_1.5.1        testthat_3.0.2      </span>
<span class="co">#&gt;  [93] gtable_0.3.0         purrr_0.3.4          tidyr_1.1.3          clue_0.3-59         </span>
<span class="co">#&gt;  [97] assertthat_0.2.1     cachem_1.0.5         xfun_0.22            survival_3.2-7      </span>
<span class="co">#&gt; [101] seriation_1.3.0      tibble_3.1.2         iterators_1.0.13     registry_0.5-1      </span>
<span class="co">#&gt; [105] memoise_2.0.0        IRanges_2.24.1       cluster_2.1.0        ellipsis_0.3.2</span></code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/david-barnett/microViz/">https://‚Äãgithub.com/‚Äãdavid-barnett/‚ÄãmicroViz/‚Äã</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small>GPL (&gt;= 3)</small></li>
</ul>
</div>
<div class="community">
<h2>Community</h2>
<ul class="list-unstyled">
<li><a href="CODE_OF_CONDUCT.html">Code of conduct</a></li>
</ul>
</div>
<div class="citation">
<h2>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html">Citing microViz</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>David Barnett <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0003-1961-7206" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/david-barnett/microViz/actions"><img src="https://github.com/david-barnett/microViz/workflows/R-CMD-check/badge.svg" alt="R-CMD-check"></a></li>
<li><a href="https://codecov.io/gh/david-barnett/microViz"><img src="https://codecov.io/gh/david-barnett/microViz/branch/main/graph/badge.svg?token=C1EoVkhnxA" alt="codecov"></a></li>
<li><a href="https://zenodo.org/badge/latestdoi/307119750"><img src="https://zenodo.org/badge/307119750.svg" alt="DOI"></a></li>
<li><a href="https://hub.docker.com/r/barnettdavid/microviz-rocker-verse"><img src="https://img.shields.io/docker/cloud/build/barnettdavid/microviz-rocker-verse" alt="Docker Cloud Build Status"></a></li>
<li><a href="https://joss.theoj.org/papers/4547b492f224a26d96938ada81fee3fa"><img src="https://joss.theoj.org/papers/4547b492f224a26d96938ada81fee3fa/status.svg" alt="status"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by David Barnett.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: '22e6f0a75b7b1ffae9bf5bc5b58864d7',
    indexName: 'microViz',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
