<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Microbiome Data Analysis and Visualization • microViz</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><!-- docsearch --><script src="docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><meta property="og:title" content="Microbiome Data Analysis and Visualization">
<meta property="og:description" content="Microbiome data visualization and statistics tools built upon phyloseq.">
<meta property="og:image" content="https://david-barnett.github.io/microViz/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-2YQ669S4V5"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-2YQ669S4V5');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">microViz</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.9.1.9003</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li>
  <a href="articles/learn/exercises.html">Learn</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/web-only/phyloseq.html">Working with phyloseq objects</a>
    </li>
    <li>
      <a href="articles/web-only/tax-fixing.html">Fixing your tax_table</a>
    </li>
    <li>
      <a href="articles/web-only/ordination.html">Ordination plots</a>
    </li>
    <li>
      <a href="articles/web-only/ordination-interactive.html">Interactive ordination plots</a>
    </li>
    <li>
      <a href="articles/web-only/compositions.html">Visualising compositions with barplots</a>
    </li>
    <li>
      <a href="articles/web-only/heatmaps.html">Heatmaps of microbiome composition or correlation</a>
    </li>
    <li>
      <a href="articles/web-only/modelling-taxa.html">Statistical modelling of individual taxa</a>
    </li>
    <li>
      <a href="articles/web-only/videos.html">Tutorial video</a>
    </li>
    <li>
      <a href="articles/web-only/atlas1006.html">Example analyses with atlas1006 data</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">News</a>
</li>
<li class="dropdown-header">
      </li>
</ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/david-barnett/microViz/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="microviz-">microViz <a href="https://david-barnett.github.io/microViz/index.html"><img src="reference/figures/logo.png" align="right" height="180" width="156"></a>
<a class="anchor" aria-label="anchor" href="#microviz-"></a>
</h1></div>
<!-- badges: start -->
<!-- badges: end --><div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p><span class="emoji" data-emoji="package">📦</span> <code>microViz</code> is an R package for analysis and visualization of microbiome sequencing data.</p>
<p><span class="emoji" data-emoji="hammer">🔨</span> <code>microViz</code> functions are intended to be easy to use and flexible.</p>
<p><span class="emoji" data-emoji="microscope">🔬</span> <code>microViz</code> extends and complements popular microbial ecology packages like <code>phyloseq</code>, <code>vegan</code>, &amp; <code>microbiome</code>.</p>
</div>
<div class="section level2">
<h2 id="learn-more">Learn more<a class="anchor" aria-label="anchor" href="#learn-more"></a>
</h2>
<p><span class="emoji" data-emoji="paperclip">📎</span> This website is the best place for documentation and examples: <a href="https://david-barnett.github.io/microViz/" class="uri">https://david-barnett.github.io/microViz/</a></p>
<ul>
<li><p><a href="https://david-barnett.github.io/microViz/"><strong>This ReadMe</strong></a> shows a few example analyses</p></li>
<li><p><strong>The <a href="https://david-barnett.github.io/microViz/reference/index.html">Reference</a> page</strong> lists all functions and links to help pages and examples</p></li>
<li><p><strong>The <a href="https://david-barnett.github.io/microViz/news/index.html">Changelog</a></strong> describes important changes in new microViz package versions</p></li>
<li>
<p><strong>The Articles pages</strong> give tutorials and further examples</p>
<ul>
<li><p><a href="https://david-barnett.github.io/microViz/articles/web-only/phyloseq.html">Working with phyloseq objects</a></p></li>
<li><p><a href="https://david-barnett.github.io/microViz/articles/web-only/tax-fixing.html">Fixing your taxa table with tax_fix</a></p></li>
<li><p><a href="https://david-barnett.github.io/microViz/articles/web-only/ordination.html">Creating ordination plots</a> (e.g. PCA or PCoA)</p></li>
<li><p><a href="https://david-barnett.github.io/microViz/articles/web-only/ordination-interactive.html">Interactive ordination plots with ord_explore</a></p></li>
<li><p><a href="https://david-barnett.github.io/microViz/articles/web-only/compositions.html">Visualising taxonomic compositions with comp_barplot</a></p></li>
<li><p><a href="https://david-barnett.github.io/microViz/articles/web-only/heatmaps.html">Heatmaps of microbiome composition and correlation</a></p></li>
<li><p><a href="https://david-barnett.github.io/microViz/articles/web-only/modelling-taxa.html">Modelling and plotting individual taxon associations with taxatrees</a></p></li>
<li><p>More coming soon(ish)! Post on <a href="https://github.com/david-barnett/microViz/discussions" class="external-link">GitHub discussions</a> if you have questions/requests</p></li>
</ul>
</li>
</ul>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>You can install the latest available microViz package version using the following instructions.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Installing from github requires the devtools package</span>
<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span> 

<span class="co"># To install the latest "released" version of this package</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org/reference/remote-reexports.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"david-barnett/microViz@0.9.1"</span><span class="op">)</span> <span class="co"># check 0.9.1 is the latest release</span>

<span class="co"># To install the very latest version:</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org/reference/remote-reexports.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"david-barnett/microViz"</span><span class="op">)</span>
<span class="co"># If you encounter a bug please try the latest version &amp; let me know if the bug persists!</span>

<span class="co"># If the Bioconductor dependencies don't automatically install you can install</span>
<span class="co"># them yourself like this:</span>
<span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span> 
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"phyloseq"</span>, <span class="st">"microbiome"</span>, <span class="st">"ComplexHeatmap"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><span class="emoji" data-emoji="computer">💻</span> <strong>Windows users</strong> - will need to have RTools installed so that your computer can build this package (follow instructions here: <a href="http://jtleek.com/modules/01_DataScientistToolbox/02_10_rtools/" class="external-link uri">http://jtleek.com/modules/01_DataScientistToolbox/02_10_rtools/</a>)</p>
<p><strong><span class="emoji" data-emoji="apple">🍎</span> macOS</strong> <strong>users</strong> - might need to install <a href="https://www.xquartz.org/" class="external-link">xquartz</a> to make the heatmaps work (to do this with homebrew, run the following command in your mac’s Terminal: <code>brew install --cask xquartz</code></p>
<p><span class="emoji" data-emoji="package">📦</span> I highly recommend using <a href="https://rstudio.github.io/renv/index.html" class="external-link">renv</a> for managing your R package installations across multiple projects.</p>
<p><span class="emoji" data-emoji="whale">🐳</span> Alternatively, for Docker users an image with the main branch installed is available at: <a href="https://hub.docker.com/r/barnettdavid/microviz-rocker-verse" class="external-link uri">https://hub.docker.com/r/barnettdavid/microviz-rocker-verse</a></p>
<p><span class="emoji" data-emoji="date">📅</span> microViz is tested to work with R version 4 on Windows, MacOS, and Ubuntu 18 and 20. R version 3.6.* should probably work, but I don’t formally test this.</p>
</div>
<div class="section level2">
<h2 id="interactive-ordination-exploration">Interactive ordination exploration<a class="anchor" aria-label="anchor" href="#interactive-ordination-exploration"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://david-barnett.github.io/microViz">microViz</a></span><span class="op">)</span></code></pre></div>
<p>microViz provides a Shiny app for an easy way to start exploring your microbiome data: all you need is a phyloseq object.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># example data from corncob package</span>
<span class="va">pseq</span> <span class="op">&lt;-</span> <span class="fu">corncob</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/corncob/man/ibd_phylo.html" class="external-link">ibd_phylo</a></span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="reference/tax_fix.html">tax_fix</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="reference/phyloseq_validate.html">phyloseq_validate</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="reference/ord_explore.html">ord_explore</a></span><span class="op">(</span><span class="va">pseq</span><span class="op">)</span> <span class="co"># gif generated with microViz version 0.7.4 (plays at 1.75x speed)</span></code></pre></div>
<p><img src="articles/web-only/images/20210429_ord_explore_x175.gif"></p>
</div>
<div class="section level2">
<h2 id="example-analyses">Example analyses<a class="anchor" aria-label="anchor" href="#example-analyses"></a>
</h2>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://dx.plos.org/10.1371/journal.pone.0061217" class="external-link">phyloseq</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span>
<span class="co">#&gt; Warning: package 'dplyr' was built under R version 4.1.2</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'dplyr'</span>
<span class="co">#&gt; The following objects are masked from 'package:stats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     filter, lag</span>
<span class="co">#&gt; The following objects are masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     intersect, setdiff, setequal, union</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get some example data</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"dietswap"</span>, package <span class="op">=</span> <span class="st">"microbiome"</span><span class="op">)</span>

<span class="co"># create a couple of numerical variables to use as constraints or conditions</span>
<span class="va">dietswap</span> <span class="op">&lt;-</span> <span class="va">dietswap</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="reference/ps_mutate.html">ps_mutate</a></span><span class="op">(</span>
    weight <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/recode.html" class="external-link">recode</a></span><span class="op">(</span><span class="va">bmi_group</span>, obese <span class="op">=</span> <span class="fl">3</span>, overweight <span class="op">=</span> <span class="fl">2</span>, lean <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>,
    female <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html" class="external-link">if_else</a></span><span class="op">(</span><span class="va">sex</span> <span class="op">==</span> <span class="st">"female"</span>, true <span class="op">=</span> <span class="fl">1</span>, false <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>,
    african <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html" class="external-link">if_else</a></span><span class="op">(</span><span class="va">nationality</span> <span class="op">==</span> <span class="st">"AFR"</span>, true <span class="op">=</span> <span class="fl">1</span>, false <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>
  <span class="op">)</span>
<span class="co"># add a couple of missing values to show how microViz handles missing data</span>
<span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/sample_data-methods.html" class="external-link">sample_data</a></span><span class="op">(</span><span class="va">dietswap</span><span class="op">)</span><span class="op">$</span><span class="va">african</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">4</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">NA</span></code></pre></div>
<div class="section level3">
<h3 id="looking-at-your-data">Looking at your data<a class="anchor" aria-label="anchor" href="#looking-at-your-data"></a>
</h3>
<p>You have quite a few samples in your phyloseq object, and would like to visualise their compositions. Perhaps these example data differ participant nationality?</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dietswap</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="reference/comp_barplot.html">comp_barplot</a></span><span class="op">(</span>
    tax_level <span class="op">=</span> <span class="st">"Genus"</span>, n_taxa <span class="op">=</span> <span class="fl">15</span>, other_name <span class="op">=</span> <span class="st">"Other"</span>,
    taxon_renamer <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_remove.html" class="external-link">str_remove</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">" [ae]t rel."</span><span class="op">)</span>,
    palette <span class="op">=</span> <span class="fu"><a href="reference/distinct_palette.html">distinct_palette</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">15</span>, add <span class="op">=</span> <span class="st">"grey90"</span><span class="op">)</span>,
    merge_other <span class="op">=</span> <span class="cn">FALSE</span>, bar_outline_colour <span class="op">=</span> <span class="st">"darkgrey"</span>
  <span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="st">"nationality"</span>, nrow <span class="op">=</span> <span class="fl">1</span>, scales <span class="op">=</span> <span class="st">"free"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="cn">NULL</span>, y <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>, axis.ticks.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; Registered S3 method overwritten by 'seriation':</span>
<span class="co">#&gt;   method         from </span>
<span class="co">#&gt;   reorder.hclust vegan</span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-2-1.png" width="100%"></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">htmp</span> <span class="op">&lt;-</span> <span class="va">dietswap</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="reference/ps_mutate.html">ps_mutate</a></span><span class="op">(</span>nationality <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">nationality</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="reference/tax_transform.html">tax_transform</a></span><span class="op">(</span><span class="st">"log2"</span>, add <span class="op">=</span> <span class="fl">1</span>, chain <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="reference/comp_heatmap.html">comp_heatmap</a></span><span class="op">(</span>
    taxa <span class="op">=</span> <span class="fu"><a href="reference/tax_top.html">tax_top</a></span><span class="op">(</span><span class="va">dietswap</span>, n <span class="op">=</span> <span class="fl">30</span><span class="op">)</span>, grid_col <span class="op">=</span> <span class="cn">NA</span>, name <span class="op">=</span> <span class="st">"Log2p"</span>,
    taxon_renamer <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_remove.html" class="external-link">str_remove</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">" [ae]t rel."</span><span class="op">)</span>,
    colors <span class="op">=</span> <span class="fu"><a href="reference/heat_palette.html">heat_palette</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="fu">viridis</span><span class="fu">::</span><span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">turbo</a></span><span class="op">(</span><span class="fl">11</span><span class="op">)</span><span class="op">)</span>,
    row_names_side <span class="op">=</span> <span class="st">"left"</span>, row_dend_side <span class="op">=</span> <span class="st">"right"</span>, sample_side <span class="op">=</span> <span class="st">"bottom"</span>,
    sample_anno <span class="op">=</span> <span class="fu"><a href="reference/sampleAnnotation.html">sampleAnnotation</a></span><span class="op">(</span>
      Nationality <span class="op">=</span> <span class="fu"><a href="reference/anno_sample_cat.html">anno_sample_cat</a></span><span class="op">(</span>
        var <span class="op">=</span> <span class="st">"nationality"</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>AAM <span class="op">=</span> <span class="st">"grey35"</span>, AFR <span class="op">=</span> <span class="st">"grey85"</span><span class="op">)</span>,
        box_col <span class="op">=</span> <span class="cn">NA</span>, legend_title <span class="op">=</span> <span class="st">"Nationality"</span>, size <span class="op">=</span> <span class="fu">grid</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">4</span>, <span class="st">"mm"</span><span class="op">)</span>
      <span class="op">)</span>
    <span class="op">)</span>
  <span class="op">)</span>

<span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/draw-dispatch.html" class="external-link">draw</a></span><span class="op">(</span>
  object <span class="op">=</span> <span class="va">htmp</span>, annotation_legend_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">htmp</span>, <span class="st">"AnnoLegends"</span><span class="op">)</span>,
  merge_legends <span class="op">=</span> <span class="cn">TRUE</span>
<span class="op">)</span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-3-1.png" width="100%"></p>
</div>
<div class="section level3">
<h3 id="example-ordination-plot-workflow">Example ordination plot workflow<a class="anchor" aria-label="anchor" href="#example-ordination-plot-workflow"></a>
</h3>
<p>Ordination methods can also help you to visualise if overall microbial ecosystem composition differs markedly between groups, e.g. BMI.</p>
<p>Here is one option as an example:</p>
<ol style="list-style-type: decimal">
<li>Filter out rare taxa (e.g. remove Genera not present in at least 10% of samples) - use <code><a href="reference/tax_filter.html">tax_filter()</a></code>
</li>
<li>Aggregate the taxa into bacterial families (for example) - use <code><a href="reference/tax_agg.html">tax_agg()</a></code>
</li>
<li>Transform the microbial data with the centre-log-ratio transformation - use <code><a href="reference/tax_transform.html">tax_transform()</a></code>
</li>
<li>Perform PCA with the clr-transformed features (equivalent to aitchison distance PCoA) - use <code><a href="reference/ord_calc.html">ord_calc()</a></code>
</li>
<li>Plot the first 2 axes of this PCA ordination, colouring samples by group and adding taxon loading arrows to visualise which taxa generally differ across your samples - use <code><a href="reference/ord_plot.html">ord_plot()</a></code>
</li>
<li>Customise the theme of the ggplot as you like and/or add features like ellipses or annotations</li>
</ol>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># perform ordination</span>
<span class="va">unconstrained_aitchison_pca</span> <span class="op">&lt;-</span> <span class="va">dietswap</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="reference/tax_filter.html">tax_filter</a></span><span class="op">(</span>min_prevalence <span class="op">=</span> <span class="fl">0.1</span>, tax_level <span class="op">=</span> <span class="st">"Genus"</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="reference/tax_agg.html">tax_agg</a></span><span class="op">(</span><span class="st">"Family"</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="reference/tax_transform.html">tax_transform</a></span><span class="op">(</span><span class="st">"clr"</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="reference/ord_calc.html">ord_calc</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; Proportional min_prevalence given: 0.1 --&gt; min 23/222 samples.</span>
<span class="co"># ord_calc will automatically infer you want a "PCA" here</span>
<span class="co"># specify explicitly with method = "PCA", or you can pick another method</span>

<span class="co"># create plot</span>
<span class="va">pca_plot</span> <span class="op">&lt;-</span> <span class="va">unconstrained_aitchison_pca</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="reference/ord_plot.html">ord_plot</a></span><span class="op">(</span>
    plot_taxa <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">6</span>, colour <span class="op">=</span> <span class="st">"bmi_group"</span>, size <span class="op">=</span> <span class="fl">1.5</span>,
    tax_vec_length <span class="op">=</span> <span class="fl">0.325</span>,
    tax_lab_style <span class="op">=</span> <span class="fu"><a href="reference/Ordination-labels.html">tax_lab_style</a></span><span class="op">(</span>max_angle <span class="op">=</span> <span class="fl">90</span>, aspect_ratio <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span>,
    auto_caption <span class="op">=</span> <span class="fl">8</span>
  <span class="op">)</span>

<span class="co"># customise plot</span>
<span class="va">customised_plot</span> <span class="op">&lt;-</span> <span class="va">pca_plot</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/stat_ellipse.html" class="external-link">stat_ellipse</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>linetype <span class="op">=</span> <span class="va">bmi_group</span>, colour <span class="op">=</span> <span class="va">bmi_group</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_colour_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Set1"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span>ratio <span class="op">=</span> <span class="fl">0.5</span>, clip <span class="op">=</span> <span class="st">"off"</span><span class="op">)</span> <span class="co"># makes rotated labels align correctly</span>

<span class="co"># show plot</span>
<span class="va">customised_plot</span></code></pre></div>
<p><img src="reference/figures/README-ordination-plot-1.png" width="100%"></p>
</div>
<div class="section level3">
<h3 id="permanova">PERMANOVA<a class="anchor" aria-label="anchor" href="#permanova"></a>
</h3>
<p>You visualised your ordinated data in the plot above. Below you can see how to perform a PERMANOVA to test the significance of BMI’s association with overall microbial composition. This example uses the Family-level aitchison distance to correspond with the plot above.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># calculate distances</span>
<span class="va">aitchison_dists</span> <span class="op">&lt;-</span> <span class="va">dietswap</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="reference/tax_filter.html">tax_filter</a></span><span class="op">(</span>min_prevalence <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="reference/tax_transform.html">tax_transform</a></span><span class="op">(</span><span class="st">"identity"</span>, rank <span class="op">=</span> <span class="st">"Family"</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="reference/dist_calc.html">dist_calc</a></span><span class="op">(</span><span class="st">"aitchison"</span><span class="op">)</span>
<span class="co">#&gt; Proportional min_prevalence given: 0.1 --&gt; min 23/222 samples.</span>

<span class="co"># the more permutations you request, the longer it takes</span>
<span class="co"># but also the more stable and precise your p-values become</span>
<span class="va">aitchison_perm</span> <span class="op">&lt;-</span> <span class="va">aitchison_dists</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span> 
  <span class="fu"><a href="reference/dist_permanova.html">dist_permanova</a></span><span class="op">(</span>
    seed <span class="op">=</span> <span class="fl">1234</span>, <span class="co"># for set.seed to ensure reproducibility of random process</span>
    n_processes <span class="op">=</span> <span class="fl">1</span>, n_perms <span class="op">=</span> <span class="fl">99</span>, <span class="co"># you should use at least 999!</span>
    variables <span class="op">=</span> <span class="st">"bmi_group"</span>
  <span class="op">)</span>
<span class="co">#&gt; 2022-04-28 12:21:50 - Starting PERMANOVA with 99 perms with 1 processes</span>
<span class="co">#&gt; 2022-04-28 12:21:50 - Finished PERMANOVA</span>

<span class="co"># view the permanova results</span>
<span class="fu"><a href="reference/ps_extra-accessors.html">perm_get</a></span><span class="op">(</span><span class="va">aitchison_perm</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt;            Df  SumOfSqs         R2        F Pr(&gt;F)</span>
<span class="co">#&gt; bmi_group   2  104.0678 0.04177157 4.773379   0.01</span>
<span class="co">#&gt; Residual  219 2387.2862 0.95822843       NA     NA</span>
<span class="co">#&gt; Total     221 2491.3540 1.00000000       NA     NA</span>

<span class="co"># view the info stored about the distance calculation</span>
<span class="fu"><a href="reference/ps_extra-accessors.html">info_get</a></span><span class="op">(</span><span class="va">aitchison_perm</span><span class="op">)</span>
<span class="co">#&gt; ps_extra info:</span>
<span class="co">#&gt; tax_agg = Family </span>
<span class="co">#&gt; tax_transform = identity </span>
<span class="co">#&gt; tax_scale = NA </span>
<span class="co">#&gt; distMethod = aitchison </span>
<span class="co">#&gt; ordMethod = NA </span>
<span class="co">#&gt; constraints = NA </span>
<span class="co">#&gt; conditions = NA</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="constrained-partial-ordination">Constrained partial ordination<a class="anchor" aria-label="anchor" href="#constrained-partial-ordination"></a>
</h3>
<p>You could visualise the effect of the (numeric/logical) variables in your permanova directly using the <code>ord_plot</code> function with constraints (and conditions).</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">perm2</span> <span class="op">&lt;-</span> <span class="va">aitchison_dists</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span> 
  <span class="fu"><a href="reference/dist_permanova.html">dist_permanova</a></span><span class="op">(</span>variables <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"weight"</span>, <span class="st">"african"</span>, <span class="st">"sex"</span><span class="op">)</span>, seed <span class="op">=</span> <span class="fl">321</span><span class="op">)</span>
<span class="co">#&gt; Dropping samples with missings: 2</span>
<span class="co">#&gt; 2022-04-28 12:21:50 - Starting PERMANOVA with 999 perms with 1 processes</span>
<span class="co">#&gt; 2022-04-28 12:21:53 - Finished PERMANOVA</span></code></pre></div>
<p>We’ll visualise the effect of nationality and bodyweight on sample composition, after first removing the effect of sex.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">perm2</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="reference/ord_calc.html">ord_calc</a></span><span class="op">(</span>constraints <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"weight"</span>, <span class="st">"african"</span><span class="op">)</span>, conditions <span class="op">=</span> <span class="st">"female"</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="reference/ord_plot.html">ord_plot</a></span><span class="op">(</span>
    colour <span class="op">=</span> <span class="st">"nationality"</span>, size <span class="op">=</span> <span class="fl">2.5</span>, alpha <span class="op">=</span> <span class="fl">0.35</span>,
    auto_caption <span class="op">=</span> <span class="fl">7</span>,
    constraint_vec_length <span class="op">=</span> <span class="fl">1</span>,
    constraint_vec_style <span class="op">=</span> <span class="fu"><a href="reference/ord_arrows.html">vec_constraint</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">1.5</span>, colour <span class="op">=</span> <span class="st">"grey15"</span><span class="op">)</span>,
    constraint_lab_style <span class="op">=</span> <span class="fu"><a href="reference/Ordination-labels.html">constraint_lab_style</a></span><span class="op">(</span>
      max_angle <span class="op">=</span> <span class="fl">90</span>, size <span class="op">=</span> <span class="fl">3</span>, aspect_ratio <span class="op">=</span> <span class="fl">0.35</span>, colour <span class="op">=</span> <span class="st">"black"</span>
    <span class="op">)</span>
  <span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/stat_ellipse.html" class="external-link">stat_ellipse</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="va">nationality</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_color_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Set1"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span>ratio <span class="op">=</span> <span class="fl">0.35</span>, clip <span class="op">=</span> <span class="st">"off"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.9</span>, <span class="fl">0.1</span><span class="op">)</span>, legend.background <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_rect</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Centering (mean) and scaling (sd) the constraints and conditions:</span>
<span class="co">#&gt;  weight</span>
<span class="co">#&gt;  african</span>
<span class="co">#&gt;  female</span></code></pre></div>
<p><img src="reference/figures/README-constrained-ord-plot-1.png" width="100%"></p>
</div>
<div class="section level3">
<h3 id="correlation-heatmaps">Correlation Heatmaps<a class="anchor" aria-label="anchor" href="#correlation-heatmaps"></a>
</h3>
<p>microViz heatmaps are powered by <code>ComplexHeatmap</code> and annotated with taxa prevalence and/or abundance.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># set up the data with numerical variables and filter to top taxa</span>
<span class="va">psq</span> <span class="op">&lt;-</span> <span class="va">dietswap</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="reference/ps_mutate.html">ps_mutate</a></span><span class="op">(</span>
    weight <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/recode.html" class="external-link">recode</a></span><span class="op">(</span><span class="va">bmi_group</span>, obese <span class="op">=</span> <span class="fl">3</span>, overweight <span class="op">=</span> <span class="fl">2</span>, lean <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>,
    female <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html" class="external-link">if_else</a></span><span class="op">(</span><span class="va">sex</span> <span class="op">==</span> <span class="st">"female"</span>, true <span class="op">=</span> <span class="fl">1</span>, false <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>,
    african <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html" class="external-link">if_else</a></span><span class="op">(</span><span class="va">nationality</span> <span class="op">==</span> <span class="st">"AFR"</span>, true <span class="op">=</span> <span class="fl">1</span>, false <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>
  <span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="reference/tax_filter.html">tax_filter</a></span><span class="op">(</span>
    tax_level <span class="op">=</span> <span class="st">"Genus"</span>, min_prevalence <span class="op">=</span> <span class="fl">1</span> <span class="op">/</span> <span class="fl">10</span>, min_sample_abundance <span class="op">=</span> <span class="fl">1</span> <span class="op">/</span> <span class="fl">10</span>
  <span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span>
  <span class="fu"><a href="reference/tax_transform.html">tax_transform</a></span><span class="op">(</span><span class="st">"identity"</span>, rank <span class="op">=</span> <span class="st">"Genus"</span><span class="op">)</span>
<span class="co">#&gt; Proportional min_prevalence given: 0.1 --&gt; min 23/222 samples.</span>

<span class="co"># randomly select 30 taxa from the 50 most abundant taxa (just for an example)</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span>
<span class="va">taxa</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="reference/tax_top.html">tax_top</a></span><span class="op">(</span><span class="va">psq</span>, n <span class="op">=</span> <span class="fl">50</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">30</span><span class="op">)</span>
<span class="co"># actually draw the heatmap</span>
<span class="fu"><a href="reference/cor_heatmap.html">cor_heatmap</a></span><span class="op">(</span>
  data <span class="op">=</span> <span class="va">psq</span>, taxa <span class="op">=</span> <span class="va">taxa</span>,
  taxon_renamer <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_remove.html" class="external-link">str_remove</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">" [ae]t rel."</span><span class="op">)</span>,
  tax_anno <span class="op">=</span> <span class="fu"><a href="reference/taxAnnotation.html">taxAnnotation</a></span><span class="op">(</span>
    Prev. <span class="op">=</span> <span class="fu"><a href="reference/anno_tax_prev.html">anno_tax_prev</a></span><span class="op">(</span>undetected <span class="op">=</span> <span class="fl">50</span><span class="op">)</span>,
    Log2 <span class="op">=</span> <span class="fu"><a href="reference/anno_tax_box.html">anno_tax_box</a></span><span class="op">(</span>undetected <span class="op">=</span> <span class="fl">50</span>, trans <span class="op">=</span> <span class="st">"log2"</span>, zero_replace <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
  <span class="op">)</span>
<span class="op">)</span></code></pre></div>
<p><img src="reference/figures/README-heatmap-1.png" width="100%"></p>
</div>
</div>
<div class="section level2">
<h2 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h2>
<p><span class="emoji" data-emoji="innocent">😇</span> If you find any part of microViz useful to your work, please consider citing the JOSS article:</p>
<p>Barnett et al., (2021). microViz: an R package for microbiome data visualization and statistics. Journal of Open Source Software, 6(63), 3201, <a href="https://doi.org/10.21105/joss.03201" class="external-link uri">https://doi.org/10.21105/joss.03201</a></p>
</div>
<div class="section level2">
<h2 id="contributing">Contributing<a class="anchor" aria-label="anchor" href="#contributing"></a>
</h2>
<p>Bug reports, questions, suggestions for new features, and other contributions are all welcome. Feel free to create a <a href="https://github.com/david-barnett/microViz/issues" class="external-link">GitHub Issue</a> or write on the <a href="https://github.com/david-barnett/microViz/discussions" class="external-link">Discussions</a> page. Alternatively you could also contact me (David) on Twitter <a href="https://twitter.com/_david_barnett_" class="external-link">@_david_barnett_</a> .</p>
<p>This project is released with a <a href="https://david-barnett.github.io/microViz/CODE_OF_CONDUCT.html">Contributor Code of Conduct</a> and by participating in this project you agree to abide by its terms.</p>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R version 4.1.1 (2021-08-10)</span>
<span class="co">#&gt; Platform: x86_64-w64-mingw32/x64 (64-bit)</span>
<span class="co">#&gt; Running under: Windows 10 x64 (build 19044)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt; [1] LC_COLLATE=English_United States.1252 </span>
<span class="co">#&gt; [2] LC_CTYPE=English_United States.1252   </span>
<span class="co">#&gt; [3] LC_MONETARY=English_United States.1252</span>
<span class="co">#&gt; [4] LC_NUMERIC=C                          </span>
<span class="co">#&gt; [5] LC_TIME=English_United States.1252    </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt; [1] ggplot2_3.3.5       dplyr_1.0.8         phyloseq_1.38.0    </span>
<span class="co">#&gt; [4] microViz_0.9.0.9009</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;   [1] ggtext_0.1.1           nlme_3.1-155           matrixStats_0.61.0    </span>
<span class="co">#&gt;   [4] bitops_1.0-7           RColorBrewer_1.1-3     doParallel_1.0.17     </span>
<span class="co">#&gt;   [7] GenomeInfoDb_1.30.1    tools_4.1.1            utf8_1.2.2            </span>
<span class="co">#&gt;  [10] R6_2.5.1               vegan_2.5-7            DBI_1.1.2             </span>
<span class="co">#&gt;  [13] BiocGenerics_0.40.0    mgcv_1.8-38            colorspace_2.0-3      </span>
<span class="co">#&gt;  [16] GetoptLong_1.0.5       permute_0.9-7          rhdf5filters_1.6.0    </span>
<span class="co">#&gt;  [19] ade4_1.7-18            withr_2.5.0            tidyselect_1.1.2      </span>
<span class="co">#&gt;  [22] gridExtra_2.3          compiler_4.1.1         microbiome_1.16.0     </span>
<span class="co">#&gt;  [25] cli_3.2.0              Biobase_2.54.0         Cairo_1.5-15          </span>
<span class="co">#&gt;  [28] TSP_1.2-0              xml2_1.3.3             labeling_0.4.2        </span>
<span class="co">#&gt;  [31] scales_1.2.0           stringr_1.4.0          digest_0.6.29         </span>
<span class="co">#&gt;  [34] rmarkdown_2.13         XVector_0.34.0         pkgconfig_2.0.3       </span>
<span class="co">#&gt;  [37] htmltools_0.5.2        fastmap_1.1.0          highr_0.9             </span>
<span class="co">#&gt;  [40] GlobalOptions_0.1.2    rlang_1.0.2            rstudioapi_0.13       </span>
<span class="co">#&gt;  [43] shape_1.4.6            generics_0.1.2         farver_2.1.0          </span>
<span class="co">#&gt;  [46] jsonlite_1.8.0         RCurl_1.98-1.6         magrittr_2.0.2        </span>
<span class="co">#&gt;  [49] GenomeInfoDbData_1.2.7 biomformat_1.22.0      Matrix_1.4-0          </span>
<span class="co">#&gt;  [52] Rcpp_1.0.8.3           munsell_0.5.0          S4Vectors_0.32.3      </span>
<span class="co">#&gt;  [55] Rhdf5lib_1.16.0        fansi_1.0.2            ape_5.6-2             </span>
<span class="co">#&gt;  [58] viridis_0.6.2          lifecycle_1.0.1        stringi_1.7.6         </span>
<span class="co">#&gt;  [61] yaml_2.3.5             MASS_7.3-55            zlibbioc_1.40.0       </span>
<span class="co">#&gt;  [64] rhdf5_2.38.1           Rtsne_0.15             plyr_1.8.7            </span>
<span class="co">#&gt;  [67] grid_4.1.1             parallel_4.1.1         crayon_1.5.1          </span>
<span class="co">#&gt;  [70] lattice_0.20-45        Biostrings_2.62.0      splines_4.1.1         </span>
<span class="co">#&gt;  [73] gridtext_0.1.4         multtest_2.50.0        circlize_0.4.14       </span>
<span class="co">#&gt;  [76] magick_2.7.3           ComplexHeatmap_2.10.0  knitr_1.37            </span>
<span class="co">#&gt;  [79] pillar_1.7.0           igraph_1.2.11          rjson_0.2.21          </span>
<span class="co">#&gt;  [82] markdown_1.1           reshape2_1.4.4         codetools_0.2-18      </span>
<span class="co">#&gt;  [85] stats4_4.1.1           glue_1.6.1             corncob_0.2.0         </span>
<span class="co">#&gt;  [88] evaluate_0.14          data.table_1.14.2      png_0.1-7             </span>
<span class="co">#&gt;  [91] vctrs_0.3.8            foreach_1.5.2          gtable_0.3.0          </span>
<span class="co">#&gt;  [94] purrr_0.3.4            tidyr_1.2.0            clue_0.3-60           </span>
<span class="co">#&gt;  [97] assertthat_0.2.1       xfun_0.30              survival_3.2-13       </span>
<span class="co">#&gt; [100] viridisLite_0.4.0      seriation_1.3.5        tibble_3.1.6          </span>
<span class="co">#&gt; [103] iterators_1.0.14       registry_0.5-1         IRanges_2.28.0        </span>
<span class="co">#&gt; [106] cluster_2.1.2          ellipsis_0.3.2</span></code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/david-barnett/microViz/" class="external-link">Browse source code</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small>GPL (&gt;= 3)</small></li>
</ul>
</div>

<div class="community">
<h2 data-toc-skip>Community</h2>
<ul class="list-unstyled">
<li><a href="CODE_OF_CONDUCT.html">Code of conduct</a></li>
</ul>
</div>

<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing microViz</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>David Barnett <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0003-1961-7206" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/david-barnett/microViz/actions" class="external-link"><img src="https://github.com/david-barnett/microViz/workflows/R-CMD-check/badge.svg" alt="R-CMD-check"></a></li>
<li><a href="https://codecov.io/gh/david-barnett/microViz" class="external-link"><img src="https://codecov.io/gh/david-barnett/microViz/branch/main/graph/badge.svg?token=C1EoVkhnxA" alt="codecov"></a></li>
<li><a href="https://hub.docker.com/r/barnettdavid/microviz-rocker-verse" class="external-link"><img src="https://img.shields.io/docker/cloud/build/barnettdavid/microviz-rocker-verse" alt="Docker Cloud Build Status"></a></li>
<li><a href="https://joss.theoj.org/papers/4547b492f224a26d96938ada81fee3fa" class="external-link"><img src="https://joss.theoj.org/papers/4547b492f224a26d96938ada81fee3fa/status.svg" alt="status"></a></li>
<li><a href="https://zenodo.org/badge/latestdoi/307119750" class="external-link"><img src="https://zenodo.org/badge/307119750.svg" alt="DOI"></a></li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by David Barnett.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.5.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: '22e6f0a75b7b1ffae9bf5bc5b58864d7',
    indexName: 'microViz',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
