<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Microbiome Data Analysis and Visualization • microViz</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Microbiome Data Analysis and Visualization">
<meta property="og:description" content="Extends phyloseq and microbiome with further visualization and data manipulation functionality.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">microViz</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.7.0.9001</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="articles/articles/microViz.html">Get started</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/articles/Interactive-Ordination.html">Interactive Ordination Plot</a>
    </li>
    <li>
      <a href="articles/articles/Visualising-compositions.html">Visualising compositions</a>
    </li>
    <li>
      <a href="articles/articles/phyloseq.html">Working with phyloseq objects</a>
    </li>
    <li>
      <a href="articles/articles/tax-fixing.html">Fixing your tax_table</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/david-barnett/microViz/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="microviz-" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#microviz-" class="anchor"></a>microViz <a href="https://david-barnett.github.io/microViz/index.html"><img src="reference/figures/microViz-smaller.png" align="right" height="139"></a>
</h1></div>
<!-- badges: start -->

<!-- badges: end -->
<p>microViz provides functions for analysis and visualization of microbiome sequencing data. These functions are intended to be easy to use and flexible. microViz wraps, extends and complements popular microbial ecology packages like phyloseq, vegan, and microbiome.</p>
<p><strong>See the documentation website for full details and examples:</strong> <a href="https://david-barnett.github.io/microViz/" class="uri">https://david-barnett.github.io/microViz/</a></p>
<ul>
<li><p>The <a href="https://david-barnett.github.io/microViz/articles/articles/microViz.html">“Get started”</a> page, and also <a href="https://david-barnett.github.io/microViz/index.html">this ReadMe</a>, show a few example analyses.</p></li>
<li><p>The <a href="https://david-barnett.github.io/microViz/reference/index.html">reference</a> page lists all functions and links to their documentation and examples</p></li>
<li><p>Example video of <a href="https://david-barnett.github.io/microViz/articles/articles/Interactive-Ordination.html">Interactive Shiny app</a> for exploring your ordination plots!</p></li>
<li>
<p>Other articles pages give tutorials and more complicated examples</p>
<ul>
<li><p><a href="https://david-barnett.github.io/microViz/articles/articles/phyloseq.html">Working with phyloseq objects</a></p></li>
<li><p><a href="https://david-barnett.github.io/microViz/articles/articles/tax-fixing.html">Fixing your taxa table with tax_fix</a></p></li>
<li><p><a href="https://david-barnett.github.io/microViz/articles/articles/Visualising-compositions.html">Visualising taxonomic compositions with customised barplots</a></p></li>
<li><p>(more coming soon! raise an issue on github if you have questions/requests)</p></li>
</ul>
</li>
<li><p>The <a href="https://david-barnett.github.io/microViz/news/index.html">changelog</a> describes important changes in new microViz package versions</p></li>
</ul>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>You can install the latest available microViz package version using the following instructions.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># If you are on Windows you will need to install RTools so that your computer can build this package</span>
<span class="co"># Follow instructions here: http://jtleek.com/modules/01_DataScientistToolbox/02_10_rtools/</span>

<span class="co"># If you are on macOS, you might need to install xquartz to make the heatmaps work (ComplexHeatmaps package)</span>
<span class="co"># You can do this with homebrew, running the following command in your mac's Terminal: brew install --cask xquartz </span>

<span class="co"># for installing from github you'll need the devtools package</span>
<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span>, type <span class="op">=</span> <span class="st">"binary"</span><span class="op">)</span> <span class="co"># (not binary if you're on linux)</span>

<span class="co"># If you don't already have the latest versions of phyloseq and microbiome, you should install these from bioconductor:</span>

<span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
 <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, type <span class="op">=</span> <span class="st">"binary"</span><span class="op">)</span> <span class="co"># (not binary if you're on linux)</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"phyloseq"</span>, <span class="st">"microbiome"</span><span class="op">)</span><span class="op">)</span>

<span class="co"># # Installing the latest version of this package # #</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span><span class="op">(</span><span class="st">"david-barnett/microViz@0.7.0"</span><span class="op">)</span> <span class="co"># check 0.7.0 is the latest version?</span>
<span class="co"># advanced tip: add @&lt;commit-hash&gt; after microViz to install a version from a particular commit</span></code></pre></div>
</div>
<div id="examples-below" class="section level2">
<h2 class="hasAnchor">
<a href="#examples-below" class="anchor"></a>Examples below</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/david-barnett/microViz">microViz</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://dx.plos.org/10.1371/journal.pone.0061217">phyloseq</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get some example data</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"dietswap"</span>, package <span class="op">=</span> <span class="st">"microbiome"</span><span class="op">)</span>

<span class="co"># create a couple of numerical variables to use as constraints or conditions</span>
<span class="va">dietswap</span> <span class="op">&lt;-</span> <span class="va">dietswap</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/ps_mutate.html">ps_mutate</a></span><span class="op">(</span>
    weight <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/recode.html">recode</a></span><span class="op">(</span><span class="va">bmi_group</span>, obese <span class="op">=</span> <span class="fl">3</span>, overweight <span class="op">=</span> <span class="fl">2</span>, lean <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>,
    female <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html">if_else</a></span><span class="op">(</span><span class="va">sex</span> <span class="op">==</span> <span class="st">"female"</span>, true <span class="op">=</span> <span class="fl">1</span>, false <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>
  <span class="op">)</span>
<span class="co"># add a couple of missing values to show how microViz handles missing data</span>
<span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/sample_data-methods.html">sample_data</a></span><span class="op">(</span><span class="va">dietswap</span><span class="op">)</span><span class="op">$</span><span class="va">female</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">4</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">NA</span></code></pre></div>
</div>
<div id="looking-at-your-data" class="section level2">
<h2 class="hasAnchor">
<a href="#looking-at-your-data" class="anchor"></a>Looking at your data</h2>
<p>You have quite a few samples in your phyloseq object, and would like to visualise their compositions. Perhaps these example data differ across BMI groups?</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dietswap</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/tax_filter.html">tax_filter</a></span><span class="op">(</span>min_prevalence <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/comp_barplot.html">comp_barplot</a></span><span class="op">(</span>
    tax_level <span class="op">=</span> <span class="st">"Genus"</span>,  n_taxa <span class="op">=</span> <span class="fl">12</span>, 
    merge_other <span class="op">=</span> <span class="cn">FALSE</span> 
    <span class="co"># set merge_other = TRUE (the default) to remove outlines from inside "other" category</span>
  <span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="st">"bmi_group"</span>, nrow <span class="op">=</span> <span class="fl">3</span>, scales <span class="op">=</span> <span class="st">"free"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="cn">NULL</span>, y <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>
    axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,
    axis.ticks.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>
  <span class="op">)</span>
<span class="co">#&gt; Registered S3 method overwritten by 'seriation':</span>
<span class="co">#&gt;   method         from </span>
<span class="co">#&gt;   reorder.hclust vegan</span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-2-1.png" width="100%"></p>
</div>
<div id="example-ordination-plot-workflow" class="section level2">
<h2 class="hasAnchor">
<a href="#example-ordination-plot-workflow" class="anchor"></a>Example ordination plot workflow</h2>
<p>Maybe visually inspecting all your samples isn’t quite what you want. Ordination methods can also help you to visualise if overall microbial ecosystem composition differs markedly between groups, e.g. BMI.</p>
<p>Here is one option to try first:</p>
<ol>
<li>Filter out rare taxa (e.g. remove Genera not present in at least 10% of samples) - use <code><a href="reference/tax_filter.html">tax_filter()</a></code>
</li>
<li>Aggregate the taxa into bacterial families (for example) - use <code><a href="reference/tax_agg.html">tax_agg()</a></code>
</li>
<li>Transform the microbial data with the centre-log-ratio transformation - use <code><a href="reference/tax_transform.html">tax_transform()</a></code>
</li>
<li>Perform PCA with the clr-transformed features (equivalent to aitchison distance PCoA) - use <code><a href="reference/ord_calc.html">ord_calc()</a></code>
</li>
<li>Plot the first 2 axes of this PCA ordination, colouring samples by group and adding taxon loading arrows to visualise which taxa generally differ across your samples - use <code><a href="reference/ord_plot.html">ord_plot()</a></code>
</li>
<li>Customise the theme of the ggplot as you like and/or add features like ellipses or annotations</li>
</ol>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># perform ordination</span>
<span class="va">unconstrained_aitchison_pca</span> <span class="op">&lt;-</span>
  <span class="va">dietswap</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/tax_filter.html">tax_filter</a></span><span class="op">(</span>min_prevalence <span class="op">=</span> <span class="fl">0.1</span>, tax_level <span class="op">=</span> <span class="st">"Genus"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/tax_agg.html">tax_agg</a></span><span class="op">(</span><span class="st">"Family"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/tax_transform.html">tax_transform</a></span><span class="op">(</span><span class="st">"clr"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/ord_calc.html">ord_calc</a></span><span class="op">(</span><span class="op">)</span> 
<span class="co">#&gt; Proportional min_prevalence given: 0.1 --&gt; min 23/222 samples.</span>
<span class="co"># ord_calc will automatically infer you want a "PCA" here</span>
<span class="co"># specify explicitly with method = "PCA", or you can pick another method</span>

<span class="co"># create plot</span>
<span class="va">pca_plot</span> <span class="op">&lt;-</span> <span class="va">unconstrained_aitchison_pca</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/ord_plot.html">ord_plot</a></span><span class="op">(</span>
    plot_taxa <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">6</span>, colour <span class="op">=</span> <span class="st">"bmi_group"</span>, size <span class="op">=</span> <span class="fl">2</span>,
    tax_vec_length <span class="op">=</span> <span class="fl">0.5</span>,
    auto_caption <span class="op">=</span> <span class="cn">TRUE</span>
  <span class="op">)</span>

<span class="co"># customise plot</span>
<span class="va">customised_plot</span> <span class="op">&lt;-</span> <span class="va">pca_plot</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/stat_ellipse.html">stat_ellipse</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>linetype <span class="op">=</span> <span class="va">bmi_group</span>, colour <span class="op">=</span> <span class="va">bmi_group</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_colour_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Set1"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span>

<span class="co"># show plot</span>
<span class="va">customised_plot</span></code></pre></div>
<p><img src="reference/figures/README-ordination-plot-1.png" width="100%"></p>
</div>
<div id="permanova" class="section level2">
<h2 class="hasAnchor">
<a href="#permanova" class="anchor"></a>PERMANOVA</h2>
<p>You visualised your ordinated data in the plot above. Below you can see how to perform a PERMANOVA to test the significance of BMI’s association with overall microbial composition. This example uses the Family-level aitchison distance to correspond with the plot above.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># calculate distances</span>
<span class="va">aitchison_dists</span> <span class="op">&lt;-</span>
  <span class="va">dietswap</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/tax_filter.html">tax_filter</a></span><span class="op">(</span>min_prevalence <span class="op">=</span> <span class="fl">0.1</span>, tax_level <span class="op">=</span> <span class="st">"Genus"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/tax_agg.html">tax_agg</a></span><span class="op">(</span><span class="st">"Family"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/tax_transform.html">tax_transform</a></span><span class="op">(</span><span class="st">"identity"</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="reference/dist_calc.html">dist_calc</a></span><span class="op">(</span><span class="st">"aitchison"</span><span class="op">)</span>
<span class="co">#&gt; Proportional min_prevalence given: 0.1 --&gt; min 23/222 samples.</span>

<span class="co"># the more permutations you request, the longer it takes</span>
<span class="co"># but also the more stable and precise your p-values become</span>
<span class="va">aitchison_perm</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/dist_permanova.html">dist_permanova</a></span><span class="op">(</span>
  data <span class="op">=</span> <span class="va">aitchison_dists</span>,
  seed <span class="op">=</span> <span class="fl">1234</span>, <span class="co"># for set.seed to ensure reproducibility of random process</span>
  n_perms <span class="op">=</span> <span class="fl">99</span>, <span class="co"># you should use at least 999!</span>
  n_processes <span class="op">=</span> <span class="fl">1</span>,
  variables <span class="op">=</span> <span class="st">"bmi_group + female"</span>
<span class="op">)</span>
<span class="co">#&gt; Dropping samples with missings: 2</span>
<span class="co">#&gt; 2021-04-12 12:46:53 - Starting PERMANOVA with 99 perms with 1 processes</span>
<span class="co">#&gt; 2021-04-12 12:46:54 - Finished PERMANOVA</span>
<span class="co"># view the permanova results</span>
<span class="fu"><a href="reference/ps_extra-accessors.html">perm_get</a></span><span class="op">(</span><span class="va">aitchison_perm</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt;            Df   SumOfSqs         R2        F Pr(&gt;F)</span>
<span class="co">#&gt; bmi_group   2  106.32182 0.04332620 5.008911   0.01</span>
<span class="co">#&gt; female      1   53.22231 0.02168812 5.014697   0.01</span>
<span class="co">#&gt; Residual  216 2292.46559 0.93418103       NA     NA</span>
<span class="co">#&gt; Total     219 2453.98430 1.00000000       NA     NA</span>
<span class="co"># view the info stored about the distance calculation</span>
<span class="fu"><a href="reference/ps_extra-accessors.html">info_get</a></span><span class="op">(</span><span class="va">aitchison_perm</span><span class="op">)</span>
<span class="co">#&gt; phyloseq info:</span>
<span class="co">#&gt; tax_agg = Family tax_transform = identity</span></code></pre></div>
</div>
<div id="constrained-ordination" class="section level2">
<h2 class="hasAnchor">
<a href="#constrained-ordination" class="anchor"></a>Constrained ordination</h2>
<p>You could visualise the effect of the (numeric/logical) variables in your permanova directly using the ord_plot function with constraints.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">perm2</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/dist_permanova.html">dist_permanova</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">aitchison_dists</span>, variables <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"weight"</span>, <span class="st">"female"</span><span class="op">)</span>, seed <span class="op">=</span> <span class="fl">321</span><span class="op">)</span>
<span class="co">#&gt; Dropping samples with missings: 2</span>
<span class="co">#&gt; 2021-04-12 12:46:54 - Starting PERMANOVA with 999 perms with 1 processes</span>
<span class="co">#&gt; 2021-04-12 12:46:55 - Finished PERMANOVA</span>
<span class="fu"><a href="reference/ps_extra-accessors.html">perm_get</a></span><span class="op">(</span><span class="va">perm2</span><span class="op">)</span>
<span class="co">#&gt; Permutation test for adonis under reduced model</span>
<span class="co">#&gt; Marginal effects of terms</span>
<span class="co">#&gt; Permutation: free</span>
<span class="co">#&gt; Number of permutations: 999</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; vegan::adonis2(formula = formula, data = metadata, permutations = n_perms, by = by, parallel = parall)</span>
<span class="co">#&gt;           Df SumOfSqs      R2      F Pr(&gt;F)    </span>
<span class="co">#&gt; weight     1    56.06 0.02284 5.1927  0.001 ***</span>
<span class="co">#&gt; female     1    55.83 0.02275 5.1714  0.001 ***</span>
<span class="co">#&gt; Residual 217  2342.73 0.95466                  </span>
<span class="co">#&gt; Total    219  2453.98 1.00000                  </span>
<span class="co">#&gt; ---</span>
<span class="co">#&gt; Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span>

<span class="fu"><a href="reference/ord_calc.html">ord_calc</a></span><span class="op">(</span><span class="va">perm2</span>, constraints <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"weight"</span>, <span class="st">"female"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="reference/ord_plot.html">ord_plot</a></span><span class="op">(</span>
    colour <span class="op">=</span> <span class="st">"sex"</span>, size <span class="op">=</span> <span class="fl">2</span>,
    constraint_vec_style <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="st">"black"</span>, size <span class="op">=</span> <span class="fl">1.15</span><span class="op">)</span>,
    constraint_lab_style <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span>
  <span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/stat_ellipse.html">stat_ellipse</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="va">sex</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_color_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Dark2"</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Centering (mean) and scaling (sd) the constraints and conditions:</span>
<span class="co">#&gt;  weight</span>
<span class="co">#&gt;  female</span></code></pre></div>
<p><img src="reference/figures/README-constrained-ord-1.png" width="100%"></p>
</div>
<div id="heatmaps" class="section level2">
<h2 class="hasAnchor">
<a href="#heatmaps" class="anchor"></a>Heatmaps</h2>
<p>microViz heatmaps are powered by <code>ComplexHeatmap</code> and annotated with taxa prevalence and/or abundance.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># set up the data with numerical variables and filter to top taxa</span>
<span class="va">psq</span> <span class="op">&lt;-</span> <span class="va">dietswap</span> <span class="op">%&gt;%</span>
 <span class="fu"><a href="reference/ps_mutate.html">ps_mutate</a></span><span class="op">(</span>
   weight <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/recode.html">recode</a></span><span class="op">(</span><span class="va">bmi_group</span>, obese <span class="op">=</span> <span class="fl">3</span>, overweight <span class="op">=</span> <span class="fl">2</span>, lean <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>,
   female <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html">if_else</a></span><span class="op">(</span><span class="va">sex</span> <span class="op">==</span> <span class="st">"female"</span>, true <span class="op">=</span> <span class="fl">1</span>, false <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>,
   african <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html">if_else</a></span><span class="op">(</span><span class="va">nationality</span> <span class="op">==</span> <span class="st">"AFR"</span>, true <span class="op">=</span> <span class="fl">1</span>, false <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>
 <span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="reference/tax_filter.html">tax_filter</a></span><span class="op">(</span>
    tax_level <span class="op">=</span> <span class="st">"Genus"</span>, min_prevalence <span class="op">=</span> <span class="fl">1</span> <span class="op">/</span> <span class="fl">10</span>, min_sample_abundance <span class="op">=</span> <span class="fl">1</span> <span class="op">/</span> <span class="fl">10</span>
  <span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="reference/tax_transform.html">tax_transform</a></span><span class="op">(</span><span class="st">"identity"</span>, rank <span class="op">=</span> <span class="st">"Genus"</span><span class="op">)</span>
<span class="co">#&gt; Proportional min_prevalence given: 0.1 --&gt; min 23/222 samples.</span>

<span class="co"># randomly select 30 taxa from the 50 most abundant taxa (just for an example)</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span>
<span class="va">taxa</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="fu">microbiome</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/microbiome/man/top_taxa.html">top_taxa</a></span><span class="op">(</span><span class="fu"><a href="reference/ps_extra-accessors.html">ps_get</a></span><span class="op">(</span><span class="va">psq</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">50</span><span class="op">]</span>, size <span class="op">=</span> <span class="fl">30</span><span class="op">)</span>
<span class="co"># actually draw the heatmap</span>
<span class="fu"><a href="reference/cor_heatmap.html">cor_heatmap</a></span><span class="op">(</span><span class="va">psq</span>, <span class="va">taxa</span>, anno_tax <span class="op">=</span> <span class="fu"><a href="reference/heatmap-annotations.html">tax_anno</a></span><span class="op">(</span>undetected <span class="op">=</span> <span class="fl">50</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="reference/figures/README-heatmap-1.png" width="100%"></p>
</div>
<div id="session-info" class="section level2">
<h2 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session info</h2>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/sessioninfo/man/session_info.html">session_info</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; ─ Session info ───────────────────────────────────────────────────────────────────────────────────</span>
<span class="co">#&gt;  setting  value                       </span>
<span class="co">#&gt;  version  R version 4.0.3 (2020-10-10)</span>
<span class="co">#&gt;  os       Ubuntu 18.04.5 LTS          </span>
<span class="co">#&gt;  system   x86_64, linux-gnu           </span>
<span class="co">#&gt;  ui       X11                         </span>
<span class="co">#&gt;  language en_GB:en                    </span>
<span class="co">#&gt;  collate  en_GB.UTF-8                 </span>
<span class="co">#&gt;  ctype    en_GB.UTF-8                 </span>
<span class="co">#&gt;  tz       Europe/Amsterdam            </span>
<span class="co">#&gt;  date     2021-04-12                  </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; ─ Packages ───────────────────────────────────────────────────────────────────────────────────────</span>
<span class="co">#&gt;  package        * version    date       lib source        </span>
<span class="co">#&gt;  ade4             1.7-16     2020-10-28 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  ape              5.4-1      2020-08-13 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  Biobase          2.50.0     2020-10-27 [1] Bioconductor  </span>
<span class="co">#&gt;  BiocGenerics     0.36.0     2020-10-27 [1] Bioconductor  </span>
<span class="co">#&gt;  biomformat       1.18.0     2020-10-27 [1] Bioconductor  </span>
<span class="co">#&gt;  Biostrings       2.58.0     2020-10-27 [1] Bioconductor  </span>
<span class="co">#&gt;  cachem           1.0.4      2021-02-13 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  Cairo            1.5-12.2   2020-07-07 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  callr            3.6.0      2021-03-28 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  circlize         0.4.12     2021-01-08 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  cli              2.4.0      2021-04-05 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  clue             0.3-58     2020-12-03 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  cluster          2.1.0      2019-06-19 [2] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  codetools        0.2-18     2020-11-04 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  colorspace       2.0-0      2020-11-11 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  ComplexHeatmap   2.6.2      2020-11-12 [1] Bioconductor  </span>
<span class="co">#&gt;  crayon           1.4.1      2021-02-08 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  data.table       1.14.0     2021-02-21 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  debugme          1.1.0      2017-10-22 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  desc             1.3.0      2021-03-05 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  devtools       * 2.3.2      2020-09-18 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  digest           0.6.27     2020-10-24 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  dplyr          * 1.0.5      2021-03-05 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  ellipsis         0.3.1      2020-05-15 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  evaluate         0.14       2019-05-28 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  fansi            0.4.2      2021-01-15 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  farver           2.1.0      2021-02-28 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  fastmap          1.1.0      2021-01-25 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  foreach          1.5.1      2020-10-15 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  fs               1.5.0      2020-07-31 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  generics         0.1.0      2020-10-31 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  GetoptLong       1.0.5      2020-12-15 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  ggplot2        * 3.3.3      2020-12-30 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  GlobalOptions    0.1.2      2020-06-10 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  glue             1.4.2      2020-08-27 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  gtable           0.3.0      2019-03-25 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  highr            0.8        2019-03-20 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  hms              1.0.0      2021-01-13 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  htmltools        0.5.1.1    2021-01-22 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  igraph           1.2.6      2020-10-06 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  IRanges          2.24.1     2020-12-12 [1] Bioconductor  </span>
<span class="co">#&gt;  iterators        1.0.13     2020-10-15 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  jsonlite         1.7.2      2020-12-09 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  knitr            1.31       2021-01-27 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  labeling         0.4.2      2020-10-20 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  lattice          0.20-41    2020-04-02 [2] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  lifecycle        1.0.0      2021-02-15 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  magick           2.7.1      2021-03-20 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  magrittr         2.0.1      2020-11-17 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  MASS             7.3-53     2020-09-09 [2] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  Matrix           1.3-2      2021-01-06 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  matrixStats      0.58.0     2021-01-29 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  memoise          2.0.0      2021-01-26 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  mgcv             1.8-33     2020-08-27 [2] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  microbiome       1.12.0     2020-10-27 [1] Bioconductor  </span>
<span class="co">#&gt;  microViz       * 0.7.0.9001 2021-04-12 [1] local         </span>
<span class="co">#&gt;  multtest         2.46.0     2020-10-27 [1] Bioconductor  </span>
<span class="co">#&gt;  munsell          0.5.0      2018-06-12 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  nlme             3.1-149    2020-08-23 [2] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  permute          0.9-5      2019-03-12 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  phyloseq       * 1.34.0     2020-10-27 [1] Bioconductor  </span>
<span class="co">#&gt;  pillar           1.5.1      2021-03-05 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  pkgbuild         1.2.0      2020-12-15 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  pkgconfig        2.0.3      2019-09-22 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  pkgdown        * 1.6.1      2020-09-12 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  pkgload          1.2.0      2021-02-23 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  plyr             1.8.6      2020-03-03 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  png              0.1-7      2013-12-03 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  prettyunits      1.1.1      2020-01-24 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  processx         3.5.0      2021-03-23 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  progress         1.2.2      2019-05-16 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  ps               1.6.0      2021-02-28 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  purrr            0.3.4      2020-04-17 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  R6               2.5.0      2020-10-28 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  RColorBrewer     1.1-2      2014-12-07 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  Rcpp             1.0.6      2021-01-15 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  registry         0.5-1      2019-03-05 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  remotes          2.2.0      2020-07-21 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  reshape2         1.4.4      2020-04-09 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  rhdf5            2.34.0     2020-10-27 [1] Bioconductor  </span>
<span class="co">#&gt;  rhdf5filters     1.2.0      2020-10-27 [1] Bioconductor  </span>
<span class="co">#&gt;  Rhdf5lib         1.12.1     2021-01-26 [1] Bioconductor  </span>
<span class="co">#&gt;  rjson            0.2.20     2018-06-08 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  rlang            0.4.10     2020-12-30 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  rmarkdown        2.7        2021-02-19 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  rprojroot        2.0.2      2020-11-15 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  Rtsne            0.15       2018-11-10 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  S4Vectors        0.28.1     2020-12-09 [1] Bioconductor  </span>
<span class="co">#&gt;  scales           1.1.1      2020-05-11 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  seriation        1.2-9      2020-10-01 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  sessioninfo      1.1.1      2018-11-05 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  shape            1.4.5      2020-09-13 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  stringi          1.5.3      2020-09-09 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  stringr          1.4.0      2019-02-10 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  survival         3.2-7      2020-09-28 [2] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  testthat         3.0.2      2021-02-14 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  tibble           3.1.0      2021-02-25 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  tidyr            1.1.3      2021-03-03 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  tidyselect       1.1.0      2020-05-11 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  TSP              1.1-10     2020-04-17 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  usethis        * 2.0.1      2021-02-10 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  utf8             1.2.1      2021-03-12 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  vctrs            0.3.7      2021-03-29 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  vegan            2.5-7      2020-11-28 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  withr            2.4.1      2021-01-26 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  xfun             0.22       2021-03-11 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  XVector          0.30.0     2020-10-27 [1] Bioconductor  </span>
<span class="co">#&gt;  yaml             2.2.1      2020-02-01 [1] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  zlibbioc         1.36.0     2020-10-27 [1] Bioconductor  </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [1] /home/david/R/x86_64-pc-linux-gnu-library/4.0</span>
<span class="co">#&gt; [2] /opt/R/4.0.3/lib/R/library</span></code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/david-barnett/microViz/">https://​github.com/​david-barnett/​microViz/​</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small>GPL (&gt;= 3)</small></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>David Barnett <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0003-1961-7206" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/david-barnett/microViz/actions"><img src="https://github.com/david-barnett/microViz/workflows/R-CMD-check/badge.svg" alt="R-CMD-check"></a></li>
<li><a href="https://codecov.io/gh/david-barnett/microViz"><img src="https://codecov.io/gh/david-barnett/microViz/branch/main/graph/badge.svg?token=C1EoVkhnxA" alt="codecov"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by David Barnett.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
