---
title: "microViz: an R package for microbiome data visualization and statistics"
tags: 
  - R
  - microbiome
  - visualization
author: "David Barnett"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Summary 

microViz is an R package for the statistical analysis and visualisation of microbiota data. This package extends the functionality of popular microbial ecosystem data analysis R packages phyloseq, vegan and microbiome.
microViz provides a selection of powerful additions to the toolbox of researchers already familiar with phyloseq and other ubiquitous microbiome R packages, as well as assisting researchers with less R programming experience to independently explore and analyse their data and generate publication-ready figures.

microViz offers tools including:

- Taxonomic association trees: a visualisation approach for compact and intuitive visualization of numerous microbe-metadata associations derived from (multivariable) statistical models;

- Interactive exploration of microbiota data: using paired ordination and abundance barchart visualizations in R Shiny;

- Iris plots: a novel approach pairing ordination and polar barcharts for comprehensive, intuitive and compact visualization of the similarity and composition of hundreds of microbial ecosystems;

- Functions for generating publication-ready ordination plots with ggplot2, accommodating constrained ordination, biplots and triplots, and automatic captioning designed to promote methodological transparency and reproducibility.

# Statement of need

Modern microbiome research typically involves the use of next-generation-sequencing methods to profile the relative abundance of hundreds of microbial taxa across tens or hundreds of samples. Alongside increasing sample sizes, the amount of relevant metadata collected is growing, particularly in human cohort studies. These trends all increase the size and complexity of the resulting dataset, which makes its exploration, statistical analysis, and presentation  increasingly challenging.

ps_extra class for ordination

- simple list object to contain phyloseq and associated info relating to the ordination produced.
- auto captioning (for development/internal communication and promoting good reporting)
- reliability, harder to make mistakes if objects kept together. (also permanova and ordination on same data)

dplyr toolset for phyloseq:

Beyond the main visualization functionality, microViz provides a suite of powerful tools for working with phyloseq objects including wrappers that bring functions from the popular dplyr package to phyloseq, to easily filter, select join, mutate and arrange sample data. All microViz functions are designed to work with Râ€™s pipe operator (%>%) , to chain successive functions together and improve code readability. For user convenience, microViz documentation is hosted online via a pkgdown website on Github Pages, with extensive examples of code and output using example datasets. 

taxatree plots:

- spurious findings at lower taxonomic ranks? sensitivity of methods?
- intuitive presentation of the balance inherent in compositional data
- metacoder similar but inflexible?
- ability to attach multivariable model results important.
- graphlan ? microbiomeViz ? connection to R and phyloseq useful here and detached from phylogenetic tree which is not always available (does graphlan really need phylogenetic tree?)?




