---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# microViz

<!-- badges: start -->
<!-- badges: end -->

The goal of microViz is to make wrangling, stats and visualisation of microbiome (16S) sequencing readcount data easier, modular and reproducible.
microViz wraps and extends core microbial ecology packages like phyloseq, vegan, and microbiome.

## Installation

You can install the latest available package version using the following code.

``` r
# # How to authenticate your R to install from this private github repo # #
# install.packages("usethis")
# usethis::browse_github_pat() 
# # click "generate token" on the webpage and then follow instructions in the R console.
# # For more help see: https://happygitwithr.com/github-pat.html#step-by-step

# # Installing the latest version of this package # #
# install.packages("devtools")
devtools::install_github("MUMC-MEDMIC/microViz")
# # advanced tip: add @<commit-hash> after microViz to install a particular version
```

## Setup

```{r load, message=FALSE, warning=FALSE}
library(microViz)
library(phyloseq)
library(vegan)
library(microbiome)
library(dplyr)
library(ggplot2)
```

```{r data setup}
# get some example data
data("dietswap", package = "microbiome")

# create a couple of numerical variables to use as constraints or conditions
sample_data(dietswap)$weight <-
  recode(sample_data(dietswap)$bmi_group, obese = 3, overweight = 2, lean = 1)
# add a couple of missing values to demo automated dropping of observations with missings
sample_data(dietswap)$female <-
  if_else(sample_data(dietswap)$sex == "female", true = 1, false = 0)
sample_data(dietswap)$female[c(3, 4)] <- NA
```


## Example: ordination plot workflow

You want to visually see if overall microbial ecosystem composition differs markedly between groups, e.g. BMI.

Here is one option to try first:

1. Filter out rare taxa (e.g. remove Genera not present in at least 10% of samples) - use `tax_filter()`
2. Aggregate the taxa into bacterial families (for example) - use `tax_agg()`
3. Transform the microbial data with the centre-log-ratio transformation - use `tax_transform()`
4. Perform PCA with the clr-transformed features (equivalent to aitchison distance PCoA) - use `ordin8()`
5. Plot the first 2 axes of this PCA ordination, colouring samples by group and adding taxon loading arrows to visualise which taxa generally differ across your samples - use `plot_ordin8()`
6. Customise the theme of the ggplot as you like and/or add features like ellipses or annotations

```{r ordination-plot, dpi=300}
# perform ordination
unconstrained_aitchison_pca <- 
  dietswap %>%
  tax_filter(min_prevalence = 0.1, tax_level = 'Genus') %>% 
  tax_agg("Family") %>%
  tax_transform('clr') %>%
  ordin8(method = 'RDA')

# create plot
pca_plot <- unconstrained_aitchison_pca %>%
  plot_ordin8(plot_taxa = 1:5, colour = 'bmi_group', auto_title = TRUE)

# customise plot
customised_plot <- pca_plot + 
  stat_ellipse(aes(linetype = bmi_group, colour = bmi_group)) + 
  scale_colour_brewer(palette = 'Set1') + 
  xlim(c(-1.5,2.75)) + theme(legend.position = 'bottom')

# show plot
customised_plot
  
```

# Session info
```{r}
options(width = 100)
devtools::session_info()
```

