---
output: github_document
editor_options: 
  markdown: 
    wrap: sentence
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
options(width = 100)
```

# microViz <a href='https://david-barnett.github.io/microViz/index.html'><img src="man/figures/microViz-smaller.png" align="right" height="139"/></a>

<!-- badges: start -->

[![R-CMD-check](https://github.com/david-barnett/microViz/workflows/R-CMD-check/badge.svg)](https://github.com/david-barnett/microViz/actions) [![codecov](https://codecov.io/gh/david-barnett/microViz/branch/main/graph/badge.svg?token=C1EoVkhnxA)](https://codecov.io/gh/david-barnett/microViz) [![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.4680616.svg)](https://doi.org/10.5281/zenodo.4680616) ![Docker Cloud Build Status](https://img.shields.io/docker/cloud/build/barnettdavid/microviz-rocker-verse)

<!-- badges: end -->

:package: microViz is an R package for analysis and visualization of microbiome sequencing data.
microViz functions are intended to be easy to use and flexible.
microViz extends and complements popular microbial ecology packages like phyloseq, vegan, and microbiome.

**See the documentation website for full details and examples:** <https://david-barnett.github.io/microViz/>

-   The ["Get started"](https://david-barnett.github.io/microViz/articles/articles/microViz.html) page, and also [this ReadMe](https://david-barnett.github.io/microViz/index.html), show a few example analyses.

-   The [reference](https://david-barnett.github.io/microViz/reference/index.html) page lists all functions and links to their documentation and examples

-   Example video of [Interactive Shiny app](https://david-barnett.github.io/microViz/articles/articles/Interactive-Ordination.html) for exploring your ordination plots!

-   Other articles pages give tutorials and more complicated examples

    -   [Working with phyloseq objects](https://david-barnett.github.io/microViz/articles/articles/phyloseq.html)

    -   [Fixing your taxa table with tax_fix](https://david-barnett.github.io/microViz/articles/articles/tax-fixing.html)

    -   [Visualising taxonomic compositions with customised barplots](https://david-barnett.github.io/microViz/articles/articles/Visualising-compositions.html)

    -   (more coming soon! raise an issue on github if you have questions/requests)

-   The [changelog](https://david-barnett.github.io/microViz/news/index.html) describes important changes in new microViz package versions

## Installation

You can install the latest available microViz package version using the following instructions.

``` r
# If you are on Windows you will need to install RTools so that your computer can build this package
# Follow instructions here: http://jtleek.com/modules/01_DataScientistToolbox/02_10_rtools/

# If you are on macOS, you might need to install xquartz to make the heatmaps work (ComplexHeatmaps package)
# You can do this with homebrew, running the following command in your mac's Terminal: brew install --cask xquartz 

# for installing from github you'll need the devtools package
install.packages("devtools", type = "binary") # (not binary if you're on linux)

# If you don't already have the latest versions of phyloseq and microbiome, you can install these from Bioconductor:

if (!requireNamespace("BiocManager", quietly = TRUE))
 install.packages("BiocManager", type = "binary") # (not binary if you're on linux)
BiocManager::install(c("phyloseq", "microbiome"))

# # Installing the latest "released" version of this package # #
devtools::install_github("david-barnett/microViz@0.7.2") # check 0.7.2 is the latest release

# # If you encounter bugs: please try installing the very latest development version:
devtools::install_github("david-barnett/microViz")

# advanced tip: add @<commit-hash> after microViz to install a version from a particular commit
```

:package: I highly recommend using [renv](https://rstudio.github.io/renv/index.html) for managing your R package installations across multiple projects.

:whale: Alternatively, for Docker users an image with the main branch installed is available at: <https://hub.docker.com/r/barnettdavid/microviz-rocker-verse>

:date: microViz is tested to work with R version 4.0.0 and higher, on Windows, MacOS, and Ubuntu 18 and 20.
R version 3.6.\* should probably work, but I don't formally test this.

## A few examples

```{r load, message=FALSE}
library(microViz)
library(phyloseq)
library(dplyr)
library(ggplot2)
```

```{r data setup}
# get some example data
data("dietswap", package = "microbiome")

# create a couple of numerical variables to use as constraints or conditions
dietswap <- dietswap %>%
  ps_mutate(
    weight = recode(bmi_group, obese = 3, overweight = 2, lean = 1),
    female = if_else(sex == "female", true = 1, false = 0)
  )
# add a couple of missing values to show how microViz handles missing data
sample_data(dietswap)$female[c(3, 4)] <- NA
```

### Looking at your data

You have quite a few samples in your phyloseq object, and would like to visualise their compositions.
Perhaps these example data differ across BMI groups?

```{r fig.height=8, fig.width=10, dpi=300}
dietswap %>%
  tax_filter(min_prevalence = 1) %>%
  comp_barplot(
    tax_level = "Genus",  n_taxa = 12, 
    merge_other = FALSE 
    # set merge_other = TRUE (the default) to remove outlines from inside "other" category
  ) +
  facet_wrap("bmi_group", nrow = 3, scales = "free") +
  labs(x = NULL, y = NULL) +
  theme(
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank()
  )
```

### Example ordination plot workflow

Maybe visually inspecting all your samples isn't quite what you want.
Ordination methods can also help you to visualise if overall microbial ecosystem composition differs markedly between groups, e.g.
BMI.

Here is one option to try first:

1.  Filter out rare taxa (e.g. remove Genera not present in at least 10% of samples) - use `tax_filter()`
2.  Aggregate the taxa into bacterial families (for example) - use `tax_agg()`
3.  Transform the microbial data with the centre-log-ratio transformation - use `tax_transform()`
4.  Perform PCA with the clr-transformed features (equivalent to aitchison distance PCoA) - use `ord_calc()`
5.  Plot the first 2 axes of this PCA ordination, colouring samples by group and adding taxon loading arrows to visualise which taxa generally differ across your samples - use `ord_plot()`
6.  Customise the theme of the ggplot as you like and/or add features like ellipses or annotations

```{r ordination-plot, dpi=300}
# perform ordination
unconstrained_aitchison_pca <-
  dietswap %>%
  tax_filter(min_prevalence = 0.1, tax_level = "Genus") %>%
  tax_agg("Family") %>%
  tax_transform("clr") %>%
  ord_calc() 
# ord_calc will automatically infer you want a "PCA" here
# specify explicitly with method = "PCA", or you can pick another method

# create plot
pca_plot <- unconstrained_aitchison_pca %>%
  ord_plot(
    plot_taxa = 1:6, colour = "bmi_group", size = 2,
    tax_vec_length = 0.5,
    auto_caption = TRUE
  )

# customise plot
customised_plot <- pca_plot +
  stat_ellipse(aes(linetype = bmi_group, colour = bmi_group)) +
  scale_colour_brewer(palette = "Set1") +
  theme(legend.position = "bottom")

# show plot
customised_plot
```

### PERMANOVA

You visualised your ordinated data in the plot above.
Below you can see how to perform a PERMANOVA to test the significance of BMI's association with overall microbial composition.
This example uses the Family-level aitchison distance to correspond with the plot above.

```{r}
# calculate distances
aitchison_dists <-
  dietswap %>%
  tax_filter(min_prevalence = 0.1, tax_level = "Genus") %>%
  tax_agg("Family") %>%
  tax_transform("identity") %>% 
  dist_calc("aitchison")

# the more permutations you request, the longer it takes
# but also the more stable and precise your p-values become
aitchison_perm <- dist_permanova(
  data = aitchison_dists,
  seed = 1234, # for set.seed to ensure reproducibility of random process
  n_perms = 99, # you should use at least 999!
  n_processes = 1,
  variables = "bmi_group + female"
)
# view the permanova results
perm_get(aitchison_perm) %>% as.data.frame()
# view the info stored about the distance calculation
info_get(aitchison_perm)
```

### Constrained ordination

You could visualise the effect of the (numeric/logical) variables in your permanova directly using the ord_plot function with constraints.

```{r constrained-ord, fig.height=5, fig.width=7, dpi=300}
perm2 <- dist_permanova(data = aitchison_dists, variables = c("weight", "female"), seed = 321)
perm_get(perm2)

ord_calc(perm2, constraints = c("weight", "female")) %>%
  ord_plot(
    colour = "sex", size = 2,
    constraint_vec_style = list(colour = "black", size = 1.15),
    constraint_lab_style = list(colour = "black")
  ) +
  stat_ellipse(aes(colour = sex)) + scale_color_brewer(palette = "Dark2")
```

### Heatmaps

microViz heatmaps are powered by `ComplexHeatmap` and annotated with taxa prevalence and/or abundance.

```{r heatmap, dpi=300}
# set up the data with numerical variables and filter to top taxa
psq <- dietswap %>%
 ps_mutate(
   weight = recode(bmi_group, obese = 3, overweight = 2, lean = 1),
   female = if_else(sex == "female", true = 1, false = 0),
   african = if_else(nationality == "AFR", true = 1, false = 0)
 ) %>% 
  tax_filter(
    tax_level = "Genus", min_prevalence = 1 / 10, min_sample_abundance = 1 / 10
  ) %>% 
  tax_transform("identity", rank = "Genus")

# randomly select 30 taxa from the 50 most abundant taxa (just for an example)
set.seed(123)
taxa <- sample(microbiome::top_taxa(ps_get(psq))[1:50], size = 30)
# actually draw the heatmap
cor_heatmap(psq, taxa, anno_tax = tax_anno(undetected = 50))
```

## Citation

:innocent: If you find any part of microViz useful to your work, please consider citing the Zenodo archive DOI: <https://doi.org/10.5281/zenodo.4644057>

Publication pending: microViz has been submitted to a peer-reviewed journal.

## Session info

```{r}
devtools::session_info()
```
