---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
options(width = 100)
```

# microViz

<!-- badges: start -->
<!-- badges: end -->

microViz package provides functions for wrangling, stats and visualisation of microbiome (16S) sequencing readcount data. These functions are intended to be easy to use (with clear documentation) and modular (for extensibility and flexibility). 
microViz wraps and extends popular microbial ecology packages like phyloseq, vegan, and microbiome.

**See the documentation website for full details and examples:** https://david-barnett.github.io/microViz/

- The [reference](https://david-barnett.github.io/microViz/reference/index.html) page lists all functions and links to their documentation and examples
- The articles pages discuss and give more complicated examples (often under development, e.g. [polar composition plots](https://david-barnett.github.io/microViz/articles/PCA-sorted-polar-composition-plots.html))
- The [changelog](https://david-barnett.github.io/microViz/news/index.html) describe important changes in new microViz package versions

## Installation

You can install the latest available microViz package version using the following instructions.

``` r
# If you are on windows you will need to install RTools so that your computer can build this package
# Follow instructions here: http://jtleek.com/modules/01_DataScientistToolbox/02_10_rtools/

# # How to authenticate your R to install from this private github repo # #
install.packages("usethis")
install.packages("credentials")
usethis::create_github_token()
# a github webpage will open (you might need to log in)
# Enter a PAT name that reminds you which computer you are currently on
# Click "generate token" at the bottom of the webpage, then run the following code in R, then copy/paste the PAT when requested:
credentials::set_github_pat()
# More help/instructions may be found on: https://happygitwithr.com/credential-caching.html#how-to-get-a-pat

# If you don't already have the latest versions of phyloseq and microbiome, you should install these from bioconductor:

if (!requireNamespace("BiocManager", quietly = TRUE))
 install.packages("BiocManager")
BiocManager::install(c("phyloseq", "microbiome"))

# # Installing the latest version of this package # #
install.packages("devtools")
devtools::install_github("MUMC-MEDMIC/microViz@0.1.2") # change 0.1.2 to the latest release version
# advanced tip: add @<commit-hash> after microViz to install a version from a particular commit
# If you get an error including something like: 
# "check that you have the required permissions" 
# then setting your credentials/PAT may have failed

# optionally install the corncob package from github (for beta binomial models)
devtools::install_github("bryandmartin/corncob@338323e9")

```

## Examples below

```{r load, message=FALSE}
library(microViz)
library(phyloseq)
library(vegan)
library(microbiome)
library(dplyr)
library(ggplot2)
```

```{r data setup}
# get some example data
data("dietswap", package = "microbiome")

# create a couple of numerical variables to use as constraints or conditions
dietswap <- dietswap %>% 
  ps_mutate(
    weight = recode(bmi_group, obese = 3, overweight = 2, lean = 1),
    female = if_else(sex == "female", true = 1, false = 0)
  )

sample_data(dietswap)$female[c(3, 4)] <- NA
```

## Looking at your data

You have quite a few samples in your phyloseq object, and would like to visualise their compositions.
Perhaps these example data differ across BMI groups?
```{r fig.height=8, fig.width=10, dpi=300}
dietswap %>%
  plot_comp_bar(
    tax_level = "Genus",
    n_taxa = 10
  ) + 
  facet_wrap("bmi_group", nrow = 3, scales = "free") +
  labs(x = NULL, y = NULL) + 
  theme(
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank()
  )
```


## Example ordination plot workflow

Maybe visually inspecting all your samples isn't quite what you want.
Ordination methods can also help you to visualise if overall microbial ecosystem composition differs markedly between groups, e.g. BMI.

Here is one option to try first:

1. Filter out rare taxa (e.g. remove Genera not present in at least 10% of samples) - use `tax_filter()`
2. Aggregate the taxa into bacterial families (for example) - use `tax_agg()`
3. Transform the microbial data with the centre-log-ratio transformation - use `tax_transform()`
4. Perform PCA with the clr-transformed features (equivalent to aitchison distance PCoA) - use `ord_calc()`
5. Plot the first 2 axes of this PCA ordination, colouring samples by group and adding taxon loading arrows to visualise which taxa generally differ across your samples - use `ord_plot()`
6. Customise the theme of the ggplot as you like and/or add features like ellipses or annotations

```{r ordination-plot, dpi=300}
# perform ordination
unconstrained_aitchison_pca <-
  dietswap %>%
  tax_filter(min_prevalence = 0.1, tax_level = "Genus") %>%
  tax_agg("Family") %>%
  tax_transform("clr") %>%
  ord_calc(method = "RDA")

# create plot
pca_plot <- unconstrained_aitchison_pca %>%
  ord_plot(
    plot_taxa = 1:6, colour = "bmi_group", 
    size = 2,
    tax_vec_length = 0.8,
    auto_caption = TRUE)

# customise plot
customised_plot <- pca_plot +
  stat_ellipse(aes(linetype = bmi_group, colour = bmi_group)) +
  scale_colour_brewer(palette = "Set1") +
  xlim(c(-1.5, 2.75)) + theme(legend.position = "bottom")

# show plot
customised_plot
```

## PERMANOVA

You visualised your ordinated data in the plot above. 
Below you can see how to perform a PERMANOVA to test the significance of BMI.
This example uses the Family-level aitchison distance to correspond with the plot above.
```{r}
# calculate distances
aitchison_dists <-
  dietswap %>%
  tax_filter(min_prevalence = 0.1, tax_level = "Genus") %>%
  tax_agg("Family") %>%
  dist_calc("aitchison")

# the more permutations you request, the longer it takes,
# but also the more stable and precise your p-values become
aitchison_perm <- permanova(
  data = aitchison_dists,
  seed = 1234, # for set.seed to ensure reproducibility of random process
  n_perms = 99, n_processes = 1, 
  variables = c("bmi_group")
)
# view the permanova results
aitchison_perm$permanova
# view the info stored about the distance calculation
aitchison_perm$info
```
## Constrained ordination
You could visualise the effect of the (numeric/logical) variables in your permanova directly using the ord_plot function with constraints.

```{r constrained-ord, fig.height=5, fig.width=7, dpi=300}
perm2 <- permanova(data = aitchison_dists, variables = c("weight", "female"), seed = 321)
perm2$permanova

ord_calc(perm2, constraints = c("weight", "female")) %>%
  ord_plot(
    colour = "sex", size = 2,
    constraint_vec_style = list(colour = "black", size = 1.15),
    constraint_lab_style = list(colour = "black"),
    constraint_vec_length = 1.5, constraint_lab_length = 1.8) +
  stat_ellipse(aes(colour = sex)) + scale_color_brewer(palette = "Dark2")
```



# Session info
```{r}
devtools::session_info()
```
