<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Plot (grouped and ordered) compositional barplots — comp_barplot • microViz</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Plot (grouped and ordered) compositional barplots — comp_barplot" />
<meta property="og:description" content="Stacked barplots showing composition of phyloseq samples for a specified number of coloured taxa.
Normally your phyloseq object should contain counts data,
as by default comp_barplot() performs the &quot;compositional&quot; taxa transformation for you,
and requires count input for some sample_order methods!" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">microViz</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.7.3.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/articles/microViz.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/articles/Interactive-Ordination.html">Interactive Ordination Plot</a>
    </li>
    <li>
      <a href="../articles/articles/Visualising-compositions.html">Visualising compositions</a>
    </li>
    <li>
      <a href="../articles/articles/phyloseq.html">Working with phyloseq objects</a>
    </li>
    <li>
      <a href="../articles/articles/tax-fixing.html">Fixing your tax_table</a>
    </li>
    <li>
      <a href="../articles/ordination.html">Ordination plots</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/david-barnett/microViz/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Plot (grouped and ordered) compositional barplots</h1>
    <small class="dont-index">Source: <a href='https://github.com/david-barnett/microViz/blob/master/R/comp_barplot.R'><code>R/comp_barplot.R</code></a></small>
    <div class="hidden name"><code>comp_barplot.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Stacked barplots showing composition of phyloseq samples for a specified number of coloured taxa.
Normally your phyloseq object should contain counts data,
as by default <code>comp_barplot()</code> performs the "compositional" taxa transformation for you,
and requires count input for some sample_order methods!</p>
    </div>

    <pre class="usage"><span class='fu'>comp_barplot</span><span class='op'>(</span>
  <span class='va'>ps</span>,
  <span class='va'>tax_level</span>,
  n_taxa <span class='op'>=</span> <span class='fl'>8</span>,
  tax_order <span class='op'>=</span> <span class='va'>sum</span>,
  merge_other <span class='op'>=</span> <span class='cn'>TRUE</span>,
  taxon_renamer <span class='op'>=</span> <span class='kw'>function</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span> <span class='fu'><a href='https://rdrr.io/r/base/identity.html'>identity</a></span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span>,
  sample_order <span class='op'>=</span> <span class='st'>"aitchison"</span>,
  order_with_all_taxa <span class='op'>=</span> <span class='cn'>FALSE</span>,
  label <span class='op'>=</span> <span class='st'>"SAMPLE"</span>,
  group_by <span class='op'>=</span> <span class='cn'>NA</span>,
  facet_by <span class='op'>=</span> <span class='cn'>NA</span>,
  bar_width <span class='op'>=</span> <span class='fl'>1</span>,
  bar_outline_colour <span class='op'>=</span> <span class='st'>"grey5"</span>,
  bar_outline_width <span class='op'>=</span> <span class='fl'>0.1</span>,
  palette <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"lightgrey"</span>, <span class='fu'><a href='https://rdrr.io/r/base/rev.html'>rev</a></span><span class='op'>(</span><span class='fu'><a href='distinct_palette.html'>distinct_palette</a></span><span class='op'>(</span><span class='va'>n_taxa</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>,
  tax_transform_for_ordering <span class='op'>=</span> <span class='st'>"identity"</span>,
  tax_transform_for_plot <span class='op'>=</span> <span class='st'>"compositional"</span>,
  seriate_method <span class='op'>=</span> <span class='st'>"OLO_ward"</span>,
  keep_all_vars <span class='op'>=</span> <span class='cn'>TRUE</span>,
  interactive <span class='op'>=</span> <span class='cn'>FALSE</span>,
  max_taxa <span class='op'>=</span> <span class='fl'>10000</span>,
  <span class='va'>...</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>ps</th>
      <td><p>phyloseq object</p></td>
    </tr>
    <tr>
      <th>tax_level</th>
      <td><p>taxonomic aggregation level (from rank_names(ps))</p></td>
    </tr>
    <tr>
      <th>n_taxa</th>
      <td><p>how many taxa to show distinct colours for (all others grouped into "Other")</p></td>
    </tr>
    <tr>
      <th>tax_order</th>
      <td><p>order of taxa within the bars, either a function for tax_sort (e.g. sum),
or a vector of (all) taxa names at tax_level to set order manually</p></td>
    </tr>
    <tr>
      <th>merge_other</th>
      <td><p>if FALSE, taxa coloured/filled as "other" remain distinct,
and so can have bar outlines drawn around them</p></td>
    </tr>
    <tr>
      <th>taxon_renamer</th>
      <td><p>function that takes taxon names and returns modified names for legend</p></td>
    </tr>
    <tr>
      <th>sample_order</th>
      <td><p>vector of sample names;
or any distance measure in dist_calc that doesn't require phylogenetic tree;
or "default" for the order returned by phyloseq::sample_names(ps)</p></td>
    </tr>
    <tr>
      <th>order_with_all_taxa</th>
      <td><p>if TRUE, this will always use all taxa (not just the top n_taxa)
to calculate distances for sample ordering</p></td>
    </tr>
    <tr>
      <th>label</th>
      <td><p>sample label variable name</p></td>
    </tr>
    <tr>
      <th>group_by</th>
      <td><p>splits dataset by this variable (must be categorical)</p><ul>
<li><p>resulting in a list of plots, one for each level of the group_by variable.</p></li>
</ul></td>
    </tr>
    <tr>
      <th>facet_by</th>
      <td><p>facets plots by this variable (must be categorical). If group_by is also set
the faceting will occur separately in the plot for each group.</p></td>
    </tr>
    <tr>
      <th>bar_width</th>
      <td><p>default 1 avoids random gapping otherwise seen with many samples
(set to something less than 1 to introduce gaps between fewer samples)</p></td>
    </tr>
    <tr>
      <th>bar_outline_colour</th>
      <td><p>line colour separating taxa and samples
(use NA for none)</p></td>
    </tr>
    <tr>
      <th>bar_outline_width</th>
      <td><p>width of line separating taxa and samples
(for no outlines set bar_outline_colour = NA)</p></td>
    </tr>
    <tr>
      <th>palette</th>
      <td><p>palette for taxa fill colours</p></td>
    </tr>
    <tr>
      <th>tax_transform_for_ordering</th>
      <td><p>transformation of taxa values used before ordering samples by similarity</p></td>
    </tr>
    <tr>
      <th>tax_transform_for_plot</th>
      <td><p>default "compositional" draws proportions of total counts per sample,
but you could reasonably use another transformation,
e.g. "identity", if you have truly quantitative microbiome profiling data</p></td>
    </tr>
    <tr>
      <th>seriate_method</th>
      <td><p>name of any ordering method suitable for distance matrices
(see ?seriation::seriate)</p></td>
    </tr>
    <tr>
      <th>keep_all_vars</th>
      <td><p>FALSE may speed up internal melting with ps_melt for large phyloseq objects
but TRUE is required for some post-hoc plot customisation</p></td>
    </tr>
    <tr>
      <th>interactive</th>
      <td><p>creates plot suitable for use with ggiraph</p></td>
    </tr>
    <tr>
      <th>max_taxa</th>
      <td><p>maximum distinct taxa groups to show
(only really useful for limiting complexity of interactive plots
e.g. within ord_explore)</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>extra arguments passed to facet_wrap() (if facet_by is not NA)</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>ggplot or list of harmonised ggplots</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    
<ul>
<li><p>sample_order: Either specify a list of sample names to order manually, or the bars/samples can/will be sorted by similarity, according to a specified distance measure (default aitchison),</p></li>
<li><p>seriate_method specifies a seriation/ordering algorithm (default Ward hierarchical clustering with optimal leaf ordering, see seriation::list_seriation_methods())</p></li>
<li><p>group_by: You can group the samples on distinct plots by levels of a variable in the phyloseq object. The list of ggplots produced can be arranged flexibly with the patchwork package functions. If you want to group by several variables you can create an interaction variable with interaction(var1, var2) in the phyloseq sample_data BEFORE using comp_barplot.</p></li>
<li><p>facet_by can allow faceting of your plot(s) by a grouping variable. Using this approach is less flexible than using group_by but means you don't have to arrange a list of plots yourself like with the group_by argument. Using facet_by is equivalent to adding a call to facet_wrap(facets = facet_by, scales = "free") to your plot(s). Calling facet_wrap() yourself is itself a more flexible option as you can add other arguments like the number of rows etc. However you must use keep_all_vars = TRUE if you will add faceting manually.</p></li>
<li><p>bar_width: No gaps between bars, unless you want them (decrease width argument to add gaps between bars).</p></li>
<li><p>bar_outline_colour: Bar outlines default to "grey5" for almost black outlines. Use NA if you don't want outlines.</p></li>
<li><p>merge_other: controls whether bar outlines can be drawn around individual (lower abundance) taxa that are grouped in "other" category. If you want to see the diversity of taxa in "other" use merge_taxa = FALSE, or use TRUE if you prefer the cleaner merged look</p></li>
<li><p>palette: Default colouring is consistent across multiple plots if created with the group_by argument, and the defaults scheme retains the colouring of the most abundant taxa irrespective of n_taxa</p></li>
</ul>




    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://ggplot2.tidyverse.org'>ggplot2</a></span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='va'>dietswap</span>, package <span class='op'>=</span> <span class='st'>"microbiome"</span><span class='op'>)</span>

<span class='co'># illustrative simple customised example</span>
<span class='va'>dietswap</span> <span class='op'>%&gt;%</span>
  <span class='fu'><a href='ps_filter.html'>ps_filter</a></span><span class='op'>(</span><span class='va'>timepoint</span> <span class='op'>==</span> <span class='fl'>1</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>comp_barplot</span><span class='op'>(</span>
    tax_level <span class='op'>=</span> <span class='st'>"Family"</span>, n_taxa <span class='op'>=</span> <span class='fl'>8</span>,
    bar_outline_colour <span class='op'>=</span> <span class='cn'>NA</span>,
    sample_order <span class='op'>=</span> <span class='st'>"bray"</span>,
    bar_width <span class='op'>=</span> <span class='fl'>0.7</span>,
    taxon_renamer <span class='op'>=</span> <span class='va'>toupper</span>
  <span class='op'>)</span> <span class='op'>+</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/coord_flip.html'>coord_flip</a></span><span class='op'>(</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>Registered S3 method overwritten by 'seriation':</span>
#&gt; <span class='message'>  method         from </span>
#&gt; <span class='message'>  reorder.hclust vegan</span></div><div class='img'><img src='comp_barplot-1.png' alt='' width='700' height='433' /></div><div class='input'>
<span class='co'># Order samples by the value of one of more sample_data variables.</span>
<span class='co'># Use ps_arrange and set sample_order = "default" in comp_barplot.</span>
<span class='co'># ps_mutate is also used here to create an informative variable for axis labelling</span>
<span class='va'>dietswap</span> <span class='op'>%&gt;%</span>
  <span class='fu'><a href='ps_mutate.html'>ps_mutate</a></span><span class='op'>(</span>subject_timepoint <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/interaction.html'>interaction</a></span><span class='op'>(</span><span class='va'>subject</span>, <span class='va'>timepoint</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'><a href='ps_filter.html'>ps_filter</a></span><span class='op'>(</span><span class='va'>nationality</span> <span class='op'>==</span> <span class='st'>"AAM"</span>, <span class='va'>group</span> <span class='op'>==</span> <span class='st'>"DI"</span>, <span class='va'>sex</span> <span class='op'>==</span> <span class='st'>"female"</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'><a href='ps_arrange.html'>ps_arrange</a></span><span class='op'>(</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/desc.html'>desc</a></span><span class='op'>(</span><span class='va'>subject</span><span class='op'>)</span>, <span class='fu'><a href='https://dplyr.tidyverse.org/reference/desc.html'>desc</a></span><span class='op'>(</span><span class='va'>timepoint</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>comp_barplot</span><span class='op'>(</span>
    tax_level <span class='op'>=</span> <span class='st'>"Genus"</span>, n_taxa <span class='op'>=</span> <span class='fl'>12</span>,
    sample_order <span class='op'>=</span> <span class='st'>"default"</span>,
    bar_width <span class='op'>=</span> <span class='fl'>0.7</span>,
    bar_outline_colour <span class='op'>=</span> <span class='st'>"black"</span>,
    order_with_all_taxa <span class='op'>=</span> <span class='cn'>TRUE</span>,
    label <span class='op'>=</span> <span class='st'>"subject_timepoint"</span>
  <span class='op'>)</span> <span class='op'>+</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/coord_flip.html'>coord_flip</a></span><span class='op'>(</span><span class='op'>)</span>
</div><div class='img'><img src='comp_barplot-2.png' alt='' width='700' height='433' /></div><div class='input'>
<span class='co'># how many taxa are in those light grey "other" bars?</span>
<span class='co'># set merge_other, to find out (&amp; remember to set a bar_outline_colour)</span>
<span class='va'>dietswap</span> <span class='op'>%&gt;%</span>
  <span class='fu'><a href='ps_mutate.html'>ps_mutate</a></span><span class='op'>(</span>subject_timepoint <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/interaction.html'>interaction</a></span><span class='op'>(</span><span class='va'>subject</span>, <span class='va'>timepoint</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'><a href='ps_filter.html'>ps_filter</a></span><span class='op'>(</span><span class='va'>nationality</span> <span class='op'>==</span> <span class='st'>"AAM"</span>, <span class='va'>group</span> <span class='op'>==</span> <span class='st'>"DI"</span>, <span class='va'>sex</span> <span class='op'>==</span> <span class='st'>"female"</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'><a href='ps_arrange.html'>ps_arrange</a></span><span class='op'>(</span><span class='fu'><a href='https://dplyr.tidyverse.org/reference/desc.html'>desc</a></span><span class='op'>(</span><span class='va'>subject</span><span class='op'>)</span>, <span class='fu'><a href='https://dplyr.tidyverse.org/reference/desc.html'>desc</a></span><span class='op'>(</span><span class='va'>timepoint</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>comp_barplot</span><span class='op'>(</span>
    tax_level <span class='op'>=</span> <span class='st'>"Genus"</span>, n_taxa <span class='op'>=</span> <span class='fl'>12</span>,
    sample_order <span class='op'>=</span> <span class='st'>"default"</span>,
    merge_other <span class='op'>=</span> <span class='cn'>FALSE</span>,
    bar_width <span class='op'>=</span> <span class='fl'>0.7</span>,
    bar_outline_colour <span class='op'>=</span> <span class='st'>"black"</span>,
    order_with_all_taxa <span class='op'>=</span> <span class='cn'>TRUE</span>,
    label <span class='op'>=</span> <span class='st'>"subject_timepoint"</span>
  <span class='op'>)</span> <span class='op'>+</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/coord_flip.html'>coord_flip</a></span><span class='op'>(</span><span class='op'>)</span>
</div><div class='img'><img src='comp_barplot-3.png' alt='' width='700' height='433' /></div><div class='input'>

<span class='co'># Often to compare groups, average compositions are presented</span>
<span class='va'>p1</span> <span class='op'>&lt;-</span> <span class='fu'>phyloseq</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/phyloseq/man/merge_samples-methods.html'>merge_samples</a></span><span class='op'>(</span><span class='va'>dietswap</span>, group <span class='op'>=</span> <span class='st'>"group"</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>comp_barplot</span><span class='op'>(</span>
    tax_level <span class='op'>=</span> <span class='st'>"Genus"</span>, n_taxa <span class='op'>=</span> <span class='fl'>12</span>,
    sample_order <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"ED"</span>, <span class='st'>"HE"</span>, <span class='st'>"DI"</span><span class='op'>)</span>,
    bar_width <span class='op'>=</span> <span class='fl'>0.8</span>
  <span class='op'>)</span> <span class='op'>+</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/coord_flip.html'>coord_flip</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>labs</a></span><span class='op'>(</span>x <span class='op'>=</span> <span class='cn'>NULL</span>, y <span class='op'>=</span> <span class='cn'>NULL</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='warning'>Warning: NAs introduced by coercion</span></div><div class='input'><span class='va'>p1</span>
</div><div class='img'><img src='comp_barplot-4.png' alt='' width='700' height='433' /></div><div class='input'>
<span class='co'># However that "group-averaging" approach hides a lot of within-group variation</span>
<span class='va'>p2</span> <span class='op'>&lt;-</span> <span class='fu'>comp_barplot</span><span class='op'>(</span><span class='va'>dietswap</span>,
  tax_level <span class='op'>=</span> <span class='st'>"Genus"</span>, n_taxa <span class='op'>=</span> <span class='fl'>12</span>, group_by <span class='op'>=</span> <span class='st'>"group"</span>,
  sample_order <span class='op'>=</span> <span class='st'>"euclidean"</span>, bar_outline_colour <span class='op'>=</span> <span class='cn'>NA</span>
<span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'>patchwork</span><span class='fu'>::</span><span class='fu'><a href='https://patchwork.data-imaginist.com/reference/wrap_plots.html'>wrap_plots</a></span><span class='op'>(</span>nrow <span class='op'>=</span> <span class='fl'>3</span>, guides <span class='op'>=</span> <span class='st'>"collect"</span><span class='op'>)</span> <span class='op'>&amp;</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/coord_flip.html'>coord_flip</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>&amp;</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>labs</a></span><span class='op'>(</span>x <span class='op'>=</span> <span class='cn'>NULL</span>, y <span class='op'>=</span> <span class='cn'>NULL</span><span class='op'>)</span> <span class='op'>&amp;</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/theme.html'>theme</a></span><span class='op'>(</span>axis.text.y <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_blank</a></span><span class='op'>(</span><span class='op'>)</span>, axis.ticks.y <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_blank</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>p2</span>
</div><div class='img'><img src='comp_barplot-5.png' alt='' width='700' height='433' /></div><div class='input'>
<span class='co'># Only from p2 you can see that the apparently higher average relative abundance</span>
<span class='co'># of Oscillospira in group DI is probably driven largely by a subgroup</span>
<span class='co'># of DI samples with relatively high Oscillospira.</span>

<span class='co'># make a list of 2 harmonised composition plots (grouped by sex)</span>
<span class='va'>p</span> <span class='op'>&lt;-</span> <span class='fu'>comp_barplot</span><span class='op'>(</span><span class='va'>dietswap</span>,
  n_taxa <span class='op'>=</span> <span class='fl'>15</span>, tax_level <span class='op'>=</span> <span class='st'>"Genus"</span>,
  bar_outline_colour <span class='op'>=</span> <span class='st'>"black"</span>, merge_other <span class='op'>=</span> <span class='cn'>TRUE</span>,
  sample_order <span class='op'>=</span> <span class='st'>"aitchison"</span>, group_by <span class='op'>=</span> <span class='st'>"sex"</span>
<span class='op'>)</span>

<span class='co'># plot them side by side with patchwork package</span>
<span class='va'>patch</span> <span class='op'>&lt;-</span> <span class='fu'>patchwork</span><span class='fu'>::</span><span class='fu'><a href='https://patchwork.data-imaginist.com/reference/wrap_plots.html'>wrap_plots</a></span><span class='op'>(</span><span class='va'>p</span>, ncol <span class='op'>=</span> <span class='fl'>2</span>, guides <span class='op'>=</span> <span class='st'>"collect"</span><span class='op'>)</span>
<span class='va'>patch</span> <span class='op'>&amp;</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/coord_flip.html'>coord_flip</a></span><span class='op'>(</span><span class='op'>)</span> <span class='co'># make bars in all plots horizontal (note: use &amp; instead of +)</span>
</div><div class='img'><img src='comp_barplot-6.png' alt='' width='700' height='433' /></div><div class='input'>
<span class='co'># beautifying tweak #</span>
<span class='co'># modify one plot in place (flip the order of the samples in the 2nd plot)</span>
<span class='co'># notice that the scaling is for the x-axis</span>
<span class='co'># (that's because coord_flip is used afterwards when displaying the plots</span>
<span class='va'>patch</span><span class='op'>[[</span><span class='fl'>2</span><span class='op'>]</span><span class='op'>]</span> <span class='op'>&lt;-</span> <span class='va'>patch</span><span class='op'>[[</span><span class='fl'>2</span><span class='op'>]</span><span class='op'>]</span> <span class='op'>+</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_discrete.html'>scale_x_discrete</a></span><span class='op'>(</span>limits <span class='op'>=</span> <span class='va'>rev</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>Scale for 'x' is already present. Adding another scale for 'x', which will</span>
#&gt; <span class='message'>replace the existing scale.</span></div><div class='input'><span class='co'># Explainer: rev() function takes current limits and reverses them.</span>
<span class='co'># You could also pass a completely arbitrary order, naming all samples</span>

<span class='co'># you can theme all plots with the &amp; operator</span>
<span class='va'>patch</span> <span class='op'>&amp;</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/coord_flip.html'>coord_flip</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>&amp;</span>
  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/theme.html'>theme</a></span><span class='op'>(</span>axis.text.y <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_text</a></span><span class='op'>(</span>size <span class='op'>=</span> <span class='fl'>5</span><span class='op'>)</span>, legend.text <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_text</a></span><span class='op'>(</span>size <span class='op'>=</span> <span class='fl'>6</span><span class='op'>)</span><span class='op'>)</span>
</div><div class='img'><img src='comp_barplot-7.png' alt='' width='700' height='433' /></div><div class='input'><span class='co'># See https://patchwork.data-imaginist.com/index.html</span>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by David Barnett.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


