<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>De-duplicate phyloseq samples — ps_dedupe • microViz</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>


<!-- docsearch -->
<script src="../docsearch.js"></script>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous" />
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script>



<meta property="og:title" content="De-duplicate phyloseq samples — ps_dedupe" />
<meta property="og:description" content="Use 1 or more variables in the sample_data to identify and remove duplicate samples (leaving 1 per category).
methods:
method = &quot;readcount&quot; keeps the one sample in each duplicate group with the highest total number of reads according to sample_sums
method = &quot;first&quot; keeps the first sample in each duplicate group encountered in the row order of the sample_data
method = &quot;last&quot; keeps the last sample in each duplicate group encountered in the row order of the sample_data

" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">microViz</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.7.10.9033</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="divider"></li>
    <li class="dropdown-header">Tutorials</li>
    <li>
      <a href="../articles/web-only/atlas1006.html">Example analyses with atlas1006 data</a>
    </li>
    <li>
      <a href="../articles/web-only/phyloseq.html">Working with phyloseq objects</a>
    </li>
    <li>
      <a href="../articles/web-only/tax-fixing.html">Fixing your tax_table</a>
    </li>
    <li>
      <a href="../articles/web-only/ordination.html">Ordination plots</a>
    </li>
    <li>
      <a href="../articles/web-only/ordination-interactive.html">Interactive Ordination Plot</a>
    </li>
    <li>
      <a href="../articles/web-only/compositions.html">Visualising compositions</a>
    </li>
    <li>
      <a href="../articles/web-only/videos.html">Videos</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/david-barnett/microViz/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
      <form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
        </div>
      </form>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>De-duplicate phyloseq samples</h1>
    <small class="dont-index">Source: <a href='https://github.com/david-barnett/microViz/blob/master/R/ps_dedupe.R'><code>R/ps_dedupe.R</code></a></small>
    <div class="hidden name"><code>ps_dedupe.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Use 1 or more variables in the sample_data to identify and remove duplicate samples (leaving 1 per category).</p>
<p><strong>methods:</strong></p><ul>
<li><p>method = "readcount" keeps the one sample in each duplicate group with the highest total number of reads according to sample_sums</p></li>
<li><p>method = "first" keeps the first sample in each duplicate group encountered in the row order of the sample_data</p></li>
<li><p>method = "last" keeps the last sample in each duplicate group encountered in the row order of the sample_data</p></li>
</ul>

    </div>

    <pre class="usage"><span class='fu'>ps_dedupe</span><span class='op'>(</span><span class='va'>ps</span>, <span class='va'>vars</span>, method <span class='op'>=</span> <span class='st'>"readcount"</span>, verbose <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>ps</th>
      <td><p>phyloseq with sample data</p></td>
    </tr>
    <tr>
      <th>vars</th>
      <td><p>names of variables, whose (combined) levels identify groups from which only 1 sample is desired</p></td>
    </tr>
    <tr>
      <th>method</th>
      <td><p>keep 1 sample with max "readcount" or the "first" or "last" samples encountered in given sample_data order for each dupe group</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>message names of samples removed if TRUE (or "debug" for more info, when method = "readcount")</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>phyloseq object</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>What happens when duplicated samples have exactly equal readcounts in method = "readcount"?
The first encountered maximum is kept (in sample_data row order, like method = "first")</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='ps_filter.html'>ps_filter</a></code> for filtering samples by sample_data variables</p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='http://dx.plos.org/10.1371/journal.pone.0061217'>phyloseq</a></span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='http://microbiome.github.io/microbiome'>microbiome</a></span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='st'>"dietswap"</span><span class='op'>)</span>

<span class='va'>dietswap</span>
</div><div class='output co'>#&gt; phyloseq-class experiment-level object
#&gt; otu_table()   OTU Table:         [ 130 taxa and 222 samples ]
#&gt; sample_data() Sample Data:       [ 222 samples by 8 sample variables ]
#&gt; tax_table()   Taxonomy Table:    [ 130 taxa by 3 taxonomic ranks ]</div><div class='input'><span class='co'># let's pretend the dietswap data contains technical replicates from each subject</span>
<span class='co'># we want to keep only one of them</span>
<span class='fu'>ps_dedupe</span><span class='op'>(</span><span class='va'>dietswap</span>, vars <span class='op'>=</span> <span class='st'>"subject"</span>, method <span class='op'>=</span> <span class='st'>"readcount"</span>, verbose <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>184 samples being removed:</span>
#&gt; <span class='message'>Sample-1; Sample-4; Sample-5; Sample-7; Sample-9; Sample-10; Sample-11; Sample-13; Sample-14; Sample-15; Sample-16; Sample-17; Sample-18; Sample-19; Sample-20; Sample-21; Sample-22; Sample-24; Sample-25; Sample-26; Sample-27; Sample-28; Sample-29; Sample-30; Sample-31; Sample-32; Sample-33; Sample-34; Sample-35; Sample-36; Sample-37; Sample-39; Sample-40; Sample-41; Sample-42; Sample-43; Sample-44; Sample-45; Sample-47; Sample-48; Sample-50; Sample-51; Sample-52; Sample-53; Sample-55; Sample-56; Sample-57; Sample-58; Sample-59; Sample-60; Sample-61; Sample-62; Sample-63; Sample-64; Sample-65; Sample-66; Sample-67; Sample-68; Sample-69; Sample-70; Sample-71; Sample-72; Sample-73; Sample-74; Sample-75; Sample-76; Sample-78; Sample-79; Sample-80; Sample-82; Sample-83; Sample-84; Sample-85; Sample-86; Sample-87; Sample-88; Sample-90; Sample-92; Sample-93; Sample-94; Sample-95; Sample-97; Sample-99; Sample-100; Sample-101; Sample-102; Sample-103; Sample-104; Sample-105; Sample-108; Sample-109; Sample-112; Sample-113; Sample-114; Sample-115; Sample-117; Sample-119; Sample-120; Sample-121; Sample-122; Sample-124; Sample-125; Sample-126; Sample-127; Sample-128; Sample-129; Sample-131; Sample-132; Sample-133; Sample-134; Sample-135; Sample-136; Sample-137; Sample-138; Sample-139; Sample-140; Sample-141; Sample-142; Sample-144; Sample-145; Sample-146; Sample-147; Sample-148; Sample-151; Sample-152; Sample-153; Sample-154; Sample-155; Sample-158; Sample-159; Sample-160; Sample-162; Sample-163; Sample-164; Sample-166; Sample-168; Sample-170; Sample-171; Sample-172; Sample-173; Sample-174; Sample-175; Sample-176; Sample-177; Sample-178; Sample-179; Sample-180; Sample-182; Sample-183; Sample-184; Sample-185; Sample-186; Sample-187; Sample-188; Sample-189; Sample-190; Sample-191; Sample-192; Sample-193; Sample-194; Sample-195; Sample-196; Sample-197; Sample-198; Sample-199; Sample-201; Sample-202; Sample-204; Sample-205; Sample-206; Sample-207; Sample-208; Sample-209; Sample-211; Sample-212; Sample-213; Sample-214; Sample-215; Sample-216; Sample-217; Sample-218; Sample-219; Sample-220; Sample-221</span></div><div class='output co'>#&gt; phyloseq-class experiment-level object
#&gt; otu_table()   OTU Table:         [ 130 taxa and 38 samples ]
#&gt; sample_data() Sample Data:       [ 38 samples by 8 sample variables ]
#&gt; tax_table()   Taxonomy Table:    [ 130 taxa by 3 taxonomic ranks ]</div><div class='input'>
<span class='co'># contrived example to show identifying "duplicates" via the interaction of multiple columns</span>
<span class='va'>ps1</span> <span class='op'>&lt;-</span> <span class='fu'>ps_dedupe</span><span class='op'>(</span><span class='va'>dietswap</span>,
  method <span class='op'>=</span> <span class='st'>"readcount"</span>, verbose <span class='op'>=</span> <span class='cn'>TRUE</span>,
  vars <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"timepoint"</span>, <span class='st'>"group"</span>, <span class='st'>"bmi_group"</span><span class='op'>)</span>
<span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>204 samples being removed:</span>
#&gt; <span class='message'>Sample-1; Sample-4; Sample-5; Sample-6; Sample-7; Sample-8; Sample-9; Sample-10; Sample-13; Sample-14; Sample-15; Sample-16; Sample-18; Sample-19; Sample-21; Sample-22; Sample-23; Sample-24; Sample-26; Sample-27; Sample-29; Sample-30; Sample-31; Sample-32; Sample-33; Sample-34; Sample-35; Sample-36; Sample-37; Sample-39; Sample-40; Sample-41; Sample-42; Sample-43; Sample-44; Sample-45; Sample-47; Sample-48; Sample-49; Sample-50; Sample-51; Sample-52; Sample-53; Sample-55; Sample-56; Sample-57; Sample-58; Sample-59; Sample-60; Sample-61; Sample-62; Sample-63; Sample-64; Sample-65; Sample-66; Sample-67; Sample-68; Sample-69; Sample-70; Sample-71; Sample-72; Sample-73; Sample-74; Sample-75; Sample-76; Sample-78; Sample-79; Sample-80; Sample-81; Sample-82; Sample-83; Sample-84; Sample-85; Sample-86; Sample-87; Sample-88; Sample-89; Sample-90; Sample-91; Sample-92; Sample-93; Sample-94; Sample-95; Sample-96; Sample-97; Sample-98; Sample-99; Sample-100; Sample-102; Sample-103; Sample-104; Sample-105; Sample-106; Sample-107; Sample-108; Sample-109; Sample-110; Sample-111; Sample-112; Sample-113; Sample-114; Sample-115; Sample-116; Sample-117; Sample-118; Sample-119; Sample-120; Sample-121; Sample-122; Sample-123; Sample-124; Sample-125; Sample-126; Sample-127; Sample-128; Sample-129; Sample-130; Sample-131; Sample-132; Sample-133; Sample-134; Sample-135; Sample-136; Sample-137; Sample-138; Sample-139; Sample-140; Sample-141; Sample-142; Sample-144; Sample-145; Sample-146; Sample-147; Sample-148; Sample-149; Sample-150; Sample-151; Sample-152; Sample-153; Sample-154; Sample-155; Sample-157; Sample-158; Sample-159; Sample-160; Sample-161; Sample-162; Sample-163; Sample-164; Sample-165; Sample-166; Sample-167; Sample-168; Sample-169; Sample-170; Sample-171; Sample-172; Sample-173; Sample-174; Sample-175; Sample-176; Sample-177; Sample-178; Sample-179; Sample-180; Sample-181; Sample-182; Sample-183; Sample-184; Sample-185; Sample-186; Sample-187; Sample-188; Sample-189; Sample-190; Sample-191; Sample-192; Sample-193; Sample-194; Sample-195; Sample-196; Sample-197; Sample-198; Sample-199; Sample-200; Sample-201; Sample-202; Sample-203; Sample-204; Sample-205; Sample-206; Sample-207; Sample-209; Sample-210; Sample-211; Sample-212; Sample-213; Sample-215; Sample-216; Sample-218; Sample-219; Sample-220; Sample-221; Sample-222</span></div><div class='input'><span class='fu'><a href='https://rdrr.io/pkg/phyloseq/man/sample_data-methods.html'>sample_data</a></span><span class='op'>(</span><span class='va'>ps1</span><span class='op'>)</span>
</div><div class='output co'>#&gt; Sample Data:        [18 samples by 8 sample variables]:
#&gt;            subject    sex nationality group     sample timepoint
#&gt; Sample-2       nms   male         AFR    HE   Sample-2         2
#&gt; Sample-3       olt   male         AFR    HE   Sample-3         2
#&gt; Sample-11      olt   male         AFR    HE  Sample-11         3
#&gt; Sample-12      pku female         AFR    HE  Sample-12         3
#&gt; Sample-17      ufm   male         AFR    HE  Sample-17         3
#&gt; Sample-20      pku female         AFR    DI  Sample-20         4
#&gt; Sample-25      ufm   male         AFR    DI  Sample-25         4
#&gt; Sample-28      pku female         AFR    DI  Sample-28         5
#&gt; Sample-38      nmz   male         AAM    ED  Sample-38         1
#&gt; Sample-46      mni female         AAM    ED  Sample-46         6
#&gt; Sample-54      dwc   male         AFR    ED  Sample-54         6
#&gt; Sample-77      zaq female         AFR    DI  Sample-77         4
#&gt; Sample-101     zaq female         AFR    DI Sample-101         5
#&gt; Sample-143     kpp   male         AAM    DI Sample-143         5
#&gt; Sample-156     gtd   male         AAM    HE Sample-156         2
#&gt; Sample-208     olt   male         AFR    ED Sample-208         1
#&gt; Sample-214     ufm   male         AFR    ED Sample-214         1
#&gt; Sample-217     pku female         AFR    ED Sample-217         6
#&gt;            timepoint.within.group  bmi_group
#&gt; Sample-2                        1       lean
#&gt; Sample-3                        1 overweight
#&gt; Sample-11                       2 overweight
#&gt; Sample-12                       2      obese
#&gt; Sample-17                       2       lean
#&gt; Sample-20                       1      obese
#&gt; Sample-25                       1       lean
#&gt; Sample-28                       2      obese
#&gt; Sample-38                       1      obese
#&gt; Sample-46                       2 overweight
#&gt; Sample-54                       2       lean
#&gt; Sample-77                       1 overweight
#&gt; Sample-101                      2 overweight
#&gt; Sample-143                      2       lean
#&gt; Sample-156                      1      obese
#&gt; Sample-208                      1 overweight
#&gt; Sample-214                      1       lean
#&gt; Sample-217                      2      obese</div><div class='input'>
<span class='va'>ps2</span> <span class='op'>&lt;-</span> <span class='fu'>ps_dedupe</span><span class='op'>(</span><span class='va'>dietswap</span>,
  method <span class='op'>=</span> <span class='st'>"first"</span>, verbose <span class='op'>=</span> <span class='cn'>TRUE</span>,
  vars <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"timepoint"</span>, <span class='st'>"group"</span>, <span class='st'>"bmi_group"</span><span class='op'>)</span>
<span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>204 samples being removed:</span>
#&gt; <span class='message'>Sample-5; Sample-6; Sample-7; Sample-8; Sample-9; Sample-13; Sample-14; Sample-15; Sample-16; Sample-17; Sample-20; Sample-21; Sample-22; Sample-23; Sample-24; Sample-25; Sample-29; Sample-30; Sample-31; Sample-33; Sample-34; Sample-37; Sample-38; Sample-39; Sample-40; Sample-42; Sample-43; Sample-46; Sample-47; Sample-48; Sample-49; Sample-50; Sample-51; Sample-52; Sample-53; Sample-54; Sample-55; Sample-56; Sample-57; Sample-58; Sample-59; Sample-60; Sample-61; Sample-62; Sample-63; Sample-64; Sample-65; Sample-66; Sample-67; Sample-68; Sample-69; Sample-70; Sample-71; Sample-72; Sample-73; Sample-74; Sample-75; Sample-76; Sample-77; Sample-78; Sample-79; Sample-80; Sample-81; Sample-82; Sample-83; Sample-84; Sample-85; Sample-86; Sample-87; Sample-88; Sample-89; Sample-90; Sample-91; Sample-92; Sample-93; Sample-94; Sample-95; Sample-96; Sample-97; Sample-98; Sample-99; Sample-100; Sample-101; Sample-102; Sample-103; Sample-104; Sample-105; Sample-106; Sample-107; Sample-108; Sample-109; Sample-110; Sample-111; Sample-112; Sample-113; Sample-114; Sample-115; Sample-116; Sample-117; Sample-118; Sample-119; Sample-120; Sample-121; Sample-122; Sample-123; Sample-124; Sample-125; Sample-126; Sample-127; Sample-128; Sample-129; Sample-130; Sample-131; Sample-132; Sample-133; Sample-134; Sample-135; Sample-136; Sample-137; Sample-138; Sample-139; Sample-140; Sample-141; Sample-142; Sample-143; Sample-144; Sample-145; Sample-146; Sample-147; Sample-148; Sample-149; Sample-150; Sample-151; Sample-152; Sample-153; Sample-154; Sample-155; Sample-156; Sample-157; Sample-158; Sample-159; Sample-160; Sample-161; Sample-162; Sample-163; Sample-164; Sample-165; Sample-166; Sample-167; Sample-168; Sample-169; Sample-170; Sample-171; Sample-172; Sample-173; Sample-174; Sample-175; Sample-176; Sample-177; Sample-178; Sample-179; Sample-180; Sample-181; Sample-182; Sample-183; Sample-184; Sample-185; Sample-186; Sample-187; Sample-188; Sample-189; Sample-190; Sample-191; Sample-192; Sample-193; Sample-194; Sample-195; Sample-196; Sample-197; Sample-198; Sample-199; Sample-200; Sample-201; Sample-202; Sample-203; Sample-204; Sample-205; Sample-206; Sample-207; Sample-208; Sample-209; Sample-210; Sample-211; Sample-212; Sample-213; Sample-214; Sample-215; Sample-216; Sample-217; Sample-218; Sample-219; Sample-220; Sample-221; Sample-222</span></div><div class='input'><span class='fu'><a href='https://rdrr.io/pkg/phyloseq/man/sample_data-methods.html'>sample_data</a></span><span class='op'>(</span><span class='va'>ps2</span><span class='op'>)</span>
</div><div class='output co'>#&gt; Sample Data:        [18 samples by 8 sample variables]:
#&gt;           subject    sex nationality group    sample timepoint
#&gt; Sample-1      byn   male         AAM    DI  Sample-1         4
#&gt; Sample-2      nms   male         AFR    HE  Sample-2         2
#&gt; Sample-3      olt   male         AFR    HE  Sample-3         2
#&gt; Sample-4      pku female         AFR    HE  Sample-4         2
#&gt; Sample-10     nms   male         AFR    HE Sample-10         3
#&gt; Sample-11     olt   male         AFR    HE Sample-11         3
#&gt; Sample-12     pku female         AFR    HE Sample-12         3
#&gt; Sample-18     nms   male         AFR    DI Sample-18         4
#&gt; Sample-19     olt   male         AFR    DI Sample-19         4
#&gt; Sample-26     nms   male         AFR    DI Sample-26         5
#&gt; Sample-27     olt   male         AFR    DI Sample-27         5
#&gt; Sample-28     pku female         AFR    DI Sample-28         5
#&gt; Sample-32     azh female         AAM    ED Sample-32         1
#&gt; Sample-35     kpp   male         AAM    ED Sample-35         1
#&gt; Sample-36     lot female         AAM    ED Sample-36         1
#&gt; Sample-41     azh female         AAM    ED Sample-41         6
#&gt; Sample-44     kpp   male         AAM    ED Sample-44         6
#&gt; Sample-45     lot female         AAM    ED Sample-45         6
#&gt;           timepoint.within.group  bmi_group
#&gt; Sample-1                       1      obese
#&gt; Sample-2                       1       lean
#&gt; Sample-3                       1 overweight
#&gt; Sample-4                       1      obese
#&gt; Sample-10                      2       lean
#&gt; Sample-11                      2 overweight
#&gt; Sample-12                      2      obese
#&gt; Sample-18                      1       lean
#&gt; Sample-19                      1 overweight
#&gt; Sample-26                      2       lean
#&gt; Sample-27                      2 overweight
#&gt; Sample-28                      2      obese
#&gt; Sample-32                      1 overweight
#&gt; Sample-35                      1       lean
#&gt; Sample-36                      1      obese
#&gt; Sample-41                      2 overweight
#&gt; Sample-44                      2       lean
#&gt; Sample-45                      2      obese</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by David Barnett.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script>
<script>
  docsearch({
    
    
    apiKey: '22e6f0a75b7b1ffae9bf5bc5b58864d7',
    indexName: 'microViz',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>



  </body>
</html>


