<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>De-duplicate phyloseq samples — ps_dedupe • microViz</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><!-- docsearch --><script src="../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous"><link href="../docsearch.css" rel="stylesheet"><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><meta property="og:title" content="De-duplicate phyloseq samples — ps_dedupe"><meta property="og:description" content='Use one or more variables in the sample_data to identify and
remove duplicate samples (leaving one sample per group).
methods:
method = "readcount" keeps the one sample in each duplicate group
with the highest total number of reads (phyloseq::sample_sums)
method = "first" keeps the first sample in each duplicate group
encountered in the row order of the sample_data
method = "last" keeps the last sample in each duplicate group
encountered in the row order of the sample_data
method = "random" keeps a random sample from each duplicate group
(set.seed for reproducibility)


More than one "duplicate" sample can be kept per group by setting n samples &amp;gt; 1.'><meta property="og:image" content="https://david-barnett.github.io/microViz/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-2YQ669S4V5"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-2YQ669S4V5');
</script></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">microViz</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.12.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/shao19-analyses.html">Getting Started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/web-only/phyloseq.html">Working with phyloseq objects</a>
    </li>
    <li>
      <a href="../articles/web-only/tax-fixing.html">Fixing your tax_table</a>
    </li>
    <li>
      <a href="../articles/web-only/ordination.html">Ordination plots</a>
    </li>
    <li>
      <a href="../articles/web-only/ordination-interactive.html">Interactive ordination plots</a>
    </li>
    <li>
      <a href="../articles/web-only/compositions.html">Visualising compositions with barplots</a>
    </li>
    <li>
      <a href="../articles/web-only/heatmaps.html">Heatmaps of microbiome composition or correlation</a>
    </li>
    <li>
      <a href="../articles/web-only/modelling-taxa.html">Statistical modelling of individual taxa</a>
    </li>
    <li>
      <a href="../articles/web-only/videos.html">Tutorial video</a>
    </li>
    <li>
      <a href="../articles/ps_extra-replaced.html">Upgrading to microViz version 0.10.0</a>
    </li>
    <li>
      <a href="../articles/web-only/atlas1006.html">Example analyses with atlas1006 data</a>
    </li>
  </ul></li>
<li>
  <a href="../articles/learn/exercises.html">Learn</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
<li class="dropdown-header">
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/david-barnett/microViz/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul><form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off"></div>
      </form>

    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>De-duplicate phyloseq samples</h1>
    <small class="dont-index">Source: <a href="https://github.com/david-barnett/microViz/blob/main/R/ps_dedupe.R" class="external-link"><code>R/ps_dedupe.R</code></a></small>
    <div class="hidden name"><code>ps_dedupe.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Use one or more variables in the sample_data to identify and
remove duplicate samples (leaving one sample per group).</p>
<p><strong>methods:</strong></p><ul><li><p>method = "readcount" keeps the one sample in each duplicate group
with the highest total number of reads (phyloseq::sample_sums)</p></li>
<li><p>method = "first" keeps the first sample in each duplicate group
encountered in the row order of the sample_data</p></li>
<li><p>method = "last" keeps the last sample in each duplicate group
encountered in the row order of the sample_data</p></li>
<li><p>method = "random" keeps a random sample from each duplicate group
(set.seed for reproducibility)</p></li>
</ul><p>More than one "duplicate" sample can be kept per group by setting <code>n</code> samples &gt; 1.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">ps_dedupe</span><span class="op">(</span></span>
<span>  <span class="va">ps</span>,</span>
<span>  <span class="va">vars</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"readcount"</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  n <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  .keep_group_var <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  .keep_readcount <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  .message_IDs <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  .label_only <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  .keep_all_taxa <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-ps">ps<a class="anchor" aria-label="anchor" href="#arg-ps"></a></dt>
<dd><p>phyloseq object</p></dd>


<dt id="arg-vars">vars<a class="anchor" aria-label="anchor" href="#arg-vars"></a></dt>
<dd><p>names of variables, whose (combined) levels identify groups
from which only 1 sample is desired</p></dd>


<dt id="arg-method">method<a class="anchor" aria-label="anchor" href="#arg-method"></a></dt>
<dd><p>keep sample with max "readcount" or the "first" or "last" or "random"
samples encountered in given sample_data order for each duplicate group</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>message about number of groups, and number of samples dropped?</p></dd>


<dt id="arg-n">n<a class="anchor" aria-label="anchor" href="#arg-n"></a></dt>
<dd><p>number of 'duplicates' to keep per group, defaults to 1</p></dd>


<dt id="arg--keep-group-var">.keep_group_var<a class="anchor" aria-label="anchor" href="#arg--keep-group-var"></a></dt>
<dd><p>keep grouping variable .GROUP. in phyloseq object?</p></dd>


<dt id="arg--keep-readcount">.keep_readcount<a class="anchor" aria-label="anchor" href="#arg--keep-readcount"></a></dt>
<dd><p>keep readcount variable .READCOUNT. in phyloseq object?</p></dd>


<dt id="arg--message-ids">.message_IDs<a class="anchor" aria-label="anchor" href="#arg--message-ids"></a></dt>
<dd><p>message sample names of dropped variables?</p></dd>


<dt id="arg--label-only">.label_only<a class="anchor" aria-label="anchor" href="#arg--label-only"></a></dt>
<dd><p>if TRUE, the samples will NOT be filtered, just labelled with a new logical
variable .KEEP_SAMPLE.</p></dd>


<dt id="arg--keep-all-taxa">.keep_all_taxa<a class="anchor" aria-label="anchor" href="#arg--keep-all-taxa"></a></dt>
<dd><p>keep all taxa after removing duplicates?
If FALSE, the default, taxa are removed if they never occur in any of
the retained samples</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>phyloseq object</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>What happens when duplicated samples have exactly equal readcounts in method = "readcount"?
The first encountered maximum is kept (in sample_data row order, like method = "first")</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="ps_filter.html">ps_filter</a></code> for filtering samples by sample_data variables</p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"dietswap"</span>, package <span class="op">=</span> <span class="st">"microbiome"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">dietswap</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> phyloseq-class experiment-level object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> otu_table()   OTU Table:         [ 130 taxa and 222 samples ]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> sample_data() Sample Data:       [ 222 samples by 8 sample variables ]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> tax_table()   Taxonomy Table:    [ 130 taxa by 3 taxonomic ranks ]</span>
<span class="r-in"><span><span class="co"># let's pretend the dietswap data contains technical replicates from each subject</span></span></span>
<span class="r-in"><span><span class="co"># we want to keep only one of them</span></span></span>
<span class="r-in"><span><span class="fu">ps_dedupe</span><span class="op">(</span><span class="va">dietswap</span>, vars <span class="op">=</span> <span class="st">"subject"</span>, method <span class="op">=</span> <span class="st">"readcount"</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 38 groups: with 4 to 6 samples each</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Dropped 184 samples.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> phyloseq-class experiment-level object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> otu_table()   OTU Table:         [ 120 taxa and 38 samples ]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> sample_data() Sample Data:       [ 38 samples by 8 sample variables ]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> tax_table()   Taxonomy Table:    [ 120 taxa by 3 taxonomic ranks ]</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># contrived example to show identifying "duplicates" via the interaction of multiple columns</span></span></span>
<span class="r-in"><span><span class="va">ps1</span> <span class="op">&lt;-</span> <span class="fu">ps_dedupe</span><span class="op">(</span></span></span>
<span class="r-in"><span>  ps <span class="op">=</span> <span class="va">dietswap</span>, method <span class="op">=</span> <span class="st">"readcount"</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>  vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"timepoint"</span>, <span class="st">"group"</span>, <span class="st">"bmi_group"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 18 groups: with 8 to 15 samples each</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Dropped 204 samples.</span>
<span class="r-in"><span><span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/sample_data-methods.html" class="external-link">sample_data</a></span><span class="op">(</span><span class="va">ps1</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>            subject    sex nationality group     sample timepoint</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-2       nms   male         AFR    HE   Sample-2         2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-3       olt   male         AFR    HE   Sample-3         2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-11      olt   male         AFR    HE  Sample-11         3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-12      pku female         AFR    HE  Sample-12         3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-17      ufm   male         AFR    HE  Sample-17         3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-20      pku female         AFR    DI  Sample-20         4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-25      ufm   male         AFR    DI  Sample-25         4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-28      pku female         AFR    DI  Sample-28         5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-38      nmz   male         AAM    ED  Sample-38         1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-46      mni female         AAM    ED  Sample-46         6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-54      dwc   male         AFR    ED  Sample-54         6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-77      zaq female         AFR    DI  Sample-77         4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-101     zaq female         AFR    DI Sample-101         5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-143     kpp   male         AAM    DI Sample-143         5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-156     gtd   male         AAM    HE Sample-156         2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-208     olt   male         AFR    ED Sample-208         1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-214     ufm   male         AFR    ED Sample-214         1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-217     pku female         AFR    ED Sample-217         6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>            timepoint.within.group  bmi_group</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-2                        1       lean</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-3                        1 overweight</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-11                       2 overweight</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-12                       2      obese</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-17                       2       lean</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-20                       1      obese</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-25                       1       lean</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-28                       2      obese</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-38                       1      obese</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-46                       2 overweight</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-54                       2       lean</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-77                       1 overweight</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-101                      2 overweight</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-143                      2       lean</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-156                      1      obese</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-208                      1 overweight</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-214                      1       lean</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-217                      2      obese</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">ps2</span> <span class="op">&lt;-</span> <span class="fu">ps_dedupe</span><span class="op">(</span></span></span>
<span class="r-in"><span>  ps <span class="op">=</span> <span class="va">dietswap</span>, method <span class="op">=</span> <span class="st">"first"</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>  vars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"timepoint"</span>, <span class="st">"group"</span>, <span class="st">"bmi_group"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 18 groups: with 8 to 15 samples each</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Dropped 204 samples.</span>
<span class="r-in"><span><span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/sample_data-methods.html" class="external-link">sample_data</a></span><span class="op">(</span><span class="va">ps2</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           subject    sex nationality group    sample timepoint</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-1      byn   male         AAM    DI  Sample-1         4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-2      nms   male         AFR    HE  Sample-2         2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-3      olt   male         AFR    HE  Sample-3         2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-4      pku female         AFR    HE  Sample-4         2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-10     nms   male         AFR    HE Sample-10         3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-11     olt   male         AFR    HE Sample-11         3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-12     pku female         AFR    HE Sample-12         3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-18     nms   male         AFR    DI Sample-18         4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-19     olt   male         AFR    DI Sample-19         4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-26     nms   male         AFR    DI Sample-26         5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-27     olt   male         AFR    DI Sample-27         5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-28     pku female         AFR    DI Sample-28         5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-32     azh female         AAM    ED Sample-32         1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-35     kpp   male         AAM    ED Sample-35         1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-36     lot female         AAM    ED Sample-36         1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-41     azh female         AAM    ED Sample-41         6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-44     kpp   male         AAM    ED Sample-44         6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-45     lot female         AAM    ED Sample-45         6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           timepoint.within.group  bmi_group</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-1                       1      obese</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-2                       1       lean</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-3                       1 overweight</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-4                       1      obese</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-10                      2       lean</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-11                      2 overweight</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-12                      2      obese</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-18                      1       lean</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-19                      1 overweight</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-26                      2       lean</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-27                      2 overweight</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-28                      2      obese</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-32                      1 overweight</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-35                      1       lean</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-36                      1      obese</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-41                      2 overweight</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-44                      2       lean</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sample-45                      2      obese</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by David Barnett.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>


<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({


    apiKey: '22e6f0a75b7b1ffae9bf5bc5b58864d7',
    indexName: 'microViz',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script><script data-goatcounter="https://microviz.goatcounter.com/count" async src="//gc.zgo.at/count.js"></script></body></html>

