<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Melt phyloseq data object into large data.frame (tibble) — ps_melt • microViz</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><!-- docsearch --><script src="../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous"><link href="../docsearch.css" rel="stylesheet"><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><meta property="og:title" content="Melt phyloseq data object into large data.frame (tibble) — ps_melt"><meta property="og:description" content='The ps_melt function is a specialized melt function for melting phyloseq objects
(instances of the phyloseq class), usually for producing graphics
with ggplot2.
The naming conventions used in downstream phyloseq graphics functions
have reserved the following variable names that should not be used
as the names of sample_variables
or taxonomic rank_names.
These reserved names are c("Sample", "Abundance", "OTU").
Also, you should not have identical names for
sample variables and taxonomic ranks.
That is, the intersection of the output of the following two functions
sample_variables, rank_names
should be an empty vector
(e.g. intersect(sample_variables(ps), rank_names(ps))).
All of these potential name collisions are checked-for
and renamed automatically with a warning.
However, if you (re)name your variables accordingly ahead of time,
it will reduce confusion and eliminate the warnings.
NOTE: Code and documentation copied (and very slightly modified) from an old version of speedyseq by Michael McLaren.
speedyseq reimplements psmelt from phyloseq to use functions from the tidyr
and dplyr packages. The name in microViz is changed to ps_melt for consistency with other functions.'><meta property="og:image" content="https://david-barnett.github.io/microViz/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-2YQ669S4V5"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-2YQ669S4V5');
</script></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">microViz</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.12.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/shao19-analyses.html">Getting Started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/web-only/phyloseq.html">Working with phyloseq objects</a>
    </li>
    <li>
      <a href="../articles/web-only/tax-fixing.html">Fixing your tax_table</a>
    </li>
    <li>
      <a href="../articles/web-only/ordination.html">Ordination plots</a>
    </li>
    <li>
      <a href="../articles/web-only/ordination-interactive.html">Interactive ordination plots</a>
    </li>
    <li>
      <a href="../articles/web-only/compositions.html">Visualising compositions with barplots</a>
    </li>
    <li>
      <a href="../articles/web-only/heatmaps.html">Heatmaps of microbiome composition or correlation</a>
    </li>
    <li>
      <a href="../articles/web-only/modelling-taxa.html">Statistical modelling of individual taxa</a>
    </li>
    <li>
      <a href="../articles/web-only/videos.html">Tutorial video</a>
    </li>
    <li>
      <a href="../articles/ps_extra-replaced.html">Upgrading to microViz version 0.10.0</a>
    </li>
    <li>
      <a href="../articles/web-only/atlas1006.html">Example analyses with atlas1006 data</a>
    </li>
  </ul></li>
<li>
  <a href="../articles/learn/exercises.html">Learn</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
<li class="dropdown-header">
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/david-barnett/microViz/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul><form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off"></div>
      </form>

    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Melt phyloseq data object into large data.frame (tibble)</h1>
    <small class="dont-index">Source: <a href="https://github.com/david-barnett/microViz/blob/main/R/ps_melt.R" class="external-link"><code>R/ps_melt.R</code></a></small>
    <div class="hidden name"><code>ps_melt.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>The ps_melt function is a specialized melt function for melting phyloseq objects
(instances of the phyloseq class), usually for producing graphics
with <code><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a>2</code>.
The naming conventions used in downstream phyloseq graphics functions
have reserved the following variable names that should not be used
as the names of <code><a href="https://rdrr.io/pkg/phyloseq/man/sample_variables.html" class="external-link">sample_variables</a></code>
or taxonomic <code><a href="https://rdrr.io/pkg/phyloseq/man/rank_names.html" class="external-link">rank_names</a></code>.
These reserved names are <code>c("Sample", "Abundance", "OTU")</code>.
Also, you should not have identical names for
sample variables and taxonomic ranks.
That is, the intersection of the output of the following two functions
<code><a href="https://rdrr.io/pkg/phyloseq/man/sample_variables.html" class="external-link">sample_variables</a></code>, <code><a href="https://rdrr.io/pkg/phyloseq/man/rank_names.html" class="external-link">rank_names</a></code>
should be an empty vector
(e.g. <code>intersect(sample_variables(ps), rank_names(ps))</code>).
All of these potential name collisions are checked-for
and renamed automatically with a warning.
However, if you (re)name your variables accordingly ahead of time,
it will reduce confusion and eliminate the warnings.</p>
<p>NOTE: Code and documentation copied (and very slightly modified) from an old version of speedyseq by Michael McLaren.
speedyseq reimplements <code>psmelt</code> from <code>phyloseq</code> to use functions from the <code>tidyr</code>
and <code>dplyr</code> packages. The name in microViz is changed to <code>ps_melt</code> for consistency with other functions.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">ps_melt</span><span class="op">(</span><span class="va">ps</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-ps">ps<a class="anchor" aria-label="anchor" href="#arg-ps"></a></dt>
<dd><p>(Required). An <code><a href="https://rdrr.io/pkg/phyloseq/man/otu_table-class.html" class="external-link">otu_table-class</a></code> or
<code><a href="https://rdrr.io/pkg/phyloseq/man/phyloseq-class.html" class="external-link">phyloseq-class</a></code>. Function most useful for phyloseq-class.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A <code><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></code> class data frame.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>Note that “melted” phyloseq data is stored much less efficiently, and so
RAM storage issues could arise with a smaller dataset (smaller number of
samples/OTUs/variables) than one might otherwise expect.  For common sizes
of graphics-ready datasets, however, this should not be a problem.  Because
the number of OTU entries has a large effect on the RAM requirement, methods
to reduce the number of separate OTU entries – for instance by
agglomerating OTUs based on phylogenetic distance using
<code><a href="https://rdrr.io/pkg/phyloseq/man/tip_glom.html" class="external-link">tip_glom</a></code> – can help alleviate RAM usage problems.  This
function is made user-accessible for flexibility, but is also used
extensively by plot functions in phyloseq.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="https://rdrr.io/pkg/phyloseq/man/psmelt.html" class="external-link">psmelt</a></code></p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://dx.plos.org/10.1371/journal.pone.0061217" class="external-link">phyloseq</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"GlobalPatterns"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">gp_ch</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/subset_taxa-methods.html" class="external-link">subset_taxa</a></span><span class="op">(</span><span class="va">GlobalPatterns</span>, <span class="va">Phylum</span> <span class="op">==</span> <span class="st">"Chlamydiae"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">mdf</span> <span class="op">&lt;-</span> <span class="fu">ps_melt</span><span class="op">(</span><span class="va">gp_ch</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">mdf2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/psmelt.html" class="external-link">psmelt</a></span><span class="op">(</span><span class="va">gp_ch</span><span class="op">)</span> <span class="co"># slower</span></span></span>
<span class="r-in"><span><span class="co"># same dataframe, except with somewhat different row orders</span></span></span>
<span class="r-in"><span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/all_equal.html" class="external-link">all_equal</a></span><span class="op">(</span><span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="va">mdf</span><span class="op">)</span>, <span class="va">mdf2</span>, convert <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="co"># TRUE</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>`all_equal()` was deprecated in dplyr 1.1.0.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Please use `all.equal()` instead.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> And manually order the rows/cols as needed</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] TRUE</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">mdf2</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 546</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">mdf</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 17</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">mdf</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1] "OTU"                      "Sample"                  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [3] "Abundance"                "X.SampleID"              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [5] "Primer"                   "Final_Barcode"           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [7] "Barcode_truncated_plus_T" "Barcode_full_length"     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [9] "SampleType"               "Description"             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [11] "Kingdom"                  "Phylum"                  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [13] "Class"                    "Order"                   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [15] "Family"                   "Genus"                   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [17] "Species"                 </span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">mdf</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "1" "2" "3" "4" "5" "6"</span>
<span class="r-in"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">mdf</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">SampleType</span>, y <span class="op">=</span> <span class="va">Abundance</span>, fill <span class="op">=</span> <span class="va">Genus</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="va">p</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"black"</span>, stat <span class="op">=</span> <span class="st">"identity"</span>, position <span class="op">=</span> <span class="st">"stack"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># This example plot doesn't make any sense</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">p</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="ps_melt-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="co"># TODO replace this...</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by David Barnett.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>


<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({


    apiKey: '22e6f0a75b7b1ffae9bf5bc5b58864d7',
    indexName: 'microViz',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script><script data-goatcounter="https://microviz.goatcounter.com/count" async src="//gc.zgo.at/count.js"></script></body></html>

