<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Filter rare and/or low abundance taxa from a phyloseq object — tax_filter • microViz</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>


<!-- docsearch -->
<script src="../docsearch.js"></script>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous" />
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script>



<meta property="og:title" content="Filter rare and/or low abundance taxa from a phyloseq object — tax_filter" />
<meta property="og:description" content="Removes taxa (from all samples) that do not meet a given criterion or combination of criteria.
If a value for min_prevalence, min_total_abundance or min_sample_abundance is 1 or greater, then
it is treated as an absolute minimum number of samples/reads. If &amp;lt;1, it is treated as proportion of all samples/reads.
This function is designed to work with counts. otu_table must contain counts particularly if you want to set a non-zero value for min_total_abundance." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">microViz</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.7.6.9002</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/articles/microViz.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/articles/Interactive-Ordination.html">Interactive Ordination Plot</a>
    </li>
    <li>
      <a href="../articles/articles/Visualising-compositions.html">Visualising compositions</a>
    </li>
    <li>
      <a href="../articles/articles/phyloseq.html">Working with phyloseq objects</a>
    </li>
    <li>
      <a href="../articles/articles/tax-fixing.html">Fixing your tax_table</a>
    </li>
    <li>
      <a href="../articles/ordination.html">Ordination plots</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/david-barnett/microViz/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
      <form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
        </div>
      </form>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Filter rare and/or low abundance taxa from a phyloseq object</h1>
    <small class="dont-index">Source: <a href='https://github.com/david-barnett/microViz/blob/master/R/tax_filter.R'><code>R/tax_filter.R</code></a></small>
    <div class="hidden name"><code>tax_filter.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Removes taxa (from all samples) that do not meet a given criterion or combination of criteria.
If a value for min_prevalence, min_total_abundance or min_sample_abundance is 1 or greater, then
it is treated as an absolute minimum number of samples/reads. If &lt;1, it is treated as proportion of all samples/reads.
This function is designed to work with counts. otu_table must contain counts particularly if you want to set a non-zero value for min_total_abundance.</p>
    </div>

    <pre class="usage"><span class='fu'>tax_filter</span><span class='op'>(</span>
  <span class='va'>ps</span>,
  min_prevalence <span class='op'>=</span> <span class='fl'>1</span>,
  prev_detection_threshold <span class='op'>=</span> <span class='fl'>1</span>,
  min_total_abundance <span class='op'>=</span> <span class='fl'>0</span>,
  min_sample_abundance <span class='op'>=</span> <span class='fl'>0</span>,
  tax_level <span class='op'>=</span> <span class='cn'>NA</span>,
  names_only <span class='op'>=</span> <span class='cn'>FALSE</span>,
  is_counts <span class='op'>=</span> <span class='cn'>TRUE</span>,
  undetected <span class='op'>=</span> <span class='cn'>NULL</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>ps</th>
      <td><p>phyloseq object (ideally with COUNTS in otu_table)</p></td>
    </tr>
    <tr>
      <th>min_prevalence</th>
      <td><p>number or proportion of samples that a taxon must be present in (alternatively see undetected)</p></td>
    </tr>
    <tr>
      <th>prev_detection_threshold</th>
      <td><p>min required counts (or value) for a taxon to be considered present in that sample (or set undetected arg)</p></td>
    </tr>
    <tr>
      <th>min_total_abundance</th>
      <td><p>minimum total readcount of a taxon, summed across all samples (can be proportion of all counts)</p></td>
    </tr>
    <tr>
      <th>min_sample_abundance</th>
      <td><p>taxa must have at least this many reads in one or more samples (or proportion of that sample's reads)</p></td>
    </tr>
    <tr>
      <th>tax_level</th>
      <td><p>if given, aggregates data at named taxonomic rank before filtering, but returns phyloseq at the ORIGINAL level of aggregation!</p></td>
    </tr>
    <tr>
      <th>names_only</th>
      <td><p>if names_only is true return only names of taxa, not the phyloseq</p></td>
    </tr>
    <tr>
      <th>is_counts</th>
      <td><p>expect count data in phyloseq otu_table? default is TRUE</p></td>
    </tr>
    <tr>
      <th>undetected</th>
      <td><p>e.g. 0, value at (or below) which a taxon is considered not present in that sample. If set, this overrides prev_detection_threshold.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>filtered phyloseq object AT ORIGINAL LEVEL OF AGGREGATION (not at the level in tax_level)</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='st'>"dietswap"</span>, package <span class='op'>=</span> <span class='st'>"microbiome"</span><span class='op'>)</span>
<span class='co'># Dropping rare and low abundance taxa #</span>
<span class='co'># Filter at unique taxa level, keeping only those with a prevalence of 10% or more</span>
<span class='co'># and at least 10 thousand reads when summed across all samples.</span>
<span class='co'># Then aggregate to Family level taxonomy.</span>
<span class='va'>dietswap</span> <span class='op'>%&gt;%</span>
  <span class='fu'>tax_filter</span><span class='op'>(</span>min_prevalence <span class='op'>=</span> <span class='fl'>0.1</span>, min_total_abundance <span class='op'>=</span> <span class='fl'>10000</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'><a href='tax_agg.html'>tax_agg</a></span><span class='op'>(</span><span class='st'>"Family"</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>Proportional min_prevalence given: 0.1 --&gt; min 23/222 samples.</span></div><div class='output co'>#&gt; ps_extra object - a list with phyloseq and extras:
#&gt; 
#&gt; phyloseq-class experiment-level object
#&gt; otu_table()   OTU Table:         [ 9 taxa and 222 samples ]
#&gt; sample_data() Sample Data:       [ 222 samples by 8 sample variables ]
#&gt; tax_table()   Taxonomy Table:    [ 9 taxa by 2 taxonomic ranks ]
#&gt; 
#&gt; ps_extra info:
#&gt; tax_agg = Family tax_transform = NA</div><div class='input'>
<span class='co'># Keeping ubiquitous families #</span>
<span class='co'># keep only families that have at least 1000 counts present in 90% of samples</span>
<span class='co'># then aggregate the remaining taxa at 'Genus' level</span>
<span class='va'>dietswap</span> <span class='op'>%&gt;%</span>
  <span class='fu'>tax_filter</span><span class='op'>(</span>tax_level <span class='op'>=</span> <span class='st'>"Family"</span>, min_prevalence <span class='op'>=</span> <span class='fl'>0.90</span>, prev_detection_threshold <span class='op'>=</span> <span class='fl'>1000</span><span class='op'>)</span> <span class='op'>%&gt;%</span>
  <span class='fu'><a href='tax_agg.html'>tax_agg</a></span><span class='op'>(</span><span class='st'>"Genus"</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>Proportional min_prevalence given: 0.9 --&gt; min 200/222 samples.</span></div><div class='output co'>#&gt; ps_extra object - a list with phyloseq and extras:
#&gt; 
#&gt; phyloseq-class experiment-level object
#&gt; otu_table()   OTU Table:         [ 28 taxa and 222 samples ]
#&gt; sample_data() Sample Data:       [ 222 samples by 8 sample variables ]
#&gt; tax_table()   Taxonomy Table:    [ 28 taxa by 3 taxonomic ranks ]
#&gt; 
#&gt; ps_extra info:
#&gt; tax_agg = Genus tax_transform = NA</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by David Barnett.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script>
<script>
  docsearch({
    
    
    apiKey: '22e6f0a75b7b1ffae9bf5bc5b58864d7',
    indexName: 'microViz',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>



  </body>
</html>


