---
title: "Ordination plots"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Ordination plots}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Intro

This article will show you how to create and customise ordination plots, like [PCA](#pca) and [RDA](#rda), with microViz.

For an even quicker start with ordinating your phyloseq data, check out the `ord_explore` [Shiny app](https://david-barnett.github.io/microViz/articles/articles/Interactive-Ordination.html), which allows you to create ordination plots with a point and click interface, and generates `ord_plot` code for you to copy. Return to this article to understand more about creating and customising your ordination plotting script.

```{r setup}
library(phyloseq)
library(ggplot2)
library(microViz)
knitr::opts_chunk$set(fig.width = 7, fig.height = 6)
```

We will use example data from stool samples from an inflammatory bowel disease (IBD) study, borrowed from the great `corncob` package. See the article about [working with phyloseq objects](https://david-barnett.github.io/microViz/articles/articles/phyloseq.html) if you want to get started with your own data, or just to learn more about manipulating phyloseq objects with microViz.

```{r get data}
ibd <- corncob::ibd_phylo
ibd
```

First we fix any uninformative tax_table entries and check the phyloseq object (see the article on [fixing your tax_table](https://david-barnett.github.io/microViz/articles/articles/tax-fixing.html) for more info).

```{r fix it}
ibd <- tax_fix(ibd) # try tax_fix_interactive if you have problems with your own data
ibd <- phyloseq_validate(ibd, remove_undetected = TRUE)
```

## Motivation

Ordination plots are a great way to see any clustering or other patterns of microbiota (dis)similarity in (many) samples. Ordinations like PCA or PCoA show the largest patterns of variation in your data, and constrained ordination techniques like RDA or CCA can show you microbial variation that could be explained by other variables in your sample_data (but interpret constrained ordinations with care, and ideally test for the statistical significance of any hypothesised associations using a method like PERMANOVA, see `dist_permanova()`).

Check out the GUide to STatistical Analysis in Microbial Ecology (GUSTA ME)[website](https://sites.google.com/site/mb3gustame/) for a gentle theoretical introduction to [PCA](https://sites.google.com/site/mb3gustame/indirect-gradient-analysis/pca), [PCoA](https://sites.google.com/site/mb3gustame/dissimilarity-based-methods/principal-coordinates-analysis), [RDA](https://sites.google.com/site/mb3gustame/constrained-analyses/rda), [CCA](https://sites.google.com/site/mb3gustame/constrained-analyses/cca) and more.

Ordination plots can also be paired with barplots for greater insight into microbial compositions, e.g. see `ord_plot_iris()` and the `ord_explore()` interactive [Shiny app](https://david-barnett.github.io/microViz/articles/articles/Interactive-Ordination.html).

## Prepare your microbes

When creating an ordination plot, you first need to prepare the microbiota variables.

-   Decide at which taxonomic rank to aggregate your data, e.g. "Genus"

-   Consider transforming the microbial counts, e.g. using the "clr" (centred log ratio) transformation, which is often recommended for compositional data ([like sequencing data](https://doi.org/10.3389/fmicb.2017.02224 "Gloor 2017"))

```{r clr}
ibd %>% 
  tax_transform(transformation = "clr", rank = "Genus")
```

-   Some methods, such as [PCoA](#pcoa), require a matrix of pairwise distances between samples, which you can easily calculate with `dist_calc()`. Normally you should NOT transform your data when using a distance-based method, but it is useful to record an "identity" transformation anyway, to make it clear you have not transformed your data.

```{r identity}
ibd %>% 
  tax_transform(transformation = "identity", rank = "Genus") %>% 
  dist_calc("bray") # bray curtis distance
```

-   Some dissimilarity measures, such as unweighted UniFrac, do not consider the abundance of each taxon when calculating dissimilarity, and so may be (overly) sensitive to differences in rare/low-abundance taxa. So you might want to filter out very rare taxa, with `tax_filter()` before using `dist_calc(ps, dist = "unifrac")`. Distances that are (implicitly) abundance weighted, including Generalised UniFrac, Bray-Curtis and Aitchison distance, should be less sensitive to rare taxa / filtering threshold choices.

### ps_extra

Notice that the objects created above are of class "ps_extra". This is just a simple S3 list object that holds your phyloseq object and additional stuff created from this phyloseq object, such as a distance matrix, as well as info on any transformation and aggregation applied to your taxa. microViz uses this to automatically create plot captions, to help you and your collaborators remember how you made each plot! You can access the phyloseq object, distance matrix and other parts of a ps_extra object with `ps_get()`, `dist_get()`, and friends.

## PCA {#pca}

Principle ***Components*** Analysis is an unconstrained method that does not use a distance matrix. PCA directly uses the (transformed) microbial variables, so you do not need `dist_calc()`. `ord_calc` performs the ordination (adding it to the ps_extra object) and `ord_plot()` creates the ggplot2 scatterplot (which you can customise like other ggplot objects).

Each point is a sample, and samples that appear closer together are typically more similar to each other than samples which are further apart. So by colouring the points by IBD status you can see that the microbiota from people with IBD is often, but not always, highly distinct from people without IBD.

```{r pca}
ibd %>% 
  tax_transform("clr", rank = "Genus") %>% 
  # when no distance matrix or constraints are supplied, PCA is the default/auto ordination method
  ord_calc() %>% 
  ord_plot(color = "ibd", shape = "DiseaseState", size = 2) +
  scale_colour_brewer(palette = "Dark2")
```

One benefit of not using a distance matrix, is that you can plot taxa "loadings" onto your PCA axes, using the plot_taxa argument. microViz plots all of the taxa loading vectors in light grey, and you choose how many of the vectors to label, starting with the longest arrows (alternatively you can name particular taxa to label).

The relative length of each loading vector indicates its contribution to each PCA axis shown, and allows you to roughly estimate which samples will contain more of that taxon e.g. samples on the left of the plot below, will typically contain more *Escherichia*/*Shigella* than samples on the right, and this taxon contributes heavily to the PC1 axis.

```{r pca loadings}
ibd %>% 
  tax_transform("clr", rank = "Genus") %>% 
  # when no distance matrix or constraints are supplied, PCA is the default/auto ordination method
  ord_calc(method = "PCA") %>% 
  ord_plot(color = "ibd", shape = "DiseaseState", plot_taxa = 1:5, size = 2) +
  scale_colour_brewer(palette = "Dark2")
```

microViz also allows you directly visualize the sample compositions on a circular barplot or "iris plot" (named because it looks kinda like an eyeball) alongside the PCA plot. The samples on the iris plot are automatically arranged by their rotational position around the center/origin of the PCA plot.

```{r iris, fig.height=8, fig.width=7}
ibd %>% 
  tax_transform("clr", rank = "Genus") %>% 
  # when no distance matrix or constraints are supplied, PCA is the default/auto ordination method
  ord_calc() %>% 
  ord_plot_iris(tax_level = "Genus", ord_plot = "above", anno_colour = "ibd")
```

Here we created the ordination plot as a quick accompaniment to the circular barchart, but it is more flexible to create and customise the ordination plot and iris plot separately, and then pair them afterwards with patchwork. See the `ord_plot_iris` [docs](https://david-barnett.github.io/microViz/reference/ord_plot_iris.html) for examples.

## PCoA {#pcoa}

Principle ***Co-ordinates*** Analysis is also an unconstrained method, but it does require a distance matrix. In an ecological context, a distance (or more generally a "dissimilarity") measure indicates how different a pair of (microbial) ecosystems are. This can be calculated in many ways.

### Aitchison distance

The [Euclidean distance](https://en.wikipedia.org/wiki/Euclidean_distance) is similar to the distance we humans are familiar with in the physical world. The results of a PCA is practically equivalent to a PCoA with Euclidean distances. The Aitchison distance is a dissimilarity measure calculated as the Euclidean distance between observations (samples) after performing a centred log ratio ("clr") transformation. That is why the Aitchison distance PCoA, below, looks the same as the PCA we made earlier. However, we cannot use plot_taxa, as the taxa loadings are only available for PCA (and related methods like RDA).

```{r aitchison pcoa}
ibd %>% 
  tax_transform("identity", rank = "Genus") %>% # don't transform!
  dist_calc("aitchison") %>% 
  ord_calc("PCoA") %>% 
  ord_plot(color = "ibd", shape = "DiseaseState", size = 2) +
  scale_colour_brewer(palette = "Dark2")
```

Note that PCoA is also known as MDS, for (metric) Multi-Dimensional Scaling, hence the axes names.

### Ecological dissimilarities

Over the years, ecologists have invented numerous ways of quantifying dissimilarity between pairs of ecosystems. One ubiquitous example is the [Bray-Curtis](https://en.wikipedia.org/wiki/Brayâ€“Curtis_dissimilarity) dissimilarity measure, shown below.

```{r bray}
ibd %>% 
  tax_transform("identity", rank = "Genus") %>% # don't transform!
  dist_calc("bray") %>% 
  ord_calc("PCoA") %>% 
  ord_plot(color = "ibd", shape = "DiseaseState", size = 2) +
  scale_colour_brewer(palette = "Dark2")
```

Beyond Bray-Curtis, microViz `dist_calc()` can also help you calculate all the other ecological distances listed in `phyloseq::distanceMethodList` such the Jensen-Shannon Divergence, `"jsd"`, or Jaccard dissimilarity `"jaccard"`. Beware that if you want a binary dissimilarity measure from `vegan::vegdist()` (i.e. only using presence/absence info, and noting all the caveats about sensitivity to low abundance taxa) you will need to pass `binary = TRUE`, as below.

```{r binary jaccard}
ibd %>% 
  tax_transform("identity", rank = "Genus") %>% 
  dist_calc(dist = "jaccard", binary = TRUE) %>% 
  ord_calc("PCoA") %>% 
  ord_plot(color = "ibd", shape = "DiseaseState", size = 2) +
  scale_colour_brewer(palette = "Dark2")
```

### UniFrac distances

If you have a phylogenetic tree available, and [attached to your phyloseq object](https://david-barnett.github.io/microViz/articles/articles/phyloseq.html#getting-your-data-into-phyloseq). You can calculate dissimilarities from the [UniFrac family](https://en.wikipedia.org/wiki/UniFrac) of methods, which take into account the phylogenetic relatedness of the taxa / sequences in your samples when calculating dissimilarity. Un-weighted UniFrac, `dist_calc(dist = "unifrac")`, does not consider the relative abundance of taxa, only their presence (detection) or absence, which can make it (overly) sensitive to rare taxa, sequencing artefacts, and abundance filtering choices. Conversely, weighted UniFrac, `"wunifrac"`, does put (perhaps too much) more importance on highly abundant taxa, when determining dissimilarities. The Generalised UniFrac, `"gunifrac"`, finds a balance between these two extremes, and by adjusting the `gunifrac_alpha` argument of `dist_calc()`, you can tune this balance to your liking (although the 0.5 default should be fine!).

Below is a Generalised UniFrac example using a different, and tiny, example dataset from the phyloseq package that has a phylogenetic tree.

You should **not** aggregate taxa before using a phylogenetic distance measure, but you can and probably should register the choice not to transform or aggregate, as below.

```{r gunifrac}
data("esophagus", package = "phyloseq")
esophagus %>%
  phyloseq_validate(verbose = FALSE) %>% 
  tax_transform("identity", rank = "unique") %>% 
  dist_calc("gunifrac", gunifrac_alpha = 0.5)

```

## Further dimensions

You can show other dimensions / axes of an ordination than just the first two, by setting the axes argument. You can judge from the variation explained by each successive axis (on a scree plot) whether this is worthwhile information to show, e.g. in the example below, it could be interesting to also show the 3rd axis, but not any others.

```{r scree, fig.width=5, fig.height=3}
ibd %>% 
  tax_transform("identity", rank = "Genus") %>% # don't transform!
  dist_calc("bray") %>% 
  ord_calc("PCoA") %>% 
  ord_get() %>% 
  phyloseq::plot_scree() + theme(axis.text.x = element_text(size = 6))
```

Let us view the 1st and 3rd axes.

```{r bray 3}
ibd %>% 
  tax_transform("identity", rank = "Genus") %>% # don't transform!
  dist_calc("bray") %>% 
  ord_calc("PCoA") %>% 
  ord_plot(axes = c(1, 3), color = "ibd", shape = "DiseaseState", size = 2) +
  scale_colour_brewer(palette = "Dark2") 
```

## RDA {#rda}

Redundancy analysis is a constrained ordination method. It displays the microbial variation that can also be explained by selected constraint variables.

*Behind the scenes, a linear regression model is created for each microbial abundance variable (using the constraints as the explanatory variables) and a PCA is performed using the fitted values of the microbial abundances.*

Starting from the same phyloseq object `ibd` the code below first creates a couple of binary (0/1) numeric variables to use a constraint variables. This is easy enough with ps_mutate.

Then we aggregate and transform our taxa, and like PCA we skip the dist_calc step.

```{r rda, fig.height=7, fig.width=7}
ibd %>%
  ps_mutate(
    IBD = as.numeric(ibd == "ibd"),
    Female = as.numeric(gender == "female"),
    Abx. = as.numeric(abx == "abx")
  ) %>%
  tax_transform("clr", rank = "Genus") %>%
  ord_calc(
    constraints = c("IBD", "Female", "Abx."),
    # method = "RDA", # Note: you can specify RDA explicitly, and it is good practice to do so, but microViz can guess automatically that you want an RDA here (helpful if you don't remember the name?)
    scale_cc = FALSE # doesn't make a difference
  ) %>%
  ord_plot(
    colour = "DiseaseState", size = 2, alpha = 0.5, shape = "active",
    plot_taxa = 1:8
  )
```

### Customising your ordination plot

The plot above looks okay by default, but it is fairly easy to tweak ord_plot further to get the style just how you want it. The code below has comments to explain which part makes which changes to the plot.

```{r rda custom, fig.width=7, fig.height=8}
# first we make a function that replaces any unwanted "_" in our taxa labels with spaces
library(stringr) 
renamer <- function(x) str_replace(x, pattern = "_", replacement = " ")

ibd %>%
  ps_mutate(
    IBD = as.numeric(ibd == "ibd"),
    Female = as.numeric(gender == "female"),
    Abx. = as.numeric(abx == "abx")
  ) %>%
  tax_transform("clr", rank = "Genus") %>%
  ord_calc(
    constraints = c("IBD", "Female", "Abx."),
    method = "RDA",
    scale_cc = FALSE # doesn't make a difference
  ) %>%
  ord_plot(
    colour = "DiseaseState", size = 2, alpha = 0.5, shape = "active",
    auto_caption = NA, # remove the helpful automatic caption
    plot_taxa = 1:8, taxon_renamer = renamer, # renamer is the function we made earlier
    tax_lab_length = 3.4, # this multiplier moves the labels, independently of the arrowheads
    tax_lab_style = list(size = 1.8, alpha = 0.5), # give a list of aesthetics to tweak the taxa labels' default style
    constraint_vec_length = 3, # this adjusts the length of the constraint arrows, and the labels track these lengths by default
    constraint_vec_style = list(size = 1.5, alpha = 0.5), # this styles the constraint arrows
    constraint_lab_style = list(size = 3) # this styles the constraint labels
  ) +
  # the functions below are from ggplot2:
  # You can pick a different colour scale, such as a color_brewer palette
  scale_colour_brewer(palette = "Set1") +
  # You can set any scale's values manually, such as the shapes used
  scale_shape_manual(values = c(
    active = "circle", mild = "circle cross",
    inactive = "circle open", control = "square open"
  )) +
  # this is how you add a title and subtitle
  ggtitle(
    label = "[Insert your exciting interpretations here?]",
    subtitle = "RDA with clr-transformed genera: constraints in red, taxa in black"
  ) +
  # and this is how you make your own caption
  labs(caption = "91 samples, 178 genera. Type 2 scaling.") +
  # this is one way to set the aspect ratio of the plot
  coord_fixed(ratio = 1, clip = "off")
```

### Partial ordinations

Tutorial coming soon, for now see `ord_plot()` for examples.

## Session info

```{r session}
devtools::session_info()
```
