<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>tree-development • microViz</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="tree-development">
<meta property="og:description" content="microViz">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">microViz</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/PCA-sorted-polar-composition-plots.html">PCA-sorted polar composition plots</a>
    </li>
    <li>
      <a href="../articles/tree-development.html">tree-development</a>
    </li>
    <li>
      <a href="../articles/Visualising-compositions.html">Visualising-compositions</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/david-barnett/microViz/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="tree-development_files/accessible-code-block-0.0.1/empty-anchor.js"></script><link href="tree-development_files/anchor-sections-1.0/anchor-sections.css" rel="stylesheet">
<script src="tree-development_files/anchor-sections-1.0/anchor-sections.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>tree-development</h1>
                        <h4 class="author">David Barnett</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/david-barnett/microViz/blob/master/vignettes/tree-development.Rmd"><code>vignettes/tree-development.Rmd</code></a></small>
      <div class="hidden name"><code>tree-development.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">100</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/david-barnett/microViz">microViz</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://microbiome.github.io/microbiome">microbiome</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://igraph.org">igraph</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidygraph.data-imaginist.com">tidygraph</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggraph.data-imaginist.com">ggraph</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com">patchwork</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tibble.tidyverse.org/">tibble</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">corncob</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/HenrikBengtsson/future">future</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/HenrikBengtsson/future.apply">future.apply</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://colorspace.R-Forge.R-project.org">colorspace</a></span><span class="op">)</span>
<span class="fu">knitr</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/knitr/man/opts_chunk.html">opts_chunk</a></span><span class="op">$</span><span class="fu">set</span><span class="op">(</span>fig.width <span class="op">=</span> <span class="fl">9</span>, fig.height <span class="op">=</span> <span class="fl">6</span><span class="op">)</span></code></pre></div>
<div id="intro" class="section level1">
<h1 class="hasAnchor">
<a href="#intro" class="anchor"></a>Intro</h1>
<p>This is a work in progress script about visualising taxon-specific statistics on a taxonomic tree (kinda like a phylogenetic tree with arbitrary branch lengths and with connections not directly defined by phylogenetic similarity).</p>
<p>In the future I will turn this process into a (couple of) function(s) with documentation for visualising any statistic/numbers as the colour and size of the nodes and edges of this tree.</p>
<p>The layout should be somewhat like a circular tree, rooted at e.g. kingdom, but not uniformly arranged. The characteristic/quasi-stochastic visual arrangement of the tree actually will be a feature that allows easy visual orientation across multiple copies of the same tree.</p>
<p>Much of this can be done with <code><a href="https://rdrr.io/pkg/metacoder/man/heat_tree.html">metacoder::heat_tree()</a></code> but this package has several limitations in my opinion:</p>
<ul>
<li>it generates a kind of ugly “legend” for which the breaks and sig.figs etc cannot be customised</li>
<li>it draws the nodes with <code>geom_polygon</code>
<ul>
<li>preventing addition of customised legends</li>
<li>preventing flexible (re)styling of the ggplot output</li>
<li>preventing the straightforward addition of further (filtered) geoms (e.g. for highlighting points)</li>
</ul>
</li>
<li>the <code>metacoder</code> package is very complicated, meaning:
<ul>
<li>debugging errors is just frustrating</li>
<li>if it fails in the future/or becomes incompatible with a future R or dependency version, I cannot fix it (it currently already uses deprecated tidyverse functions that throw warnings…)</li>
</ul>
</li>
<li>it is slow to draw multiple plots in a grid because it has to calculate the layout again for each one</li>
</ul>
</div>
<div id="setting-up-some-example-data" class="section level1">
<h1 class="hasAnchor">
<a href="#setting-up-some-example-data" class="anchor"></a>Setting up some example data</h1>
<p>Getting example dataset <code>dietswap</code>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"dietswap"</span>, package <span class="op">=</span> <span class="st">"microbiome"</span><span class="op">)</span>
<span class="va">ps</span> <span class="op">&lt;-</span> <span class="va">dietswap</span>

<span class="co"># filter out rare taxa</span>
<span class="va">ps</span> <span class="op">&lt;-</span> <span class="va">ps</span> <span class="op">%&gt;%</span> <span class="fu"><a href="../reference/tax_filter.html">tax_filter</a></span><span class="op">(</span>min_prevalence <span class="op">=</span> <span class="fl">0.2</span>, min_total_abundance <span class="op">=</span> <span class="fl">10000</span><span class="op">)</span>
<span class="co">#&gt; Proportional min_prevalence given: 0.2 --&gt; min 45/222 samples.</span>

<span class="co"># introduce some unclear taxonomic classifications</span>
<span class="co"># (to demonstrate fixing these with helper functions)</span>
<span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/tax_table-methods.html">tax_table</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">8</span>, <span class="fl">9</span><span class="op">)</span>, <span class="st">"Genus"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"unknown"</span>
<span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/tax_table-methods.html">tax_table</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span><span class="op">[</span><span class="fl">8</span>, <span class="st">"Family"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"unknown"</span></code></pre></div>
<p>When preparing a phyloseq dataset for stats and tree-based plotting it is critically important that the taxonomic names at each rank are fully identifiable/unique both within and across ranks, as they will function as the unique node IDs for the taxonomic tree graph structure.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># fix problem names</span>
<span class="va">ps</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tax_table_helpers.html">tax_fill_unknowns</a></span><span class="op">(</span><span class="va">ps</span>, unknowns <span class="op">=</span> <span class="st">"unknown"</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'purrr'</span>
<span class="co">#&gt; The following objects are masked from 'package:igraph':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     compose, simplify</span>

<span class="co"># This example dataset has some taxa with the same name for phylum and family... this needs to be fixed</span>
<span class="va">ps</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tax_table_helpers.html">tax_prepend_ranks</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span>
<span class="co"># this example dataset also has no root level: this is unusual and needs to be fixed</span>
<span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/tax_table-methods.html">tax_table</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span>root <span class="op">=</span> <span class="st">"root"</span>, <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/tax_table-methods.html">tax_table</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/tax_table-methods.html">tax_table</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span><span class="op">[</span><span class="fl">7</span><span class="op">:</span><span class="fl">11</span>, <span class="op">]</span>
<span class="co">#&gt; Taxonomy Table:     [5 taxa by 4 taxonomic ranks]:</span>
<span class="co">#&gt;                                  root   Phylum              Family                       </span>
<span class="co">#&gt; Bacteroides splachnicus et rel.  "root" "P: Bacteroidetes"  "F: Bacteroidetes"           </span>
<span class="co">#&gt; Bacteroides uniformis et rel.    "root" "P: Bacteroidetes"  "F: Bacteroidetes.Family"    </span>
<span class="co">#&gt; Bacteroides vulgatus et rel.     "root" "P: Bacteroidetes"  "F: Bacteroidetes"           </span>
<span class="co">#&gt; Bifidobacterium                  "root" "P: Actinobacteria" "F: Actinobacteria"          </span>
<span class="co">#&gt; Bryantella formatexigens et rel. "root" "P: Firmicutes"     "F: Clostridium cluster XIVa"</span>
<span class="co">#&gt;                                  Genus                                 </span>
<span class="co">#&gt; Bacteroides splachnicus et rel.  "G: Bacteroides splachnicus et rel."  </span>
<span class="co">#&gt; Bacteroides uniformis et rel.    "G: Bacteroidetes.unknown.Genus"      </span>
<span class="co">#&gt; Bacteroides vulgatus et rel.     "G: Bacteroidetes.Bacteroidetes.Genus"</span>
<span class="co">#&gt; Bifidobacterium                  "G: Bifidobacterium"                  </span>
<span class="co">#&gt; Bryantella formatexigens et rel. "G: Bryantella formatexigens et rel."</span></code></pre></div>
<div id="stats-to-decorate-trees" class="section level2">
<h2 class="hasAnchor">
<a href="#stats-to-decorate-trees" class="anchor"></a>Stats to decorate trees</h2>
<p>To get stats for every node of a taxonomic tree we must statistically model taxa at all taxonomic levels (except root).</p>
<p>It is first important to set up binary variables for the stats models, as heat_trees or other heatmap-type visualisations cannot be used with unordered categorical variables.</p>
<p><strong>Note:</strong> <em>If you have continuous variables and wish to visualise the size of their microbial associations (e.g. regression coefs) on a similar (colour) scale as for associations with binary categorical variables, then you should scale the range of your continuous variable to between 0 and 1 (with e.g. min-max normalisation) if it is not already on a 0 to 1 scale.</em></p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/sample_data-methods.html">sample_data</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span><span class="op">$</span><span class="va">female</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html">if_else</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/sample_data-methods.html">sample_data</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span><span class="op">$</span><span class="va">sex</span> <span class="op">==</span> <span class="st">"female"</span>, <span class="fl">1</span>, <span class="fl">0</span>, <span class="cn">NaN</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/sample_data-methods.html">sample_data</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span><span class="op">$</span><span class="va">overweight</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html">if_else</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/sample_data-methods.html">sample_data</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span><span class="op">$</span><span class="va">bmi_group</span> <span class="op">==</span> <span class="st">"overweight"</span>, <span class="fl">1</span>, <span class="fl">0</span>, <span class="cn">NaN</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/sample_data-methods.html">sample_data</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span><span class="op">$</span><span class="va">obese</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html">if_else</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/sample_data-methods.html">sample_data</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span><span class="op">$</span><span class="va">bmi_group</span> <span class="op">==</span> <span class="st">"obese"</span>, <span class="fl">1</span>, <span class="fl">0</span>, <span class="cn">NaN</span><span class="op">)</span>
<span class="va">variables</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"female"</span>, <span class="st">"overweight"</span>, <span class="st">"obese"</span><span class="op">)</span>

<span class="va">tax_models_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span>
  X <span class="op">=</span> <span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/rank_names.html">rank_names</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span>,
  <span class="kw">function</span><span class="op">(</span><span class="va">r</span><span class="op">)</span> <span class="op">{</span>
    <span class="fu"><a href="https://rdrr.io/r/base/message.html">message</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html">Sys.time</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">" - modelling at level: "</span>, <span class="va">r</span><span class="op">)</span>
    <span class="va">models</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/model_tax_corncob.html">model_tax_corncob</a></span><span class="op">(</span>
      ps <span class="op">=</span> <span class="va">ps</span>,
      tax_level <span class="op">=</span> <span class="va">r</span>, taxa <span class="op">=</span> <span class="st">"all"</span>,
      variables <span class="op">=</span> <span class="va">variables</span>
    <span class="op">)</span>
    <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="va">models</span><span class="op">)</span>
  <span class="op">}</span>
<span class="op">)</span>
<span class="co">#&gt; 2021-01-11 09:30:22 - modelling at level: Phylum</span>
<span class="co">#&gt; 2021-01-11 09:30:24 - modelling at level: Family</span>
<span class="co">#&gt; 2021-01-11 09:30:25 - modelling at level: Genus</span>
<span class="co"># Flatten all taxonomic level lists into one list of models</span>
<span class="co"># (this is why taxon names must be completely identifiable across ranks)</span>
<span class="va">flat_models_list</span> <span class="op">&lt;-</span> <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/flatten.html">flatten</a></span><span class="op">(</span><span class="va">tax_models_list</span><span class="op">)</span>
<span class="va">var_stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/model_tax_corncob.html">models2stats_corncob</a></span><span class="op">(</span><span class="va">flat_models_list</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="from-phyloseq-to-treegraph" class="section level1">
<h1 class="hasAnchor">
<a href="#from-phyloseq-to-treegraph" class="anchor"></a>From phyloseq to tree/graph</h1>
<div id="nodes-and-edges" class="section level2">
<h2 class="hasAnchor">
<a href="#nodes-and-edges" class="anchor"></a>Nodes and edges</h2>
<ol style="list-style-type: decimal">
<li>Create a dataframe of taxon nodes (nodes are the points on the tree graph)</li>
<li>Join stats about the associations with the variable <code>obese</code> to the augmented nodes dataframe</li>
<li>Create a dataframe of edges from the nodes (edges join the nodes in the tree together)</li>
</ol>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">nodes_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/taxatree_funs.html">taxatree_nodes</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span>
<span class="va">nodes_df</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span><span class="va">nodes_df</span>, <span class="va">var_stats</span><span class="op">$</span><span class="va">obese</span>, by <span class="op">=</span> <span class="st">"taxon_name"</span><span class="op">)</span>
<span class="va">edge_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/taxatree_funs.html">taxatree_edges</a></span><span class="op">(</span><span class="va">nodes_df</span><span class="op">)</span>
<span class="va">graph</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tidygraph.data-imaginist.com/reference/tbl_graph.html">tbl_graph</a></span><span class="op">(</span>nodes <span class="op">=</span> <span class="va">nodes_df</span>, node_key <span class="op">=</span> <span class="st">"taxon_name"</span>, edges <span class="op">=</span> <span class="va">edge_df</span>, directed <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
</div>
<div id="creating-a-circular-tree-graph-layout" class="section level2">
<h2 class="hasAnchor">
<a href="#creating-a-circular-tree-graph-layout" class="anchor"></a>Creating a circular tree graph layout</h2>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># create layout</span>
<span class="va">layout</span> <span class="op">&lt;-</span> <span class="fu">ggraph</span><span class="fu">::</span><span class="fu"><a href="https://ggraph.data-imaginist.com/reference/ggraph.html">create_layout</a></span><span class="op">(</span><span class="va">graph</span>, layout <span class="op">=</span> <span class="st">"tree"</span>, circular <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="co"># plot example with this layout</span>
<span class="fu"><a href="https://ggraph.data-imaginist.com/reference/ggraph.html">ggraph</a></span><span class="op">(</span>graph <span class="op">=</span> <span class="va">layout</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/theme_graph.html">theme_graph</a></span><span class="op">(</span>base_family <span class="op">=</span> <span class="st">"sans"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/geom_edge_link.html">geom_edge_link</a></span><span class="op">(</span>
    mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>
      edge_width <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">sqrt</a></span><span class="op">(</span><span class="va">taxon_mean</span><span class="op">)</span>,
      edge_colour <span class="op">=</span> <span class="va">b_mu</span>
    <span class="op">)</span>,
    show.legend <span class="op">=</span> <span class="cn">FALSE</span>
  <span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/geom_node_point.html">geom_node_point</a></span><span class="op">(</span>
    mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>
      size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">sqrt</a></span><span class="op">(</span><span class="va">taxon_mean</span><span class="op">)</span>,
      fill <span class="op">=</span> <span class="va">b_mu</span>
    <span class="op">)</span>,
    shape <span class="op">=</span> <span class="fl">21</span>,
    show.legend <span class="op">=</span> <span class="cn">FALSE</span>
  <span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/scale_edge_width.html">scale_edge_width_continuous</a></span><span class="op">(</span>range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/scale_edge_colour.html">scale_edge_colour_gradient2</a></span><span class="op">(</span>low <span class="op">=</span> <span class="st">"#a6611a"</span>, mid <span class="op">=</span> <span class="st">"#DDDDDD"</span>, high <span class="op">=</span> <span class="st">"#018571"</span>, midpoint <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html">scale_fill_gradient2</a></span><span class="op">(</span>low <span class="op">=</span> <span class="st">"#a6611a"</span>, mid <span class="op">=</span> <span class="st">"#DDDDDD"</span>, high <span class="op">=</span> <span class="st">"#018571"</span>, midpoint <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span>
  <span class="cn">NULL</span></code></pre></div>
<p><img src="tree-development_files/figure-html/unnamed-chunk-7-1.png" width="384"></p>
</div>
</div>
<div id="adjusting-the-layout" class="section level1">
<h1 class="hasAnchor">
<a href="#adjusting-the-layout" class="anchor"></a>Adjusting the layout</h1>
<p>Here I use the davidson-harel algorithm, with trial-and-error weights, to adjust the tree layout above, using a trial-and-error number of iterations.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Davidson-Harel allows weighting crossover avoidance and edge lengths etc.</span>
<span class="va">adjust_tree_layout</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">graph</span>, <span class="va">layout</span><span class="op">)</span> <span class="op">{</span>
  <span class="va">e_density</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/edge_density.html">edge_density</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span>
  <span class="va">original_layout_coords</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">layout</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"x"</span>, <span class="st">"y"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>

  <span class="va">new_layout_coords</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/layout_with_dh.html">layout_with_dh</a></span><span class="op">(</span>
    graph <span class="op">=</span> <span class="va">graph</span>,
    coords <span class="op">=</span> <span class="va">original_layout_coords</span>,
    maxiter <span class="op">=</span> <span class="fl">3</span>, fineiter <span class="op">=</span> <span class="fl">10</span>,
    weight.edge.lengths <span class="op">=</span> <span class="va">e_density</span> <span class="op">*</span> <span class="fl">100</span>,
    weight.edge.crossings <span class="op">=</span> <span class="va">e_density</span> <span class="op">*</span> <span class="fl">100000</span>,
    weight.node.edge.dist <span class="op">=</span> <span class="va">e_density</span> <span class="op">*</span> <span class="fl">10</span>,
    weight.node.dist <span class="op">=</span> <span class="va">e_density</span> <span class="op">*</span> <span class="fl">1</span>,
    weight.border <span class="op">=</span> <span class="va">e_density</span> <span class="op">*</span> <span class="fl">1</span>
  <span class="op">)</span>

  <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="va">new_layout_coords</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
<p>This is a stochastic process which I do not let find an optimum. So <code><a href="https://rdrr.io/r/base/Random.html">set.seed()</a></code> is important here.</p>
<p>I will expose the weights of the dh algo (with my defaults).<br>
I will document how to rerun a few times with RNG seed setting to help users find a layout they like and re-use it (as below).</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">layouts_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="op">{</span>
  <span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="va">i</span><span class="op">)</span>
  <span class="fu">adjust_tree_layout</span><span class="op">(</span>g <span class="op">=</span> <span class="va">graph</span>, l <span class="op">=</span> <span class="va">layout</span><span class="op">)</span>
<span class="op">}</span><span class="op">)</span></code></pre></div>
<div id="plotting-the-adjusted-layouts" class="section level2">
<h2 class="hasAnchor">
<a href="#plotting-the-adjusted-layouts" class="anchor"></a>Plotting the adjusted layouts</h2>
<p>These differ only randomly, some might be more aesthetically pleasing than others. I like the top right one.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">abs_sqrt</span> <span class="op">&lt;-</span> <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org//reference/trans_new.html">trans_new</a></span><span class="op">(</span>
  name <span class="op">=</span> <span class="st">"abs_sqrt"</span>,
  transform <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span>
    <span class="fu"><a href="https://rdrr.io/r/base/sign.html">sign</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">sqrt</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span>
  <span class="op">}</span>,
  inverse <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span>
    <span class="fu"><a href="https://rdrr.io/r/base/sign.html">sign</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">*</span> <span class="va">x</span><span class="op">^</span><span class="fl">2</span>
  <span class="op">}</span>
<span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">luminance_l2</span> <span class="op">&lt;-</span> <span class="fl">80</span>
<span class="va">colour_palette</span> <span class="op">&lt;-</span> <span class="st">"Green-Brown"</span>
<span class="va">reverse_colours</span> <span class="op">&lt;-</span> <span class="cn">FALSE</span>
<span class="va">luminance_l2</span> <span class="op">&lt;-</span> <span class="fl">80</span>
<span class="va">colour_lims</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">2</span>, <span class="fl">2</span><span class="op">)</span>

<span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">layouts_list</span>, <span class="kw">function</span><span class="op">(</span><span class="va">layout_coords</span><span class="op">)</span> <span class="op">{</span>
  <span class="fu"><a href="../reference/taxatree_plot.html">taxatree_plot</a></span><span class="op">(</span>graph_object <span class="op">=</span> <span class="va">graph</span>, layout_coords <span class="op">=</span> <span class="va">layout_coords</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">colorspace</span><span class="fu">::</span><span class="fu"><a href="http://colorspace.R-Forge.R-project.org//reference/scale_colour_continuous_diverging.html">scale_fill_continuous_diverging</a></span><span class="op">(</span>
      palette <span class="op">=</span> <span class="va">colour_palette</span>,
      l2 <span class="op">=</span> <span class="va">luminance_l2</span>,
      aesthetics <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"edge_colour"</span>, <span class="st">"fill"</span><span class="op">)</span>,
      limits <span class="op">=</span> <span class="va">colour_lims</span>,
      oob <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org//reference/oob.html">oob_squish</a></span>,
      rev <span class="op">=</span> <span class="va">reverse_colours</span>,
      trans <span class="op">=</span> <span class="va">abs_sqrt</span>,
      guide <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_colourbar.html">guide_colorbar</a></span><span class="op">(</span>
        frame.colour <span class="op">=</span> <span class="st">"black"</span>,
        ticks.colour <span class="op">=</span> <span class="st">"black"</span>,
        barwidth <span class="op">=</span> <span class="fu">grid</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grid/unit.html">unit</a></span><span class="op">(</span><span class="fl">0.01</span>, <span class="st">"npc"</span><span class="op">)</span>
      <span class="op">)</span>
    <span class="op">)</span> <span class="op">+</span>
    <span class="co"># colour scale for sig. points stroke (which has no legend)</span>
    <span class="fu">colorspace</span><span class="fu">::</span><span class="fu"><a href="http://colorspace.R-Forge.R-project.org//reference/scale_colour_continuous_diverging.html">scale_color_continuous_diverging</a></span><span class="op">(</span>
      palette <span class="op">=</span> <span class="va">colour_palette</span>,
      l2 <span class="op">=</span> <span class="va">luminance_l2</span>,
      aesthetics <span class="op">=</span> <span class="st">"color"</span>,
      rev <span class="op">=</span> <span class="va">reverse_colours</span>,
      trans <span class="op">=</span> <span class="va">abs_sqrt</span>,
      guide <span class="op">=</span> <span class="st">"none"</span>
    <span class="op">)</span>
<span class="op">}</span><span class="op">)</span>

<span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html">wrap_plots</a></span><span class="op">(</span><span class="va">plots</span>, ncol <span class="op">=</span> <span class="fl">2</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span>, guides <span class="op">=</span> <span class="st">"keep"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_annotation.html">plot_annotation</a></span><span class="op">(</span>tag_levels <span class="op">=</span> <span class="st">"1"</span><span class="op">)</span></code></pre></div>
<p><img src="tree-development_files/figure-html/unnamed-chunk-12-1.png" width="3000"></p>
</div>
<div id="selecting-1-of-the-adjusted-layouts" class="section level2">
<h2 class="hasAnchor">
<a href="#selecting-1-of-the-adjusted-layouts" class="anchor"></a>Selecting 1 of the adjusted layouts</h2>
<p>Besides just selecting the layout number 3 with <code>layouts_list[[3]]</code>, the adjustment of the original layout can be rerun exactly by setting the seed as equal to 3 again, as demonstrated below. You could use a larger grey tree version with labels as a key, as shown below.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span>
<span class="va">al</span> <span class="op">&lt;-</span> <span class="fu">adjust_tree_layout</span><span class="op">(</span>graph <span class="op">=</span> <span class="va">graph</span>, layout <span class="op">=</span> <span class="va">layout</span><span class="op">)</span>

<span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/ggraph.html">ggraph</a></span><span class="op">(</span>graph <span class="op">=</span> <span class="va">graph</span>, layout <span class="op">=</span> <span class="va">al</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/theme_graph.html">theme_graph</a></span><span class="op">(</span>base_family <span class="op">=</span> <span class="st">"sans"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/geom_edge_link.html">geom_edge_link</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>edge_width <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log</a></span><span class="op">(</span><span class="va">taxon_mean</span><span class="op">)</span><span class="op">)</span>,
    edge_colour <span class="op">=</span> <span class="st">"grey"</span>,
    alpha <span class="op">=</span> <span class="fl">0.5</span>,
    show.legend <span class="op">=</span> <span class="cn">FALSE</span>
  <span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/geom_node_point.html">geom_node_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log</a></span><span class="op">(</span><span class="va">taxon_mean</span><span class="op">)</span><span class="op">)</span>,
    fill <span class="op">=</span> <span class="st">"grey"</span>,
    shape <span class="op">=</span> <span class="fl">21</span>,
    show.legend <span class="op">=</span> <span class="cn">FALSE</span>
  <span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_size.html">scale_size_continuous</a></span><span class="op">(</span>range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">9</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/scale_edge_width.html">scale_edge_width_continuous</a></span><span class="op">(</span>range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">6</span><span class="op">)</span><span class="op">)</span>

<span class="co"># simpler way to add labels to families</span>
<span class="co"># p &lt;- p +</span>
<span class="co">#   geom_node_label(</span>
<span class="co">#     mapping = aes(</span>
<span class="co">#       label = taxon_name,</span>
<span class="co">#       size = log(log(log(taxon_mean))),</span>
<span class="co">#       filter = grepl("^F:", taxon_name)</span>
<span class="co">#     ),</span>
<span class="co">#     repel = TRUE,</span>
<span class="co">#     show.legend = FALSE</span>
<span class="co">#   )</span>

<span class="co"># alternative way of setting labels with more control</span>
<span class="va">p</span> <span class="op">&lt;-</span> <span class="va">p</span> <span class="op">+</span>
  <span class="fu">ggrepel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggrepel/man/geom_text_repel.html">geom_label_repel</a></span><span class="op">(</span>
    data <span class="op">=</span> <span class="op">~</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">.</span>, <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span><span class="op">(</span><span class="st">"^F:"</span>, <span class="va">taxon_name</span><span class="op">)</span><span class="op">)</span>,
    mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>
      x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, label <span class="op">=</span> <span class="va">taxon_name</span>,
      size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log</a></span><span class="op">(</span><span class="va">taxon_mean</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
    <span class="op">)</span>,
    xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="cn">Inf</span>, <span class="cn">Inf</span><span class="op">)</span>,
    min.segment.length <span class="op">=</span> <span class="fl">0</span>,
    show.legend <span class="op">=</span> <span class="cn">FALSE</span>
  <span class="op">)</span>

<span class="co"># add a black central node to mark the rook</span>
<span class="va">p</span> <span class="op">&lt;-</span> <span class="va">p</span> <span class="op">+</span> <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/geom_node_point.html">geom_node_point</a></span><span class="op">(</span>
  mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>
    filter <span class="op">=</span> <span class="va">taxon_level</span> <span class="op">==</span> <span class="st">"root"</span>,
    size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log</a></span><span class="op">(</span><span class="va">taxon_mean</span><span class="op">)</span>
  <span class="op">)</span>,
  show.legend <span class="op">=</span> <span class="cn">FALSE</span>
<span class="op">)</span>

<span class="va">p</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span>caption <span class="op">=</span> <span class="st">"Note: Most of these are not really Family-level names, but\nthey are listed as such in the example dataset (dietswap)."</span><span class="op">)</span></code></pre></div>
<p><img src="tree-development_files/figure-html/unnamed-chunk-13-1.png" width="672"></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/sessioninfo/man/session_info.html">session_info</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; - Session info -----------------------------------------------------------------------------------</span>
<span class="co">#&gt;  setting  value                       </span>
<span class="co">#&gt;  version  R version 4.0.3 (2020-10-10)</span>
<span class="co">#&gt;  os       Windows 10 x64              </span>
<span class="co">#&gt;  system   x86_64, mingw32             </span>
<span class="co">#&gt;  ui       RTerm                       </span>
<span class="co">#&gt;  language (EN)                        </span>
<span class="co">#&gt;  collate  English_United Kingdom.1252 </span>
<span class="co">#&gt;  ctype    English_United Kingdom.1252 </span>
<span class="co">#&gt;  tz       Europe/Berlin               </span>
<span class="co">#&gt;  date     2021-01-11                  </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; - Packages ---------------------------------------------------------------------------------------</span>
<span class="co">#&gt;  package            * version      date       lib source                               </span>
<span class="co">#&gt;  ade4                 1.7-16       2020-10-28 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  ape                  5.4-1        2020-08-13 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  assertthat           0.2.1        2019-03-21 [3] CRAN (R 4.0.0)                       </span>
<span class="co">#&gt;  Biobase              2.48.0       2020-04-27 [3] Bioconductor                         </span>
<span class="co">#&gt;  BiocGenerics         0.34.0       2020-04-27 [3] Bioconductor                         </span>
<span class="co">#&gt;  biomformat           1.16.0       2020-04-27 [3] Bioconductor                         </span>
<span class="co">#&gt;  Biostrings           2.56.0       2020-04-27 [3] Bioconductor                         </span>
<span class="co">#&gt;  callr                3.5.1        2020-10-13 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  cli                  2.2.0        2020-11-20 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  cluster              2.1.0        2019-06-19 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  codetools            0.2-18       2020-11-04 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  colorspace         * 2.0-0        2020-11-11 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  corncob            * 0.1.0        2020-10-08 [3] Github (bryandmartin/corncob@338323e)</span>
<span class="co">#&gt;  crayon               1.3.4        2017-09-16 [3] CRAN (R 4.0.0)                       </span>
<span class="co">#&gt;  data.table           1.13.2       2020-10-19 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  desc                 1.2.0        2018-05-01 [3] CRAN (R 4.0.0)                       </span>
<span class="co">#&gt;  detectseparation     0.1          2020-03-25 [3] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  devtools             2.3.2        2020-09-18 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  digest               0.6.27       2020-10-24 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  dplyr              * 1.0.2        2020-08-18 [3] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  ellipsis             0.3.1        2020-05-15 [3] CRAN (R 4.0.0)                       </span>
<span class="co">#&gt;  evaluate             0.14         2019-05-28 [3] CRAN (R 4.0.0)                       </span>
<span class="co">#&gt;  fansi                0.4.1        2020-01-08 [3] CRAN (R 4.0.0)                       </span>
<span class="co">#&gt;  farver               2.0.3        2020-01-16 [3] CRAN (R 4.0.0)                       </span>
<span class="co">#&gt;  foreach              1.5.1        2020-10-15 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  fs                   1.5.0        2020-07-31 [3] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  furrr                0.2.1        2020-10-21 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  future             * 1.21.0       2020-12-10 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  future.apply       * 1.7.0        2021-01-04 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  generics             0.1.0        2020-10-31 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  ggforce              0.3.2        2020-06-23 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  ggplot2            * 3.3.3        2020-12-30 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  ggraph             * 2.0.4        2020-11-16 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  ggrepel              0.9.0        2020-12-16 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  globals              0.14.0       2020-11-22 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  glue                 1.4.2        2020-08-27 [3] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  graphlayouts         0.7.1        2020-10-26 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  gridExtra            2.3          2017-09-09 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  gtable               0.3.0        2019-03-25 [3] CRAN (R 4.0.0)                       </span>
<span class="co">#&gt;  hms                  0.5.3        2020-01-08 [3] CRAN (R 4.0.0)                       </span>
<span class="co">#&gt;  htmltools            0.5.0        2020-06-16 [3] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  igraph             * 1.2.6        2020-10-06 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  IRanges              2.22.2       2020-05-21 [3] Bioconductor                         </span>
<span class="co">#&gt;  iterators            1.0.13       2020-10-15 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  jsonlite             1.7.2        2020-12-09 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  knitr                1.30         2020-09-22 [3] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  labeling             0.4.2        2020-10-20 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  lattice              0.20-41      2020-04-02 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  lifecycle            0.2.0        2020-03-06 [3] CRAN (R 4.0.0)                       </span>
<span class="co">#&gt;  listenv              0.8.0        2019-12-05 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  lpSolveAPI           5.5.2.0-17.7 2020-06-14 [3] CRAN (R 4.0.0)                       </span>
<span class="co">#&gt;  magrittr             2.0.1        2020-11-17 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  MASS                 7.3-53       2020-09-09 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  Matrix               1.2-18       2019-11-27 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  memoise              1.1.0        2017-04-21 [3] CRAN (R 4.0.0)                       </span>
<span class="co">#&gt;  mgcv                 1.8-33       2020-08-27 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  microbiome         * 1.10.0       2020-04-27 [3] Bioconductor                         </span>
<span class="co">#&gt;  microViz           * 0.0.4        2021-01-11 [1] local                                </span>
<span class="co">#&gt;  multtest             2.44.0       2020-04-27 [3] Bioconductor                         </span>
<span class="co">#&gt;  munsell              0.5.0        2018-06-12 [3] CRAN (R 4.0.0)                       </span>
<span class="co">#&gt;  nlme                 3.1-150      2020-10-24 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  numDeriv             2016.8-1.1   2019-06-06 [3] CRAN (R 4.0.0)                       </span>
<span class="co">#&gt;  parallelly           1.23.0       2021-01-04 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  patchwork          * 1.1.1        2020-12-17 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  permute              0.9-5        2019-03-12 [3] CRAN (R 4.0.0)                       </span>
<span class="co">#&gt;  phyloseq           * 1.32.0       2020-04-27 [3] Bioconductor                         </span>
<span class="co">#&gt;  pillar               1.4.7        2020-11-20 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  pkgbuild             1.2.0        2020-12-15 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  pkgconfig            2.0.3        2019-09-22 [3] CRAN (R 4.0.0)                       </span>
<span class="co">#&gt;  pkgdown              1.6.1        2020-09-12 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  pkgload              1.1.0        2020-05-29 [3] CRAN (R 4.0.0)                       </span>
<span class="co">#&gt;  plyr                 1.8.6        2020-03-03 [3] CRAN (R 4.0.0)                       </span>
<span class="co">#&gt;  polyclip             1.10-0       2019-03-14 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  prettyunits          1.1.1        2020-01-24 [3] CRAN (R 4.0.0)                       </span>
<span class="co">#&gt;  processx             3.4.5        2020-11-30 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  progress             1.2.2        2019-05-16 [3] CRAN (R 4.0.0)                       </span>
<span class="co">#&gt;  ps                   1.5.0        2020-12-05 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  purrr              * 0.3.4        2020-04-17 [3] CRAN (R 4.0.0)                       </span>
<span class="co">#&gt;  R6                   2.5.0        2020-10-28 [3] CRAN (R 4.0.1)                       </span>
<span class="co">#&gt;  ragg                 0.4.0        2020-10-05 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  Rcpp                 1.0.5        2020-07-06 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  registry             0.5-1        2019-03-05 [3] CRAN (R 4.0.0)                       </span>
<span class="co">#&gt;  remotes              2.2.0        2020-07-21 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  reshape2             1.4.4        2020-04-09 [3] CRAN (R 4.0.0)                       </span>
<span class="co">#&gt;  rhdf5                2.32.4       2020-10-05 [3] Bioconductor                         </span>
<span class="co">#&gt;  Rhdf5lib             1.10.1       2020-07-09 [3] Bioconductor                         </span>
<span class="co">#&gt;  rlang                0.4.10       2020-12-30 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  rmarkdown            2.5          2020-10-21 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  ROI                  1.0-0        2020-08-31 [3] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  ROI.plugin.lpsolve   1.0-0        2020-08-29 [3] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  rprojroot            2.0.2        2020-11-15 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  rstudioapi           0.13         2020-11-12 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  Rtsne                0.15         2018-11-10 [3] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  S4Vectors            0.26.1       2020-05-16 [3] Bioconductor                         </span>
<span class="co">#&gt;  scales               1.1.1        2020-05-11 [3] CRAN (R 4.0.0)                       </span>
<span class="co">#&gt;  sessioninfo          1.1.1        2018-11-05 [3] CRAN (R 4.0.0)                       </span>
<span class="co">#&gt;  slam                 0.1-47       2019-12-21 [3] CRAN (R 4.0.0)                       </span>
<span class="co">#&gt;  stringi              1.5.3        2020-09-09 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  stringr              1.4.0        2019-02-10 [3] CRAN (R 4.0.0)                       </span>
<span class="co">#&gt;  survival             3.2-7        2020-09-28 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  systemfonts          0.3.2        2020-09-29 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  testthat             3.0.1        2020-12-17 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  textshaping          0.1.2        2020-10-08 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  tibble             * 3.0.4        2020-10-12 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  tidygraph          * 1.2.0        2020-05-12 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  tidyr                1.1.2        2020-08-27 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  tidyselect           1.1.0        2020-05-11 [3] CRAN (R 4.0.0)                       </span>
<span class="co">#&gt;  trust                0.1-8        2020-01-10 [3] CRAN (R 4.0.0)                       </span>
<span class="co">#&gt;  tweenr               1.0.1        2018-12-14 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  usethis              1.6.3        2020-09-17 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  vctrs                0.3.6        2020-12-17 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  vegan                2.5-7        2020-11-28 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  VGAM                 1.1-4        2020-10-23 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  viridis              0.5.1        2018-03-29 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  viridisLite          0.3.0        2018-02-01 [3] CRAN (R 4.0.0)                       </span>
<span class="co">#&gt;  withr                2.3.0        2020-09-22 [3] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  xfun                 0.19         2020-10-30 [3] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  XVector              0.28.0       2020-04-27 [3] Bioconductor                         </span>
<span class="co">#&gt;  yaml                 2.2.1        2020-02-01 [3] CRAN (R 4.0.0)                       </span>
<span class="co">#&gt;  zlibbioc             1.34.0       2020-04-27 [3] Bioconductor                         </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [1] C:/Users/david.barnett/AppData/Local/Temp/RtmpgZa9qP/temp_libpatha98460d3573</span>
<span class="co">#&gt; [2] C:/Users/david.barnett/AppData/Local/Temp/RtmpgZa9qP/temp_libpatha9879507e4c</span>
<span class="co">#&gt; [3] C:/Program Files/R/R-4.0.3/library</span></code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by David Barnett.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
