<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>tree-development • microViz</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="tree-development">
<meta property="og:description" content="microViz">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">microViz</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/tree-development.html">tree-development</a>
    </li>
    <li>
      <a href="../articles/Visualising-compositions.html">Visualising-compositions</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/david-barnett/microViz/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="tree-development_files/accessible-code-block-0.0.1/empty-anchor.js"></script><link href="tree-development_files/anchor-sections-1.0/anchor-sections.css" rel="stylesheet">
<script src="tree-development_files/anchor-sections-1.0/anchor-sections.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>tree-development</h1>
                        <h4 class="author">David Barnett</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/david-barnett/microViz/blob/master/vignettes/tree-development.Rmd"><code>vignettes/tree-development.Rmd</code></a></small>
      <div class="hidden name"><code>tree-development.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/david-barnett/microViz">microViz</a></span><span class="op">)</span></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://microbiome.github.io/microbiome">microbiome</a></span><span class="op">)</span>
<span class="co">#&gt; Loading required package: phyloseq</span>
<span class="co">#&gt; Loading required package: ggplot2</span>
<span class="co">#&gt; Warning: package 'ggplot2' was built under R version 4.0.2</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; microbiome R package (microbiome.github.com)</span>
<span class="co">#&gt;     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; </span>
<span class="co">#&gt;  Copyright (C) 2011-2020 Leo Lahti, </span>
<span class="co">#&gt;     Sudarshan Shetty et al. &lt;microbiome.github.io&gt;</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'microbiome'</span>
<span class="co">#&gt; The following object is masked from 'package:ggplot2':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     alpha</span>
<span class="co">#&gt; The following object is masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     transform</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://igraph.org">igraph</a></span><span class="op">)</span>
<span class="co">#&gt; Warning: package 'igraph' was built under R version 4.0.3</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'igraph'</span>
<span class="co">#&gt; The following object is masked from 'package:microbiome':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     diversity</span>
<span class="co">#&gt; The following objects are masked from 'package:stats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     decompose, spectrum</span>
<span class="co">#&gt; The following object is masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     union</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidygraph.data-imaginist.com">tidygraph</a></span><span class="op">)</span>
<span class="co">#&gt; Warning: package 'tidygraph' was built under R version 4.0.3</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'tidygraph'</span>
<span class="co">#&gt; The following object is masked from 'package:igraph':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     groups</span>
<span class="co">#&gt; The following object is masked from 'package:stats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     filter</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggraph.data-imaginist.com">ggraph</a></span><span class="op">)</span>
<span class="co">#&gt; Warning: package 'ggraph' was built under R version 4.0.3</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com">patchwork</a></span><span class="op">)</span>
<span class="co">#&gt; Warning: package 'patchwork' was built under R version 4.0.3</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span>
<span class="co">#&gt; Warning: package 'dplyr' was built under R version 4.0.2</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'dplyr'</span>
<span class="co">#&gt; The following objects are masked from 'package:igraph':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     as_data_frame, groups, union</span>
<span class="co">#&gt; The following objects are masked from 'package:stats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     filter, lag</span>
<span class="co">#&gt; The following objects are masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     intersect, setdiff, setequal, union</span></pre></div>
<div id="intro" class="section level1">
<h1 class="hasAnchor">
<a href="#intro" class="anchor"></a>Intro</h1>
<p>This is a work in progress script about visualising taxon-specific statistics on a taxonomic tree (kinda like a phylogenetic tree with arbitrary branch lengths and with connections not directly defined by phylogenetic similarity).</p>
<p>In the future I will turn this process into a (couple of) function(s) with documentation for visualising any statistic/numbers as the colour and size of the nodes and edges of this tree.</p>
<p>The layout should be somewhat like a circular tree, rooted at e.g. kingdom, but not uniformly arranged. The characteristic/quasi-stochastic visual arrangement of the tree actually will be a feature that allows easy visual orientation across multiple copies of the same tree.</p>
<p>Much of this can be done with <code><a href="https://rdrr.io/pkg/metacoder/man/heat_tree.html">metacoder::heat_tree()</a></code> but this package has several limitations in my opinion:</p>
<ul>
<li>it generates a kind of ugly “legend” for which the breaks and sig.figs etc cannot be customised</li>
<li>it draws the nodes with <code>geom_polygon</code>
<ul>
<li>preventing addition of customised legends</li>
<li>preventing flexible (re)styling of the ggplot output</li>
</ul>
</li>
<li>the <code>metacoder</code> package is very complicated, meaning:
<ul>
<li>debugging errors is just frustrating</li>
<li>if it fails in the future/or becomes incompatible with a future R or dependency version, I cannot fix it (it currently already uses deprecated tidyverse functions that throw warnings…)</li>
</ul>
</li>
</ul>
</div>
<div id="setting-up-some-example-data" class="section level1">
<h1 class="hasAnchor">
<a href="#setting-up-some-example-data" class="anchor"></a>Setting up some example data</h1>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">dietswap</span><span class="op">)</span>
<span class="va">ps</span> <span class="op">&lt;-</span> <span class="va">dietswap</span>

<span class="va">ps</span> <span class="op">&lt;-</span> <span class="va">ps</span> <span class="op">%&gt;%</span> <span class="fu"><a href="../reference/tax_filter.html">tax_filter</a></span><span class="op">(</span>min_prevalence <span class="op">=</span> <span class="fl">0.1</span>, min_total_abundance <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span>
<span class="co">#&gt; Proportional min_prevalence given: 0.1 --&gt; min 23/222 samples.</span>

<span class="co"># This example dataset has some taxa with the same name for phylum and family... this needs to be fixed</span>
<span class="va">tt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/rank_names.html">rank_names</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">r</span><span class="op">)</span><span class="op">{</span>
  <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/tax_table-methods.html">tax_table</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span><span class="op">[</span>, <span class="va">r</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html">as.vector</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/tax_table-methods.html">tax_table</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span><span class="op">[</span>, <span class="va">r</span><span class="op">]</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/substr.html">substr</a></span><span class="op">(</span><span class="va">r</span>, <span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>, <span class="st">': '</span>, <span class="va">.</span><span class="op">)</span>
<span class="op">}</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">tt</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/taxa_names-methods.html">taxa_names</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span>
<span class="co"># this example dataset also has no root, this is unusual and needs to be fixed</span>
<span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/tax_table-methods.html">tax_table</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span>root <span class="op">=</span> <span class="st">'root'</span>, <span class="va">tt</span><span class="op">)</span>

<span class="co"># test non-unique genera problem handling</span>
<span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/tax_table-methods.html">tax_table</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">3</span>,<span class="fl">4</span>,<span class="fl">56</span><span class="op">)</span>, <span class="st">'Genus'</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"G: unknown"</span></pre></div>
</div>
<div id="from-phyloseq-to-treegraph" class="section level1">
<h1 class="hasAnchor">
<a href="#from-phyloseq-to-treegraph" class="anchor"></a>From phyloseq to tree/graph</h1>
<div id="creating-a-circular-tree-graph-layout" class="section level2">
<h2 class="hasAnchor">
<a href="#creating-a-circular-tree-graph-layout" class="anchor"></a>Creating a circular tree graph layout</h2>
<p>Creating a graph/network object from a phyloseq will be incorporated into a function.</p>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="co"># create taxon nodes and calculate basic info about each taxon (at all ranks)</span>
<span class="va">ranks</span> <span class="op">&lt;-</span> <span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/rank_names.html">rank_names</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span>
<span class="va">n_samples</span> <span class="op">&lt;-</span> <span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/nsamples-methods.html">nsamples</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span>

<span class="va">node_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">ranks</span><span class="op">)</span>, FUN <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">r</span><span class="op">)</span><span class="op">{</span>
  
  <span class="va">ps</span> <span class="op">&lt;-</span> <span class="fu">microbiome</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/microbiome/man/aggregate_taxa.html">aggregate_taxa</a></span><span class="op">(</span><span class="va">ps</span>, level <span class="op">=</span> <span class="va">ranks</span><span class="op">[[</span><span class="va">r</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>
  <span class="va">out</span> <span class="op">&lt;-</span> 
    <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>
      taxon_name_unique <span class="op">=</span> <span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/taxa_names-methods.html">taxa_names</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span>,
      taxon_name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html">as.vector</a></span><span class="op">(</span><span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/tax_table-methods.html">tax_table</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span><span class="op">[</span>, <span class="va">ranks</span><span class="op">[[</span><span class="va">r</span><span class="op">]</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>,
      taxon_parent <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html">as.vector</a></span><span class="op">(</span><span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/tax_table-methods.html">tax_table</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span><span class="op">[</span>, <span class="va">ranks</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="va">r</span> <span class="op">-</span> <span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span><span class="op">]</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>,
      taxon_count <span class="op">=</span> <span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/taxa_sums.html">taxa_sums</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span>
    <span class="op">)</span>
  <span class="va">out</span><span class="op">$</span><span class="va">taxon_mean</span> <span class="op">&lt;-</span> <span class="va">out</span><span class="op">$</span><span class="va">taxon_count</span><span class="op">/</span><span class="va">n_samples</span>
  <span class="va">out</span>
<span class="op">}</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">node_info</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">ranks</span>
<span class="va">node_df</span> <span class="op">&lt;-</span> <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/reduce.html">reduce</a></span><span class="op">(</span><span class="va">node_info</span>, <span class="va">rbind.data.frame</span><span class="op">)</span>

<span class="co"># create edges df from node df</span>
<span class="va">edge_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span>
  <span class="va">node_df</span><span class="op">$</span><span class="va">taxon_name_unique</span>, 
  <span class="kw">function</span><span class="op">(</span><span class="va">unique_name</span><span class="op">)</span><span class="op">{</span>
    <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>
      from <span class="op">=</span> <span class="va">node_df</span><span class="op">[</span><span class="va">node_df</span><span class="op">$</span><span class="va">taxon_name_unique</span> <span class="op">==</span> <span class="va">unique_name</span>, <span class="st">"taxon_parent"</span><span class="op">]</span>,
      to <span class="op">=</span> <span class="va">unique_name</span>
    <span class="op">)</span>  
  <span class="op">}</span><span class="op">)</span>
<span class="va">edge_df</span> <span class="op">&lt;-</span> <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/reduce.html">reduce</a></span><span class="op">(</span><span class="va">edge_list</span>, <span class="va">rbind.data.frame</span><span class="op">)</span>
<span class="va">edge_df</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html">distinct</a></span><span class="op">(</span><span class="va">edge_df</span>, <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/across.html">across</a></span><span class="op">(</span><span class="fu">tidyselect</span><span class="fu">::</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/everything.html">everything</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="va">edge_df</span> <span class="op">&lt;-</span> <span class="va">edge_df</span><span class="op">[</span><span class="va">edge_df</span><span class="op">$</span><span class="va">from</span> <span class="op">!=</span> <span class="va">edge_df</span><span class="op">$</span><span class="va">to</span>,<span class="op">]</span>
<span class="co"># edge_df gets all attributes from the "to" node </span>
<span class="va">edge_df</span> <span class="op">&lt;-</span> <span class="va">edge_df</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span><span class="va">node_df</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'to'</span> <span class="op">=</span> <span class="st">'taxon_name_unique'</span><span class="op">)</span><span class="op">)</span>

<span class="co"># make a graph</span>
<span class="va">graph</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tidygraph.data-imaginist.com/reference/tbl_graph.html">tbl_graph</a></span><span class="op">(</span>nodes <span class="op">=</span> <span class="va">node_df</span>, node_key <span class="op">=</span> <span class="st">'taxon_name_unique'</span>, edges <span class="op">=</span> <span class="va">edge_df</span>, directed <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">layout</span> <span class="op">&lt;-</span> <span class="fu">ggraph</span><span class="fu">::</span><span class="fu"><a href="https://ggraph.data-imaginist.com/reference/ggraph.html">create_layout</a></span><span class="op">(</span><span class="va">graph</span>, layout <span class="op">=</span> <span class="st">'tree'</span>, circular <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></pre></div>
</div>
<div id="plotting-the-circular-tree-layout" class="section level2">
<h2 class="hasAnchor">
<a href="#plotting-the-circular-tree-layout" class="anchor"></a>Plotting the circular tree layout</h2>
<p>Actually plotting the graph will probably be left to the user</p>
<div class="sourceCode" id="cb5"><pre class="downlit">
<span class="co"># plot a graph</span>
<span class="fu"><a href="https://ggraph.data-imaginist.com/reference/ggraph.html">ggraph</a></span><span class="op">(</span><span class="va">layout</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/theme_graph.html">theme_graph</a></span><span class="op">(</span>base_family <span class="op">=</span> <span class="st">'sans'</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/geom_edge_link.html">geom_edge_link</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>edge_width <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">sqrt</a></span><span class="op">(</span><span class="va">taxon_mean</span><span class="op">)</span>,
                     edge_colour <span class="op">=</span> <span class="va">taxon_mean</span><span class="op">)</span>,
                 show.legend <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/geom_node_point.html">geom_node_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">sqrt</a></span><span class="op">(</span><span class="va">taxon_mean</span><span class="op">)</span>,
                      fill <span class="op">=</span> <span class="va">taxon_mean</span><span class="op">)</span>, 
                  shape <span class="op">=</span> <span class="fl">21</span>,
                  show.legend <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/scale_edge_width.html">scale_edge_width_continuous</a></span><span class="op">(</span>range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/scale_edge_colour.html">scale_edge_colour_viridis</a></span><span class="op">(</span>direction <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>, end <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://rdrr.io/pkg/viridis/man/scale_viridis.html">scale_fill_viridis</a></span><span class="op">(</span>direction <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>, end <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span></pre></div>
<p><img src="tree-development_files/figure-html/unnamed-chunk-5-1.png" width="384"></p>
</div>
</div>
<div id="adjusting-the-layout" class="section level1">
<h1 class="hasAnchor">
<a href="#adjusting-the-layout" class="anchor"></a>Adjusting the layout</h1>
<p>Here I use the davidson-harel algorithm, with trial-and-error weights, to adjust the tree layout above, using a trial-and-error number of iterations.</p>
<div class="sourceCode" id="cb6"><pre class="downlit">
<span class="co"># adjust layout of graph with davidson-harel </span>
<span class="co"># (allows weighting crossover avoidance and edge lengths etc)</span>
<span class="va">adjust_layout</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">g</span>, <span class="va">l</span><span class="op">)</span> <span class="op">{</span>
  <span class="co">#</span>
  <span class="va">adjusted_coords</span> <span class="op">&lt;-</span>
    <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/layout_with_dh.html">layout_with_dh</a></span><span class="op">(</span>
      graph <span class="op">=</span> <span class="va">g</span>,
      coords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">l</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"x"</span>, <span class="st">"y"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>,
      maxiter <span class="op">=</span> <span class="fl">3</span>, fineiter <span class="op">=</span> <span class="fl">10</span>,
      weight.edge.lengths <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/igraph/man/edge_density.html">edge_density</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span> <span class="op">*</span> <span class="fl">100</span>,
      weight.edge.crossings <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/igraph/man/edge_density.html">edge_density</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span> <span class="op">*</span> <span class="fl">100000</span>,
      weight.node.edge.dist <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/igraph/man/edge_density.html">edge_density</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span> <span class="op">*</span> <span class="fl">10</span>,
      weight.node.dist <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/igraph/man/edge_density.html">edge_density</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span> <span class="op">*</span> <span class="fl">1</span>,
      weight.border <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/igraph/man/edge_density.html">edge_density</a></span><span class="op">(</span><span class="va">graph</span><span class="op">)</span> <span class="op">*</span> <span class="fl">1</span>
    <span class="op">)</span>
  
  <span class="va">adjusted_layout</span> <span class="op">&lt;-</span> <span class="va">layout</span>
  <span class="va">adjusted_layout</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"x"</span>, <span class="st">"y"</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">adjusted_coords</span>
  
  <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="va">adjusted_layout</span><span class="op">)</span>
<span class="op">}</span></pre></div>
<p>This is a stochastic process which I do not let find an optimum. So <code>set.seed</code> is important here.</p>
<p>I will expose the weights of the dh algo (with my defaults).<br>
I will document how to rerun a few times with RNG seed setting to help users find a layout they like and re-use it (as below).</p>
<div class="sourceCode" id="cb7"><pre class="downlit">
<span class="va">layouts_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">9</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span><span class="op">{</span>
  <span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="va">i</span><span class="op">)</span>
  <span class="fu">adjust_layout</span><span class="op">(</span>g <span class="op">=</span> <span class="va">graph</span>, l <span class="op">=</span> <span class="va">layout</span><span class="op">)</span>
<span class="op">}</span><span class="op">)</span></pre></div>
<div id="plotting-the-adjusted-layouts" class="section level2">
<h2 class="hasAnchor">
<a href="#plotting-the-adjusted-layouts" class="anchor"></a>Plotting the adjusted layouts</h2>
<p>These differ only randomly, some are clearly more aesthetically pleasing than others. I like the top right one.</p>
<div class="sourceCode" id="cb8"><pre class="downlit">
<span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">layouts_list</span>, <span class="kw">function</span><span class="op">(</span><span class="va">al</span><span class="op">)</span><span class="op">{</span>
  <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/ggraph.html">ggraph</a></span><span class="op">(</span>graph <span class="op">=</span> <span class="va">al</span><span class="op">)</span> <span class="op">+</span> 
    <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/theme_graph.html">theme_graph</a></span><span class="op">(</span>base_family <span class="op">=</span> <span class="st">'sans'</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/geom_edge_link.html">geom_edge_link</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>edge_width <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">sqrt</a></span><span class="op">(</span><span class="va">taxon_mean</span><span class="op">)</span>,
                       edge_colour <span class="op">=</span> <span class="va">taxon_mean</span><span class="op">)</span>,
                   show.legend <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/geom_node_point.html">geom_node_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">sqrt</a></span><span class="op">(</span><span class="va">taxon_mean</span><span class="op">)</span>,
                        fill <span class="op">=</span> <span class="va">taxon_mean</span><span class="op">)</span>, 
                    shape <span class="op">=</span> <span class="fl">21</span>,
                    show.legend <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span>
    <span class="co"># geom_node_label(</span>
    <span class="co"># aes(label = taxon_name_unique), position = 'identity',</span>
    <span class="co"># size = 1</span>
    <span class="co"># ) +</span>
    <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/scale_edge_width.html">scale_edge_width_continuous</a></span><span class="op">(</span>range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/scale_edge_colour.html">scale_edge_colour_viridis</a></span><span class="op">(</span>direction <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>, end <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://rdrr.io/pkg/viridis/man/scale_viridis.html">scale_fill_viridis</a></span><span class="op">(</span>direction <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>, end <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span> <span class="op">+</span>
    <span class="cn">NULL</span>
<span class="op">}</span><span class="op">)</span>
  
<span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html">wrap_plots</a></span><span class="op">(</span><span class="va">plots</span>, ncol <span class="op">=</span> <span class="fl">3</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></pre></div>
<p><img src="tree-development_files/figure-html/unnamed-chunk-8-1.png" width="1152"></p>
</div>
<div id="selecting-1-of-the-adjusted-layouts" class="section level2">
<h2 class="hasAnchor">
<a href="#selecting-1-of-the-adjusted-layouts" class="anchor"></a>Selecting 1 of the adjusted layouts</h2>
<p>Besides just selecting the layout number 3 with <code>layouts_list[[3]]</code>, the whole generation can be rerun by setting the seed as equal to 3 again, as demonstrated below. You could use such a larger grey version with labels as a key.</p>
<div class="sourceCode" id="cb9"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span>
<span class="va">al</span> <span class="op">&lt;-</span> <span class="fu">adjust_layout</span><span class="op">(</span><span class="va">graph</span>, l <span class="op">=</span> <span class="va">layout</span><span class="op">)</span>

<span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/ggraph.html">ggraph</a></span><span class="op">(</span>graph <span class="op">=</span> <span class="va">al</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/theme_graph.html">theme_graph</a></span><span class="op">(</span>base_family <span class="op">=</span> <span class="st">'sans'</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/geom_edge_link.html">geom_edge_link</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>edge_width <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">sqrt</a></span><span class="op">(</span><span class="va">taxon_mean</span><span class="op">)</span><span class="op">)</span>,
                 edge_colour <span class="op">=</span> <span class="st">"grey"</span>,
                 alpha <span class="op">=</span> <span class="fl">0.5</span>,
                 show.legend <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/geom_node_point.html">geom_node_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">sqrt</a></span><span class="op">(</span><span class="va">taxon_mean</span><span class="op">)</span><span class="op">)</span>,
                  fill <span class="op">=</span> <span class="st">"grey"</span>, 
                  shape <span class="op">=</span> <span class="fl">21</span>,
                  show.legend <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/geom_node_text.html">geom_node_label</a></span><span class="op">(</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html">if_else</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span><span class="op">(</span><span class="st">'^F'</span>, <span class="va">taxon_name_unique</span><span class="op">)</span>, <span class="va">taxon_name_unique</span>, <span class="cn">NA_character_</span><span class="op">)</span>,
        size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">taxon_mean</span><span class="op">)</span><span class="op">)</span>,
    repel <span class="op">=</span> <span class="cn">TRUE</span>,
    show.legend <span class="op">=</span> <span class="cn">FALSE</span>
  <span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_size.html">scale_size_continuous</a></span><span class="op">(</span>range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">9</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/scale_edge_width.html">scale_edge_width_continuous</a></span><span class="op">(</span>range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">6</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="cn">NULL</span>
<span class="va">p</span>
<span class="co">#&gt; Warning: Removed 82 rows containing missing values (geom_label_repel).</span></pre></div>
<p><img src="tree-development_files/figure-html/unnamed-chunk-9-1.png" width="864"></p>
<div class="sourceCode" id="cb10"><pre class="downlit">
<span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/sessioninfo/man/session_info.html">session_info</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; - Session info ---------------------------------------------------------------</span>
<span class="co">#&gt;  setting  value                       </span>
<span class="co">#&gt;  version  R version 4.0.1 (2020-06-06)</span>
<span class="co">#&gt;  os       Windows 10 x64              </span>
<span class="co">#&gt;  system   x86_64, mingw32             </span>
<span class="co">#&gt;  ui       RTerm                       </span>
<span class="co">#&gt;  language (EN)                        </span>
<span class="co">#&gt;  collate  English_United Kingdom.1252 </span>
<span class="co">#&gt;  ctype    English_United Kingdom.1252 </span>
<span class="co">#&gt;  tz       Europe/Berlin               </span>
<span class="co">#&gt;  date     2020-11-03                  </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; - Packages -------------------------------------------------------------------</span>
<span class="co">#&gt;  package      * version    date       lib source        </span>
<span class="co">#&gt;  ade4           1.7-15     2020-02-13 [3] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  ape            5.4-1      2020-08-13 [3] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  assertthat     0.2.1      2019-03-21 [3] CRAN (R 4.0.0)</span>
<span class="co">#&gt;  backports      1.1.8      2020-06-17 [3] CRAN (R 4.0.1)</span>
<span class="co">#&gt;  Biobase        2.48.0     2020-04-27 [3] Bioconductor  </span>
<span class="co">#&gt;  BiocGenerics   0.34.0     2020-04-27 [3] Bioconductor  </span>
<span class="co">#&gt;  biomformat     1.16.0     2020-04-27 [3] Bioconductor  </span>
<span class="co">#&gt;  Biostrings     2.56.0     2020-04-27 [3] Bioconductor  </span>
<span class="co">#&gt;  callr          3.5.1      2020-10-13 [3] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  cli            2.1.0      2020-10-12 [3] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  cluster        2.1.0      2019-06-19 [3] CRAN (R 4.0.1)</span>
<span class="co">#&gt;  codetools      0.2-16     2018-12-24 [3] CRAN (R 4.0.1)</span>
<span class="co">#&gt;  colorspace     1.4-1      2019-03-18 [3] CRAN (R 4.0.0)</span>
<span class="co">#&gt;  crayon         1.3.4      2017-09-16 [3] CRAN (R 4.0.0)</span>
<span class="co">#&gt;  data.table     1.13.2     2020-10-19 [3] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  desc           1.2.0      2018-05-01 [3] CRAN (R 4.0.0)</span>
<span class="co">#&gt;  devtools       2.3.2      2020-09-18 [3] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  digest         0.6.25     2020-02-23 [3] CRAN (R 4.0.0)</span>
<span class="co">#&gt;  dplyr        * 1.0.2      2020-08-18 [3] CRAN (R 4.0.2)</span>
<span class="co">#&gt;  ellipsis       0.3.1      2020-05-15 [3] CRAN (R 4.0.0)</span>
<span class="co">#&gt;  evaluate       0.14       2019-05-28 [3] CRAN (R 4.0.0)</span>
<span class="co">#&gt;  fansi          0.4.1      2020-01-08 [3] CRAN (R 4.0.0)</span>
<span class="co">#&gt;  farver         2.0.3      2020-01-16 [3] CRAN (R 4.0.0)</span>
<span class="co">#&gt;  foreach        1.5.1      2020-10-15 [3] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  fs             1.5.0      2020-07-31 [3] CRAN (R 4.0.2)</span>
<span class="co">#&gt;  generics       0.0.2      2018-11-29 [3] CRAN (R 4.0.0)</span>
<span class="co">#&gt;  ggforce        0.3.2      2020-06-23 [3] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  ggplot2      * 3.3.2      2020-06-19 [3] CRAN (R 4.0.2)</span>
<span class="co">#&gt;  ggraph       * 2.0.3      2020-05-20 [3] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  ggrepel        0.8.2      2020-03-08 [3] CRAN (R 4.0.2)</span>
<span class="co">#&gt;  glue           1.4.2      2020-08-27 [3] CRAN (R 4.0.2)</span>
<span class="co">#&gt;  graphlayouts   0.7.1      2020-10-26 [3] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  gridExtra      2.3        2017-09-09 [3] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  gtable         0.3.0      2019-03-25 [3] CRAN (R 4.0.0)</span>
<span class="co">#&gt;  htmltools      0.5.0      2020-06-16 [3] CRAN (R 4.0.2)</span>
<span class="co">#&gt;  igraph       * 1.2.6      2020-10-06 [3] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  IRanges        2.22.2     2020-05-21 [3] Bioconductor  </span>
<span class="co">#&gt;  iterators      1.0.13     2020-10-15 [3] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  jsonlite       1.7.1      2020-09-07 [3] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  knitr          1.30       2020-09-22 [3] CRAN (R 4.0.2)</span>
<span class="co">#&gt;  labeling       0.4.2      2020-10-20 [3] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  lattice        0.20-41    2020-04-02 [3] CRAN (R 4.0.1)</span>
<span class="co">#&gt;  lifecycle      0.2.0      2020-03-06 [3] CRAN (R 4.0.0)</span>
<span class="co">#&gt;  magrittr       1.5        2014-11-22 [3] CRAN (R 4.0.0)</span>
<span class="co">#&gt;  MASS           7.3-53     2020-09-09 [3] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  Matrix         1.2-18     2019-11-27 [3] CRAN (R 4.0.1)</span>
<span class="co">#&gt;  memoise        1.1.0      2017-04-21 [3] CRAN (R 4.0.0)</span>
<span class="co">#&gt;  mgcv           1.8-33     2020-08-27 [3] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  microbiome   * 1.10.0     2020-04-27 [3] Bioconductor  </span>
<span class="co">#&gt;  microViz     * 0.0.0.9000 2020-11-03 [2] local         </span>
<span class="co">#&gt;  multtest       2.44.0     2020-04-27 [3] Bioconductor  </span>
<span class="co">#&gt;  munsell        0.5.0      2018-06-12 [3] CRAN (R 4.0.0)</span>
<span class="co">#&gt;  nlme           3.1-150    2020-10-24 [3] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  patchwork    * 1.0.1      2020-06-22 [3] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  permute        0.9-5      2019-03-12 [3] CRAN (R 4.0.0)</span>
<span class="co">#&gt;  phyloseq     * 1.32.0     2020-04-27 [3] Bioconductor  </span>
<span class="co">#&gt;  pillar         1.4.6      2020-07-10 [3] CRAN (R 4.0.2)</span>
<span class="co">#&gt;  pkgbuild       1.1.0      2020-07-13 [3] CRAN (R 4.0.2)</span>
<span class="co">#&gt;  pkgconfig      2.0.3      2019-09-22 [3] CRAN (R 4.0.0)</span>
<span class="co">#&gt;  pkgdown        1.6.1      2020-09-12 [3] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  pkgload        1.1.0      2020-05-29 [3] CRAN (R 4.0.0)</span>
<span class="co">#&gt;  plyr           1.8.6      2020-03-03 [3] CRAN (R 4.0.0)</span>
<span class="co">#&gt;  polyclip       1.10-0     2019-03-14 [3] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  prettyunits    1.1.1      2020-01-24 [3] CRAN (R 4.0.0)</span>
<span class="co">#&gt;  processx       3.4.4      2020-09-03 [3] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  ps             1.3.3      2020-05-08 [3] CRAN (R 4.0.0)</span>
<span class="co">#&gt;  purrr          0.3.4      2020-04-17 [3] CRAN (R 4.0.0)</span>
<span class="co">#&gt;  R6             2.5.0      2020-10-28 [3] CRAN (R 4.0.1)</span>
<span class="co">#&gt;  ragg           0.4.0      2020-10-05 [3] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  Rcpp           1.0.5      2020-07-06 [3] CRAN (R 4.0.2)</span>
<span class="co">#&gt;  remotes        2.2.0      2020-07-21 [3] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  reshape2       1.4.4      2020-04-09 [3] CRAN (R 4.0.0)</span>
<span class="co">#&gt;  rhdf5          2.32.4     2020-10-05 [3] Bioconductor  </span>
<span class="co">#&gt;  Rhdf5lib       1.10.1     2020-07-09 [3] Bioconductor  </span>
<span class="co">#&gt;  rlang          0.4.7      2020-07-09 [3] CRAN (R 4.0.2)</span>
<span class="co">#&gt;  rmarkdown      2.5        2020-10-21 [3] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  rprojroot      1.3-2      2018-01-03 [3] CRAN (R 4.0.0)</span>
<span class="co">#&gt;  rstudioapi     0.11       2020-02-07 [3] CRAN (R 4.0.0)</span>
<span class="co">#&gt;  Rtsne          0.15       2018-11-10 [3] CRAN (R 4.0.2)</span>
<span class="co">#&gt;  S4Vectors      0.26.1     2020-05-16 [3] Bioconductor  </span>
<span class="co">#&gt;  scales         1.1.1      2020-05-11 [3] CRAN (R 4.0.0)</span>
<span class="co">#&gt;  sessioninfo    1.1.1      2018-11-05 [3] CRAN (R 4.0.0)</span>
<span class="co">#&gt;  stringi        1.5.3      2020-09-09 [3] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  stringr        1.4.0      2019-02-10 [3] CRAN (R 4.0.0)</span>
<span class="co">#&gt;  survival       3.2-7      2020-09-28 [3] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  systemfonts    0.3.2      2020-09-29 [3] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  testthat       2.3.2      2020-03-02 [3] CRAN (R 4.0.0)</span>
<span class="co">#&gt;  textshaping    0.1.2      2020-10-08 [3] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  tibble         3.0.4      2020-10-12 [3] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  tidygraph    * 1.2.0      2020-05-12 [3] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  tidyr          1.1.2      2020-08-27 [3] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  tidyselect     1.1.0      2020-05-11 [3] CRAN (R 4.0.0)</span>
<span class="co">#&gt;  tweenr         1.0.1      2018-12-14 [3] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  usethis        1.6.3      2020-09-17 [3] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  vctrs          0.3.4      2020-08-29 [3] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  vegan          2.5-6      2019-09-01 [3] CRAN (R 4.0.0)</span>
<span class="co">#&gt;  viridis        0.5.1      2018-03-29 [3] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  viridisLite    0.3.0      2018-02-01 [3] CRAN (R 4.0.0)</span>
<span class="co">#&gt;  withr          2.3.0      2020-09-22 [3] CRAN (R 4.0.2)</span>
<span class="co">#&gt;  xfun           0.18       2020-09-29 [3] CRAN (R 4.0.3)</span>
<span class="co">#&gt;  XVector        0.28.0     2020-04-27 [3] Bioconductor  </span>
<span class="co">#&gt;  yaml           2.2.1      2020-02-01 [3] CRAN (R 4.0.0)</span>
<span class="co">#&gt;  zlibbioc       1.34.0     2020-04-27 [3] Bioconductor  </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [1] C:/Users/david.barnett/AppData/Local/Temp/RtmpGAyKgk/temp_libpath140c24dd647</span>
<span class="co">#&gt; [2] C:/Users/david.barnett/AppData/Local/Temp/RtmpGAyKgk/temp_libpath140c65992ed6</span>
<span class="co">#&gt; [3] C:/Program Files/R/R-4.0.1/library</span></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by David Barnett.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
