<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Ordination plots • microViz</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><!-- docsearch --><script src="../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><meta property="og:title" content="Ordination plots">
<meta property="og:description" content="microViz">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">microViz</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.7.4.9001</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/articles/microViz.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/articles/Interactive-Ordination.html">Interactive Ordination Plot</a>
    </li>
    <li>
      <a href="../articles/articles/Visualising-compositions.html">Visualising compositions</a>
    </li>
    <li>
      <a href="../articles/articles/phyloseq.html">Working with phyloseq objects</a>
    </li>
    <li>
      <a href="../articles/articles/tax-fixing.html">Fixing your tax_table</a>
    </li>
    <li>
      <a href="../articles/ordination.html">Ordination plots</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/david-barnett/microViz/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="ordination_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Ordination plots</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/david-barnett/microViz/blob/master/vignettes/ordination.Rmd"><code>vignettes/ordination.Rmd</code></a></small>
      <div class="hidden name"><code>ordination.Rmd</code></div>

    </div>

    
    
<div id="intro" class="section level2">
<h2 class="hasAnchor">
<a href="#intro" class="anchor"></a>Intro</h2>
<p>This article will show you how to create and customise ordination plots, like <a href="#pca">PCA</a> and <a href="#rda">RDA</a>, with microViz.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://dx.plos.org/10.1371/journal.pone.0061217">phyloseq</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://david-barnett.github.io/microViz">microViz</a></span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; * microViz version 0.7.4.9001 - Copyright (C) 2021 David Barnett</span>
<span class="co">#&gt; * Website: https://david-barnett.github.io/microViz/</span>
<span class="co">#&gt; * Useful? Consider citing: https://zenodo.org/record/4644057</span>
<span class="co">#&gt; * Silence: suppressPackageStartupMessages(library(microViz))</span>
<span class="fu">knitr</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/knitr/man/opts_chunk.html">opts_chunk</a></span><span class="op">$</span><span class="fu">set</span><span class="op">(</span>fig.width <span class="op">=</span> <span class="fl">7</span>, fig.height <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></code></pre></div>
<p>We will use example data from stool samples from an inflammatory bowel disease (IBD) study, borrowed from the great <code>corncob</code> package. See the article about <a href="https://david-barnett.github.io/microViz/articles/articles/phyloseq.html">working with phyloseq objects</a> if you want to get started with your own data, or just to learn more about manipulating phyloseq objects with microViz.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ibd</span> <span class="op">&lt;-</span> <span class="fu">corncob</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/corncob/man/ibd_phylo.html">ibd_phylo</a></span>
<span class="va">ibd</span>
<span class="co">#&gt; phyloseq-class experiment-level object</span>
<span class="co">#&gt; otu_table()   OTU Table:         [ 36349 taxa and 91 samples ]</span>
<span class="co">#&gt; sample_data() Sample Data:       [ 91 samples by 15 sample variables ]</span>
<span class="co">#&gt; tax_table()   Taxonomy Table:    [ 36349 taxa by 7 taxonomic ranks ]</span></code></pre></div>
<p>First we check the phyloseq object and fix any uninformative tax_table entries (see the article on <a href="https://david-barnett.github.io/microViz/articles/articles/tax-fixing.html">fixing your tax_table</a> for more info).</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ibd</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/phyloseq_validate.html">phyloseq_validate</a></span><span class="op">(</span><span class="va">ibd</span>, remove_undetected <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">#&gt; Short values detected in phyloseq tax_table (nchar&lt;4) :</span>
<span class="co">#&gt; Consider using tax_fix() to make taxa uniquely identifiable</span>
<span class="va">ibd</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tax_fix.html">tax_fix</a></span><span class="op">(</span><span class="va">ibd</span><span class="op">)</span> <span class="co"># try tax_fix_interactive if you have problems with your own data</span></code></pre></div>
</div>
<div id="motivation" class="section level2">
<h2 class="hasAnchor">
<a href="#motivation" class="anchor"></a>Motivation</h2>
<p>Ordination plots are a great way to see any clustering or other patterns of microbiota (dis)similarity in (many) samples. Ordinations like PCA or PCoA show the largest patterns of variation in your data, and constrained ordination techniques like RDA or CCA can show you microbial variation that could be explained by other variables in your sample_data (but interpret constrained ordinations with care, and ideally test for the statistical significance of any hypothesised associations using a method like PERMANOVA, see <code><a href="../reference/dist_permanova.html">dist_permanova()</a></code>).</p>
<p>Check out the <a href="https://sites.google.com/site/mb3gustame/">Gusta Me website</a> for a gentle theoretical introduction to <a href="https://sites.google.com/site/mb3gustame/indirect-gradient-analysis/pca">PCA</a>, <a href="https://sites.google.com/site/mb3gustame/dissimilarity-based-methods/principal-coordinates-analysis">PCoA</a>, <a href="https://sites.google.com/site/mb3gustame/constrained-analyses/rda">RDA</a>, <a href="https://sites.google.com/site/mb3gustame/constrained-analyses/cca">CCA</a> and more.</p>
<p>Ordination plots can also be paired with barplots for greater insight into microbial compositions, e.g. see <code><a href="../reference/ord_plot_iris.html">ord_plot_iris()</a></code> and the <code><a href="../reference/ord_explore.html">ord_explore()</a></code> interactive <a href="https://david-barnett.github.io/microViz/articles/articles/Interactive-Ordination.html">Shiny app</a>.</p>
</div>
<div id="prepare-your-microbes" class="section level2">
<h2 class="hasAnchor">
<a href="#prepare-your-microbes" class="anchor"></a>Prepare your microbes</h2>
<p>When creating an ordination plot, you first need to prepare the microbiota variables.</p>
<ul>
<li><p>Decide at which taxonomic rank to aggregate your data, e.g. “Genus”</p></li>
<li><p>Consider transforming the microbial counts, e.g. using the “clr” (centred log ratio) transformation, which is often recommended for compositional data (<a href="https://doi.org/10.3389/fmicb.2017.02224" title="Gloor 2017">like sequencing data</a>)</p></li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ibd</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="../reference/tax_transform.html">tax_transform</a></span><span class="op">(</span>transformation <span class="op">=</span> <span class="st">"clr"</span>, rank <span class="op">=</span> <span class="st">"Genus"</span><span class="op">)</span>
<span class="co">#&gt; ps_extra object - a list with phyloseq and extras:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; phyloseq-class experiment-level object</span>
<span class="co">#&gt; otu_table()   OTU Table:         [ 178 taxa and 91 samples ]</span>
<span class="co">#&gt; sample_data() Sample Data:       [ 91 samples by 15 sample variables ]</span>
<span class="co">#&gt; tax_table()   Taxonomy Table:    [ 178 taxa by 6 taxonomic ranks ]</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; ps_extra info:</span>
<span class="co">#&gt; tax_agg = Genus tax_transform = clr</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $counts OTU Table: [ 178 taxa and 91 samples ]</span></code></pre></div>
<ul>
<li>Some methods, such as <a href="#pcoa">PCoA</a>, require a matrix of pairwise distances between samples, which you can easily calculate with <code><a href="../reference/dist_calc.html">dist_calc()</a></code>. Normally you should NOT transform your data when using a distance-based method, but it is useful to record an “identity” transformation anyway, to make it clear you have not transformed your data.</li>
</ul>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ibd</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="../reference/tax_transform.html">tax_transform</a></span><span class="op">(</span>transformation <span class="op">=</span> <span class="st">"identity"</span>, rank <span class="op">=</span> <span class="st">"Genus"</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="../reference/dist_calc.html">dist_calc</a></span><span class="op">(</span><span class="st">"bray"</span><span class="op">)</span> <span class="co"># bray curtis distance</span>
<span class="co">#&gt; ps_extra object - a list with phyloseq and extras:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; phyloseq-class experiment-level object</span>
<span class="co">#&gt; otu_table()   OTU Table:         [ 178 taxa and 91 samples ]</span>
<span class="co">#&gt; sample_data() Sample Data:       [ 91 samples by 15 sample variables ]</span>
<span class="co">#&gt; tax_table()   Taxonomy Table:    [ 178 taxa by 6 taxonomic ranks ]</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; ps_extra info:</span>
<span class="co">#&gt; tax_agg = Genus tax_transform = identity</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; bray distance matrix of size 91 </span>
<span class="co">#&gt; 0.8294664 0.7156324 0.5111652 0.5492537 0.8991251 ...</span></code></pre></div>
<ul>
<li>Some dissimilarity measures, such as unweighted UniFrac, do not consider the abundance of each taxon when calculating dissimilarity, and so may be (overly) sensitive to differences in rare/low-abundance taxa. So you might want to filter out very rare taxa, with <code><a href="../reference/tax_filter.html">tax_filter()</a></code> before using <code><a href="../reference/dist_calc.html">dist_calc(ps, dist = "unifrac")</a></code>. Distances that are (implicitly) abundance weighted, including Generalised UniFrac, Bray-Curtis and Aitchison distance, should be less sensitive to rare taxa / filtering threshold choices.</li>
</ul>
<div id="ps_extra" class="section level3">
<h3 class="hasAnchor">
<a href="#ps_extra" class="anchor"></a>ps_extra</h3>
<p>Notice that the objects created above are of class “ps_extra”. This is just a simple S3 list object that holds your phyloseq object and additional stuff created from this phyloseq object, such as a distance matrix, as well as info on any transformation and aggregation applied to your taxa. microViz uses this to automatically create plot captions, to help you and your collaborators remember how you made each plot! You can access the phyloseq object, distance matrix and other parts of a ps_extra object with <code><a href="../reference/ps_extra-accessors.html">ps_get()</a></code>, <code><a href="../reference/ps_extra-accessors.html">dist_get()</a></code>, and friends.</p>
</div>
</div>
<div id="pca" class="section level2">
<h2 class="hasAnchor">
<a href="#pca" class="anchor"></a>PCA</h2>
<p>Principle <strong><em>Components</em></strong> Analysis is an unconstrained method that does not use a distance matrix. PCA directly uses the (transformed) microbial variables, so you do not need <code><a href="../reference/dist_calc.html">dist_calc()</a></code>. <code>ord_calc</code> performs the ordination (adding it to the ps_extra object) and <code><a href="../reference/ord_plot.html">ord_plot()</a></code> creates the ggplot2 scatterplot (which you can customise like other ggplot objects).</p>
<p>Each point is a sample, and samples that appear closer together are typically more similar to each other than samples which are further apart. So by colouring the points by IBD status you can see that the microbiota from people with IBD is often, but not always, highly distinct from people without IBD.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ibd</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="../reference/tax_transform.html">tax_transform</a></span><span class="op">(</span><span class="st">"clr"</span>, rank <span class="op">=</span> <span class="st">"Genus"</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="co"># when no distance matrix or constraints are supplied, PCA is the default/auto ordination method</span>
  <span class="fu"><a href="../reference/ord_calc.html">ord_calc</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="../reference/ord_plot.html">ord_plot</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"ibd"</span>, shape <span class="op">=</span> <span class="st">"DiseaseState"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_colour_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Dark2"</span><span class="op">)</span></code></pre></div>
<p><img src="ordination_files/figure-html/pca-1.png" width="672"></p>
<p>One benefit of not using a distance matrix, is that you can plot taxa “loadings” onto your PCA axes, using the plot_taxa argument. microViz plots all of the taxa loading vectors in light grey, and you choose how many of the vectors to label, starting with the longest arrows (alternatively you can name particular taxa to label).</p>
<p>The relative length of each loading vector indicates its contribution to each PCA axis shown, and allows you to roughly estimate which samples will contain more of that taxon e.g. samples on the left of the plot below, will typically contain more <em>Escherichia</em>/<em>Shigella</em> than samples on the right, and this taxon contributes heavily to the PC1 axis.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ibd</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="../reference/tax_transform.html">tax_transform</a></span><span class="op">(</span><span class="st">"clr"</span>, rank <span class="op">=</span> <span class="st">"Genus"</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="co"># when no distance matrix or constraints are supplied, PCA is the default/auto ordination method</span>
  <span class="fu"><a href="../reference/ord_calc.html">ord_calc</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="../reference/ord_plot.html">ord_plot</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"ibd"</span>, shape <span class="op">=</span> <span class="st">"DiseaseState"</span>, plot_taxa <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_colour_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Dark2"</span><span class="op">)</span></code></pre></div>
<p><img src="ordination_files/figure-html/pca%20loadings-1.png" width="672"></p>
<p>microViz also allows you directly visualize the sample compositions on a circular barplot or “iris plot” (named because it looks kinda like an eyeball) alongside the PCA plot. The samples on the iris plot are automatically arranged by their rotational position around the center/origin of the PCA plot.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ibd</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="../reference/tax_transform.html">tax_transform</a></span><span class="op">(</span><span class="st">"clr"</span>, rank <span class="op">=</span> <span class="st">"Genus"</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="co"># when no distance matrix or constraints are supplied, PCA is the default/auto ordination method</span>
  <span class="fu"><a href="../reference/ord_calc.html">ord_calc</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="../reference/ord_plot_iris.html">ord_plot_iris</a></span><span class="op">(</span>tax_level <span class="op">=</span> <span class="st">"Genus"</span>, ord_plot <span class="op">=</span> <span class="st">"above"</span>, anno_colour <span class="op">=</span> <span class="st">"ibd"</span><span class="op">)</span>
<span class="co">#&gt; Loading required namespace: patchwork</span></code></pre></div>
<p><img src="ordination_files/figure-html/iris-1.png" width="672"></p>
</div>
<div id="pcoa" class="section level2">
<h2 class="hasAnchor">
<a href="#pcoa" class="anchor"></a>PCoA</h2>
<p>Principle <strong><em>Co-ordinates</em></strong> Analysis is also an unconstrained method, but it does require a distance matrix. In an ecological context, a distance (or more generally a “dissimilarity”) measure indicates how different a pair of (microbial) ecosystems are. This can be calculated in many ways.</p>
<div id="aitchison-distance" class="section level3">
<h3 class="hasAnchor">
<a href="#aitchison-distance" class="anchor"></a>Aitchison distance</h3>
<p>The <a href="https://en.wikipedia.org/wiki/Euclidean_distance">Euclidean distance</a> is similar to the distance we humans are familiar with in the physical world. The results of a PCA is practically equivalent to a PCoA with Euclidean distances. The Aitchison distance is a dissimilarity measure calculated as the Euclidean distance between observations (samples) after performing a centred log ratio (“clr”) transformation. That is why the Aitchison distance PCoA, below, looks the same as the PCA we made earlier. However, we cannot use plot_taxa, as the taxa loadings are only available for PCA (and related methods like RDA).</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ibd</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="../reference/tax_transform.html">tax_transform</a></span><span class="op">(</span><span class="st">"identity"</span>, rank <span class="op">=</span> <span class="st">"Genus"</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="co"># don't transform!</span>
  <span class="fu"><a href="../reference/dist_calc.html">dist_calc</a></span><span class="op">(</span><span class="st">"aitchison"</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="../reference/ord_calc.html">ord_calc</a></span><span class="op">(</span><span class="st">"PCoA"</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="../reference/ord_plot.html">ord_plot</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"ibd"</span>, shape <span class="op">=</span> <span class="st">"DiseaseState"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_colour_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Dark2"</span><span class="op">)</span></code></pre></div>
<p><img src="ordination_files/figure-html/aitchison%20pcoa-1.png" width="672"></p>
<p>Note that PCoA is also known as MDS, for (metric) Multi-Dimensional Scaling, hence the axes names.</p>
</div>
<div id="ecological-dissimilarities" class="section level3">
<h3 class="hasAnchor">
<a href="#ecological-dissimilarities" class="anchor"></a>Ecological dissimilarities</h3>
<p>Over the years, ecologists have invented numerous ways of quantifying dissimilarity between pairs of ecosystems. One ubiquitous example is the <a href="https://en.wikipedia.org/wiki/Bray%E2%80%93Curtis_dissimilarity">Bray-Curtis</a> dissimilarity measure, shown below.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ibd</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="../reference/tax_transform.html">tax_transform</a></span><span class="op">(</span><span class="st">"identity"</span>, rank <span class="op">=</span> <span class="st">"Genus"</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="co"># don't transform!</span>
  <span class="fu"><a href="../reference/dist_calc.html">dist_calc</a></span><span class="op">(</span><span class="st">"bray"</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="../reference/ord_calc.html">ord_calc</a></span><span class="op">(</span><span class="st">"PCoA"</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="../reference/ord_plot.html">ord_plot</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"ibd"</span>, shape <span class="op">=</span> <span class="st">"DiseaseState"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_colour_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Dark2"</span><span class="op">)</span></code></pre></div>
<p><img src="ordination_files/figure-html/bray-1.png" width="672"></p>
<p>Beyond Bray-Curtis, microViz <code><a href="../reference/dist_calc.html">dist_calc()</a></code> can also help you calculate all the other ecological distances listed in <code><a href="https://rdrr.io/pkg/phyloseq/man/distanceMethodList.html">phyloseq::distanceMethodList</a></code> such the Jensen-Shannon Divergence, <code>"jsd"</code>, or Jaccard dissimilarity <code>"jaccard"</code>. Beware that if you want a binary dissimilarity measure from <code><a href="https://rdrr.io/pkg/vegan/man/vegdist.html">vegan::vegdist()</a></code> (i.e. only using presence/absence info, and noting all the caveats about sensitivity to low abundance taxa) you will need to pass <code>binary = TRUE</code>, as below.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ibd</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="../reference/tax_transform.html">tax_transform</a></span><span class="op">(</span><span class="st">"identity"</span>, rank <span class="op">=</span> <span class="st">"Genus"</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="../reference/dist_calc.html">dist_calc</a></span><span class="op">(</span>dist <span class="op">=</span> <span class="st">"jaccard"</span>, binary <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="../reference/ord_calc.html">ord_calc</a></span><span class="op">(</span><span class="st">"PCoA"</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="../reference/ord_plot.html">ord_plot</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"ibd"</span>, shape <span class="op">=</span> <span class="st">"DiseaseState"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_colour_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Dark2"</span><span class="op">)</span></code></pre></div>
<p><img src="ordination_files/figure-html/binary%20jaccard-1.png" width="672"></p>
</div>
<div id="unifrac-distances" class="section level3">
<h3 class="hasAnchor">
<a href="#unifrac-distances" class="anchor"></a>UniFrac distances</h3>
<p>If you have a phylogenetic tree available, and <a href="https://david-barnett.github.io/microViz/articles/articles/phyloseq.html#getting-your-data-into-phyloseq">attached to your phyloseq object</a>. You can calculate dissimilarities from the <a href="https://en.wikipedia.org/wiki/UniFrac">UniFrac family</a> of methods, which take into account the phylogenetic relatedness of the taxa / sequences in your samples when calculating dissimilarity. Un-weighted UniFrac, <code><a href="../reference/dist_calc.html">dist_calc(dist = "unifrac")</a></code>, does not consider the relative abundance of taxa, only their presence (detection) or absence, which can make it (overly) sensitive to rare taxa, sequencing artefacts, and abundance filtering choices. Conversely, weighted UniFrac, <code>"wunifrac"</code>, does put (perhaps too much) more importance on highly abundant taxa, when determining dissimilarities. The Generalised UniFrac, <code>"gunifrac"</code>, finds a balance between these two extremes, and by adjusting the <code>gunifrac_alpha</code> argument of <code><a href="../reference/dist_calc.html">dist_calc()</a></code>, you can tune this balance to your liking (although the 0.5 default should be fine!).</p>
<p>Below is a Generalised UniFrac example using a different, and tiny, example dataset from the phyloseq package that has a phylogenetic tree.</p>
<p>You should <strong>not</strong> aggregate taxa before using a phylogenetic distance measure, but you can and probably should register the choice not to transform or aggregate, as below.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"esophagus"</span>, package <span class="op">=</span> <span class="st">"phyloseq"</span><span class="op">)</span>
<span class="va">esophagus</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../reference/phyloseq_validate.html">phyloseq_validate</a></span><span class="op">(</span>verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="../reference/tax_transform.html">tax_transform</a></span><span class="op">(</span><span class="st">"identity"</span>, rank <span class="op">=</span> <span class="st">"unique"</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="../reference/dist_calc.html">dist_calc</a></span><span class="op">(</span><span class="st">"gunifrac"</span>, gunifrac_alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span>
<span class="co">#&gt; ps_extra object - a list with phyloseq and extras:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; phyloseq-class experiment-level object</span>
<span class="co">#&gt; otu_table()   OTU Table:         [ 58 taxa and 3 samples ]</span>
<span class="co">#&gt; sample_data() Sample Data:       [ 3 samples by 1 sample variables ]</span>
<span class="co">#&gt; tax_table()   Taxonomy Table:    [ 58 taxa by 1 taxonomic ranks ]</span>
<span class="co">#&gt; phy_tree()    Phylogenetic Tree: [ 58 tips and 57 internal nodes ]</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; ps_extra info:</span>
<span class="co">#&gt; tax_agg = unique tax_transform = identity</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; gunifrac_0.5 distance matrix of size 3 </span>
<span class="co">#&gt; 0.4404284 0.4332325 0.4969773 ...</span></code></pre></div>
</div>
</div>
<div id="further-dimensions" class="section level2">
<h2 class="hasAnchor">
<a href="#further-dimensions" class="anchor"></a>Further dimensions</h2>
<p>You can show other dimensions / axes of an ordination than just the first two, by setting the axes argument. You can judge from the variation explained by each successive axis (on a scree plot) whether this is worthwhile information to show, e.g. in the example below, it could be interesting to also show the 3rd axis, but not any others.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ibd</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="../reference/tax_transform.html">tax_transform</a></span><span class="op">(</span><span class="st">"identity"</span>, rank <span class="op">=</span> <span class="st">"Genus"</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="co"># don't transform!</span>
  <span class="fu"><a href="../reference/dist_calc.html">dist_calc</a></span><span class="op">(</span><span class="st">"bray"</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="../reference/ord_calc.html">ord_calc</a></span><span class="op">(</span><span class="st">"PCoA"</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="../reference/ps_extra-accessors.html">ord_get</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/plot_scree.html">plot_scree</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">6</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="ordination_files/figure-html/scree-1.png" width="480"></p>
<p>Let us view the 1st and 3rd axes, and add IBD/non-IBD group ellipses.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ibd</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="../reference/tax_transform.html">tax_transform</a></span><span class="op">(</span><span class="st">"identity"</span>, rank <span class="op">=</span> <span class="st">"Genus"</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="co"># don't transform!</span>
  <span class="fu"><a href="../reference/dist_calc.html">dist_calc</a></span><span class="op">(</span><span class="st">"bray"</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="../reference/ord_calc.html">ord_calc</a></span><span class="op">(</span><span class="st">"PCoA"</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="../reference/ord_plot.html">ord_plot</a></span><span class="op">(</span>axes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span>, color <span class="op">=</span> <span class="st">"ibd"</span>, shape <span class="op">=</span> <span class="st">"DiseaseState"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_colour_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Dark2"</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/stat_ellipse.html">stat_ellipse</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">ibd</span>, linetype <span class="op">=</span> <span class="va">ibd</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="ordination_files/figure-html/bray%203-1.png" width="672"></p>
</div>
<div id="rda" class="section level2">
<h2 class="hasAnchor">
<a href="#rda" class="anchor"></a>RDA</h2>
<p>Redundancy analysis is a constrained ordination method.</p>
<p>Tutorial coming soon, for now see <code><a href="../reference/ord_plot.html">ord_plot()</a></code> for examples.</p>
<div id="partial-ordinations" class="section level3">
<h3 class="hasAnchor">
<a href="#partial-ordinations" class="anchor"></a>Partial ordinations</h3>
<p>Tutorial coming soon, for now see <code><a href="../reference/ord_plot.html">ord_plot()</a></code> for examples.</p>
</div>
</div>
<div id="customising-your-ordination-plot" class="section level2">
<h2 class="hasAnchor">
<a href="#customising-your-ordination-plot" class="anchor"></a>Customising your ordination plot</h2>
<p>It is possible to modify the aesthetic style of the taxon loading and constraint vectors of your ordination plots. e.g. making them smaller or more transparent.</p>
<p>It is also possible to rename the taxa labels using a function passed to the <code>taxon_renamer</code> argument of <code><a href="../reference/ord_plot.html">ord_plot()</a></code>.</p>
<p>Tutorial coming soon!</p>
</div>
<div id="session-info" class="section level2">
<h2 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session info</h2>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/sessioninfo/man/session_info.html">session_info</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; ─ Session info ───────────────────────────────────────────────────────────────</span>
<span class="co">#&gt;  setting  value                       </span>
<span class="co">#&gt;  version  R version 4.0.5 (2021-03-31)</span>
<span class="co">#&gt;  os       macOS Catalina 10.15.7      </span>
<span class="co">#&gt;  system   x86_64, darwin17.0          </span>
<span class="co">#&gt;  ui       X11                         </span>
<span class="co">#&gt;  language (EN)                        </span>
<span class="co">#&gt;  collate  en_US.UTF-8                 </span>
<span class="co">#&gt;  ctype    en_US.UTF-8                 </span>
<span class="co">#&gt;  tz       UTC                         </span>
<span class="co">#&gt;  date     2021-05-06                  </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; ─ Packages ───────────────────────────────────────────────────────────────────</span>
<span class="co">#&gt;  package      * version    date       lib source                               </span>
<span class="co">#&gt;  ade4           1.7-16     2020-10-28 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  ape            5.5        2021-04-25 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  Biobase        2.50.0     2020-10-27 [1] Bioconductor                         </span>
<span class="co">#&gt;  BiocGenerics   0.36.1     2021-04-16 [1] Bioconductor                         </span>
<span class="co">#&gt;  biomformat     1.18.0     2020-10-27 [1] Bioconductor                         </span>
<span class="co">#&gt;  Biostrings     2.58.0     2020-10-27 [1] Bioconductor                         </span>
<span class="co">#&gt;  bslib          0.2.4      2021-01-25 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  cachem         1.0.4      2021-02-13 [1] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  callr          3.7.0      2021-04-20 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  cli            2.5.0      2021-04-26 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  cluster        2.1.1      2021-02-14 [2] CRAN (R 4.0.5)                       </span>
<span class="co">#&gt;  codetools      0.2-18     2020-11-04 [2] CRAN (R 4.0.5)                       </span>
<span class="co">#&gt;  colorspace     2.0-1      2021-05-04 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  corncob        0.1.0      2021-02-10 [1] Github (bryandmartin/corncob@338323e)</span>
<span class="co">#&gt;  crayon         1.4.1      2021-02-08 [1] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  data.table     1.14.0     2021-02-21 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  desc           1.3.0      2021-03-05 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  devtools       2.4.0      2021-04-07 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  digest         0.6.27     2020-10-24 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  dplyr          1.0.5      2021-03-05 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  ellipsis       0.3.2      2021-04-29 [1] CRAN (R 4.0.5)                       </span>
<span class="co">#&gt;  evaluate       0.14       2019-05-28 [1] CRAN (R 4.0.1)                       </span>
<span class="co">#&gt;  fansi          0.4.2      2021-01-15 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  farver         2.1.0      2021-02-28 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  fastmap        1.1.0      2021-01-25 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  foreach        1.5.1      2020-10-15 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  fs             1.5.0      2020-07-31 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  generics       0.1.0      2020-10-31 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  ggplot2      * 3.3.3      2020-12-30 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  glue           1.4.2      2020-08-27 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  gtable         0.3.0      2019-03-25 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  GUniFrac       1.2        2021-05-03 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  highr          0.9        2021-04-16 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  hms            1.0.0      2021-01-13 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  htmltools      0.5.1.1    2021-01-22 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  igraph         1.2.6      2020-10-06 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  IRanges        2.24.1     2020-12-12 [1] Bioconductor                         </span>
<span class="co">#&gt;  iterators      1.0.13     2020-10-15 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  jquerylib      0.1.4      2021-04-26 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  jsonlite       1.7.2      2020-12-09 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  knitr          1.33       2021-04-24 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  labeling       0.4.2      2020-10-20 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  lattice        0.20-41    2020-04-02 [2] CRAN (R 4.0.5)                       </span>
<span class="co">#&gt;  lifecycle      1.0.0      2021-02-15 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  magrittr       2.0.1      2020-11-17 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  MASS           7.3-53.1   2021-02-12 [2] CRAN (R 4.0.5)                       </span>
<span class="co">#&gt;  Matrix         1.3-2      2021-01-06 [2] CRAN (R 4.0.5)                       </span>
<span class="co">#&gt;  matrixStats    0.58.0     2021-01-29 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  memoise        2.0.0      2021-01-26 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  mgcv           1.8-34     2021-02-16 [2] CRAN (R 4.0.5)                       </span>
<span class="co">#&gt;  microbiome     1.12.0     2020-10-27 [1] Bioconductor                         </span>
<span class="co">#&gt;  microViz     * 0.7.4.9001 2021-05-06 [1] local                                </span>
<span class="co">#&gt;  multtest       2.46.0     2020-10-27 [1] Bioconductor                         </span>
<span class="co">#&gt;  munsell        0.5.0      2018-06-12 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  nlme           3.1-152    2021-02-04 [2] CRAN (R 4.0.5)                       </span>
<span class="co">#&gt;  patchwork      1.1.1      2020-12-17 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  permute        0.9-5      2019-03-12 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  phyloseq     * 1.34.0     2020-10-27 [1] Bioconductor                         </span>
<span class="co">#&gt;  pillar         1.6.0      2021-04-13 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  pkgbuild       1.2.0      2020-12-15 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  pkgconfig      2.0.3      2019-09-22 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  pkgdown        1.6.1      2020-09-12 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  pkgload        1.2.1      2021-04-06 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  plyr           1.8.6      2020-03-03 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  prettyunits    1.1.1      2020-01-24 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  processx       3.5.2      2021-04-30 [1] CRAN (R 4.0.5)                       </span>
<span class="co">#&gt;  progress       1.2.2      2019-05-16 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  ps             1.6.0      2021-02-28 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  purrr          0.3.4      2020-04-17 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  R6             2.5.0      2020-10-28 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  ragg           0.4.1      2021-01-11 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  RColorBrewer   1.1-2      2014-12-07 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  Rcpp           1.0.6      2021-01-15 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  remotes        2.3.0      2021-04-01 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  reshape2       1.4.4      2020-04-09 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  rhdf5          2.34.0     2020-10-27 [1] Bioconductor                         </span>
<span class="co">#&gt;  rhdf5filters   1.2.1      2021-05-03 [1] Bioconductor                         </span>
<span class="co">#&gt;  Rhdf5lib       1.12.1     2021-01-26 [1] Bioconductor                         </span>
<span class="co">#&gt;  rlang          0.4.11     2021-04-30 [1] CRAN (R 4.0.5)                       </span>
<span class="co">#&gt;  rmarkdown      2.7        2021-02-19 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  rprojroot      2.0.2      2020-11-15 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  Rtsne          0.15       2018-11-10 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  S4Vectors      0.28.1     2020-12-09 [1] Bioconductor                         </span>
<span class="co">#&gt;  sass           0.3.1      2021-01-24 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  scales         1.1.1      2020-05-11 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  sessioninfo    1.1.1      2018-11-05 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  stringi        1.5.3      2020-09-09 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  stringr        1.4.0      2019-02-10 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  survival       3.2-10     2021-03-16 [2] CRAN (R 4.0.5)                       </span>
<span class="co">#&gt;  systemfonts    1.0.1      2021-02-09 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  testthat       3.0.2      2021-02-14 [1] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  textshaping    0.2.1      2020-11-13 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  tibble         3.1.1      2021-04-18 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  tidyr          1.1.3      2021-03-03 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  tidyselect     1.1.1      2021-04-30 [1] CRAN (R 4.0.5)                       </span>
<span class="co">#&gt;  usethis        2.0.1      2021-02-10 [1] CRAN (R 4.0.3)                       </span>
<span class="co">#&gt;  utf8           1.2.1      2021-03-12 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  vctrs          0.3.8      2021-04-29 [1] CRAN (R 4.0.5)                       </span>
<span class="co">#&gt;  vegan          2.5-7      2020-11-28 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  withr          2.4.2      2021-04-18 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  xfun           0.22       2021-03-11 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  XVector        0.30.0     2020-10-28 [1] Bioconductor                         </span>
<span class="co">#&gt;  yaml           2.2.1      2020-02-01 [1] CRAN (R 4.0.2)                       </span>
<span class="co">#&gt;  zlibbioc       1.36.0     2020-10-28 [1] Bioconductor                         </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [1] /Users/runner/work/_temp/Library</span>
<span class="co">#&gt; [2] /Library/Frameworks/R.framework/Versions/4.0/Resources/library</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by David Barnett.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: '22e6f0a75b7b1ffae9bf5bc5b58864d7',
    indexName: 'microViz',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
