<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Example analyses: Shao 2019 data ‚Ä¢ microViz</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><!-- docsearch --><script src="../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><meta property="og:title" content="Example analyses: Shao 2019 data">
<meta property="og:description" content="microViz">
<meta property="og:image" content="https://david-barnett.github.io/microViz/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=G-2YQ669S4V5"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-2YQ669S4V5');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">microViz</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.11.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/shao19-analyses.html">Getting Started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/web-only/phyloseq.html">Working with phyloseq objects</a>
    </li>
    <li>
      <a href="../articles/web-only/tax-fixing.html">Fixing your tax_table</a>
    </li>
    <li>
      <a href="../articles/web-only/ordination.html">Ordination plots</a>
    </li>
    <li>
      <a href="../articles/web-only/ordination-interactive.html">Interactive ordination plots</a>
    </li>
    <li>
      <a href="../articles/web-only/compositions.html">Visualising compositions with barplots</a>
    </li>
    <li>
      <a href="../articles/web-only/heatmaps.html">Heatmaps of microbiome composition or correlation</a>
    </li>
    <li>
      <a href="../articles/web-only/modelling-taxa.html">Statistical modelling of individual taxa</a>
    </li>
    <li>
      <a href="../articles/web-only/videos.html">Tutorial video</a>
    </li>
    <li>
      <a href="../articles/ps_extra-replaced.html">Upgrading to microViz version 0.10.0</a>
    </li>
    <li>
      <a href="../articles/web-only/atlas1006.html">Example analyses with atlas1006 data</a>
    </li>
  </ul>
</li>
<li>
  <a href="../articles/learn/exercises.html">Learn</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
<li class="dropdown-header">
      </li>
</ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/david-barnett/microViz/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Example analyses: Shao 2019 data</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/david-barnett/microViz/blob/HEAD/vignettes/articles/shao19-analyses.Rmd" class="external-link"><code>vignettes/articles/shao19-analyses.Rmd</code></a></small>
      <div class="hidden name"><code>shao19-analyses.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This article will demonstrate some of the data processing steps,
statistical analyses, and visualisations that can be performed with the
microViz package. It will also provide some tips for how to use microViz
with your own data, and point you to other articles for more
details.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span> <span class="co"># for reproducible stochastic processes</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://dx.plos.org/10.1371/journal.pone.0061217" class="external-link">phyloseq</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span> <span class="co"># for combining multiple plots</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://david-barnett.github.io/microViz">microViz</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">microViz version 0.11.0 - Copyright (C) 2023 David Barnett</span></span></span>
<span><span class="co">#&gt; <span style="color: #BBBB00;">!</span> Website: https://david-barnett.github.io/microViz</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">‚úî</span> Useful?  For citation details, run: `citation("microViz")`</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">‚úñ</span> Silence? `suppressPackageStartupMessages(library(microViz))`</span></span></code></pre></div>
<p>The data used in this article are derived from faecal samples
obtained from infants and mothers participating in a large birth cohort
study.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">shao19</span></span>
<span><span class="co">#&gt; phyloseq-class experiment-level object</span></span>
<span><span class="co">#&gt; otu_table()   OTU Table:         [ 819 taxa and 1644 samples ]</span></span>
<span><span class="co">#&gt; sample_data() Sample Data:       [ 1644 samples by 11 sample variables ]</span></span>
<span><span class="co">#&gt; tax_table()   Taxonomy Table:    [ 819 taxa by 6 taxonomic ranks ]</span></span>
<span><span class="co">#&gt; phy_tree()    Phylogenetic Tree: [ 819 tips and 818 internal nodes ]</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">?</span><span class="va">shao19</span> <span class="co"># for more details on the dataset</span></span></code></pre></div>
<p><strong>Note:</strong> These human gut microbiota data were generated
from shotgun metagenomic sequencing, but microViz can be (and has been)
used with microbiota data from various other sources, including 16S and
ITS marker gene amplicon sequencing techniques, as well as HITChip
profiling. Environmental and <em>in vitro</em> microbiota datasets are
all also welcome in microViz, not just human gut bugs.</p>
<p>If you can put your data in a phyloseq object, then you can use
microViz with it. If you need guidance on how to create a phyloseq
object from your own data: <a href="https://david-barnett.github.io/microViz/articles/web-only/phyloseq.html#getting-your-data-into-phyloseq" title="Getting your data into phyloseq">see this article for resource
links</a>.</p>
</div>
<div class="section level2">
<h2 id="checking-your-data">Checking your data<a class="anchor" aria-label="anchor" href="#checking-your-data"></a>
</h2>
<p>You can check the basic characteristics of your phyloseq dataset
using standard phyloseq functions. microViz also provides a few helper
functions.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/sample_names-methods.html" class="external-link">sample_names</a></span><span class="op">(</span><span class="va">shao19</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "B01042_mo"    "B01042_ba_10" "B01042_ba_7"  "B01089_mo"    "B01089_ba_4" </span></span>
<span><span class="co">#&gt; [6] "B01089_ba_7"</span></span></code></pre></div>
<p>Note: These taxa already have informative unique names, but if your
taxa_names are just numbers or sequences, look at the
<code><a href="../reference/tax_rename.html">tax_rename()</a></code> function for one way to replace them with more
readable/informative unique names.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/taxa_names-methods.html" class="external-link">taxa_names</a></span><span class="op">(</span><span class="va">shao19</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Escherichia coli"             "Bacteroides caccae"          </span></span>
<span><span class="co">#&gt; [3] "Bacteroides stercoris"        "Ruminococcus bromii"         </span></span>
<span><span class="co">#&gt; [5] "[Eubacterium] rectale"        "Bifidobacterium adolescentis"</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/sample_variables.html" class="external-link">sample_variables</a></span><span class="op">(</span><span class="va">shao19</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "subject_id"              "family_id"              </span></span>
<span><span class="co">#&gt;  [3] "sex"                     "family_role"            </span></span>
<span><span class="co">#&gt;  [5] "age"                     "infant_age"             </span></span>
<span><span class="co">#&gt;  [7] "birth_weight"            "birth_mode"             </span></span>
<span><span class="co">#&gt;  [9] "c_section_type"          "antibiotics_current_use"</span></span>
<span><span class="co">#&gt; [11] "number_reads"</span></span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/psExtra-accessors.html">samdat_tbl</a></span><span class="op">(</span><span class="va">shao19</span><span class="op">)</span> <span class="co"># retrieve sample_data as a tibble</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1,644 √ó 12</span></span></span>
<span><span class="co">#&gt;    .sample_name subject_id family_id sex    family_role   age infant_age</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> B01042_mo    B01042_mo  193       male   mother         32         <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> B01042_ba_10 B01042_ba  193       male   child           0         10</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> B01042_ba_7  B01042_ba  193       male   child           0          7</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> B01089_mo    B01089_mo  194       female mother         38         <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> B01089_ba_4  B01089_ba  194       female child           0          4</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> B01089_ba_7  B01089_ba  194       female child           0          7</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> B01089_ba_21 B01089_ba  194       female child           0         21</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> B01128_ba_7  B01128_ba  195       male   child           0          7</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> B01128_mo    B01128_mo  195       male   mother         32         <span style="color: #BB0000;">NA</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> B01190_ba_21 B01190_ba  196       male   child           0         21</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ‚Ñπ 1,634 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ‚Ñπ 5 more variables: birth_weight &lt;dbl&gt;, birth_mode &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   c_section_type &lt;chr&gt;, antibiotics_current_use &lt;chr&gt;, number_reads &lt;int&gt;</span></span></span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/psExtra-accessors.html">otu_get</a></span><span class="op">(</span><span class="va">shao19</span>, taxa <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, samples <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span> <span class="co"># look at a tiny part of the otu_table</span></span>
<span><span class="co">#&gt; OTU Table:          [3 taxa and 5 samples]</span></span>
<span><span class="co">#&gt;                      taxa are columns</span></span>
<span><span class="co">#&gt;              Escherichia coli Bacteroides caccae Bacteroides stercoris</span></span>
<span><span class="co">#&gt; B01042_mo             5787899            2237960               1225392</span></span>
<span><span class="co">#&gt; B01042_ba_10           130453                  0                     0</span></span>
<span><span class="co">#&gt; B01042_ba_7                 0                  0                     0</span></span>
<span><span class="co">#&gt; B01089_mo             7371972             595532                746435</span></span>
<span><span class="co">#&gt; B01089_ba_4                 0                  0                     0</span></span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/rank_names.html" class="external-link">rank_names</a></span><span class="op">(</span><span class="va">shao19</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "phylum"  "class"   "order"   "family"  "genus"   "species"</span></span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/tax_table-methods.html" class="external-link">tax_table</a></span><span class="op">(</span><span class="va">shao19</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span></span>
<span><span class="co">#&gt; Taxonomy Table:     [3 taxa by 6 taxonomic ranks]:</span></span>
<span><span class="co">#&gt;                       phylum           class                 order             </span></span>
<span><span class="co">#&gt; Escherichia coli      "Proteobacteria" "Gammaproteobacteria" "Enterobacterales"</span></span>
<span><span class="co">#&gt; Bacteroides caccae    "Bacteroidetes"  "Bacteroidia"         "Bacteroidales"   </span></span>
<span><span class="co">#&gt; Bacteroides stercoris "Bacteroidetes"  "Bacteroidia"         "Bacteroidales"   </span></span>
<span><span class="co">#&gt;                       family               genus        </span></span>
<span><span class="co">#&gt; Escherichia coli      "Enterobacteriaceae" "Escherichia"</span></span>
<span><span class="co">#&gt; Bacteroides caccae    "Bacteroidaceae"     "Bacteroides"</span></span>
<span><span class="co">#&gt; Bacteroides stercoris "Bacteroidaceae"     "Bacteroides"</span></span>
<span><span class="co">#&gt;                       species                </span></span>
<span><span class="co">#&gt; Escherichia coli      "Escherichia coli"     </span></span>
<span><span class="co">#&gt; Bacteroides caccae    "Bacteroides caccae"   </span></span>
<span><span class="co">#&gt; Bacteroides stercoris "Bacteroides stercoris"</span></span></code></pre></div>
<p>The function <code><a href="../reference/phyloseq_validate.html">phyloseq_validate()</a></code> can be used to check
for common problems with phyloseq objects, so I suggest running it on
your data before trying to start your analyses.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">shao19</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/phyloseq_validate.html">phyloseq_validate</a></span><span class="op">(</span><span class="va">shao19</span><span class="op">)</span> <span class="co"># no messages or warnings means no detected problems</span></span></code></pre></div>
<div class="section level3">
<h3 id="fixing-your-tax_table">Fixing your tax_table<a class="anchor" aria-label="anchor" href="#fixing-your-tax_table"></a>
</h3>
<p>One common problem you will encounter either when you run
phyloseq_validate or shortly after, are problematic entries in the
taxonomy table. For example if many or all of the ‚Äúspecies‚Äù rank entries
are <code>"s__"</code> or <code>"unknown_species"</code> or
<code>NA</code> etc. The same species name should not appear under
multiple genera, so these duplicated or uninformative entries need to be
replaced before you can proceed.</p>
<p><a href="https://david-barnett.github.io/microViz/articles/web-only/tax-fixing.html" title="Fixing your tax_table">See this article</a> for a discussion of
how to fix these problems using <code><a href="../reference/tax_fix.html">tax_fix()</a></code> and maybe
<code><a href="../reference/tax_fix_interactive.html">tax_fix_interactive()</a></code> and/or <code><a href="../reference/tax_filter.html">tax_filter()</a></code>. As
a last resort you could also try deleting entirely unwanted taxa by
using <code><a href="../reference/tax_select.html">tax_select()</a></code>.</p>
</div>
</div>
<div class="section level2">
<h2 id="modify-your-sample_data">Modify your sample_data<a class="anchor" aria-label="anchor" href="#modify-your-sample_data"></a>
</h2>
<p>Later in these example analyses we will need modified version of the
sample variables stored in the sample_data slot of the phyloseq object.
The <code><a href="../reference/ps_mutate.html">ps_mutate()</a></code> function provides an easy way to modify
your phyloseq sample_data (ps is short for phyloseq). You can use
<code><a href="../reference/ps_mutate.html">ps_mutate()</a></code> in a similar way to
<code><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">dplyr::mutate()</a></code>. If you are unfamiliar with the
<a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a> package, I highly recommend you look at the <a href="https://dplyr.tidyverse.org/" class="external-link">dplyr website</a>, to learn about
some incredibly handy tools for data transformation, and because several
of the microViz data transformation functions are used in a similar
way.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">shao19</span> <span class="op">&lt;-</span> <span class="va">shao19</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/ps_mutate.html">ps_mutate</a></span><span class="op">(</span></span>
<span>    Csection <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">birth_mode</span> <span class="op">==</span> <span class="st">"c_section"</span>, yes <span class="op">=</span> <span class="fl">1</span>, no <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>,</span>
<span>    birth_mode3 <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/recode.html" class="external-link">recode</a></span><span class="op">(</span></span>
<span>      <span class="va">c_section_type</span>,</span>
<span>      Elective_CS <span class="op">=</span> <span class="st">"CS Elective"</span>, Emergency_CS <span class="op">=</span> <span class="st">"CS Emergency"</span>, .missing <span class="op">=</span> <span class="st">"Vaginal"</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="subset-your-samples">Subset your samples<a class="anchor" aria-label="anchor" href="#subset-your-samples"></a>
</h2>
<p>For the first part of this example analyses we will look at only one
sample per infant, from the timepoint when they were 4 days old. The
microViz function <code><a href="../reference/ps_filter.html">ps_filter()</a></code> makes this easy. You use
<code><a href="../reference/ps_filter.html">ps_filter()</a></code> in a similar way to
<code><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">dplyr::filter()</a></code>, to filter samples using variables in the
sample_data.</p>
<p>We will also use <code><a href="../reference/ps_dedupe.html">ps_dedupe()</a></code> to ‚Äúdeduplicate‚Äù samples,
to ensure that we definitely only keep one sample per family (e.g.¬†if
any infant has more than one sample at age 4, or if there are
twins).</p>
<ul>
<li><p><strong>Note 1:</strong> Observe that <code>==</code> is used
here, not <code>=</code></p></li>
<li><p><strong>Note 2</strong>: By default, <code><a href="../reference/ps_filter.html">ps_filter()</a></code>
also removes taxa that no longer appear in the filtered dataset (zero
total counts). This is different to
<code><a href="https://rdrr.io/pkg/phyloseq/man/subset_samples-methods.html" class="external-link">phyloseq::subset_samples()</a></code> and
<code><a href="https://rdrr.io/pkg/phyloseq/man/prune_samples-methods.html" class="external-link">phyloseq::prune_samples()</a></code>.</p></li>
</ul>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">shao4d</span> <span class="op">&lt;-</span> <span class="va">shao19</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/ps_filter.html">ps_filter</a></span><span class="op">(</span><span class="va">family_role</span> <span class="op">==</span> <span class="st">"child"</span>, <span class="va">infant_age</span> <span class="op">==</span> <span class="fl">4</span>, .keep_all_taxa <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/ps_dedupe.html">ps_dedupe</a></span><span class="op">(</span>vars <span class="op">=</span> <span class="st">"family_id"</span><span class="op">)</span></span>
<span><span class="co">#&gt; 306 groups: with 1 samples each</span></span>
<span><span class="co">#&gt; Dropped 0 samples.</span></span>
<span><span class="va">shao4d</span></span>
<span><span class="co">#&gt; phyloseq-class experiment-level object</span></span>
<span><span class="co">#&gt; otu_table()   OTU Table:         [ 353 taxa and 306 samples ]</span></span>
<span><span class="co">#&gt; sample_data() Sample Data:       [ 306 samples by 13 sample variables ]</span></span>
<span><span class="co">#&gt; tax_table()   Taxonomy Table:    [ 353 taxa by 6 taxonomic ranks ]</span></span>
<span><span class="co">#&gt; phy_tree()    Phylogenetic Tree: [ 353 tips and 352 internal nodes ]</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="composition-barplot">Composition barplot<a class="anchor" aria-label="anchor" href="#composition-barplot"></a>
</h2>
<p>Let us look at the compositions of infant samples from age of 4 days,
grouping the samples by birth mode. The <code><a href="../reference/comp_barplot.html">comp_barplot()</a></code>
function make generating an attractive and informative ggplot2 bar chart
easy: by default it automatically sorts the samples by microbiota
similarity, sorts the taxa by overall abundance, and applies a sensible
categorical colour palette with many colours. Taxa not assigned a colour
are merged into one light grey bar by default, but can be shown
un-merged, as in this example.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">shao4d</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/comp_barplot.html">comp_barplot</a></span><span class="op">(</span><span class="st">"genus"</span>, n_taxa <span class="op">=</span> <span class="fl">15</span>, merge_other <span class="op">=</span> <span class="cn">FALSE</span>, label <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/vars.html" class="external-link">vars</a></span><span class="op">(</span><span class="va">birth_mode</span><span class="op">)</span>, scales <span class="op">=</span> <span class="st">"free"</span><span class="op">)</span> <span class="op">+</span> <span class="co"># scales = "free" is IMPORTANT!</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span></span>
<span>    <span class="st">"Neonatal gut microbiota composition at 4 days"</span>,</span>
<span>    <span class="st">"Substanstial differences are visible in profiles of high abundance taxa"</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.ticks.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>, strip.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="shao19-analyses_files/figure-html/unnamed-chunk-14-1.png" width="672"></p>
<p>As practice, try modifying this barplot by changing some of the
<code><a href="../reference/comp_barplot.html">comp_barplot()</a></code> arguments, try for example: using a
different dissimilarity measure to sort the samples, displaying a
different taxonomic rank, colouring fewer taxa, and/or removing the bar
outlines.</p>
<p>Check out <a href="https://david-barnett.github.io/microViz/articles/web-only/compositions.html" title="Composition bar plots">the article on composition bar plots</a>
for more guidance and ideas.</p>
</div>
<div class="section level2">
<h2 id="ordination-plot">Ordination plot<a class="anchor" aria-label="anchor" href="#ordination-plot"></a>
</h2>
<p>microViz provides an easy workflow for creating ordination plots
including <a href="https://sites.google.com/site/mb3gustame/indirect-gradient-analysis/pca" title="GUSTA ME PCA introduction" class="external-link">PCA</a>, <a href="https://sites.google.com/site/mb3gustame/dissimilarity-based-methods/principal-coordinates-analysis" title="GUSTA ME PCoA introduction" class="external-link">PCoA</a> and <a href="https://sites.google.com/site/mb3gustame/dissimilarity-based-methods/nmds" title="GUSTA ME NMDS introduction" class="external-link">NMDS</a> using ggplot2, including
plotting taxa loadings arrows for PCA. Preparing for an ordination plot
requires a few steps.</p>
<ol style="list-style-type: decimal">
<li>
<code><a href="../reference/tax_filter.html">tax_filter()</a></code> to filter out rare taxa - this is an
optional step, relevant for some ordination methods</li>
<li>
<code><a href="../reference/tax_agg.html">tax_agg()</a></code> to aggregate taxa at your chosen taxonomic
rank, e.g.¬†genus</li>
<li>
<code><a href="../reference/tax_transform.html">tax_transform()</a></code> to transform the abundance counts
(important for PCA, but inappropriate for many dissimilarity-based
ordinations)</li>
<li>
<code><a href="../reference/dist_calc.html">dist_calc()</a></code> to calculate a sample-sample distance or
dissimilarity matrix (only needed for <a href="https://sites.google.com/site/mb3gustame/dissimilarity-based-methods" title="GUSTA Me dissimilarity based methods" class="external-link">dissimilarity-based
methods</a>, e.g.¬†PCoA or NMDS)</li>
<li>
<code><a href="../reference/ord_calc.html">ord_calc()</a></code> to perform the ordination analysis</li>
<li>
<code><a href="../reference/ord_plot.html">ord_plot()</a></code> to plot any two dimensions of your ordinated
data</li>
</ol>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">shao4d_psX</span> <span class="op">&lt;-</span> <span class="va">shao4d</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="co"># keep only taxa belonging to genera that have over 100 counts in at least 5% of samples</span></span>
<span>  <span class="fu"><a href="../reference/tax_filter.html">tax_filter</a></span><span class="op">(</span>min_prevalence <span class="op">=</span> <span class="fl">0.05</span>, undetected <span class="op">=</span> <span class="fl">100</span>, tax_level <span class="op">=</span> <span class="st">"genus"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="co"># aggregate counts at genus-level &amp; transform with robust CLR transformation</span></span>
<span>  <span class="fu"><a href="../reference/tax_transform.html">tax_transform</a></span><span class="op">(</span>trans <span class="op">=</span> <span class="st">"rclr"</span>, rank <span class="op">=</span> <span class="st">"genus"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="co"># no distances are needed for PCA: so skip dist_calc and go straight to ord_calc</span></span>
<span>  <span class="fu"><a href="../reference/ord_calc.html">ord_calc</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Proportional min_prevalence given: 0.05 --&gt; min 16/306 samples.</span></span>
<span></span>
<span><span class="va">shao4d_psX</span></span>
<span><span class="co">#&gt; psExtra object - a phyloseq object with extra slots:</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; phyloseq-class experiment-level object</span></span>
<span><span class="co">#&gt; otu_table()   OTU Table:         [ 29 taxa and 306 samples ]</span></span>
<span><span class="co">#&gt; sample_data() Sample Data:       [ 306 samples by 13 sample variables ]</span></span>
<span><span class="co">#&gt; tax_table()   Taxonomy Table:    [ 29 taxa by 5 taxonomic ranks ]</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; otu_get(counts = TRUE)        [ 29 taxa and 306 samples ]</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; psExtra info:</span></span>
<span><span class="co">#&gt; tax_agg = "genus" tax_trans = "rclr" </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ordination of class: rda cca </span></span>
<span><span class="co">#&gt; rda(formula = OTU ~ 1, data = data)</span></span>
<span><span class="co">#&gt; Ordination info:</span></span>
<span><span class="co">#&gt; method = 'PCA'   </span></span></code></pre></div>
<p><strong>Note:</strong> microViz will record your choices in steps 2
through 5, by adding additional information to the phyloseq object,
which is now called a ‚Äú<a href="https://david-barnett.github.io/microViz/articles/ps_extra-replaced.html" title="A note on ps_extra to psExtra transition">psExtra</a>‚Äù. By
default the psExtra object will also store: the original counts OTU
table (before any transformation); the distance matrix; and ordination
object. To retrieve each of these, use
<code>ps_get(counts = TRUE)</code>, <code><a href="../reference/psExtra-accessors.html">dist_get()</a></code>, or
<code><a href="../reference/psExtra-accessors.html">ord_get()</a></code>.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">PCA_plot</span> <span class="op">&lt;-</span> <span class="va">shao4d_psX</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/ord_plot.html">ord_plot</a></span><span class="op">(</span></span>
<span>    colour <span class="op">=</span> <span class="st">"birth_mode"</span>, shape <span class="op">=</span> <span class="st">"birth_mode"</span>,</span>
<span>    plot_taxa <span class="op">=</span> <span class="fl">10</span><span class="op">:</span><span class="fl">1</span>,</span>
<span>    tax_vec_length <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>    tax_lab_style <span class="op">=</span> <span class="fu"><a href="../reference/Ordination-labels.html">tax_lab_style</a></span><span class="op">(</span></span>
<span>      type <span class="op">=</span> <span class="st">"label"</span>, max_angle <span class="op">=</span> <span class="fl">90</span>, aspect_ratio <span class="op">=</span> <span class="fl">1</span>,</span>
<span>      size <span class="op">=</span> <span class="fl">2.5</span>, alpha <span class="op">=</span> <span class="fl">0.8</span>, fontface <span class="op">=</span> <span class="st">"bold"</span>, <span class="co"># style the labels</span></span>
<span>      label.r <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">0</span>, <span class="st">"mm"</span><span class="op">)</span> <span class="co"># square corners of labels - see ?geom_label</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span>ratio <span class="op">=</span> <span class="fl">1</span>, clip <span class="op">=</span> <span class="st">"off"</span>, xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">2</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># match coord_fixed() ratio to tax_lab_style() aspect_ratio for correct text angles</span></span>
<span></span>
<span><span class="va">PCA_plot</span></span></code></pre></div>
<p><img src="shao19-analyses_files/figure-html/unnamed-chunk-16-1.png" width="720"></p>
<p>As this PCA plot is a ggplot object, you can adjust the aesthetic
scales (colour, shape, size etc.) and theme elements in <a href="https://ggplot2.tidyverse.org/" title="ggplot2 website" class="external-link">the usual
ggplot ways</a>. However, styling the taxon loadings arrows and labels
can only be done within the <code><a href="../reference/ord_plot.html">ord_plot()</a></code> call itself.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">PCA_plot_custom</span> <span class="op">&lt;-</span> <span class="va">PCA_plot</span> <span class="op">+</span></span>
<span>  <span class="co"># add a convex hull around the points for each group, to aid the eye</span></span>
<span>  <span class="fu"><a href="../reference/stat_chull.html">stat_chull</a></span><span class="op">(</span></span>
<span>    mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="va">birth_mode</span>, linetype <span class="op">=</span> <span class="va">birth_mode</span><span class="op">)</span>,</span>
<span>    linewidth <span class="op">=</span> <span class="fl">0.5</span>, alpha <span class="op">=</span> <span class="fl">0.5</span>, show.legend <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_shape_manual</a></span><span class="op">(</span></span>
<span>    name <span class="op">=</span> <span class="st">"Birth mode"</span>,</span>
<span>    values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"circle"</span>, <span class="st">"circle open"</span><span class="op">)</span>, labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"C-section"</span>, <span class="st">"Vaginal"</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="co"># set a custom colour scale and customise the legend order and appearance</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span></span>
<span>    name <span class="op">=</span> <span class="st">"Birth mode"</span>,</span>
<span>    values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"grey45"</span><span class="op">)</span>, labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"C-section"</span>, <span class="st">"Vaginal"</span><span class="op">)</span>,</span>
<span>    guide <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">4</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="co"># add a title and delete the automatic caption</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"4-day-old gut microbiota and birth mode"</span>, caption <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="co"># put the legend at the bottom and draw a border around it</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span>, legend.background <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_rect</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">PCA_plot_custom</span></span></code></pre></div>
<p><img src="shao19-analyses_files/figure-html/unnamed-chunk-17-1.png" width="672"></p>
<p>There are many choices to make during ordination analysis and
visualisation. Try customising the ordination plot itself: change the
arguments in <code><a href="../reference/ord_plot.html">ord_plot()</a></code> (e.g.¬†map the shape aesthetic to
the infant sex, and plot the 1st &amp; 4th principal components); or use
ggplot functions to change the appearance of the plot (e.g.¬†pick a new
theme or modify the current theme to remove the the panel.grid).</p>
<p>Try out some different choices for the ordination analysis itself,
e.g.¬†make an NMDS plot using Bray-Curtis dissimilarities calculating
using class-level data. If that sounds like too much typing, you might
like to skip ahead to the section on creating and exploring <a href="#interactive-ordination-plots">Interactive ordination plots</a>
with <code><a href="../reference/ord_explore.html">ord_explore()</a></code>.</p>
<p><strong>Note 1:</strong> Most of the distance calculation and
ordination analysis methods are implemented or collected within the
brilliant <a href="https://vegandevs.github.io/vegan/articles/FAQ-vegan.html" title="What is vegan?" class="external-link">vegan package</a> developed by the statisticians
Jari Oksanen and Gavin Simpson, amongst others. microViz uses vegan
functions internally, and provides a ggplot2 approach to visualising the
ordination results.</p>
<p><strong>Note 2:</strong> Constrained and conditioned ordination
analyses, e.g.¬†RDA, are also possible. If you understand the rationale
behind these analyses, feel free to try out setting the constraints
and/or conditions arguments in <code><a href="../reference/ord_calc.html">ord_calc()</a></code>. If these
concepts are new to you, the <a href="https://sites.google.com/site/mb3gustame/" class="external-link"><strong>‚ÄúGUSTA
ME‚Äù</strong> <strong><em>GUide to STatistical Analysis in Microbial
Ecology</em></strong></a> website is one good resource for learning more
about useful methods like <a href="https://sites.google.com/site/mb3gustame/constrained-analyses/rda" title="Redundancy analysis" class="external-link">redundancy analysis</a> and <a href="https://sites.google.com/site/mb3gustame/constrained-analyses/rda/partial-redundancy-analysis" title="Partial redundancy analysis" class="external-link">partial redundancy analysis</a>. It
is easy to produce quite misleading plots through the misuse of <a href="https://sites.google.com/site/mb3gustame/constrained-analyses" title="GUSTA ME constrained analyses intro" class="external-link">constrained analyses</a>, so
be careful! üòá</p>
</div>
<div class="section level2">
<h2 id="ordination-sorted-circular-barplot">Ordination-sorted circular barplot<a class="anchor" aria-label="anchor" href="#ordination-sorted-circular-barplot"></a>
</h2>
<p>For additional insight into your ordination plot results, microViz
provides a novel approach to pairing a circular microbiota composition
bar chart, or ‚Äúiris‚Äù plot, with an ordination plot. The samples on the
circular bar chart are ordered using the rotational order of the samples
on the ordination plot axes!</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">irisPlot</span> <span class="op">&lt;-</span> <span class="va">shao4d_psX</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/ord_plot_iris.html">ord_plot_iris</a></span><span class="op">(</span></span>
<span>    axes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>, tax_level <span class="op">=</span> <span class="st">"genus"</span>, n_taxa <span class="op">=</span> <span class="fl">8</span>,</span>
<span>    anno_binary <span class="op">=</span> <span class="st">"Csection"</span>, <span class="co"># add an annotation ring indicating C-section birth</span></span>
<span>    anno_binary_style <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.5</span>, colour <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span>, <span class="co"># from ord plot colour scale</span></span>
<span>    ord_plot <span class="op">=</span> <span class="st">"none"</span> <span class="co"># we'll reuse the customised PCA plot from earlier</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">PCA_plot_custom</span>, <span class="va">irisPlot</span>, nrow <span class="op">=</span> <span class="fl">1</span>, widths <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="shao19-analyses_files/figure-html/unnamed-chunk-18-1.png" width="1152"></p>
</div>
<div class="section level2">
<h2 id="interactive-ordination-plots">Interactive ordination plots<a class="anchor" aria-label="anchor" href="#interactive-ordination-plots"></a>
</h2>
<p>Want to create an ordination plot and/or explore your dataset, but
don‚Äôt fancy much typing?</p>
<p>Starting with just a validated phyloseq object you can run
<code><a href="../reference/ord_explore.html">ord_explore()</a></code> to interactively create and explore
ordination plots.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ord_explore.html">ord_explore</a></span><span class="op">(</span><span class="va">shao19</span><span class="op">)</span></span></code></pre></div>
<p><img src="assets/ord_explore.gif"></p>
<p>As shown in the video clip above, your default web browser will open
a new window (or tab), and display an interactive <a href="https://shiny.posit.co/" class="external-link">shiny</a>
application. Select options from the menu to build an interactive
ordination plot, and click on/lasso select samples to view their
compositions. Click the ‚ÄúOptions‚Äù button to change the ordination
settings. Click on the ‚ÄúCode‚Äù button to generate code that you can
copy-paste into your script/notebook to reproduce the ordination
plot.</p>
</div>
<div class="section level2">
<h2 id="composition-heatmap">Composition heatmap<a class="anchor" aria-label="anchor" href="#composition-heatmap"></a>
</h2>
<p>Heatmaps can be a useful way to display taxonomic composition of your
samples, in addition to or instead of bar charts. You can transform the
taxa using <code><a href="../reference/tax_transform.html">tax_transform()</a></code>, various kinds of log
transformations (including clr or rclr) can be helpful to illustrate
relative abundance patterns for both low abundance and high abundance
taxa on the same plot. This is an advantage of heatmaps over
<code><a href="../reference/comp_barplot.html">comp_barplot()</a></code>, where patterns involving low relative
abundance taxa can be hard to spot.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">shao4d</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/tax_transform.html">tax_transform</a></span><span class="op">(</span>trans <span class="op">=</span> <span class="st">"rclr"</span>, rank <span class="op">=</span> <span class="st">"genus"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/tax_filter.html">tax_filter</a></span><span class="op">(</span>min_prevalence <span class="op">=</span> <span class="fl">0.1</span>, use_counts <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/comp_heatmap.html">comp_heatmap</a></span><span class="op">(</span></span>
<span>    colors <span class="op">=</span> <span class="fu"><a href="../reference/heat_palette.html">heat_palette</a></span><span class="op">(</span>sym <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, grid_col <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>    sample_side <span class="op">=</span> <span class="st">"bottom"</span>, name <span class="op">=</span> <span class="st">"Robust\nCLR"</span>,</span>
<span>    sample_anno <span class="op">=</span> <span class="fu"><a href="../reference/sampleAnnotation.html">sampleAnnotation</a></span><span class="op">(</span></span>
<span>      <span class="st">"Birth mode"</span> <span class="op">=</span> <span class="fu"><a href="../reference/anno_sample.html">anno_sample</a></span><span class="op">(</span><span class="st">"birth_mode3"</span><span class="op">)</span>,</span>
<span>      col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"Birth mode"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>        <span class="st">"CS Elective"</span> <span class="op">=</span> <span class="st">"black"</span>, <span class="st">"CS Emergency"</span> <span class="op">=</span> <span class="st">"orange"</span>, <span class="st">"Vaginal"</span> <span class="op">=</span> <span class="st">"lightgrey"</span></span>
<span>      <span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="co">#&gt; Proportional min_prevalence given: 0.1 --&gt; min 31/306 samples.</span></span></code></pre></div>
<p><img src="shao19-analyses_files/figure-html/unnamed-chunk-20-1.png" width="816"></p>
<p>See the microViz <a href="https://david-barnett.github.io/microViz/articles/web-only/heatmaps.html" title="Heatmaps">heatmaps article</a> for more guidance on making
composition heatmaps.</p>
<p><a href="https://jokergoo.github.io/ComplexHeatmap-reference/book/" title="ComplexHeatmap book" class="external-link">See the ComplexHeatmap online book</a> for
more information on controlling the display of Complex Heatmaps.</p>
</div>
<div class="section level2">
<h2 id="statistical-testing">Statistical testing<a class="anchor" aria-label="anchor" href="#statistical-testing"></a>
</h2>
<p>microViz can also help you perform and visualise the results of
statistical tests. A couple of simple examples follow.</p>
<div class="section level3">
<h3 id="permanova">PERMANOVA<a class="anchor" aria-label="anchor" href="#permanova"></a>
</h3>
<p>One common method for assessing whether a variable has is associated
with an overall difference in microbiota composition in your dataset is
permutational multivariate analysis of variance (PERMANOVA). For a great
introduction to PERMANOVA, <a href="https://doi.org/10.1002/9781118445112.stat07841" title="Permutational Multivariate Analysis of Variance (PERMANOVA)" class="external-link">see
this resource by Marti J Anderson</a>.</p>
<p>vegan provides the <code><a href="https://rdrr.io/pkg/vegan/man/adonis.html" class="external-link">vegan::adonis2()</a></code> function for
performing PERMANOVA analyses. microViz provides a convenience function,
<code><a href="../reference/dist_permanova.html">dist_permanova()</a></code> which allows you to use the distance
matrix stored in a psExtra object to perform PERMANOVA with adonis2, and
store the result in the psExtra too.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">shao4d_perm</span> <span class="op">&lt;-</span> <span class="va">shao4d</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/tax_transform.html">tax_transform</a></span><span class="op">(</span><span class="st">"identity"</span>, rank <span class="op">=</span> <span class="st">"genus"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/dist_calc.html">dist_calc</a></span><span class="op">(</span><span class="st">"aitchison"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/dist_permanova.html">dist_permanova</a></span><span class="op">(</span></span>
<span>    variables <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"birth_mode"</span>, <span class="st">"sex"</span>, <span class="st">"number_reads"</span><span class="op">)</span>,</span>
<span>    n_perms <span class="op">=</span> <span class="fl">99</span>, <span class="co"># you should use more permutations in your real analyses!</span></span>
<span>    n_processes <span class="op">=</span> <span class="fl">1</span></span>
<span>  <span class="op">)</span></span>
<span><span class="co">#&gt; Dropping samples with missings: 15</span></span>
<span><span class="co">#&gt; 2023-07-31 10:51:16.678858 - Starting PERMANOVA with 99 perms with 1 processes</span></span>
<span><span class="co">#&gt; 2023-07-31 10:51:17.09314 - Finished PERMANOVA</span></span>
<span></span>
<span><span class="va">shao4d_perm</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="../reference/psExtra-accessors.html">perm_get</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Permutation test for adonis under reduced model</span></span>
<span><span class="co">#&gt; Marginal effects of terms</span></span>
<span><span class="co">#&gt; Permutation: free</span></span>
<span><span class="co">#&gt; Number of permutations: 99</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; vegan::adonis2(formula = formula, data = metadata, permutations = n_perms, by = by, parallel = parall)</span></span>
<span><span class="co">#&gt;               Df SumOfSqs      R2       F Pr(&gt;F)   </span></span>
<span><span class="co">#&gt; birth_mode     1    10462 0.09055 29.3778   0.01 **</span></span>
<span><span class="co">#&gt; sex            1      402 0.00348  1.1296   0.31   </span></span>
<span><span class="co">#&gt; number_reads   1     1117 0.00967  3.1364   0.01 **</span></span>
<span><span class="co">#&gt; Residual     287   102209 0.88462                  </span></span>
<span><span class="co">#&gt; Total        290   115540 1.00000                  </span></span>
<span><span class="co">#&gt; ---</span></span>
<span><span class="co">#&gt; Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="differential-abundance-testing">Differential abundance testing<a class="anchor" aria-label="anchor" href="#differential-abundance-testing"></a>
</h3>
<p>microViz provides functions for applying various statistical
modelling methods to microbiome data, see the <a href="https://david-barnett.github.io/microViz/articles/web-only/modelling-taxa.html" title="Statistical modelling of individual taxa">statistical modelling
article for a longer discussion</a>.</p>
<p>microViz does not introduce any novel statistical method for
differential abundance testing, but allows you to apply various methods,
including those from other packages, on multiple taxa across multiple
taxonomic ranks of aggregation, using the <code><a href="../reference/taxatree_models.html">taxatree_models()</a></code>
function. If you want to apply statistical tests to many or all taxa in
a single rank, you could use the <code><a href="../reference/tax_model.html">tax_model()</a></code> function.</p>
<p>microViz offers a somewhat novel general approach to visualising the
results (effect estimates and significance) of many tests across
multiple ranks, ‚ÄúTaxonomic Association Trees‚Äù, a heatmap-style
visualisation with results arranged in a tree structure, following the
hierarchy of taxonomic rank relationships in the taxonomy table:
<code><a href="../reference/taxatree_plots.html">taxatree_plots()</a></code>.</p>
<div class="section level4">
<h4 id="taxonomic-association-tree-plots">Taxonomic association tree plots<a class="anchor" aria-label="anchor" href="#taxonomic-association-tree-plots"></a>
</h4>
<p>Starting with a simple approach borrowed from <a href="https://huttenhower.sph.harvard.edu/maaslin/" class="external-link">MaAsLin2</a>, let‚Äôs
perform simple linear regression using log2-transformed relative
abundances (as the dependent variable). We will test for differences in
the average relative abundance of each taxon between infants born by
C-section and vaginally delivered infants.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># First transform and filter the taxa, ready for statistical modelling #</span></span>
<span></span>
<span><span class="va">shao4d_prev10</span> <span class="op">&lt;-</span> <span class="va">shao4d</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="co"># prepend the 1st letter of the rank to each tax_table entry, to ensure all are unique</span></span>
<span>  <span class="fu"><a href="../reference/tax_prepend_ranks.html">tax_prepend_ranks</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/tax_transform.html">tax_transform</a></span><span class="op">(</span>trans <span class="op">=</span> <span class="st">"compositional"</span>, rank <span class="op">=</span> <span class="st">"genus"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="co"># for various statistical, biological, and practical reasons, let's strictly filter taxa</span></span>
<span>  <span class="fu"><a href="../reference/tax_filter.html">tax_filter</a></span><span class="op">(</span>min_prevalence <span class="op">=</span> <span class="fl">0.1</span>, use_counts <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Proportional min_prevalence given: 0.1 --&gt; min 31/306 samples.</span></span>
<span></span>
<span><span class="va">shao4d_prev10</span></span>
<span><span class="co">#&gt; psExtra object - a phyloseq object with extra slots:</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; phyloseq-class experiment-level object</span></span>
<span><span class="co">#&gt; otu_table()   OTU Table:         [ 18 taxa and 306 samples ]</span></span>
<span><span class="co">#&gt; sample_data() Sample Data:       [ 306 samples by 13 sample variables ]</span></span>
<span><span class="co">#&gt; tax_table()   Taxonomy Table:    [ 18 taxa by 5 taxonomic ranks ]</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; otu_get(counts = TRUE)        [ 18 taxa and 306 samples ]</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; psExtra info:</span></span>
<span><span class="co">#&gt; tax_agg = "genus" tax_trans = "compositional"</span></span></code></pre></div>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">shao4d_treeStats</span> <span class="op">&lt;-</span> <span class="va">shao4d_prev10</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="co"># run all the statistical models</span></span>
<span>  <span class="fu"><a href="../reference/taxatree_models.html">taxatree_models</a></span><span class="op">(</span></span>
<span>    ranks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"phylum"</span>, <span class="st">"class"</span>, <span class="st">"order"</span>, <span class="st">"family"</span>, <span class="st">"genus"</span><span class="op">)</span>,</span>
<span>    trans <span class="op">=</span> <span class="st">"log2"</span>, trans_args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>add <span class="op">=</span> <span class="st">"halfmin"</span><span class="op">)</span>,</span>
<span>    variables <span class="op">=</span> <span class="st">"Csection"</span>, type <span class="op">=</span> <span class="va">lm</span> <span class="co"># modelling function</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="co"># extract stats from the models</span></span>
<span>  <span class="fu"><a href="../reference/models2stats.html">taxatree_models2stats</a></span><span class="op">(</span>.keep_models <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="co"># adjust the p values for multiple testing, within each rank</span></span>
<span>  <span class="fu"><a href="../reference/taxatree_stats_p_adjust.html">taxatree_stats_p_adjust</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"fdr"</span>, grouping <span class="op">=</span> <span class="st">"rank"</span><span class="op">)</span></span>
<span><span class="co">#&gt; 2023-07-31 10:51:17.630655 - modelling at rank: phylum</span></span>
<span><span class="co">#&gt; 2023-07-31 10:51:17.863391 - modelling at rank: class</span></span>
<span><span class="co">#&gt; 2023-07-31 10:51:18.189435 - modelling at rank: order</span></span>
<span><span class="co">#&gt; 2023-07-31 10:51:18.616774 - modelling at rank: family</span></span>
<span><span class="co">#&gt; 2023-07-31 10:51:19.10824 - modelling at rank: genus</span></span></code></pre></div>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">shao4d_treeStats</span></span>
<span><span class="co">#&gt; psExtra object - a phyloseq object with extra slots:</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; phyloseq-class experiment-level object</span></span>
<span><span class="co">#&gt; otu_table()   OTU Table:         [ 18 taxa and 306 samples ]</span></span>
<span><span class="co">#&gt; sample_data() Sample Data:       [ 306 samples by 13 sample variables ]</span></span>
<span><span class="co">#&gt; tax_table()   Taxonomy Table:    [ 18 taxa by 5 taxonomic ranks ]</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; otu_get(counts = TRUE)        [ 18 taxa and 306 samples ]</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; psExtra info:</span></span>
<span><span class="co">#&gt; tax_agg = "genus" tax_trans = "compositional" </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; taxatree_models list:</span></span>
<span><span class="co">#&gt; Ranks: phylum/class/order/family/genus</span></span>
<span><span class="co">#&gt; taxatree_stats dataframe:</span></span>
<span><span class="co">#&gt; 57 taxa at 5 ranks: phylum, class, order, family, genus </span></span>
<span><span class="co">#&gt; 1 terms: Csection</span></span></code></pre></div>
<p>The models are stored in a nested list. A list of ranks containing
lists of taxa.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/psExtra-accessors.html">taxatree_models_get</a></span><span class="op">(</span><span class="va">shao4d_treeStats</span><span class="op">)</span><span class="op">$</span><span class="va">genus</span><span class="op">$</span><span class="va">`g: Phocaeicola`</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Call:</span></span>
<span><span class="co">#&gt; `g: Phocaeicola` ~ Csection</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Coefficients:</span></span>
<span><span class="co">#&gt; (Intercept)     Csection  </span></span>
<span><span class="co">#&gt;     -12.400       -7.057</span></span></code></pre></div>
<p>The stats extracted from each model (with the help of
<code><a href="https://generics.r-lib.org/reference/tidy.html" class="external-link">broom::tidy()</a></code>) are stored in a tibble (data frame).</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/psExtra-accessors.html">taxatree_stats_get</a></span><span class="op">(</span><span class="va">shao4d_treeStats</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 57 √ó 9</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Groups:   rank [5]</span></span></span>
<span><span class="co">#&gt;    term     taxon            rank  formula estimate std.error statistic  p.value</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> Csection p: Proteobacter‚Ä¶ phyl‚Ä¶ `p: Pr‚Ä¶    -<span style="color: #BB0000;">2.58</span>     0.767     -<span style="color: #BB0000;">3.36</span> 8.68<span style="color: #949494;">e</span><span style="color: #BB0000;">- 4</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> Csection p: Bacteroidetes phyl‚Ä¶ `p: Ba‚Ä¶    -<span style="color: #BB0000;">8.73</span>     0.698    -<span style="color: #BB0000;">12.5</span>  3.13<span style="color: #949494;">e</span><span style="color: #BB0000;">-29</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> Csection p: Actinobacter‚Ä¶ phyl‚Ä¶ `p: Ac‚Ä¶    -<span style="color: #BB0000;">4.67</span>     0.679     -<span style="color: #BB0000;">6.88</span> 3.38<span style="color: #949494;">e</span><span style="color: #BB0000;">-11</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> Csection p: Firmicutes    phyl‚Ä¶ `p: Fi‚Ä¶     2.73     0.327      8.36 2.22<span style="color: #949494;">e</span><span style="color: #BB0000;">-15</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> Csection c: Gammaproteob‚Ä¶ class `c: Ga‚Ä¶    -<span style="color: #BB0000;">2.67</span>     0.768     -<span style="color: #BB0000;">3.48</span> 5.72<span style="color: #949494;">e</span><span style="color: #BB0000;">- 4</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> Csection c: Bacteroidia   class `c: Ba‚Ä¶    -<span style="color: #BB0000;">8.73</span>     0.698    -<span style="color: #BB0000;">12.5</span>  3.13<span style="color: #949494;">e</span><span style="color: #BB0000;">-29</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> Csection c: Actinomycetia class `c: Ac‚Ä¶    -<span style="color: #BB0000;">4.47</span>     0.684     -<span style="color: #BB0000;">6.53</span> 2.68<span style="color: #949494;">e</span><span style="color: #BB0000;">-10</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> Csection c: Bacilli       class `c: Ba‚Ä¶     2.56     0.338      7.56 4.93<span style="color: #949494;">e</span><span style="color: #BB0000;">-13</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> Csection c: Coriobacteri‚Ä¶ class `c: Co‚Ä¶    -<span style="color: #BB0000;">2.52</span>     0.451     -<span style="color: #BB0000;">5.59</span> 5.12<span style="color: #949494;">e</span><span style="color: #BB0000;">- 8</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> Csection c: Clostridia    class `c: Cl‚Ä¶     1.58     0.599      2.64 8.64<span style="color: #949494;">e</span><span style="color: #BB0000;">- 3</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ‚Ñπ 47 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ‚Ñπ 1 more variable: p.adj.fdr.rank &lt;dbl&gt;</span></span></span></code></pre></div>
<p>Let‚Äôs make a labelled tree plot:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">treePlotsSimple</span> <span class="op">&lt;-</span> <span class="va">shao4d_treeStats</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="co"># specify which taxa will get labeled (adds a "label" variable to the stats tibble)</span></span>
<span>  <span class="fu"><a href="../reference/taxatree_label.html">taxatree_label</a></span><span class="op">(</span><span class="va">p.adj.fdr.rank</span> <span class="op">&lt;</span> <span class="fl">0.01</span>, <span class="va">rank</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"phylum"</span>, <span class="st">"genus"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="co"># make the plots (1 per predictor variable, so a list of 1 in this example)</span></span>
<span>  <span class="fu"><a href="../reference/taxatree_plots.html">taxatree_plots</a></span><span class="op">(</span></span>
<span>    sig_stat <span class="op">=</span> <span class="st">"p.adj.fdr.rank"</span>, sig_threshold <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>    drop_ranks <span class="op">=</span> <span class="cn">FALSE</span> <span class="co"># drop_ranks = TRUE has a bug in version 0.10.4 :(</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="va">treePlotsSimple</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span>max.level <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="co"># just a list with a single ggplot inside</span></span>
<span><span class="co">#&gt; List of 1</span></span>
<span><span class="co">#&gt;  $ Csection:List of 9</span></span>
<span><span class="co">#&gt;   ..- attr(*, "class")= chr [1:3] "ggraph" "gg" "ggplot"</span></span></code></pre></div>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">treePlotsSimple</span><span class="op">$</span><span class="va">Csection</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="co"># add labels to the plot, only for the taxa indicated earlier</span></span>
<span>  <span class="fu"><a href="../reference/taxatree_plot_labels.html">taxatree_plot_labels</a></span><span class="op">(</span></span>
<span>    taxon_renamer <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, pattern <span class="op">=</span> <span class="st">"[pg]: "</span>, replacement <span class="op">=</span> <span class="st">""</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="cn">TRUE</span>, clip <span class="op">=</span> <span class="st">"off"</span><span class="op">)</span> <span class="op">+</span> <span class="co"># allow scale expansion</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/expansion.html" class="external-link">expansion</a></span><span class="op">(</span>mult <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span><span class="op">)</span> <span class="co"># make space for labels</span></span></code></pre></div>
<p><img src="shao19-analyses_files/figure-html/unnamed-chunk-28-1.png" width="672"></p>
<p>Check out the <a href="https://david-barnett.github.io/microViz/articles/web-only/modelling-taxa.html" title="Statistical modelling of individual taxa">statistical modelling
article</a> for many more tree examples, including covariate-adjusted
regression.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="longitudinal-data">Longitudinal data?<a class="anchor" aria-label="anchor" href="#longitudinal-data"></a>
</h2>
<p>microViz doesn‚Äôt yet contain much functionality designed for
longitudinal data (more specifically, multiple timepoints of microbiome
data from the same individual/site/experimental run/etc.), but there are
already a few ways you can use microViz to help visualise your repeated
samples.</p>
<p>Let us create a dataset containing 10 infants with numerous samples
across multiple ages.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># some dplyr wrangling to get the names of the infants we want!</span></span>
<span><span class="va">repeatedInfants</span> <span class="op">&lt;-</span> <span class="va">shao19</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/psExtra-accessors.html">samdat_tbl</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">family_role</span> <span class="op">==</span> <span class="st">"child"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">subject_id</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>nSamples <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">nSamples</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">subject_id</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span><span class="va">shaoRepeated</span> <span class="op">&lt;-</span> <span class="va">shao19</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/ps_filter.html">ps_filter</a></span><span class="op">(</span><span class="va">subject_id</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">repeatedInfants</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/ps_arrange.html">ps_arrange</a></span><span class="op">(</span><span class="va">subject_id</span>, <span class="va">infant_age</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="co"># put the samples in age order, per infant</span></span>
<span>  <span class="fu"><a href="../reference/ps_mutate.html">ps_mutate</a></span><span class="op">(</span>age_factor <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">infant_age</span><span class="op">)</span><span class="op">)</span> <span class="co"># useful for plotting</span></span>
<span></span>
<span><span class="fu"><a href="../reference/psExtra-accessors.html">samdat_tbl</a></span><span class="op">(</span><span class="va">shaoRepeated</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 48 √ó 15</span></span></span>
<span><span class="co">#&gt;    .sample_name  subject_id family_id sex    family_role   age infant_age</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> A00502_ba_4   A00502_ba  27        female child           0          4</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> A00502_ba_7   A00502_ba  27        female child           0          7</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> A00502_ba_21  A00502_ba  27        female child           0         21</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> A00502_ba_226 A00502_ba  27        female child           0        226</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> A01166_ba_4   A01166_ba  47        female child           0          4</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> A01166_ba_7   A01166_ba  47        female child           0          7</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> A01166_ba_21  A01166_ba  47        female child           0         21</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> A01166_ba_420 A01166_ba  47        female child           0        420</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> A01173_ba_4   A01173_ba  48        female child           0          4</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> A01173_ba_7   A01173_ba  48        female child           0          7</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ‚Ñπ 38 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ‚Ñπ 8 more variables: birth_weight &lt;dbl&gt;, birth_mode &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   c_section_type &lt;chr&gt;, antibiotics_current_use &lt;chr&gt;, number_reads &lt;int&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   Csection &lt;dbl&gt;, birth_mode3 &lt;chr&gt;, age_factor &lt;fct&gt;</span></span></span></code></pre></div>
<p>Try <code><a href="../reference/ord_explore.html">ord_explore()</a></code> on this dataset (or any dataset with
repeated samples arranged in time order).</p>
<p>You can try: colour samples by subject_id, map the size of the points
to infant_age, and try selecting ‚Äúpaths‚Äù and infants from the ‚Äúadd‚Äù
menu, to follow infants trajectories over time.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ord_explore.html">ord_explore</a></span><span class="op">(</span><span class="va">shaoRepeated</span><span class="op">)</span></span></code></pre></div>
<p>Use the ‚ÄúCode‚Äù button to get example code, and you‚Äôll be able to
recreate something fun like this.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">shaoRepeated</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/tax_transform.html">tax_transform</a></span><span class="op">(</span>rank <span class="op">=</span> <span class="st">"genus"</span>, trans <span class="op">=</span> <span class="st">"identity"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/dist_calc.html">dist_calc</a></span><span class="op">(</span>dist <span class="op">=</span> <span class="st">"bray"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/ord_calc.html">ord_calc</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"auto"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/ord_plot.html">ord_plot</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="st">"subject_id"</span>, alpha <span class="op">=</span> <span class="fl">0.5</span>, size <span class="op">=</span> <span class="st">"infant_age"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/add_paths.html">add_paths</a></span><span class="op">(</span></span>
<span>    id_var <span class="op">=</span> <span class="st">"subject_id"</span>,</span>
<span>    id_values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A01166_ba"</span>, <span class="st">"SID523133_ba"</span>, <span class="st">"A01173_ba"</span>, <span class="st">"SID512120_ba"</span>, <span class="st">"SID519128_ba"</span><span class="op">)</span>,</span>
<span>    mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="va">subject_id</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_color_viridis_d</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">"turbo"</span><span class="op">)</span></span></code></pre></div>
<p><img src="shao19-analyses_files/figure-html/unnamed-chunk-31-1.png" width="576"></p>
<p>To learn how to sort samples by time, or in some other manual order,
such as in the example below, check out the <a href="https://david-barnett.github.io/microViz/articles/web-only/compositions.html#sorting-by-time" title="Sort barcharts by time">section on sorting samples by time in the
composition barplots article</a>.</p>
<p><a href="https://david-barnett.github.io/microViz/articles/web-only/compositions.html#sorting-by-time"><img src="https://david-barnett.github.io/microViz/articles/web-only/compositions_files/figure-html/unnamed-chunk-34-1.png"></a></p>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://r-lib.github.io/sessioninfo/reference/session_info.html" class="external-link">session_info</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB; font-weight: bold;">‚îÄ Session info ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ</span></span></span>
<span><span class="co">#&gt;  <span style="color: #555555; font-style: italic;">setting </span> <span style="color: #555555; font-style: italic;">value</span></span></span>
<span><span class="co">#&gt;  version  R version 4.3.1 (2023-06-16)</span></span>
<span><span class="co">#&gt;  os       Ubuntu 22.04.2 LTS</span></span>
<span><span class="co">#&gt;  system   x86_64, linux-gnu</span></span>
<span><span class="co">#&gt;  ui       X11</span></span>
<span><span class="co">#&gt;  language en</span></span>
<span><span class="co">#&gt;  collate  C.UTF-8</span></span>
<span><span class="co">#&gt;  ctype    C.UTF-8</span></span>
<span><span class="co">#&gt;  tz       UTC</span></span>
<span><span class="co">#&gt;  date     2023-07-31</span></span>
<span><span class="co">#&gt;  pandoc   2.19.2 @ /usr/bin/ (via rmarkdown)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB; font-weight: bold;">‚îÄ Packages ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ</span></span></span>
<span><span class="co">#&gt;  <span style="color: #555555; font-style: italic;">package         </span> <span style="color: #555555; font-style: italic;">*</span> <span style="color: #555555; font-style: italic;">version  </span> <span style="color: #555555; font-style: italic;">date (UTC)</span> <span style="color: #555555; font-style: italic;">lib</span> <span style="color: #555555; font-style: italic;">source</span></span></span>
<span><span class="co">#&gt;  ade4               1.7-22    <span style="color: #555555;">2023-02-06</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ape                5.7-1     <span style="color: #555555;">2023-03-13</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  backports          1.4.1     <span style="color: #555555;">2021-12-13</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  Biobase            2.60.0    <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  BiocGenerics       0.46.0    <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  biomformat         1.28.0    <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  Biostrings         2.68.1    <span style="color: #555555;">2023-05-16</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  bitops             1.0-7     <span style="color: #555555;">2021-04-24</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  broom              1.0.5     <span style="color: #555555;">2023-06-09</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  bslib              0.5.0     <span style="color: #555555;">2023-06-09</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ca                 0.71.1    <span style="color: #555555;">2020-01-24</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  cachem             1.0.8     <span style="color: #555555;">2023-05-01</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  callr              3.7.3     <span style="color: #555555;">2022-11-02</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  circlize           0.4.15    <span style="color: #555555;">2022-05-10</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  cli                3.6.1     <span style="color: #555555;">2023-03-23</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  clue               0.3-64    <span style="color: #555555;">2023-01-31</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  cluster            2.1.4     <span style="color: #555555;">2022-08-22</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  codetools          0.2-19    <span style="color: #555555;">2023-02-01</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  colorspace         2.1-0     <span style="color: #555555;">2023-01-23</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  commonmark         1.9.0     <span style="color: #555555;">2023-03-17</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ComplexHeatmap     2.16.0    <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  corncob            0.3.1     <span style="color: #555555;">2022-12-05</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  crayon             1.5.2     <span style="color: #555555;">2022-09-29</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  data.table         1.14.8    <span style="color: #555555;">2023-02-17</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  desc               1.4.2     <span style="color: #555555;">2022-09-08</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  devtools           2.4.5     <span style="color: #555555;">2022-10-11</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  digest             0.6.33    <span style="color: #555555;">2023-07-07</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  doParallel         1.0.17    <span style="color: #555555;">2022-02-07</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  dplyr              1.1.2     <span style="color: #555555;">2023-04-20</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ellipsis           0.3.2     <span style="color: #555555;">2021-04-29</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  evaluate           0.21      <span style="color: #555555;">2023-05-05</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  fansi              1.0.4     <span style="color: #555555;">2023-01-22</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  farver             2.1.1     <span style="color: #555555;">2022-07-06</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  fastmap            1.1.1     <span style="color: #555555;">2023-02-24</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  foreach            1.5.2     <span style="color: #555555;">2022-02-02</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  fs                 1.6.3     <span style="color: #555555;">2023-07-20</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  generics           0.1.3     <span style="color: #555555;">2022-07-05</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  GenomeInfoDb       1.36.1    <span style="color: #555555;">2023-06-21</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  GenomeInfoDbData   1.2.10    <span style="color: #555555;">2023-06-16</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  GetoptLong         1.0.5     <span style="color: #555555;">2020-12-15</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ggforce            0.4.1     <span style="color: #555555;">2022-10-04</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ggplot2          * 3.4.2     <span style="color: #555555;">2023-04-03</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ggraph             2.1.0     <span style="color: #555555;">2022-10-09</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ggrepel            0.9.3     <span style="color: #555555;">2023-02-03</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ggtext             0.1.2     <span style="color: #555555;">2022-09-16</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  GlobalOptions      0.1.2     <span style="color: #555555;">2020-06-10</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  glue               1.6.2     <span style="color: #555555;">2022-02-24</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  graphlayouts       1.0.0     <span style="color: #555555;">2023-05-01</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  gridExtra          2.3       <span style="color: #555555;">2017-09-09</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  gridtext           0.1.5     <span style="color: #555555;">2022-09-16</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  gtable             0.3.3     <span style="color: #555555;">2023-03-21</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  highr              0.10      <span style="color: #555555;">2022-12-22</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  htmltools          0.5.5     <span style="color: #555555;">2023-03-23</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  htmlwidgets        1.6.2     <span style="color: #555555;">2023-03-17</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  httpuv             1.6.11    <span style="color: #555555;">2023-05-11</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  igraph             1.5.0.1   <span style="color: #555555;">2023-07-23</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  IRanges            2.34.1    <span style="color: #555555;">2023-06-22</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  iterators          1.0.14    <span style="color: #555555;">2022-02-05</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  jquerylib          0.1.4     <span style="color: #555555;">2021-04-26</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  jsonlite           1.8.7     <span style="color: #555555;">2023-06-29</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  knitr              1.43      <span style="color: #555555;">2023-05-25</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  labeling           0.4.2     <span style="color: #555555;">2020-10-20</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  later              1.3.1     <span style="color: #555555;">2023-05-02</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  lattice            0.21-8    <span style="color: #555555;">2023-04-05</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  lifecycle          1.0.3     <span style="color: #555555;">2022-10-07</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  magrittr           2.0.3     <span style="color: #555555;">2022-03-30</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  markdown           1.7       <span style="color: #555555;">2023-05-16</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  MASS               7.3-60    <span style="color: #555555;">2023-05-04</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  Matrix             1.5-4.1   <span style="color: #555555;">2023-05-18</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  matrixStats        1.0.0     <span style="color: #555555;">2023-06-02</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  memoise            2.0.1     <span style="color: #555555;">2021-11-26</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  mgcv               1.8-42    <span style="color: #555555;">2023-03-02</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  microbiome         1.22.0    <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  microViz         * 0.11.0    <span style="color: #555555;">2023-07-31</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">local</span></span></span>
<span><span class="co">#&gt;  mime               0.12      <span style="color: #555555;">2021-09-28</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  miniUI             0.1.1.1   <span style="color: #555555;">2018-05-18</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  multtest           2.56.0    <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  munsell            0.5.0     <span style="color: #555555;">2018-06-12</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  nlme               3.1-162   <span style="color: #555555;">2023-01-31</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  patchwork        * 1.1.2     <span style="color: #555555;">2022-08-19</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  permute            0.9-7     <span style="color: #555555;">2022-01-27</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  phyloseq         * 1.44.0    <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  pillar             1.9.0     <span style="color: #555555;">2023-03-22</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  pkgbuild           1.4.2     <span style="color: #555555;">2023-06-26</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  pkgconfig          2.0.3     <span style="color: #555555;">2019-09-22</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  pkgdown            2.0.7     <span style="color: #555555;">2022-12-14</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  pkgload            1.3.2.1   <span style="color: #555555;">2023-07-08</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  plyr               1.8.8     <span style="color: #555555;">2022-11-11</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  png                0.1-8     <span style="color: #555555;">2022-11-29</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  polyclip           1.10-4    <span style="color: #555555;">2022-10-20</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  prettyunits        1.1.1     <span style="color: #555555;">2020-01-24</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  processx           3.8.2     <span style="color: #555555;">2023-06-30</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  profvis            0.3.8     <span style="color: #555555;">2023-05-02</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  promises           1.2.0.1   <span style="color: #555555;">2021-02-11</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ps                 1.7.5     <span style="color: #555555;">2023-04-18</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  purrr              1.0.1     <span style="color: #555555;">2023-01-10</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  R6                 2.5.1     <span style="color: #555555;">2021-08-19</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ragg               1.2.5     <span style="color: #555555;">2023-01-12</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  RColorBrewer       1.1-3     <span style="color: #555555;">2022-04-03</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  Rcpp               1.0.11    <span style="color: #555555;">2023-07-06</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  RCurl              1.98-1.12 <span style="color: #555555;">2023-03-27</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  registry           0.5-1     <span style="color: #555555;">2019-03-05</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  remotes            2.4.2.1   <span style="color: #555555;">2023-07-18</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  reshape2           1.4.4     <span style="color: #555555;">2020-04-09</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  rhdf5              2.44.0    <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  rhdf5filters       1.12.1    <span style="color: #555555;">2023-04-30</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  Rhdf5lib           1.22.0    <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  rjson              0.2.21    <span style="color: #555555;">2022-01-09</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  rlang              1.1.1     <span style="color: #555555;">2023-04-28</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  rmarkdown          2.23      <span style="color: #555555;">2023-07-01</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  rprojroot          2.0.3     <span style="color: #555555;">2022-04-02</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  Rtsne              0.16      <span style="color: #555555;">2022-04-17</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  S4Vectors          0.38.1    <span style="color: #555555;">2023-05-02</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  sass               0.4.7     <span style="color: #555555;">2023-07-15</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  scales             1.2.1     <span style="color: #555555;">2022-08-20</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  seriation          1.5.1     <span style="color: #555555;">2023-07-20</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  sessioninfo        1.2.2     <span style="color: #555555;">2021-12-06</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  shape              1.4.6     <span style="color: #555555;">2021-05-19</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  shiny              1.7.4.1   <span style="color: #555555;">2023-07-06</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  stringi            1.7.12    <span style="color: #555555;">2023-01-11</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  stringr            1.5.0     <span style="color: #555555;">2022-12-02</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  survival           3.5-5     <span style="color: #555555;">2023-03-12</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  systemfonts        1.0.4     <span style="color: #555555;">2022-02-11</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  textshaping        0.3.6     <span style="color: #555555;">2021-10-13</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  tibble             3.2.1     <span style="color: #555555;">2023-03-20</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  tidygraph          1.2.3     <span style="color: #555555;">2023-02-01</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  tidyr              1.3.0     <span style="color: #555555;">2023-01-24</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  tidyselect         1.2.0     <span style="color: #555555;">2022-10-10</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  TSP                1.2-4     <span style="color: #555555;">2023-04-04</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  tweenr             2.0.2     <span style="color: #555555;">2022-09-06</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  urlchecker         1.0.1     <span style="color: #555555;">2021-11-30</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  usethis            2.2.2     <span style="color: #555555;">2023-07-06</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  utf8               1.2.3     <span style="color: #555555;">2023-01-31</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  vctrs              0.6.3     <span style="color: #555555;">2023-06-14</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  vegan              2.6-4     <span style="color: #555555;">2022-10-11</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  viridis            0.6.4     <span style="color: #555555;">2023-07-22</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  viridisLite        0.4.2     <span style="color: #555555;">2023-05-02</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  withr              2.5.0     <span style="color: #555555;">2022-03-03</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  xfun               0.39      <span style="color: #555555;">2023-04-20</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  xml2               1.3.5     <span style="color: #555555;">2023-07-06</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  xtable             1.8-4     <span style="color: #555555;">2019-04-21</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  XVector            0.40.0    <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  yaml               2.3.7     <span style="color: #555555;">2023-01-23</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  zlibbioc           1.46.0    <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #555555;"> [1] /home/runner/work/_temp/Library</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;"> [2] /opt/R/4.3.1/lib/R/site-library</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;"> [3] /opt/R/4.3.1/lib/R/library</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB; font-weight: bold;">‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ</span></span></span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by David Barnett.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: '22e6f0a75b7b1ffae9bf5bc5b58864d7',
    indexName: 'microViz',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script><script data-goatcounter="https://microviz.goatcounter.com/count" async src="//gc.zgo.at/count.js"></script>
</body>
</html>
